---
title: "PLEASE analysis"
author: "Eric Z. Chen"
date: "`r format(Sys.time(), '%Y %B %d ,  %I:%M %p')`"
output:
  html_document:
    fig_width: 8
    number_sections: yes
    toc: yes
    css: style.css
---


```{r note, evel=FALSE,include=FALSE,cache=FALSE,echo=FALSE}
#### rewrite the plot_MDS function
######################################

#### The total reads table is not correct
#### 6004-02 -> 6004-03
#### 6004-03 -> 6004-04


```




```{r load_packages, echo=FALSE,include=FALSE,cache=FALSE}
####### install necessary packages
packages <- c(#'phyloseq',
              'minerva',
              'foreach', ## wilcox_test
              'doParallel', ## wilcox_test
              'pROC',## plot.roc
              ## The arm package will try to unload MASS and cause an error
              ## clean working space and re-start R to re-run the code
              'arm', ## coefplot
              'geiger', ## tips
              'quantreg', 
              'picante', ## prune.sample
              'ape', ## read.tree
              'knitr', ## kable
              'fpc', ## pamk
              'pheatmap','reshape2',
              'randomForest','devtools','ggplot2','lazyeval',
              'vegan','tidyr','dplyr','readr')
### unload ggplot2, because it conflits with arm
try(unloadNamespace("ggplot2"), silent = TRUE)
### load arm first then ggplot2
for (pk in packages) {library(pk,character.only = TRUE,quietly = TRUE)}

### !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
### How to reproduce my analysis with the same package
### Copy my packrat.lock file into your packrat folder
### Run packrat::status and packrat::restore
### !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!


### use packrat to manage the packages
### do not install them manually
### If you want to install them manually, change FALSE to TRUE
if (FALSE) {
  for (pk in packages) {
    if (!require(pk,character.only = TRUE)) {
      if (pk == 'pheatmap') {
          devtools::install_github("raivokolde/pheatmap")
      }
      else {
        install.packages(pk)
      }
    }
    require(pk,character.only = TRUE,quietly = TRUE)
  }
}

#####!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!##########
### dplyr::select() conflicts with MASS::select()
#####!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!##########

#########################
### change the working directory
try(setwd('5_Analysis/'))


#####################################################
#### load the package instead of sourcing the code in the future  
source('../../14_Metagenomic_toolkit/metagenomic_toolkit/R/merge_two_tables.R')
source('../../14_Metagenomic_toolkit/metagenomic_toolkit/R/subset_data_by_sample_info.R')
source('../../14_Metagenomic_toolkit/metagenomic_toolkit/R/calculate_distance.R')
source('../../14_Metagenomic_toolkit/metagenomic_toolkit/R/find_best_clustering.R')
source('../../14_Metagenomic_toolkit/metagenomic_toolkit/R/wilcox_test.R')
source('../../14_Metagenomic_toolkit/metagenomic_toolkit/R/random_forest.R')
source('../../14_Metagenomic_toolkit/metagenomic_toolkit/R/PERMANOVA_analysis.R')
source('../../14_Metagenomic_toolkit/metagenomic_toolkit/R/linear_regression_with_permutation.R')


####################################################
opts_chunk$set(#fig.path = '~/3_Results/',
#  root.dir = '../3_Results/',
  progress = TRUE, verbose = TRUE,
  warning=FALSE, message=FALSE,echo=TRUE)
```


----
  
----
  
```{r global_variables}
reload.data <- TRUE
#try(setwd('5_Analysis/'))
data.folder <- '../1_Data/'
#######
TLevel<- 'G'
##'Binary_Jaccard', 'Numerical_Jaccard', 'UniFrac'
bacteral.distance.method <- "Binary_Jaccard"
#Distance.Method <- "Bray"
```


```{r define_colors,include=FALSE,echo=TRUE}
cgreen <- rgb(77,175,74,maxColorValue =255)
cpurple <- rgb(152,78,163,maxColorValue =255)
cblue <- rgb(55,126,184,maxColorValue =255)
cred <- rgb(228,26,28,maxColorValue =255)
```

```{r functions}

add_back_rownames = function(df,row.var){
  df <- as.data.frame(df)
  rownames(df) <- df[,row.var]
  df <- df[,colnames(df) != row.var,drop=F]
  return(df)
}

count_samples_by_rownames = function(data.mat){
  P.T1 <- length(grep('-01',rownames(data.mat)))
  P.T2 <- length(grep('-02',rownames(data.mat)))
  P.T3 <- length(grep('-03',rownames(data.mat)))
  P.T4 <- length(grep('-04',rownames(data.mat)))
  Comb <- length(grep('-',rownames(data.mat),invert=TRUE))
  cat('COMBO samples ',Comb ,'\n')
  cat('PLEASE-T1 samples ',P.T1 ,'\n')
  cat('PLEASE-T2 samples ',P.T2 ,'\n')
  cat('PLEASE-T3 samples ',P.T3 ,'\n')
  cat('PLEASE-T4 samples ',P.T4 ,'\n')
}
```


---------------------------------------------------------------
***************************************************************
# Load raw data
***************
  
## Load bacterial data
Load COMBO (normal) and PLEASE (disease) bacterial data. Load total read counts for each sample.
```{r load_bacterial_data, cache=TRUE}
PLEASE.file <- paste('../1_Data/Raw_Data/MetaPhlAn/PLEASE/',TLevel,
                     '_Remove_unclassfied_Renormalized_Merge_Rel_MetaPhlAn_Result.xls',sep='')
PLEASE.raw <- read.table(PLEASE.file,sep='\t',header=TRUE,row.names = 1,check.names=FALSE,stringsAsFactors=FALSE)
PLEASE.raw <- t(PLEASE.raw)
cat('samples','taxa',dim(PLEASE.raw),'\n')
PLEASE.raw[1:3,1:3]


COMBO.file <- paste('../1_Data/Raw_Data/MetaPhlAn/COMBO/',TLevel,
                    '_Remove_unclassfied_Renormalized_Merge_Rel_MetaPhlAn_Result.xls',sep='')
COMBO.raw <- read.table(COMBO.file,sep='\t',header=TRUE,row.names = 1,check.names=FALSE,stringsAsFactors=FALSE)
COMBO.raw <- t(COMBO.raw)
cat('samples','taxa',dim(COMBO.raw),'\n')
COMBO.raw[1:3,1:3]
```


```{r load_human_read_information,cache=TRUE}
###################################
### load total non-human read counts
human.read.file <- '../1_Data/Raw_Data/MetaPhlAn/Human_Reads/please_combo_human_reads.xls'
human.read <- read.table(human.read.file,sep='\t',header=TRUE,row.names=1,stringsAsFactors=FALSE)
```

### Merge PLEASE data and COMBO data. 
```{r merge_please_combo_data, cache=TRUE}
taxa.raw <- merge_two_tables(COMBO.raw,PLEASE.raw,
                             fill.zero = TRUE,rnames = 'union',cnames = 'union')
cat('samples','taxa',dim(taxa.raw),'\n')
taxa.raw[1:3,1:3]
```

### Distribution of total non-human reads
```{r distribution_of_total_non_human_reads,cache=TRUE}
hist(log10(human.read$NonHumanReads),xlab='log10 non-human reads',
     main='',breaks=40)

barplot(log10(human.read$NonHumanReads[order(human.read$NonHumanReads)]),
        ylab='log10 non-human reads', main='')
```


### Filter low depth samples
Filter out low sequencing depth samples.Filter out samples with non-humand reads < 30000.
```{r filter_low_depth_samples,cache=TRUE}
low.depth.samples <- subset(human.read,NonHumanReads<30000)
low.depth.samples[,1:5]
### Delete these samples from PLEASE data.
### Due to the low sequencing depth, some of the samples have no MetaPhlAn output
### These are the samples with MetaPhlAn output but also low sequencing
rownames(taxa.raw)[which(rownames(taxa.raw) %in% rownames(low.depth.samples))]
### Before deletion
dim(taxa.raw)
### After deletion
taxa.raw <- taxa.raw[-which(rownames(taxa.raw) %in% rownames(low.depth.samples)),]
dim(taxa.raw)
## 5018-03,7001-02,7003-03,6005-04,7011-02,7003-02,7009-03, 7010-03
```

### Filter low abundant bacterial data
Filter low abundant bacteria. I'm going to filter out the low abundant species.The species must present in at least 10% of the samples and the max abundance across samples must large than 1%. Renormalize the species abundance so that the sum of the species abundance is 100 (%)
```{r filter_low_abundant_bacteria, cache=TRUE, fig.width=4,fig.height=4}
filter.index1 <- apply(taxa.raw,2,function(X){sum(X>0)>0.1*length(X)})
filter.index2 <- apply(taxa.raw,2,function(X){max(X)>1})
taxa.filter <- taxa.raw[,filter.index1 & filter.index2]
taxa.filter <- 100*sweep(taxa.filter, 1, rowSums(taxa.filter), FUN="/")
cat('after filter:','samples','taxa',dim(taxa.filter),'\n')
head(rowSums(taxa.filter))

## Did the filtering process change the abundance ?
plot(taxa.raw[,colnames(taxa.filter)],taxa.filter,
xlab='before filtering',ylab='after filtering')
```


```{r save_processed_taxa_abundance_data,cache=TRUE}
### save the processed taxa abundance
write.csv(taxa.filter,file=paste('../1_Data/Processed_Data/MetaPhlAn/',TLevel,'_Processed_MetaPhlAn_Abundance.csv',sep=''))

### Remove s__ before the taxa name
colnames(taxa.filter) <- colsplit(colnames(taxa.filter), '__', c('V1','V2'))[,'V2']

####### change the variable name
taxa.data <- as.data.frame(taxa.filter)
```

### Distribution of identified taxa 
Check the distribution of taxa called in each sample.
```{r distribution_of_identified_taxa,cache=TRUE, fig.width=4,fig.height=4}
species.count <- apply(taxa.filter,1,function(X)sum(X>0))
hist(apply(taxa.filter,1,function(X)sum(X>0)),breaks=20,
main=paste('Distribution of taxa called'),
xlab=paste('Number of taxa called in each sample'))
### What is the smallest number?
head(sort(apply(taxa.filter,1,function(X)sum(X>0))))

### Correlation between number of taxa called and sequencing depth
### total non-human reads
plot(log10(human.read[names(species.count),'NonHumanReads']),log10(species.count),ylab='log10 Number of taxa',xlab='log10 Non-human reads')
### total human reads
plot(log10(human.read[names(species.count),'HumanReads']),log10(species.count),ylab='log10 Number of taxa',xlab='log 10 Human reads')
```


## Load rarefraction data
```{r load_rarefraction_data,cache=TRUE}
PLEASE.rare.file <- paste('../1_Data/Raw_Data/Rarefraction_Data/PLEASE_MetaPhlAn/',TLevel,'_Remove_unclassfied_Renormalized_Merge_Rel_MetaPhlAn_Result.xls',sep='')
PLEASE.rare.raw <- read.table(PLEASE.rare.file,sep='\t',header=TRUE,row.names = 1,check.names=FALSE,stringsAsFactors=FALSE)

COMBO.rare.file <- paste('../1_Data/Raw_Data/Rarefraction_Data/COMBO_MetaPhlAn/',TLevel,'_Remove_unclassfied_Renormalized_Merge_Rel_MetaPhlAn_Result.xls',sep='')
COMBO.rare.raw <- read.table(COMBO.rare.file,sep='\t',header=TRUE,row.names = 1,check.names=FALSE,stringsAsFactors=FALSE)

## merge the data
taxa.rare.raw <- merge_two_tables(t(COMBO.rare.raw),t(PLEASE.rare.raw),
                             fill.zero = TRUE,rnames = 'union',cnames = 'union')
```

## Load phylogenetic tree
Phylogenetic tree information was downloaded from PhyloPhlAn website.
```{r load_phylogenetic_tree,cache=TRUE}

get_unique_leaves = function(phylo.tree){
#phylo.tree
#str(phylo.tree)
taxa.uniq <-unique(phylo.tree$tip.label)
tree.tmp <- phylo.tree 
for (taxa in taxa.uniq){
  if(sum(tree.tmp$tip.label==taxa) > 1){
    taxa.tip <- tree.tmp$tip.label[tree.tmp$tip.label==taxa]
    taxa.tip[2:length(taxa.tip)] <- paste(taxa.tip[2:length(taxa.tip)],'drop',sep='.')
    tree.tmp$tip.label[tree.tmp$tip.label==taxa] <- taxa.tip
    tree.tmp <-drop.tip(tree.tmp, paste(taxa,'drop',sep='.'))
    #break
  }
}
#plot(tree.tmp)

tips.to.be.drop <- unique(tree.tmp$tip.label)[
                    grep('__',unique(tree.tmp$tip.label),invert=TRUE)]
tree.tmp <-drop.tip(tree.tmp,tips.to.be.drop)

#plot(tree.tmp)
return(tree.tmp)
}


## This is from PhyloPhlAn, the tree has repeated leaves
phylo.tree <- read.tree(paste("../1_Data/Raw_Data/Phlogenetic_tree/",
                        TLevel,"_change_name_bs_tree.reroot.nwk",sep=''))

phylo.tree.uniq = get_unique_leaves(phylo.tree)

## This is from greengene database
#phylo.tree <- read_tree_greengenes("../1_Data/Raw_Data/Phlogenetic_tree/greengene_12_10_threshold97_otus.tree")

```

## Load kmer data
```{r load_kmer_data, cache=TRUE}
kmer.raw <- t(read.table('../1_Data/Raw_Data/Kmer/DSK_Kmer6/All_Samples_kmer6_remove_human_dna.txt',header=TRUE,row.names=1,check.names = FALSE))

###### Normalize kmer
kmer.nor <- 100*sweep(kmer.raw,1, rowSums(kmer.raw), FUN="/")
rowSums(kmer.nor)[1:5]
```


## Load nutrient data
```{r load_nutrient_data, cache=TRUE, eval=FALSE}
nutr0 <- read_delim('../1_Data/Raw_Data/Nutrient/COMBO NDS spreadsheed deidentified11-30-10-1.txt',delim='\t')
nutr1 <- read_delim('../1_Data/Raw_Data/Nutrient/PLEASE Week 0 Recalls 05-7-13.txt',delim='\t')
nutr2 <- read_delim('../1_Data/Raw_Data/Nutrient/PLEASE Week 1 Recalls 5-7-13.txt',delim='\t')
nutr3 <- read_delim('../1_Data/Raw_Data/Nutrient/PLEASE Week 4 Recalls 5-7-13.txt',delim='\t')
nutr4 <- read_delim('../1_Data/Raw_Data/Nutrient/PLEASE Week 8 Recalls NDS Please8 5-7-13.txt',delim='\t')


##remove duplicated column "Column_intentionally_left_blank"
cnames <- Reduce(intersect, list(
                       colnames(nutr0),
                       colnames(nutr1),
                       colnames(nutr2),
                       colnames(nutr3),
                       colnames(nutr4)))
cnames <- cnames[grep('intentionally',cnames,invert=TRUE)]
cnames <- cnames[grep('User',cnames,invert=TRUE)]
##bind all samples together
nutr <- bind_rows(nutr0[,cnames],nutr1[,cnames],nutr2[,cnames],nutr3[,cnames],nutr4[,cnames])
#### change the col names
cnames <- gsub('\\(|\\)|,|\\+|&|\\[|\\]|;','',cnames)
cnames <- gsub('%','Per',cnames)
cnames <- gsub('__|-|:','_',cnames)
cnames <- gsub(' ','_',cnames)
colnames(nutr) <- cnames

# ### average the three visits for each subject
# food.itm <- nutr %>% 
# tidyr::separate(Participant_ID, into=c('Subject','Time'), sep = "\\.") %>%
# as.data.frame
# 
# ### normalize by total caloric intake
# cal <- food.itm[,"Total_Grams"]
# 
# for(nu in colnames(food.itm)[-(1:5)]){
#   print(nu)
#   nu <- "Pectins_g"
#  	y <- food.itm[,nu]
    ## NA in the data
    ## different number of data points
#  	r <- resid(lm(y ~ cal))
#  	food.itm[, nu] <- r
# }
# # # Standardize food 
# food.itm <- scale(food.itm)

rm(cnames)
```

## Load fungi data
```{r load_fungi_data,cache=TRUE}
##  The plural of fungus is fungi
##  Fungal cells

### Kyle gave me only the following fungi data
#Candida albicans
#Clavispora lusitaniae (a.k.a. Candida lusitaniae)
#Cyberlindnera jadinii (a.k.a. Pichia jadinii)
#Epichloe festucae (grass fungus)
#Epichloe glyceriae (grass fungus)
#Grosmannia clavigera (tree fungus)
#Kluyveromyces marxianus
#Saccharomyces cerevisiae

### only used the following ones
fungi <- c("Candida albicans","Clavispora lusitaniae","Cyberlindnera jadinii",
"Kluyveromyces marxianus","Saccharomyces cerevisiae")
#######################################################################
######## Fungi RPKM = (read counts /1000000) / (genome length / 1000)
### RPKM = reads per kilobase per million
### = [# of mapped reads]/[length of transcript in kilo base]/[million mapped reads]
### = [# of mapped reads]/([length of transcript]/1000)/([total reads]/10^6)

### Count
PLEASE.fungi.counts <- read.csv('../1_Data/Raw_Data/Fungi/please_final_fungi_genome_readcounts.csv',stringsAsFactors=FALSE)
COMBO.fungi.counts <- read.csv('../1_Data/Raw_Data/Fungi//combo_final_fungi_genome_readcounts.csv',stringsAsFactors=FALSE)
fungi.counts <- rbind(PLEASE.fungi.counts,COMBO.fungi.counts) %>% 
dplyr::rename(Sample=SampleID) %>%
filter(Species %in% fungi)

###### !!!!!!!!!!!!!!!!!!!!
#### we may need to remove the low depth samples
#### I did not do it for fungi data
##### !!!!!!!!!!!!!!!!!!!!!!

### RPKM
fungi.RPKM <- fungi.counts %>%
mutate(count.len = NumReads/(GenomeLength/1000)) %>% 
dplyr::select(-GenomeLength,-NumReads) %>% 
left_join(dplyr::select(add_rownames(human.read,var='Sample'),Sample,NonHumanReads),by=c('Sample')) %>% 
mutate(Abundance=count.len/(NonHumanReads/1000000)) %>%
dplyr::select(Sample,Species,Abundance) %>% 
spread(Species,Abundance) %>%
na.omit  %>%
add_back_rownames(row.var='Sample')

### Total fungi percentage
fungi.per <- fungi.counts %>% 
group_by(Sample) %>%
summarise(total.counts=sum(NumReads,na.rm=T)) %>% 
left_join(dplyr::select(add_rownames(human.read,var='Sample'),Sample,NonHumanReads),by=c('Sample')) %>% 
mutate(Fungi.Per=100*total.counts/NonHumanReads) %>%
dplyr::select(Sample,Fungi.Per) %>%
na.omit() %>%
add_back_rownames(row.var='Sample')

```


```{r save_processed_fungi_data}
write.csv(fungi.RPKM,file=paste('../1_Data/Processed_Data/Fungi/Processed_Fungi_RPKM.csv',sep=''),row.names=TRUE)
```


## Load gene/pathway/module data
Only 22 COMBO samples in the raw data. Four samples (4002,4006,4007,4019) have empty output files.
```{r load_gene_pathway_module_data,cache=TRUE}
Genefile <- '../1_Data/Raw_Data/Gene_Pathway/Gene_Abundance_from_02b_mpt.txt'
Gene.raw <- t(read.table(Genefile,sep='\t',header=T,row.names = 1,check.names=F,stringsAsFactors = FALSE))
## Remove all 'hypothetical_protein'
## They all have the same names
Gene.raw <- Gene.raw[,grep('hypothetical_protein',colnames(Gene.raw),invert = TRUE)]
## Remove NA colnames
Gene.raw <- Gene.raw[,grep('__NA$',colnames(Gene.raw),invert = TRUE)]
## Remove low depth samples
Gene.raw <- Gene.raw[-which(rownames(Gene.raw) %in% rownames(low.depth.samples)),]
###
Gene.raw[1:3,1:3]
dim(Gene.raw)


Pathwayfile <- '../1_Data/Raw_Data/Gene_Pathway/Pathway_Abundance_from_04b_mpt.txt'
Pathway.raw <- t(as.matrix(read.table(Pathwayfile,sep='\t',header=T,row.names = 1,check.names=F,stringsAsFactors = FALSE)))
## Remove NA colnames
Pathway.raw <- Pathway.raw[,grep('__NA$',colnames(Pathway.raw),invert = TRUE)]
## Remove low depth samples
Pathway.raw <- Pathway.raw[-which(rownames(Pathway.raw) %in% rownames(low.depth.samples)),]
###
Pathway.raw[1:3,1:3]
dim(Pathway.raw)


Modulefile <- '../1_Data/Raw_Data/Gene_Pathway/Module_Abundance_from_04b_mpt.txt'
Module.raw <- t(as.matrix(read.table(Modulefile,sep='\t',header=T,row.names = 1,check.names=F,stringsAsFactors = FALSE)))
## Remove NA colnames
Module.raw <- Module.raw[,grep('__NA$',colnames(Module.raw),invert = TRUE)]
## Remove duplicated entries
## Module.raw <- Module.raw[,grep('Putative_ABC_transport_system',colnames(Module.raw),invert = TRUE)]
## Module.raw <- Module.raw[,grep('Manganeseiron_transport_system',colnames(Module.raw),invert = TRUE)]
## Remove low depth samples
Module.raw <- Module.raw[-which(rownames(Module.raw) %in% rownames(low.depth.samples)),]
###
Module.raw[1:3,1:3]
dim(Module.raw)


G2Pfile <- '../1_Data/Raw_Data/Gene_Pathway/GID2Pathway.txt'
G2P <- (read.table(G2Pfile,sep='\t',header=F,check.names=F,stringsAsFactors = FALSE))
```

### Filter gene/pathway/module data
```{r filter_gene_pathway_module_data,cache=TRUE}
#apply(Gene.raw,2,max)
findx1 <- apply(Gene.raw,2,function(X){sum(X>0) > 10})
findx2 <- apply(Gene.raw,2,function(X){max(X) > 0.1})
gene.data <- Gene.raw[,findx1&findx2]
gene.data <- 100*sweep(gene.data, 1, rowSums(gene.data), FUN="/")
rowSums(gene.data)[1:3]
dim(gene.data)

## renomalize to 100%?
findx1 <- apply(Pathway.raw,2,function(X){sum(X>0) > 10})
findx2 <- apply(Pathway.raw,2,function(X){max(X) > 0.1})
pathway.data <- Pathway.raw[,findx1&findx2]
pathway.data <- 100*sweep(pathway.data, 1, rowSums(pathway.data), FUN="/")
rowSums(pathway.data)[1:3]
dim(pathway.data)

module.data <- Module.raw

###
write.csv(gene.data,file='../1_Data/Processed_Data/Pathway/Gene_Filtered_Abundance.csv',row.names=TRUE)
write.csv(pathway.data,file='../1_Data/Processed_Data/Pathway/Pathway_Filtered_Abundance.csv',row.names=TRUE)

```




**************************************
## Load metabolite data and preprocess
**************************************

### Load metabolite data
```{r load_metabilte_data,cache=TRUE}
modify_header = function(dat){
  ### PLEASE samples: "PLEASE_D.5027_01" 
  ### COMBO samples: "PLEASE_C.4010_01" 
  ### remove the letters and change '_' to '-'
  colnames(dat) <- gsub('PLEASE_D\\.','',colnames(dat))
  norm.ind <- grep('PLEASE_C',colnames(dat))
  colnames(dat)[norm.ind] <- gsub('PLEASE_C\\.','',colnames(dat)[norm.ind])
  colnames(dat)[norm.ind] <- gsub('_01','',colnames(dat)[norm.ind])
  colnames(dat) <- gsub('D\\.','',colnames(dat))
  norm.ind <- grep('C\\.',colnames(dat))
  colnames(dat)[norm.ind] <- gsub('C\\.','',colnames(dat)[norm.ind])
  colnames(dat)[norm.ind] <- gsub('_01','',colnames(dat)[norm.ind])
  colnames(dat) <- gsub('_','-',colnames(dat))
  return(dat)
}

####################################
#### need to change the sample name 
#### so that they can match with those in taxa analysis
mdata1 <- read.csv('../1_Data/Raw_Data/Metabolite/14_1223_C8-pos_raw_PLEASE.csv',skip = 1,header=T,stringsAsFactors=F,na.strings = "")
mdata1 <- modify_header(mdata1)
mdata2 <- read.csv('../1_Data/Raw_Data/Metabolite/14_1223_C18-neg_raw_PLEASE.csv',skip = 1,header=T,stringsAsFactors=F,na.strings = "")
mdata2 <- modify_header(mdata2)
mdata3 <- read.csv('../1_Data/Raw_Data/Metabolite/14_1223_HILIC-pos_raw_PLEASE.csv',skip = 1,header=T,stringsAsFactors=F,na.strings = "")
mdata3 <- modify_header(mdata3)
mdata4 <- read.csv('../1_Data/Raw_Data/Metabolite/15_0130_PLEASE_stool_HILIC-neg_CMH_PLEASE.csv',skip = 1,header=T,stringsAsFactors=F,na.strings = "NA")
mdata4 <- modify_header(mdata4)
############################################
##### check the data
cat("metabolites","samples",dim(mdata1),'\n')
cat("metabolites","samples",dim(mdata2),'\n')
cat("metabolites","samples",dim(mdata3),'\n')
cat("metabolites","samples",dim(mdata4),'\n')
head(mdata1[1:3,1:10])
head(mdata2[1:3,1:10])
head(mdata3[1:3,1:10])
head(mdata4[1:3,1:10])
#### !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! ####
#### dataset 4 has different headers than the other datasets
#### !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! ####
```


```{r save_metabolite_information,cache=TRUE}
### save the metabolite infomration
### 1-6 columns are the metabolite information
### may use such information in the later analysis
minfo1 <- mdata1[,1:6]
minfo2 <- mdata2[,1:6]
minfo3 <- mdata3[,1:6]
minfo4 <- mdata4[,1:6]
### minfo4 has different colnames than the other three
colnames(minfo4) <- colnames(minfo1)
### Combine the four dataframes into one 
### metabolite annotation information
minfo <- rbind(minfo1,minfo2,minfo3,minfo4)
minfo <- minfo[!duplicated(minfo),]
rownames(minfo) <- gsub(' ','.', paste(minfo[,'Method'],minfo[,'Compound'],
                         minfo[,'HMDB.ID'],minfo[,'Metabolite'],sep='_'))
  
rm(minfo1)
rm(minfo2)
rm(minfo3)
rm(minfo4)
```


### Normalize metabolome data
```{r normalize_metabolome_data,cache=TRUE}
normalize_data = function(mdata){
  ##### mdata <- mdata3
  
  #### duplicated metablites, keep the first one
  #### duplicated() returns all except first duplicated elements
  mdata <- mdata[!duplicated(mdata[,'Compound']),]
  
  #### remove the metabolite information column 1-6
  #### use Compund information as the rownames of the data matrix
  #### Note: the current version of R allows the same row names
  dat <- as.matrix(mdata[,-c(1:6)])
  rownames(dat) <- gsub(' ','.', paste(mdata[,'Method'],mdata[,'Compound'],
                         mdata[,'HMDB.ID'],mdata[,'Metabolite'],sep='_'))
  
  #### remove metabolites with all NAs & all zeros
  dat <- dat[rowSums(dat,na.rm=T)>0,]
  #### remove metabolites with too many missing values or zeros.
  #### can not use strigent cutoff, all samples except some 
  #### small treatment group many have zeros.
  dat <- dat[apply(dat,1,function(X){sum(is.na(X)|(X==0))<0.8*length(X)}),]
  
  #### The distribution of NAs and zeros in each metabolites
  hist(apply(dat,1,function(X){sum(is.na(X) | (X==0))}),
       breaks=100,xlab='No. of NAs and zeros',ylab='No. of metabolites',
       main = 'Distribution of NA and zero')
  
  #### impute with min value
  #### consider 0 as NA
  #### some problems when using apply 
  dat.imp <- dat
  for (i in 1:nrow(dat.imp)){
    X <- dat.imp[i,]
    #if (is.infinite(min(X,na.rm=T))){print(i)}
    X[X==0] <- NA
    X[is.na(X)]<-min(X,na.rm=T)
    dat.imp[i,] <- X
  }
  
  #### remove outliers
  #### do not remove outliers here, let the downstream analysis handle this
  #### but be careful in the normalization step
  
  #### remove control samples such as "XPF.29"
  dat.imp <- dat.imp[,grep('\\.',colnames(dat.imp),invert=TRUE)]
  
  #### normalize the data: normalize by sample, use 80%, 90% or 95% quantile cumlative sum
  dat.nor <- sweep(dat.imp,2,apply(dat.imp,2,function(X){sum(X[X<quantile(X,0.90)])}),FUN='/')
  #### take the log and make the data non-negative
  dat.nor <- log10(dat.nor)+10
  
  ########## check the normalization
  ### boxplot
  boxplot(log10(dat.imp),outpch='.',
          ylab='log10 abundance',xlab='samples',main='before normalization')
  boxplot(dat.nor,outpch='.',
          ylab='normalized abundance',xlab='samples',main='after normalization')
  ### Correlation
  ran.ind <- sample(1:(nrow(dat.imp)*ncol(dat.imp)),10000)
  plot(log10(dat.imp)[ran.ind],dat.nor[ran.ind],pch=20,xlab='before normalization (log10)',ylab='after normlization (log10)',main='')
  mtext(paste('cor=',signif(cor(as.vector(log10(dat.imp)),as.vector(dat.nor)),4)))
  
  
  return(dat.nor)
}  
mdata1.nor <- normalize_data(mdata1)
mdata2.nor <- normalize_data(mdata2)
mdata3.nor <- normalize_data(mdata3)
mdata4.nor <- normalize_data(mdata4)

### something wierd about the last several samples in mdata4
### they are the samples with X.xxx

dim(mdata1)
dim(mdata1.nor)
dim(mdata2)
dim(mdata2.nor)
dim(mdata3)
dim(mdata3.nor)
dim(mdata4)
dim(mdata4.nor)
```



```{r merge_all_datasets,cache=TRUE}
snames <- intersect(intersect(intersect(colnames(mdata1.nor),colnames(mdata2.nor)),colnames(mdata3.nor)),colnames(mdata4.nor))
mdata.nor <- rbind(mdata1.nor[,snames],mdata2.nor[,snames],
                   mdata3.nor[,snames],mdata4.nor[,snames])
dim(mdata.nor)
mdata.nor[1:5,1:5]

## all metabolites
### rows:samples  columns:metabolites
allmdata.nor <- t(mdata.nor)
### four "internal standard" can be found. keep them for the following analysis

## km: known metabolite
## remove any metabolites with NA in the name
## this will also remove QI***** metabolites, it has no HMDB ID but metabolite name
kmdata.nor <- t(mdata.nor[grep('_NA',rownames(mdata.nor),invert=TRUE),])
## change the metabolite names
colnames(kmdata.nor) <- minfo[colnames(kmdata.nor),'Metabolite']
kmdata.nor[1:5,1:5]

##############################
### save the normalized metabolite abundance
write.csv(t(allmdata.nor),file='../1_Data/Processed_Data/Metabolite/All_Metabolite_Normalized_Abundance.csv',row.names=TRUE)
write.csv(t(kmdata.nor),file='../1_Data/Processed_Data/Metabolite/Known_Metabolite_Normalized_Abundance.csv',row.names=TRUE)

rm(snames)
```


```{r load_hmdb_kegg_mapping,cache=TRUE}
### Error with read.table
### Try RStudi import and get the following code
hmdb.info <- read.delim("../1_Data/Raw_Data/HMDB/hmdb_3.6.txt", quote="", stringsAsFactors=FALSE,na.strings = "")
head(hmdb.info)
dim(hmdb.info)

### merge into minfo by HMDB ID
minfo <- 
  minfo %>% add_rownames(var = "MID") %>%
  left_join(hmdb.info,by= c("HMDB.ID"="HMDB")) %>% 
  add_back_rownames(row.var = 'MID')
minfo[1:3,]

####
write.csv(minfo,file=paste('../1_Data/Processed_Data/Metabolite/Metabolites_Infor.csv',sep=''),row.names=TRUE)
```

```{r load_84_metabolites,cache=TRUE}
m84id <- read.csv('../1_Data/Raw_Data/Metabolite/20150730_84_metabolites_v2.csv',header=T,stringsAsFactors=F,na.strings = "")
head(m84id)

### KEGG has duplicated IDs in  minfo
### one KEGG ID mapps to multiple HMDB IDs
m84data.nor <- t(mdata.nor[rownames(minfo[minfo$CO %in% m84id$KEGGNumber,]),])


### Get the metabolites that in our dataset
minfo[minfo$CO %in% m84id$KEGGNumber,] %>%
add_rownames(var='MID') %>%
right_join(m84id,by=c("CO"="KEGGNumber")) %>%
dplyr::select(MID,Metabolite,ChemicalName) %>%
write.csv(file=paste('../3_Result/2015_03_05_Detectable_UCLA_Metabolites.csv',sep=''),row.names=FALSE)


## The following metabolites are not present in our metabolome data
## C00003	NAD+	Inhibitory  HMDB00902 ## Not found
## C00008	ADP	Inhibitory
## C00010	CoA	Inhibitory
## C00015	UDP	Enhance
## C00016	FAD	Inhibitory
## C00019	S-Adenosyl-L-methionine	Inhibitory
## C00020	AMP	Enhance
## C00024	Acetyl-CoA	Inhibitory
```

## Load clinical outcome information 
```{r load_clinical_information,cache=TRUE}
#### Treatment, FCP, Response
outcome.file <- '../1_Data/Raw_Data/Clinical_Information/Final_Outcome_Table_ReFormat_08_11_2014.xls'
outcome <- read.table(outcome.file,sep='\t',header=T,row.names = 1,check.names=F,stringsAsFactors=FALSE)

##############
Baseline.raw <- read.table('../1_Data/Raw_Data/Clinical_Information/PLEASE-BASELINE.txt',sep='\t',header=T,na.strings = "NULL",stringsAsFactors=FALSE,colClasses = c( "character" ))
Baseline <- Baseline.raw[,c('Pid','Bas12','Bas12a1e','Bas12a2e', 'Bas12a3e','Bas12a4e','Bas18','Bas19','Bas13','Bas14','Bas15','Bas16')]
rownames(Baseline) <-  Baseline.raw[,'Pid']

#############
DISACT.raw <- read.table('../1_Data/Raw_Data/Clinical_Information/PLEASE-DISACT-modified.txt', sep='\t',header=T,na.strings = "NULL",stringsAsFactors=FALSE,colClasses = c( "character" ))
## Pid,Vnum,Vdate,pcdai
DISACT <- DISACT.raw[,c('Pid','Vnum','Vdate','pcdai',
'AbdominalPain','Bleeding',
'StoolConsistency','Frequency',
'NocturnalBMs','Activity',
'heigthscore')]
rownames(DISACT) <- paste(DISACT.raw[,'Pid'],'-0',as.numeric(DISACT.raw[,'Vnum'])+1,sep='')

##########
Anatomy.raw <- read.table('../1_Data/Raw_Data/Clinical_Information/Anatomic_Distribution_Descriptive_data_for_Eric.txt',sep='\t',header=T,na.strings = "NULL",stringsAsFactors=FALSE,colClasses = c( "character" ))
Anatomy <- Anatomy.raw[,c('duration','esophagus','stomach',
'duod','jejun','ileum',
'cecum','colon','rectum','colorectal','upper',
'current_perianal','penetrating','stricturing')]
rownames(Anatomy) = Anatomy.raw[,'pid']

###########
antibio.raw <- read.table('../1_Data/Raw_Data/Clinical_Information/antibiotics_use_at_each_vist.txt',sep='\t',header=T,stringsAsFactors = FALSE)
rownames(antibio.raw) <- paste(antibio.raw[,1],'-0',antibio.raw[,2]+1,sep='')
```

### Merge outcome tables
```{r merge_all_sample_information, cache=TRUE,eval=TRUE}
### Creat a sample.info matrix to store all the sample information.
sample.info <- outcome %>% add_rownames(var = "Subject") %>%
dplyr::select(-(Plate:DiffStopStart),-StopReason,-(PUCAIResponse:FCP50),-(FCPGreaterRedution:MergedResponse)) %>% 
gather(Measure,Value,BristolScore_1:FCP_4) %>%
separate(col=Measure,into=c("Measure", "Time"),sep='_' ) %>%
spread(Measure,Value) %>%
## log FCP
mutate(log.FCP = log(FCP)) %>%
## sample name
unite(col=Sample,Subject,Time,sep = '-0',remove=FALSE) %>%
## Group
mutate(Group='PLEASE') %>%
## Response (same as FCPReponse, but re-code it)
mutate(Response=ifelse(FCPResponse==1,'Response','Non.Response')) %>%
## Type
unite(col=Type,Group,Time,sep='-T',remove=FALSE) %>%
## Antibiotics at each visit
left_join(dplyr::select(add_rownames(antibio.raw,var='Sample'),Sample,fcpabx7),by='Sample') %>%
mutate(Antibiotics.visit=ifelse(fcpabx7==1,'Use','Not.Use')) %>%
dplyr::select(-fcpabx7) %>%
##Baseline use of systemic or rectally administered steroids 
##from baseline form these are variables bas18 and bas19
##If either have value 1 then patient was using steroids at baseline
left_join(dplyr::select(Baseline,Pid,Bas18,Bas19),by=c('Subject'='Pid')) %>%
mutate(Steroids=ifelse(Bas18 ==1 | Bas19 == 1,'Use','Not.Use')) %>%
dplyr::select(-Bas18,-Bas19) %>% 
### EEN PEN
mutate(Treatment.Specific=ifelse(Treatment=='antiTNF','antiTNF',
ifelse(as.numeric(Subject)>=5000 & as.numeric(Subject)<6000,'PEN','EEN'))) %>%
### COMBO 
full_join(data.frame(Sample=rownames(COMBO.raw),
Subject=rownames(COMBO.raw),
Group = 'COMBO',
Type  = 'COMBO',
Antibiotics.visit='Not.Use'),
by = c('Subject','Sample','Group','Type','Antibiotics.visit')) %>%
mutate(Disease=ifelse(Group=='COMBO','Control','Crohn')) %>%
## Human DNA percentage
left_join(dplyr::select(add_rownames(human.read,var='Sample'),Sample,NonHumanReads,Human.Per=HumanPer),by=c('Sample')) %>%
### fungi percentage
left_join(add_rownames(fungi.per,var='Sample'),by=c('Sample')) %>%
### 
arrange(Subject,Time)  %>% 
###
add_back_rownames(row.var='Sample')

####
head(sample.info,n=5)

#df1 %>% inner_join(df2) %>% knitr::kable()
```

```{r for_publication_supple_table1,eval=FALSE,include=FALSE}
supple.table1 <- add_rownames(sample.info,var='Sample') %>%
left_join(add_rownames(human.read,var='Sample'),by=c('Sample')) %>%
dplyr::select(Subject,Sample,Time,Treatment.Specific, FCP, Antibiotics.visit,Steroids,TotalReads, Human.Per, Fungi.Per )
#add_back_rownames(row.var='Sample')
##############################
### save the sample information
write.csv(supple.table1,file='../3_Result/Suppl_Table1_Sample_Information.csv',row.names=FALSE)
```




### Correlation of FCP and PCDAI
First, take a look at the correlation between FCP and PCDAI
```{r correlation_of_FCP_and_PCDAI,cache=TRUE}
FCP.PCDAI.cor <- cor(outcome[,c('FCP_1','FCP_2','FCP_3','FCP_4','PCDAI_1','PCDAI_4')],use="pairwise.complete.obs",method="spearman")
FCP.PCDAI.cor
ggplot(melt(FCP.PCDAI.cor), aes(x=Var1, y=Var2, fill=value, label=round(value, 2))) + scale_fill_gradient(low="#FEE0D2", high="#FB6A4A") + geom_tile() + geom_text()+xlab("")+ylab("")
```

### FCP change across time
How FCP changed across different time points?
```{r FCP_change_arocss_time,cache=TRUE}
p <- outcome[,c('Treatment','FCPResponse','FCP_1','FCP_2','FCP_3','FCP_4')] %>%
gather(Time,FCP,FCP_1:FCP_4) %>%
mutate(FCPResponse=ifelse(FCPResponse==1,'Response','Non-Response')) %>%
na.omit %>%
ggplot(aes(y=FCP,x=Time))+geom_boxplot()+
xlab('')+ylab('FCP')+ labs(title='')+
facet_grid(Treatment ~ FCPResponse, margins=TRUE)
print(p)
```

### PCDAI change across time
How PCDAI changed across different time points?
```{r PCDAI_Change_Arocss_Time,cache=TRUE}
p <- outcome[,c('Treatment','PCDAIResponse','PCDAI_1','PCDAI_4')] %>%
gather(Time,PCDAI,PCDAI_1:PCDAI_4) %>%
mutate(PCDAIResponse=ifelse(PCDAIResponse==1,'Response','Non-Response')) %>%
na.omit %>%
ggplot(aes(y=PCDAI,x=Time))+geom_boxplot()+
xlab('')+ylab('PCDAI')+ labs(title='')+
facet_grid(Treatment ~ PCDAIResponse, margins=TRUE)
print(p)
```

### Response rate by treatment
Then, create a table for FCP and PCDAI response
```{r response_rate_by_treatment_table,cache=TRUE}
table(outcome[,c('Treatment'),drop=F],useNA='always') 
table(subset(sample.info,Time==1,select=Treatment.Specific),useNA='always') 
table(outcome[,c('FCPResponse','PCDAIResponse')],useNA='always') %>% kable(format='markdown')
table(outcome[,c('Treatment','FCPResponse')],useNA='always')  %>% kable(format='markdown')
table(outcome[,c('Treatment','PCDAIResponse')],useNA='always') %>% kable(format='markdown') 
```


************************************************************
************************************************************
# Analyze bacterial data
********************

## Normal vs. crohn-T1
```{r bacteria_normal_vs_crohn_t1}
### I don't use dplyr filter 
### because it will remove the row names
bact.cp1 <- subset_data_by_sample_info(taxa.data,sample.info,
                                       Type=='COMBO' | Type=='PLEASE-T1')
```

### Calculate distance
```{r bacteria_calculate_distance,cache=TRUE}
bact.dist <- calculate_distance(X = taxa.data, method = bacteral.distance.method, tree = NA)
```

```{r bacteria_calculate_distance_cp1,cache=TRUE}
bact.dist.cp1 <- calculate_distance(X = bact.cp1, method = bacteral.distance.method, tree = NA)
```

### Find the best clustering
```{r bacteria_find_the_best_clustering,cache=TRUE}
pamk.best <- find_best_clustering(bact.dist.cp1,plot.results=TRUE)
```


### Find the best clustering (numerical jaccard)
```{r bacteria_find_the_best_clustering_numerical_jaccard,cache=TRUE}
### Hongzhe has some concerns about the two cluster result based on the binary jaccard distance
### Here, use numerical jaccard distance to repeat the same analysis
pamk.best.numerical <- find_best_clustering(calculate_distance(X = bact.cp1, method = "Numerical_Jaccard", tree = NA),
                     plot.results=TRUE)


### MDS
set.seed(10)
meta.MDS.numerical <- metaMDS(calculate_distance(X = bact.cp1, method = "Numerical_Jaccard", tree = NA))

#### Human DNA percentage (for publication)
p <- add_rownames(as.data.frame(meta.MDS.numerical$points),var='Sample') %>%
  left_join(dplyr::select(add_rownames(sample.info,var='Sample'),
                          Sample,Disease,Cluster,Human.Per),
            by='Sample') %>% 
  ggplot(aes(x=MDS1,y=MDS2))+ 
  geom_point(aes(colour = Cluster,shape= Disease,size = Human.Per)) +
  scale_color_manual(values=c("#0066CC", "#FF3333"),name="") +
  scale_shape_manual(values=c(20,1),name="") +
  #scale_size_area(range = c(1, 10))+
  scale_size_continuous(range = c(3, 10)) +
  labs(title = 'Ordination bacterial taxa') +
  ##########################
  theme_classic()
print(p)

```


```{r update_cluster_info_and_save_sample_informaion,cache=TRUE}
### rename does not work
clustering.res <- as.data.frame(pamk.best$pamobject$clustering)
colnames(clustering.res) <- 'Cluster'
### add clustering result to sample information
### only T1 samples are in the clustering result
sample.info <- 
  add_rownames(clustering.res,var='Sample') %>%
  separate(col=Sample,into=c("Subject", "Time"),sep='-',extra='drop' ) %>%
  dplyr::select(Subject,Cluster) %>%
  right_join(add_rownames(sample.info,var='Sample'),by = c('Subject')) %>% #head  
  mutate(Cluster=ifelse(Cluster==1,'cluster 1','cluster 2')) %>% #head
  add_back_rownames(row.var='Sample') 


##############################
### save the sample information
write.csv(sample.info,file=paste('../1_Data/Processed_Data/Sample_Information/2015_02_13_Processed_Sample_Information.csv',sep=''),row.names=TRUE)
```


### MDS analysis
```{r bacteria_MDS,cache=TRUE,results='hide'}
### MDS
set.seed(10)
meta.MDS <- metaMDS(bact.dist.cp1)

### Normal/Disease
p <- add_rownames(as.data.frame(meta.MDS$points),var='Sample') %>%
  left_join(dplyr::select(add_rownames(sample.info,var='Sample'),
                          Sample,Disease),
            by='Sample') %>% 
  ggplot(aes(x=MDS1,y=MDS2))+ geom_point(aes(colour = Disease)) +
  scale_color_manual(values=c("#0066CC", "#FF3333"),name="") +
  labs(title = 'metaMDS') 
print(p)

### Normal/Disease, cluster
p <- add_rownames(as.data.frame(meta.MDS$points),var='Sample') %>%
  left_join(dplyr::select(add_rownames(sample.info,var='Sample'),
                          Sample,Disease,Cluster),
            by='Sample') %>% 
  ggplot(aes(x=MDS1,y=MDS2))+ geom_point(aes(colour = Disease,shape=Cluster)) +
  scale_color_manual(values=c("#0066CC", "#FF3333"),name="") +
  scale_shape_manual(values=c(20,4),name="") +
  labs(title = 'metaMDS') 
print(p)

#### Fugal percentage 
p <- add_rownames(as.data.frame(meta.MDS$points),var='Sample') %>%
  left_join(dplyr::select(add_rownames(sample.info,var='Sample'),
                          Sample,Disease,Cluster,Fungi.Per),
            by='Sample') %>% 
  mutate(Fungi.Per.sqrt=sqrt(Fungi.Per)) %>%
  ggplot(aes(x=MDS1,y=MDS2))+ 
  geom_point(aes(colour = Disease,shape=Cluster,size = Fungi.Per.sqrt)) +
  scale_color_manual(values=c("#0066CC", "#FF3333"),name="") +
  scale_shape_manual(values=c(20,1),name="") +
  scale_size_area()+
  labs(title = 'metaMDS') 
print(p)


#### Human DNA percentage
p <- add_rownames(as.data.frame(meta.MDS$points),var='Sample') %>%
  left_join(dplyr::select(add_rownames(sample.info,var='Sample'),
                          Sample,Disease,Cluster,Human.Per),
            by='Sample') %>% 
  ggplot(aes(x=MDS1,y=MDS2))+ 
  geom_point(aes(colour = Disease,shape=Cluster,size = Human.Per)) +
  scale_color_manual(values=c("#0066CC", "#FF3333"),name="") +
  scale_shape_manual(values=c(20,1),name="") +
  scale_size_area()+
  labs(title = 'metaMDS') 
print(p)

#### Human DNA percentage (for publication)
p <- add_rownames(as.data.frame(meta.MDS$points),var='Sample') %>%
  left_join(dplyr::select(add_rownames(sample.info,var='Sample'),
                          Sample,Disease,Cluster,Human.Per),
            by='Sample') %>% 
  ggplot(aes(x=MDS1,y=MDS2))+ 
  geom_point(aes(colour = Cluster,shape= Disease,size = Human.Per)) +
  scale_color_manual(values=c("#0066CC", "#FF3333"),name="") +
  scale_shape_manual(values=c(20,1),name="") +
  #scale_size_area(range = c(1, 10))+
  scale_size_continuous(range = c(3, 10)) +
  labs(title = 'Ordination bacterial taxa') +
  ##########################
  theme_classic()
print(p)

### Figure for Rick
### we start with the same MDS as in 1B, but don't show human DNA or near and far cluster.  Control is shown as the filled circle, Crohn's as open circles.  Current antibiotic use is shown by blue versus red coloring.  If not too hard, maybe also plot the centroid of the healties as a large yellow triangle.
p <- add_rownames(as.data.frame(meta.MDS$points),var='Sample') %>%
  left_join(dplyr::select(add_rownames(sample.info,var='Sample'),
                          Sample,Disease,Antibiotics.visit),
            by='Sample') %>% 
  ggplot(aes(x=MDS1,y=MDS2))+ 
  geom_point(aes(colour = Antibiotics.visit,shape= Disease),size=3) +
  scale_color_manual(values=c("#0066CC", "#FF3333"),name="") +
  scale_shape_manual(values=c(20,1),name="") +
  #scale_size_area(range = c(1, 10))+
  #scale_size_continuous(range = c(3, 10)) +
  labs(title = 'Ordination bacterial taxa') +
  ##########################
  theme_classic()
print(p)


#### Fugal percentage and Human DNA percentage
p <- add_rownames(as.data.frame(meta.MDS$points),var='Sample') %>%
  left_join(dplyr::select(add_rownames(sample.info,var='Sample'),
                          Sample,Disease,Cluster,Human.Per,Fungi.Per),
            by='Sample') %>% 
  mutate(Fungi.Per.sqrt=sqrt(Fungi.Per)) %>%
  ggplot(aes(x=MDS1,y=MDS2,colour=Human.Per,shape=Cluster))+ 
  geom_point(aes(size = Fungi.Per.sqrt)) + scale_size_area()+
  scale_colour_gradient(low = "#43a2ca",high="#de2d26")+ 
  scale_shape_manual(values=c(20,1),name="") + labs(title = 'metaMDS')
print(p)

### switch human and fungi 
p <- add_rownames(as.data.frame(meta.MDS$points),var='Sample') %>%
  left_join(dplyr::select(add_rownames(sample.info,var='Sample'),
                          Sample,Disease,Cluster,Human.Per,Fungi.Per),
            by='Sample') %>% 
  mutate(Fungi.Per.sqrt=sqrt(Fungi.Per)) %>%
  ggplot(aes(x=MDS1,y=MDS2,colour=Fungi.Per.sqrt,shape=Cluster))+ 
  geom_point(aes(size = Human.Per)) + scale_size_area()+
  scale_colour_gradient(low = "#43a2ca",high="#de2d26")+ 
  scale_shape_manual(values=c(20,1),name="") + labs(title = 'metaMDS')
print(p)
```

### PERMANOVA of response vs. non-response at T1
```{r bacteria_permanova_of_r_vs_nr_t1,cache=TRUE}
## R vs. NR
cov.mat <- subset(sample.info,Type=='PLEASE-T1',select=Response)
PERMANOVA_analysis(dist.mat=bact.dist.cp1,cov.mat=cov.mat)

## R vs. NR, antiTNF
cov.mat <- subset(sample.info,Type=='PLEASE-T1' & Treatment=='antiTNF',select=Response)
PERMANOVA_analysis(dist.mat=bact.dist.cp1,cov.mat=cov.mat)

## R vs. NR, Diet
cov.mat <- subset(sample.info,Type=='PLEASE-T1' & Treatment=='Diet',select=Response)
PERMANOVA_analysis(dist.mat=bact.dist.cp1,cov.mat=cov.mat)

## R vs. NR, PEN
cov.mat <- subset(sample.info,Type=='PLEASE-T1' & Treatment.Specific=='PEN',select=Response)
PERMANOVA_analysis(dist.mat=bact.dist.cp1,cov.mat=cov.mat)

## R vs. NR, EEN
cov.mat <- subset(sample.info,Type=='PLEASE-T1' & Treatment.Specific=='EEN',select=Response)
PERMANOVA_analysis(dist.mat=bact.dist.cp1,cov.mat=cov.mat)
```


### Wilcoxon rank test of normal vs. Crohn
```{r bacteria_rank_test_of_normal_vs_crohn,cache=TRUE}
### only plot the top 5 significant ones
bact.cp1.wil <- wilcox_test(bact.cp1,sample.info[,'Disease',drop=FALSE],
                            plot.heatmap=FALSE,boxplot.top=0.05,boxplot.y='Abundance(%)')
###
write.csv(bact.cp1.wil,file=paste('../3_Result/2015_03_05_Bacteria_COMBO_vs_PLEASET1_Wilcox.csv',sep=''),row.names=TRUE)
```

### Wilcoxon rank test of normal vs. Crohn without antibiotics use
```{r bacteria_rank_test_of_normal_vs_crohn_without_antibiotics_use,cache=TRUE}
### only plot the top 5 significant ones
bact.cp1.noabx.wil <- wilcox_test(bact.cp1,
              subset(sample.info,Antibiotics.visit=='Not.Use',select=Disease),
              plot.heatmap=FALSE,boxplot.top=0.05,boxplot.y='Abundance(%)')
###
write.csv(bact.cp1.noabx.wil,file=paste('../3_Result/2015_03_05_Bacteria_COMBO_vs_PLEASET1_No_Abx_Wilcox.csv',sep=''),row.names=TRUE)
```


### Heatmap of clustering results
```{r bacteria_heatmap_all_taxa,cache=TRUE,fig.width=8,fig.height=8}
bk <- unique(c(
        0,
        min(min(taxa.data[taxa.data>0])/2,0.1),
        seq(min(min(taxa.data[taxa.data>0])/2,1),0.1,length=10),
        seq(0.1,1,length=10),
        seq(1,5,length=10),
        seq(5,20,length=20),
        seq(20,50,length=20),
        seq(50,100,length=20)
        ))
hmcols<- colorRampPalette(c('white','#1f78b4','#ff7f00','#e41a1c'))(length(bk)-1)
###############
#bk = unique(c(seq(0,5, length=100),seq(5,10,length=100),seq(10,100,length=100)))
#hmcols<- colorRampPalette(c("white","firebrick3","yellow"))(length(bk)-1) 

##########
### the color lables must be columns,so I have to transform the data
pheatmap(t(bact.cp1),border_color="grey60",
         clustering_distance_rows = 'correlation',
         clustering_distance_cols = bact.dist.cp1,
         clustering_method = "complete",
         annotation=sample.info[,c("Disease","Cluster","Treatment.Specific","FCP","Response","Antibiotics.visit","Steroids")],
         color =hmcols, breaks=bk,
         show_rownames = TRUE,
         scale="none")
```



```{r bacteria_heatmap_manually_order_samples,cache=TRUE,fig.width=8,fig.height=8}
#### manually re-order the samples
data.heat <- 
  add_rownames(bact.cp1,var='Sample') %>%
  left_join(add_rownames(sample.info,var='Sample'),by='Sample') %>% 
  arrange(Disease,Cluster,Response,Treatment.Specific) %>%
  dplyr::select( 1:(ncol(bact.cp1)+1) ) %>% ## first column is the sample name
  add_back_rownames(row.var='Sample')

bk <- unique(c(
        0,
        min(min(taxa.data[taxa.data>0])/2,0.1),
        seq(min(min(taxa.data[taxa.data>0])/2,1),0.1,length=10),
        seq(0.1,1,length=10),
        seq(1,5,length=10),
        seq(5,20,length=20),
        seq(20,50,length=20),
        seq(50,100,length=20)
        ))
hmcols<- colorRampPalette(c('white','#1f78b4','#ff7f00','#e41a1c'))(length(bk)-1)

#### sometimes got an error about "factor does not match color"
#### try to install the newest pheatmap from GitHub
pheatmap(t(data.heat),border_color="grey60",
         cluster_cols = FALSE,
         clustering_distance_rows = 'correlation',
         #clustering_distance_cols = bact.dist.cp1,
         clustering_method = "complete",
         annotation=sample.info[,c("Disease","Cluster","Response","Treatment.Specific","FCP","Antibiotics.visit","Steroids")],
         color =hmcols, breaks=bk,
         show_rownames = TRUE,
         scale="none")

rm(data.heat)
```



```{r bacteria_heatmap_significant_taxa,cache=TRUE,fig.width=8,fig.height=8}
##### In the previous analysis, I plotted the heatmap for those significant taxa
##### The distance used for sample clustering was based on all taxa
##### the distance needs to be re-calcualte with those significant taxa only
```


### Random Forest 
```{r bacteria_random_forest_of_normal_vs_crohn,cache=TRUE}
random_forest(bact.cp1,sample.info[,'Disease',drop=FALSE],
              plot.MDS=TRUE,plot.importance=TRUE,print.result=FALSE,
              guess.num=50,boxplot.top=5,
              title="Normal vs. Crohn",boxplot.y="Abundance(%)",
              seed=100)
```


## Antibiotics use

### Wilcoxon rank test 
```{r bacteria_rank_test_of_antibiotics_use,cache=TRUE}
### plot all the significant ones
bact.antibio.wil <- wilcox_test(bact.cp1,
                    subset(sample.info,Group=='PLEASE',select=Antibiotics.visit),
                             plot.heatmap=FALSE,boxplot.top=0.05,boxplot.y='Abundance(%)')      
###
write.csv(bact.antibio.wil,file=paste('../3_Result/2015_03_05_Bacteria_PLEASE_Antibiotics_Use_Wilcox.csv',sep=''),row.names=TRUE)

```

### Random Forest 
```{r bacteria_random_forest_of_antibiotics_use,cache=TRUE}
random_forest(bact.cp1,
              subset(sample.info,Group=='PLEASE',select=Antibiotics.visit),
              plot.MDS=TRUE,plot.importance=TRUE,print.result=FALSE,
              guess.num=50,boxplot.top=5,
              title="Antibiotics use",boxplot.y="Abundance(%)",
              seed=100)
```



## Cluster 1 vs. cluster 2

### Wilcoxon rank test 
```{r bacteria_rank_test_of_cluster_1_vs_2,cache=TRUE}
### plot all the significant ones
bact.c1c2.wil <- wilcox_test(bact.cp1,
                             subset(sample.info,Group=='PLEASE',select=Cluster),
                             plot.heatmap=FALSE,boxplot.top=0.05,boxplot.y='Abundance(%)')      
###
write.csv(bact.c1c2.wil,file=paste('../3_Result/2015_03_05_Bacteria_PLEASE_Cluster_1_2_Wilcox.csv',sep=''),row.names=TRUE)
```

### Random Forest 
```{r bacteria_random_forest_of_cluster_1_vs_2,cache=TRUE}
random_forest(bact.cp1,subset(sample.info,Group=='PLEASE',select=Cluster),
              plot.MDS=TRUE,plot.importance=TRUE,print.result=FALSE,
              guess.num=50,boxplot.top=5,
              title="Cluster 1 vs. Cluster 2",boxplot.y="Abundance(%)",
              seed=100)
```


### Heatmap: normal vs. disease and cluster 1 vs. 2 (new label)
```{r bacteria_heatmap_normal_vs_disease_and_cluster_1_vs_2,cache=TRUE,fig.width=8,fig.height=8}

bact.cp1.heat <- add_rownames(as.data.frame(bact.cp1.wil),var='Taxa') %>%
  left_join(add_rownames(as.data.frame( bact.c1c2.wil),var='Taxa'),by='Taxa') %>%
  mutate(Disease=ifelse(Disease.qvalue<0.05,'*',' '),
         Cluster=ifelse(Cluster.qvalue<0.05,'+',' ')
  ) %>%
  mutate(Tnames=paste(Taxa,Disease,Cluster,sep=''))%>%
  dplyr::select(Taxa,Tnames) %>%
  right_join( add_rownames(as.data.frame(t(bact.cp1)),var='Taxa'),by='Taxa') %>%
  dplyr::select(-Taxa) %>%
  add_back_rownames(row.var='Tnames') 

### * disease.q < 0.05
### + cluster.q < 0.05
bk <- unique(c(
        0,
        min(min(taxa.data[taxa.data>0])/2,0.1),
        seq(min(min(taxa.data[taxa.data>0])/2,1),0.1,length=10),
        seq(0.1,1,length=10),
        seq(1,5,length=10),
        seq(5,20,length=20),
        seq(20,50,length=20),
        seq(50,100,length=20)
        ))
hmcols<- colorRampPalette(c('white','#1f78b4','#ff7f00','#e41a1c'))(length(bk)-1)
#bk = unique(c(seq(0,5, length=100),seq(5,10,length=100),seq(10,100,length=100)))
#hmcols<- colorRampPalette(c("white","firebrick3","yellow"))(length(bk)-1) 

### the color lables must be columns,so I have to transform the data
pheatmap(bact.cp1.heat,border_color="grey60",
         clustering_distance_rows = 'correlation',
         clustering_distance_cols = bact.dist.cp1,
         clustering_method = "complete",
         annotation=sample.info[,c("Disease","Cluster","Treatment.Specific","FCP","Response","Antibiotics.visit","Steroids")],
         color =hmcols, breaks=bk,
         show_rownames = TRUE,
         scale="none")

rm(bact.cp1.heat)
```

```{r bacteria_heatmap_manually_order_samples_normal_disease_cluster1_2,cache=TRUE,fig.width=8,fig.height=8}
#### manually re-order the samples
data.heat <- 
  add_rownames(as.data.frame(bact.cp1.wil),var='Taxa') %>%
  left_join(add_rownames(as.data.frame( bact.c1c2.wil),var='Taxa'),by='Taxa') %>%
  mutate(Disease=ifelse(Disease.qvalue<0.05,'*',' '),
         Cluster=ifelse(Cluster.qvalue<0.05,'+',' ')
  ) %>%
  mutate(Tnames=paste(Taxa,Disease,Cluster,sep=''))%>%
  dplyr::select(Taxa,Tnames) %>%
  right_join( add_rownames(as.data.frame(t(bact.cp1)),var='Taxa'),by='Taxa') %>%
  dplyr::select(-Taxa) %>%
  add_back_rownames(row.var='Tnames') %>%
  t() %>%
  as.data.frame %>%
  add_rownames(var='Sample') %>%
  ##
  left_join(add_rownames(sample.info,var='Sample'),by='Sample') %>% 
  arrange(Disease,Cluster,Response,Treatment.Specific,Antibiotics.visit,FCP,Steroids) %>%
  dplyr::select(1:(ncol(bact.cp1)+1)) %>%  ## first column is sample name
  add_back_rownames(row.var='Sample') 

pheatmap(t(data.heat),border_color="grey60",
         cluster_cols = FALSE,
         clustering_distance_rows = 'correlation',
         #clustering_distance_cols = bact.dist.cp1,
         clustering_method = "complete",
         annotation=sample.info[,c("Disease","Cluster","Response","Treatment.Specific","FCP","Antibiotics.visit","Steroids")],
         color =hmcols, breaks=bk,
         show_rownames = TRUE,
         scale="none")

rm(data.heat)
```

### Response rate in two clusters
```{r bacteria_response_rate_in_two_clusters,cache=TRUE}
## all
sample.info %>% dplyr::select(Response,Cluster,Type) %>%
dplyr::filter(Type=='PLEASE-T4') %>%
na.omit() %>%
group_by(Cluster,Response) %>%
summarise(Num=n()) %>%
spread(Response,Num) %>%
add_back_rownames(row.var='Cluster') %>%
(function(tdata){
  print(tdata)
  fisher.test(tdata)$p.value
})


## antiTNF
sample.info %>% 
dplyr::filter(Treatment.Specific=='antiTNF') %>%
dplyr::select(Response,Cluster,Type) %>%
dplyr::filter(Type=='PLEASE-T4') %>%
na.omit() %>%
group_by(Cluster,Response) %>%
summarise(Num=n()) %>%
spread(Response,Num) %>%
add_back_rownames(row.var='Cluster') %>%
(function(tdata){
  print(tdata)
  fisher.test(tdata)$p.value
})


## EEN
sample.info %>% 
dplyr::filter(Treatment.Specific=='EEN') %>%
dplyr::select(Response,Cluster,Type) %>%
dplyr::filter(Type=='PLEASE-T4') %>%
na.omit() %>%
group_by(Cluster,Response) %>%
summarise(Num=n()) %>%
spread(Response,Num) %>%
add_back_rownames(row.var='Cluster')# %>%
#function(tdata){
#  print(tdata)
#  fisher.test(tdata)$p.value
#}

## PEN
sample.info %>% 
dplyr::filter(Treatment.Specific=='PEN') %>%
dplyr::select(Response,Cluster,Type) %>%
dplyr::filter(Type=='PLEASE-T4') %>%
na.omit() %>%
group_by(Cluster,Response) %>%
summarise(Num=n()) %>%
spread(Response,Num) %>%
add_back_rownames(row.var='Cluster') #%>%
## NA in one cell, can not do fisher exact test
#function(tdata){
#  print(tdata)
#  fisher.test(tdata)$p.value
#}
```


## Human DNA

### Humand DNA among normal, disease cluster 1 and cluster 2
```{r bacteria_humand_dna_among_normal_disease_cluster1_and_cluster2,cache=TRUE}
### human DNA
sample.info %>% 
  dplyr::select(Type,Cluster,Human.Per) %>%
  dplyr::filter(Type=='COMBO' | Type=='PLEASE-T1') %>%
  na.omit() %>%
  unite(Group,Type,Cluster, sep = ' ')  %>%
  (function(pdata){
    boxplot(Human.Per~Group,data=pdata,ylab='Human DNA (%)')
    # Kruskal Wallis Test One Way Anova by Ranks 
    print('Kruskal Wallis test (non-parametric ANOVA) pvalue')
    kruskal.test(Human.Per~as.factor(Group),data=pdata)$p.value 
  })

### human DNA (for publication)
sample.info %>% 
  dplyr::select(Type,Cluster,Human.Per) %>%
  dplyr::filter(Type=='COMBO' | Type=='PLEASE-T1') %>%
  na.omit() %>%
  unite(Group,Type,Cluster, sep = ' ')  %>%
  ggplot(aes(factor(Group), Human.Per)) +
  geom_boxplot(aes(fill = Group),outlier.shape = NA) + 
  #geom_jitter() +
  theme_classic()


##### non-human DNA
sample.info %>% 
  dplyr::select(Type,Cluster,NonHumanReads) %>%
  dplyr::filter(Type=='COMBO' | Type=='PLEASE-T1') %>%
  na.omit() %>%
  unite(Group,Type,Cluster, sep = ' ')  %>%
  (function(pdata){
    boxplot(log10(NonHumanReads)~Group,data=pdata,ylab='log10 non-human reads')
    # Kruskal Wallis Test One Way Anova by Ranks 
    print('Kruskal Wallis test (non-parametric ANOVA) pvalue')
    kruskal.test(NonHumanReads~as.factor(Group),data=pdata)$p.value 
  })
```

### Human DNA and fungi percentage
```{r human_dna_with_fungi_percentage,cache=TRUE}
p <- subset(sample.info,Type=='PLEASE-T1' | Type=='COMBO',
            select=c(Disease,Cluster,Human.Per,Fungi.Per)) %>%
na.omit() %>%
ggplot(aes(log10(Human.Per),log10(Fungi.Per))) + 
geom_point(aes(colour = Cluster,shape= Disease),size=5) +
  scale_color_manual(values=c("#0066CC", "#FF3333"),name="") +
  scale_shape_manual(values=c(20,1),name="") +
  #scale_size_area(range = c(1, 10))+
  #scale_size_continuous(range = c(3, 10)) +
  #labs(title = 'Ordination bacterial taxa') +
  #stat_smooth_func(geom="text",method="lm",hjust=0,parse=TRUE) +
  geom_smooth(method="lm",se=FALSE) +
  theme_classic()
print(p)
```

### Human DNA with clinical outcomes
```{r human_dna_with_clinical_outcomes,cache=TRUE}
### Plot human DNA by treatment(PEN vs. EEN) and response
p<- sample.info %>% dplyr::select(Human.Per,Response,Time,Treatment.Specific) %>%
  na.omit %>%
  ggplot(aes(factor(Time), Human.Per),na.rm=T)+
  geom_boxplot(na.rm=T)+facet_grid(Treatment.Specific ~ Response, margins=TRUE)
plot(p)

####  T1 vs. T2,3,4 
sample.info %>%
  dplyr::select(Subject,Time,Treatment,Human.Per) %>% 
  na.omit() %>% 
  mutate(Time=paste('T',Time,sep='')) %>%
  spread(Time,Human.Per) %>%
  summarise_each(funs(wilcox.test(., T1, pair = TRUE)$p.value), vars = T2:T4) %>%
  rename(T1.vs.T2=vars1,T1.vs.T3=vars2,T1.vs.T4=vars3)

####  T1 vs. T2,3,4 for each treatment arm
sample.info %>%
  dplyr::select(Subject,Time,Treatment,Human.Per) %>% 
  na.omit() %>% 
  mutate(Time=paste('T',Time,sep='')) %>%
  spread(Time,Human.Per) %>%
  group_by(Treatment) %>%
  summarise_each(funs(wilcox.test(., T1, pair = TRUE)$p.value), vars = T2:T4) %>%
  rename(T1.vs.T2=vars1,T1.vs.T3=vars2,T1.vs.T4=vars3)


####  T1 vs. T2,3,4 for each treatment arm
sample.info %>%
  dplyr::select(Subject,Time,Treatment.Specific,Human.Per) %>% 
  na.omit() %>% 
  mutate(Time=paste('T',Time,sep='')) %>%
  spread(Time,Human.Per) %>% 
  group_by(Treatment.Specific) %>%
  summarise_each(funs(wilcox.test(., T1, pair = TRUE)$p.value), vars = T2:T4) %>%
  rename(T1.vs.T2=vars1,T1.vs.T3=vars2,T1.vs.T4=vars3)

####  T1 vs. T2,3,4 for each treatment arm & response
sample.info %>%
  dplyr::select(Subject,Time,Treatment,Response,Human.Per) %>% 
  na.omit() %>% 
  mutate(Time=paste('T',Time,sep='')) %>%
  spread(Time,Human.Per) %>%
  group_by(Treatment,Response) %>%
  summarise_each(funs(wilcox.test(., T1, pair = TRUE)$p.value), vars = T2:T4) %>%
  rename(T1.vs.T2=vars1,T1.vs.T3=vars2,T1.vs.T4=vars3)

####  T1 vs. T2,3,4 for each treatment arm & response
####  Not enough sample for PEN
sample.info %>%
  dplyr::select(Subject,Time,Treatment.Specific,Response,Human.Per) %>% 
  na.omit() %>% 
  mutate(Time=paste('T',Time,sep='')) %>%
  spread(Time,Human.Per) %>% 
  filter(Treatment.Specific=='EEN') %>%
  group_by(Treatment.Specific,Response) %>%
  summarise_each(funs(wilcox.test(., T1, pair = TRUE)$p.value), vars = T2:T4) %>%
  rename(T1.vs.T2=vars1,T1.vs.T3=vars2,T1.vs.T4=vars3)

###########################
##### R vs. NR
wilcoxtest.fun = function(tdata){ 
  R.vs.NR <- NA
  try(R.vs.NR <- signif(wilcox.test(Human.Per~Response,data=tdata)$p.value, 2),silent=TRUE)
  return(as.data.frame(R.vs.NR))
}


####  R vs. NR: by treatment
sample.info %>%
  dplyr::select(Subject,Time,Treatment,Treatment.Specific,Human.Per,Response) %>% 
  na.omit() %>%
  mutate(Time=paste('T',Time,sep='')) %>%
  group_by(Treatment) %>%
  do(wilcoxtest.fun(.)) %>%
  add_back_rownames(row.var='Treatment')

sample.info %>%
  dplyr::select(Subject,Time,Treatment,Treatment.Specific,Human.Per,Response) %>% 
  na.omit() %>%
  mutate(Time=paste('T',Time,sep='')) %>%
  group_by(Treatment.Specific) %>%
  do(wilcoxtest.fun(.)) %>%
  add_back_rownames(row.var='Treatment.Specific')

####  R vs. NR: by time
sample.info %>%
  dplyr::select(Subject,Time,Treatment,Treatment.Specific,Human.Per,Response) %>% 
  na.omit() %>%
  mutate(Time=paste('T',Time,sep='')) %>%
  group_by(Time) %>%
  do(wilcoxtest.fun(.)) %>%
  add_back_rownames(row.var='Time')

####  R vs. NR: by time & treatment
sample.info %>%
  dplyr::select(Subject,Time,Treatment,Treatment.Specific,Human.Per,Response) %>% 
  na.omit() %>%
  mutate(Time=paste('T',Time,sep='')) %>%
  group_by(Time,Treatment) %>%
  do(wilcoxtest.fun(.)) 


sample.info %>%
  dplyr::select(Subject,Time,Treatment,Treatment.Specific,Human.Per,Response) %>% 
  na.omit() %>%
  mutate(Time=paste('T',Time,sep='')) %>%
  group_by(Time,Treatment.Specific) %>%
  do(wilcoxtest.fun(.))

###### EEN T1 R vs. NR
sample.info %>%
  dplyr::select(Subject,Time,Treatment,Treatment.Specific,Human.Per,Response) %>% 
  dplyr::filter(Time=='1'&Treatment.Specific=='EEN') %>%
  na.omit() %>%
  ggplot(aes(factor(Response), Human.Per),na.rm=T)+
  geom_boxplot(na.rm=T)
sample.info %>%
  dplyr::select(Subject,Time,Treatment,Treatment.Specific,Human.Per,Response) %>% 
  dplyr::filter(Time=='1'&Treatment.Specific=='EEN') %>%
  wilcox.test(as.numeric(Human.Per)~as.factor(Response),data=.)
##
sample.info %>%
  dplyr::select(Subject,Time,Treatment.Specific,Human.Per,FCP) %>% 
  dplyr::filter(Treatment.Specific=='EEN') %>%
  mutate(Time=paste('T',Time,sep='')) %>%
  unite(Human.Per.FCP, Human.Per,FCP) %>%
  spread(Time,Human.Per.FCP) %>%
  separate(T1, into = c("T1.Human.Per", "T1.FCP"), sep = "_") %>%
  separate(T4, into = c("T4.Human.Per", "T4.FCP"), sep = "_") %>%
  dplyr::select(T1.Human.Per,T4.FCP) %>%
  mutate(T1.Human.Per=as.numeric(T1.Human.Per)) %>%
  mutate(T4.FCP=as.numeric(T4.FCP)) %>%
  #as.data.frame() %>%
  #na.omit() %>%
  (function(tdata){
    plot(log10(tdata$T1.Human.Per),log10(tdata$T4.FCP),main='EEN T1',
         xlab='log10(T1.Human.Per)',ylab='log10(T4.FCP)')
    mtext(paste(
          'Spearman cor:',
          signif(cor(log10(tdata$T1.Human.Per),log10(tdata$T4.FCP),method='spearman'),2),
          'Pval',
          signif(cor.test(log10(tdata$T1.Human.Per),log10(tdata$T4.FCP),method='spearman')$p.value,2)
          ))
  })
```

### Compare to healthy controls
```{r human_dna_to_healthy_controls,cache=TRUE}
#### Plot distance by treatment,response
p<- sample.info %>% dplyr::select(Type,Human.Per,Treatment.Specific,Response) %>%
  mutate(Response=ifelse(Type=='COMBO','COMBO',Response)) %>%
  mutate(Treatment.Specific=ifelse(Type=='COMBO','COMBO',Treatment.Specific)) %>%
  na.omit() %>%
  unite(Treatment.Response, Treatment.Specific,Response, sep = ".") %>%
  ggplot(aes(factor(Type),log(Human.Per))) + 
  geom_boxplot(aes(fill = factor(Treatment.Response)),outlier.shape = NA)+
  xlab('')+ylab('log Human Per') + 
  theme_classic() 
plot(p)

#### Calculate the pvalues for each box in the above figure
sample.info %>%
dplyr::select(Type,Time,Treatment.Specific,Human.Per,Response) %>% 
mutate(Treatment.Specific=ifelse(Type=='COMBO','COMBO',Treatment.Specific)) %>%
mutate(Response=ifelse(Type=='COMBO','COMBO',Response)) %>%
mutate(Time=paste('T',Time,sep='')) %>%
mutate(Time=ifelse(Type=='COMBO','COMBO',Time)) %>%
dplyr::filter(Treatment.Specific != 'PEN') %>% 
na.omit() %>%
unite(Time.Treatment.Response,Time, Treatment.Specific,Response, sep = "_") %>%
dplyr::select(Time.Treatment.Response,Human.Per) %>%
(function(tdata){
cdata <- dplyr::filter(tdata,Time.Treatment.Response=='COMBO_COMBO_COMBO') %>%
         dplyr::select(Human.Per) %>%
         mutate(Human.Per=as.numeric(Human.Per))
#print(cdata)
group_by(tdata,Time.Treatment.Response) %>%
summarise(pval.compare.to.combo=wilcox.test(Human.Per,cdata$Human.Per)$p.value) %>%
mutate(pval.compare.to.combo=signif(pval.compare.to.combo,2)) %>%
separate(Time.Treatment.Response, into=c("Time","Treatment","Response"), sep = "_")
})
```

### quantile regression on fold change
```{r human_dna_fc_quantile_regression_treatment_antibiotics_response,cache=TRUE}
## fold change T4/T1
set.seed(1)
master.table <- add_rownames(sample.info,var='Sample') %>%
  dplyr::filter(Time==1 | Time==4) %>%
  dplyr::select(Human.Per,Subject,Time,Treatment.Specific,Antibiotics.visit,log.FCP,Response) %>%
  mutate(Time=ifelse(Time==1,'T1','T4'))
### human.dna
human.temp <- master.table %>%
  mutate(Human.Per = Human.Per+min(Human.Per[Human.Per>0],na.rm=TRUE))  %>% 
  dplyr::select(Subject,Human.Per,Time,Treatment.Specific) %>%
  spread(Time,Human.Per) %>%
  mutate(Human.Per.log.fc = log(T4/T1)) %>%
  dplyr::select(Subject,Treatment.Specific,Human.Per.log.fc)
### response
response.temp <- master.table %>%
  dplyr::select(Subject,Response) 
### log FCP
log.fcp.temp <- master.table %>%
  dplyr::select(Subject,Time,log.FCP) %>%
  spread(Time,log.FCP) %>%
  mutate(log.FCP.T4 = T4) %>%
  dplyr::select(Subject,log.FCP.T4)
### antibiotics
antibiotics.temp <- master.table %>%
  dplyr::select(Subject,Time,Antibiotics.visit) %>%
  spread(Time,Antibiotics.visit) %>%
  mutate(antibiotics = ifelse(is.na(T4),T1,T4)) %>%
  dplyr::select(Subject,antibiotics)
### merge together 
rq.fc.res <- human.temp %>% full_join(antibiotics.temp,by='Subject') %>%
  full_join(response.temp,by='Subject') %>%
  ### quantile regression
  rq(Human.Per.log.fc ~ Treatment.Specific + antibiotics + Response,
     tau=0.75,data=.,na.action=na.omit) %>%
  summary.rq(.,se="boot") %>% coef %>%
  as.data.frame %>% add_rownames(var='Var') %>%
  dplyr::select(1,2,5) %>% dplyr::select(Var,Est=Value,Pval=starts_with("Pr"))%>%
  gather(value,var,Est,Pval) %>% 
  unite(Variables,Var,value, sep = ".", remove = TRUE) %>%
  add_back_rownames(row.var='Variables') %>% as.matrix


###  
rm(master.table)
rm(human.temp)
rm(antibiotics.temp)
rm(log.fcp.temp)
rm(response.temp)
#break

rq.res.table <- 
  as.data.frame(t(as.data.frame(rq.fc.res))) %>% ## must use as.data.frame again 
  add_rownames(var='Human.DNA') 

rm(rq.fc.res)
######
#### row name is a column
write.csv(rq.res.table,file=paste('../3_Result/2015_03_05_Human_DNA_Fold_Change_Quantile_Q75_Regression_Treatment_Antibiotics_Response.csv',sep=''),row.names=FALSE)

rm(rq.res.table)
```



## Responders vs. non-responders

### Random Forest:EEN T1 and EEN T4
```{r bacteria_random_forest_of_r_vs_nr_een,cache=TRUE}
### EEN T1
random_forest(taxa.data,
              subset(sample.info,Time==1 & Treatment.Specific=='EEN' & !is.na(Response),
                     select=Response),
              plot.MDS=TRUE,plot.importance=TRUE,print.result=FALSE,
              guess.num=50,boxplot.top=5,
              title="Response vs. non-response",boxplot.y="Abundance(%)",
              seed=100)
### EEN T4
random_forest(taxa.data,
              subset(sample.info,Time==4 & Treatment.Specific=='EEN' & !is.na(Response),
                     select=Response),
              plot.MDS=TRUE,plot.importance=TRUE,print.result=FALSE,
              guess.num=50,boxplot.top=5,
              title="Response vs. non-response",boxplot.y="Abundance(%)",
              seed=100)

### Prevotella
try(
  add_rownames(taxa.data[,'Prevotella',drop=F],var='Sample') %>%
  left_join(dplyr::select(add_rownames(sample.info,var='Sample'),Subject,Sample,Time,Treatment.Specific,Response), by='Sample') %>% 
  mutate(Prevotella.log = log(Prevotella+min(Prevotella[Prevotella!=0])))%>%
  #head(30)
  na.omit() %>%
  ggplot( aes(factor(Time), Prevotella.log),na.rm=T)+
  geom_boxplot(na.rm=T)+facet_grid(Treatment.Specific ~ Response, margins=TRUE)

,silent=TRUE
)

try(
  add_rownames(taxa.data[,'Prevotella',drop=F],var='Sample') %>%
  left_join(dplyr::select(add_rownames(sample.info,var='Sample'),Subject,Sample,Time), by='Sample') %>% 
  dplyr::filter(Time==1) %>% 
  left_join(dplyr::filter(dplyr::select(add_rownames(sample.info,var='Sample'),Subject,Sample,Time,log.FCP),Time==4),by='Subject') %>% 
  dplyr::select(Subject,Prevotella,log.FCP) %>%
  ggplot(aes(Prevotella,log.FCP)) + geom_point()
  #,xlab='Prevotella(T1)',ylab='log FCP (T4)')
  
,silent=TRUE
)
```

### Wilcoxon rank test: EEN T1 and EEN T4
```{r bacteria_rank_test_of_r_vs_nr_een,cache=TRUE}
### EEN T1
bact.EEN.T1 <- wilcox_test(taxa.data,
                           subset(sample.info,Time==1 & Treatment.Specific=='EEN' & !is.na(Response),
                                  select=Response),
                           plot.heatmap=FALSE,boxplot.top=5,boxplot.y='Abundance(%)')
write.csv(bact.EEN.T1,file=paste('../3_Result/2015_03_05_Bacteria_EEN_T1_R_vs_NR_Wilcox.csv',sep=''),row.names=FALSE)

### EEN T4
bact.EEN.T4 <- wilcox_test(taxa.data,
                           subset(sample.info,Time==4 & Treatment.Specific=='EEN' & !is.na(Response),
                                  select=Response),
                           plot.heatmap=FALSE,boxplot.top=5,boxplot.y='Abundance(%)')
write.csv(bact.EEN.T4,file=paste('../3_Result/2015_03_05_Bacteria_EEN_T4_R_vs_NR_Wilcox.csv',sep=''),row.names=FALSE)

```

### Random Forest:antiTNF and antiTNF T4
```{r bacteria_random_forest_of_r_vs_nr_antiTNF,cache=TRUE}
### antiTNF T1
random_forest(taxa.data,
              subset(sample.info,Time==1 & Treatment.Specific=='antiTNF' & !is.na(Response),
                     select=Response),
              plot.MDS=TRUE,plot.importance=TRUE,print.result=FALSE,
              guess.num=50,boxplot.top=5,
              title="Response vs. non-response",boxplot.y="Abundance(%)",
              seed=100)
### antiTNF T4
random_forest(taxa.data,
              subset(sample.info,Time==4 & Treatment.Specific=='antiTNF' & !is.na(Response),
                     select=Response),
              plot.MDS=TRUE,plot.importance=TRUE,print.result=FALSE,
              guess.num=50,boxplot.top=5,
              title="Response vs. non-response",boxplot.y="Abundance(%)",
              seed=100)

```

### Wilcoxon rank test: antiTNF T1 and antiTNF T4
```{r bacteria_rank_test_of_r_vs_nr_antitnf,cache=TRUE}
### antiTNF T1
bact.antiTNF.T1 <- wilcox_test(taxa.data,
                           subset(sample.info,Time==1 & Treatment.Specific=='antiTNF' & !is.na(Response),
                                  select=Response),
                           plot.heatmap=FALSE,boxplot.top=10,boxplot.y='Abundance(%)')
write.csv(bact.antiTNF.T1,file=paste('../3_Result/2015_03_05_Bacteria_antiTNF_T1_R_vs_NR_Wilcox.csv',sep=''),row.names=FALSE)

### antiTNF T4
bact.antiTNF.T4 <- wilcox_test(taxa.data,
                           subset(sample.info,Time==4 & Treatment.Specific=='antiTNF' & !is.na(Response),
                                  select=Response),
                           plot.heatmap=FALSE,boxplot.top=10,boxplot.y='Abundance(%)')
write.csv(bact.antiTNF.T4,file=paste('../3_Result/2015_03_05_Bacteria_antiTNF_T4_R_vs_NR_Wilcox.csv',sep=''),row.names=FALSE)

```

## EEN response vs. aniTNF response vs. control

### Random Forest
```{r bacteria_random_forest_of_een_r_vs_control,cache=TRUE}
### T4 EEN response vs. control
random_forest(taxa.data,
              subset(sample.info,Type=='COMBO' | (Type=='PLEASE-T4' & Treatment.Specific=='EEN' & Response=='Response'),  select=Type),
              plot.MDS=TRUE,plot.importance=TRUE,print.result=FALSE,
              guess.num=50,boxplot.top=5,
              title="",boxplot.y="Abundance(%)",
              seed=100)

### T1 EEN response vs. control
random_forest(taxa.data,
              subset(sample.info,Type=='COMBO' | (Type=='PLEASE-T1' & Treatment.Specific=='EEN' & Response=='Response'),  select=Type),
              plot.MDS=TRUE,plot.importance=TRUE,print.result=FALSE,
              guess.num=50,boxplot.top=5,
              title="",boxplot.y="Abundance(%)",
              seed=100)

### T4 aniTNF response vs. control
random_forest(taxa.data,
              subset(sample.info,Type=='COMBO' | (Type=='PLEASE-T4' & Treatment.Specific=='antiTNF' & Response=='Response'),  select=Type),
              plot.MDS=TRUE,plot.importance=TRUE,print.result=FALSE,
              guess.num=50,boxplot.top=5,
              title="",boxplot.y="Abundance(%)",
              seed=100)

### T1 aniTNF response vs. control
random_forest(taxa.data,
              subset(sample.info,Type=='COMBO' | (Type=='PLEASE-T1' & Treatment.Specific=='antiTNF' & Response=='Response'),  select=Type),
              plot.MDS=TRUE,plot.importance=TRUE,print.result=FALSE,
              guess.num=50,boxplot.top=5,
              title="",boxplot.y="Abundance(%)",
              seed=100)


### T4 EEN response vs. aniTNF response
random_forest(taxa.data,
              subset(sample.info,Type=='PLEASE-T4' & (Treatment.Specific=='EEN' | Treatment.Specific=='antiTNF') & Response=='Response',  select=Treatment.Specific),
              plot.MDS=TRUE,plot.importance=TRUE,print.result=FALSE,
              guess.num=50,boxplot.top=5,
              title="",boxplot.y="Abundance(%)",
              seed=100)


### T1 EEN response vs. aniTNF response
random_forest(taxa.data,
              subset(sample.info,Type=='PLEASE-T1' & (Treatment.Specific=='EEN' | Treatment.Specific=='antiTNF') & Response=='Response',  select=Treatment.Specific),
              plot.MDS=TRUE,plot.importance=TRUE,print.result=FALSE,
              guess.num=50,boxplot.top=5,
              title="",boxplot.y="Abundance(%)",
              seed=100)
```

### Wilcoxon rank test
```{r bacteria_rank_test_of_een_r_vs_control,cache=TRUE}
### T4 EEN response vs. control
bact.EEN.R.T4 <- wilcox_test(taxa.data,
        subset(sample.info,Type=='COMBO' | (Type=='PLEASE-T4' & Treatment.Specific=='EEN' & Response=='Response'),  select=Type),
      plot.heatmap=FALSE,boxplot.top=6,boxplot.y='Abundance(%)')
###
write.csv(bact.EEN.R.T4,file=paste('../3_Result/2015_03_05_Bacteria_EEN_T4R_vs_Control_Wilcox.csv',sep=''),row.names=TRUE)


### T1 EEN response vs. control
bact.EEN.R.T1 <- wilcox_test(taxa.data,
        subset(sample.info,Type=='COMBO' | (Type=='PLEASE-T1' & Treatment.Specific=='EEN' & Response=='Response'),  select=Type),
      plot.heatmap=FALSE,boxplot.top=6,boxplot.y='Abundance(%)')
write.csv(bact.EEN.R.T1,file=paste('../3_Result/2015_03_05_Bacteria_EEN_T1R_vs_Control_Wilcox.csv',sep=''),row.names=TRUE)

### T4 antiTNF response vs. control
bact.antiTNF.R.T4 <- wilcox_test(taxa.data,
        subset(sample.info,Type=='COMBO' | (Type=='PLEASE-T4' & Treatment.Specific=='antiTNF' & Response=='Response'),  select=Type),
      plot.heatmap=FALSE,boxplot.top=0.05,boxplot.y='Abundance(%)')
write.csv(bact.antiTNF.R.T4,file=paste('../3_Result/2015_03_05_Bacteria_anitTNF_T4R_vs_Control_Wilcox.csv',sep=''),row.names=TRUE)

### T1 antiTNF response vs. control
bact.antiTNF.R.T1 <- wilcox_test(taxa.data,
        subset(sample.info,Type=='COMBO' | (Type=='PLEASE-T1' & Treatment.Specific=='antiTNF' & Response=='Response'),  select=Type),
      plot.heatmap=FALSE,boxplot.top=0.05,boxplot.y='Abundance(%)')
write.csv(bact.antiTNF.R.T1,file=paste('../3_Result/2015_03_05_Bacteria_anitTNF_T1R_vs_Control_Wilcox.csv',sep=''),row.names=TRUE)

### T4 EEN response vs. aniTNF response
bact.EEN.antiTNF.R.T4 <- wilcox_test(taxa.data,
        subset(sample.info,Type=='PLEASE-T4' & (Treatment.Specific=='EEN' | Treatment.Specific=='antiTNF') & Response=='Response',  select=Treatment.Specific),
      plot.heatmap=FALSE,boxplot.top=6,boxplot.y='Abundance(%)')
write.csv(bact.EEN.antiTNF.R.T4,file=paste('../3_Result/2015_03_05_Bacteria_anitTNF_T4R_vs_EEN_T4R_Wilcox.csv',sep=''),row.names=TRUE)


### T1 EEN response vs. aniTNF response
bact.EEN.antiTNF.R.T1 <- wilcox_test(taxa.data,
        subset(sample.info,Type=='PLEASE-T1' & (Treatment.Specific=='EEN' | Treatment.Specific=='antiTNF') & Response=='Response',  select=Treatment.Specific),
      plot.heatmap=FALSE,boxplot.top=6,boxplot.y='Abundance(%)')
write.csv(bact.EEN.antiTNF.R.T1,file=paste('../3_Result/2015_03_05_Bacteria_anitTNF_T1R_vs_EEN_T1R_Wilcox.csv',sep=''),row.names=TRUE)

```

## Compare time points

### T1 vs. T2/3/4 by treatment
```{r bacteria_T1_vs_T234_treatment,cache=TRUE}
wil.by.time <- add_rownames(taxa.data,var='Sample') %>%
  gather(Taxa,Abu,-Sample) %>%
  left_join(dplyr::select(add_rownames(sample.info,var='Sample'),Subject,Sample,Time,Treatment.Specific,Response), by='Sample') %>%  
  filter(!is.na(Time)) %>% 
  mutate(Time=paste('T',Time,sep='')) %>%
  dplyr::select(-Sample) %>% 
  spread(Time,Abu) %>%
  group_by(Treatment.Specific,Taxa) %>%
  summarise_each(funs(wilcox.test(., T1, pair = TRUE)$p.value), vars = T2:T4) %>%
  ## raw pvalues
  rename(T1.vs.T2.p=vars1,T1.vs.T3.p=vars2,T1.vs.T4.p=vars3) %>%
  mutate(T1.vs.T2.q = p.adjust(T1.vs.T2.p,method='fdr'),
         T1.vs.T3.q = p.adjust(T1.vs.T3.p,method='fdr'),
         T1.vs.T4.q = p.adjust(T1.vs.T4.p,method='fdr')
         ) %>%
  mutate_each(funs(signif(.,digits=2)),T1.vs.T2.p:T1.vs.T4.q) %>%
  arrange(Treatment.Specific,T1.vs.T2.p,T1.vs.T3.p,T1.vs.T4.p) 

 
########### save the results
write.csv(wil.by.time,file=paste('../3_Result/2015_03_05_Bacteria_Compare_Time_Points_by_Treatment_Wilcox.csv',sep=''),row.names=FALSE)

########## print the significant results
wil.by.time %>%
  dplyr::filter(T1.vs.T2.p<0.05 | T1.vs.T3.p<0.05 | T1.vs.T4.p<0.05) %>%
  as.data.frame

```

### EEN  T1 vs. T2
```{r bacteria_T1_vs_T2_EEN,cache=TRUE}
wil.t1.t2.een <- add_rownames(taxa.data,var='Sample') %>%
  gather(Taxa,Abu,-Sample) %>%
  left_join(dplyr::select(add_rownames(sample.info,var='Sample'),Subject,Sample,Time,Treatment.Specific,Response), by='Sample') %>%  
  filter(!is.na(Time)) %>% 
  mutate(Time=paste('T',Time,sep='')) %>%
  dplyr::select(-Sample) %>% 
  spread(Time,Abu) %>% 
  dplyr::filter(Treatment.Specific=='EEN') %>% 
  group_by(Taxa) %>%
  summarise(median.T1=median(T1,na.rm=TRUE),mad.T1=mad(T1,na.rm=TRUE),
            median.T2=median(T2,na.rm=TRUE),mad.T2=mad(T2,na.rm=TRUE),
            mean.T1=mean(T1,na.rm=TRUE),sd.T1=sd(T1,na.rm=TRUE),
            mean.T2=mean(T2,na.rm=TRUE),sd.T2=sd(T2,na.rm=TRUE),
            T1.vs.T2.p=wilcox.test(T1, T2, pair = TRUE)$p.value
            ) %>%
  ## raw pvalues
  mutate(T1.vs.T2.q = p.adjust(T1.vs.T2.p,method='fdr'))
  
 
########### save the results
write.csv(wil.t1.t2.een,file=paste('../3_Result/2015_03_05_Bacteria_EEN_T1_vs_T2_Wilcox.csv',sep=''),row.names=FALSE)

```


### T1 vs. T2/3/4 by treatment and response
```{r bacteria_T1_vs_T234_treatment_response,cache=TRUE}
add_rownames(taxa.data,var='Sample') %>%
  gather(Taxa,Abu,-Sample) %>%
  left_join(dplyr::select(add_rownames(sample.info,var='Sample'),Subject,Sample,Time,Treatment.Specific,Response), by='Sample') %>%  
  filter(!is.na(Time) & Treatment.Specific != 'PEN' & !is.na(Response)) %>% 
  mutate(Time=paste('T',Time,sep='')) %>%
  dplyr::select(-Sample) %>% 
  spread(Time,Abu) %>%
  group_by(Treatment.Specific,Response,Taxa) %>%
  summarise_each(funs(wilcox.test(., T1, pair = TRUE)$p.value), vars = T2:T4) %>%
  #mutate_each(funs(p.adjust(.,method='fdr')),vars1:vars3) %>%
  ## raw pvalues
  mutate_each(funs(signif(.,digits=2)),vars1:vars3) %>%
  rename(T1.vs.T2=vars1,T1.vs.T3=vars2,T1.vs.T4=vars3) %>%
  arrange(Treatment.Specific,T1.vs.T2,T1.vs.T3,T1.vs.T4) %>%
  dplyr::filter(T1.vs.T2<0.05 | T1.vs.T3<0.05 | T1.vs.T4<0.05) %>%
  as.data.frame 


```


### Heatmap: all four time points
```{r bacteria_heatmap_all_four_time_points,cache=TRUE,fig.width=15,fig.height=10}
## Rick:
## The reviewer wondered about how the bacterial taxa changed over time.  I think we need some heat maps to summarize this, which we can then add to the supplemental information.
## How about making four heat maps, showing the two therapies, each separated by responders and non responders.  There would be four columns, one for each time point.  Rows would be bacteria, either the most abundant, or the ones that changed the most.

data.heat <- 
  add_rownames(as.data.frame(taxa.data),var='Sample') %>%
  left_join(add_rownames(sample.info,var='Sample'),by='Sample') %>% 
  arrange(Disease,Time,Treatment.Specific,Response,Cluster,Antibiotics.visit,FCP) %>%
  dplyr::select(1:(ncol(taxa.data)+1)) %>%  ## first column is sample name
  add_back_rownames(row.var='Sample') 

bk <- unique(c(
        0,
        min(min(taxa.data[taxa.data>0])/2,0.1),
        seq(min(min(taxa.data[taxa.data>0])/2,1),0.1,length=10),
        seq(0.1,1,length=10),
        seq(1,5,length=10),
        seq(5,20,length=20),
        seq(20,50,length=20),
        seq(50,100,length=20)
        ))
hmcols<- colorRampPalette(c('white','#1f78b4','#ff7f00','#e41a1c'))(length(bk)-1)
#bk = unique(c(seq(0,5, length=100),seq(5,10,length=100),seq(10,100,length=100)))
#hmcols<- colorRampPalette(c("white","firebrick3","yellow"))(length(bk)-1) 

#### distance for taxa
dist.taxa <- calculate_distance(X = t(data.heat), method = bacteral.distance.method)

#######################
 #######################
 ### use the new parameters in pheatmap to separate different treatments
 # Gaps in heatmaps
pheatmap(t(data.heat),
         border_color="grey60",
         cluster_cols = FALSE,
         clustering_distance_rows = dist.taxa,
         #clustering_distance_cols = bact.dist,
         #clustering_method = "complete",
         annotation=sample.info[,c("Disease","Time","Treatment.Specific","Response","Cluster","FCP","Antibiotics.visit","Steroids")],
         color =hmcols, breaks=bk,
         show_rownames = TRUE,
         scale="none",
         gaps_col = c(26, 111,191,273)
         )
rm(data.heat)
rm(dist.taxa)
```


```{r bacteria_heatmap_all_four_time_points_v2,cache=TRUE,fig.width=15,fig.height=10}
######
### different ways of arranging the samples (columns)
data.heat <- 
  add_rownames(as.data.frame(taxa.data),var='Sample') %>%
  left_join(add_rownames(sample.info,var='Sample'),by='Sample') %>% 
  arrange(Disease,Treatment.Specific,Response,Time,Cluster,Antibiotics.visit,FCP) %>%
  dplyr::select(1:(ncol(taxa.data)+1)) %>%  ## first column is sample name
  add_back_rownames(row.var='Sample') 

bk <- unique(c(
        0,
        min(min(taxa.data[taxa.data>0])/2,0.1),
        seq(min(min(taxa.data[taxa.data>0])/2,1),0.1,length=10),
        seq(0.1,1,length=10),
        seq(1,5,length=10),
        seq(5,20,length=20),
        seq(20,50,length=20),
        seq(50,100,length=20)
        ))
hmcols<- colorRampPalette(c('white','#1f78b4','#ff7f00','#e41a1c'))(length(bk)-1)
#bk = unique(c(seq(0,5, length=100),seq(5,10,length=100),seq(10,100,length=100)))
#hmcols<- colorRampPalette(c("white","firebrick3","yellow"))(length(bk)-1) 

#### distance for taxa
dist.taxa <- calculate_distance(X = t(data.heat), method = bacteral.distance.method)

#######################
 #######################
 ### use the new parameters in pheatmap to separate different treatments
 # Gaps in heatmaps
pheatmap(t(data.heat),
         border_color="grey60",
         cluster_cols = FALSE,
         clustering_distance_rows = dist.taxa,
         #clustering_distance_cols = bact.dist,
         #clustering_method = "complete",
         annotation=sample.info[,c("Disease","Treatment.Specific","Response","Time","Cluster","FCP","Antibiotics.visit","Steroids")],
         color =hmcols, breaks=bk,
         show_rownames = TRUE,
         scale="none",
         gaps_col = c(26, 53, 90, 94, 138, 144, 156, 211, 313, 353)
         )
rm(data.heat)
rm(dist.taxa)
```



```{r  bacteria_heatmap_all_four_time_points_v3,cache=TRUE,fig.width=15,fig.height=10}
#### manually re-order the samples
data.heat <- 
  add_rownames(as.data.frame(taxa.data),var='Sample') %>%
  left_join(add_rownames(sample.info,var='Sample'),by='Sample') %>% 
  arrange(Cluster,Treatment.Specific,Response,Time,Antibiotics.visit,FCP) %>%
  dplyr::select(1:(ncol(taxa.data)+1)) %>%  ## first column is sample name
  add_back_rownames(row.var='Sample') 

bk <- unique(c(
        0,
        min(min(taxa.data[taxa.data>0])/2,0.1),
        seq(min(min(taxa.data[taxa.data>0])/2,1),0.1,length=10),
        seq(0.1,1,length=10),
        seq(1,5,length=10),
        seq(5,20,length=20),
        seq(20,50,length=20),
        seq(50,100,length=20)
        ))
hmcols<- colorRampPalette(c('white','#1f78b4','#ff7f00','#e41a1c'))(length(bk)-1)
#bk = unique(c(seq(0,5, length=100),seq(5,10,length=100),seq(10,100,length=100)))
#hmcols<- colorRampPalette(c("white","firebrick3","yellow"))(length(bk)-1) 

#### distance for taxa
dist.taxa <- calculate_distance(X = t(data.heat), method = bacteral.distance.method)

heat.res <- pheatmap(t(data.heat),border_color="grey60",
         cluster_cols = FALSE,
         clustering_distance_rows = dist.taxa,
         #clustering_distance_cols = bact.dist,
         #clustering_method = "complete",
         annotation=sample.info[,c("Disease","Time","Response","Treatment.Specific","Cluster","FCP","Antibiotics.visit","Steroids")],
         color =hmcols, breaks=bk,
         show_rownames = TRUE,
         scale="none")
ordered.labels <- heat.res$tree_row$labels[heat.res$tree_row$order]

```

```{r bacteria_heatmap_all_four_time_points_separately,cache=TRUE,fig.width=15,fig.height=8}
#### plot each time point
bk <- unique(c(
        0,
        min(min(taxa.data[taxa.data>0])/2,0.1),
        seq(min(min(taxa.data[taxa.data>0])/2,1),0.1,length=10),
        seq(0.1,1,length=10),
        seq(1,5,length=10),
        seq(5,20,length=20),
        seq(20,50,length=20),
        seq(50,100,length=20)
        ))
hmcols<- colorRampPalette(c('white','#1f78b4','#ff7f00','#e41a1c'))(length(bk)-1)
### #1f78b4' blue
### '#e41a1c' red
### '#ff7f00' orange
### #4daf4a green

##### COMBO
pheatmap(t(subset_data_by_sample_info(data.heat,sample.info,Type=='COMBO'))[ordered.labels,],
         border_color="grey60",
         cluster_cols = FALSE,
         cluster_rows = FALSE, 
         annotation=sample.info[,c("Disease","Cluster")],
         color =hmcols, breaks=bk,
         show_rownames = TRUE,
          main='COMBO',
         cellwidth = 8, cellheight = 8,
         scale="none")

##### PLEASE
for (i in c("antiTNF","EEN","PEN")){
for (j in c("PLEASE-T1","PLEASE-T2","PLEASE-T3","PLEASE-T4")){
data.temp <- subset_data_by_sample_info(data.heat,sample.info,Treatment.Specific==i & Type==j)
#dist.temp <- calculate_distance(X = data.temp, method = bacteral.distance.method)
pheatmap(t(data.temp)[ordered.labels,],
         border_color="grey60",
         #clustering_distance_rows = 'correlation',
         #clustering_distance_cols = dist.temp,
         #clustering_method = "complete",
         cluster_cols = FALSE,
         cluster_rows = FALSE, 
         annotation=sample.info[,c("Disease","Response","Cluster","FCP","Antibiotics.visit","Steroids")],
         color =hmcols, breaks=bk,
         show_rownames = TRUE,
         main=paste(i,j),
         cellwidth = 8, cellheight = 8,
         scale="none")
}}


```

```{r bacteria_heatmap_pdf,eval=FALSE,include=FALSE}
###########
pdf('Heatmap_four_time_points.pdf',width=30,height=10)
#### plot each time point
bk <- unique(c(
        0,
        min(min(taxa.data[taxa.data>0])/2,0.1),
        seq(min(min(taxa.data[taxa.data>0])/2,1),0.1,length=10),
        seq(0.1,1,length=10),
        seq(1,5,length=10),
        seq(5,20,length=20),
        seq(20,50,length=20),
        seq(50,100,length=20)
        ))
hmcols<- colorRampPalette(c('white','#1f78b4','#ff7f00','#e41a1c'))(length(bk)-1)
##### COMBO
pheatmap(t(subset_data_by_sample_info(data.heat,sample.info,Type=='COMBO'))[ordered.labels,],
         border_color="grey60",
         cluster_cols = FALSE,
         cluster_rows = FALSE, 
         annotation=sample.info[,c("Disease","Cluster")],
         color =hmcols, breaks=bk,
         show_rownames = TRUE,
         main='COMBO',
         cellwidth = 8, cellheight = 8,
         scale="none")

##### PLEASE
for (i in c("antiTNF","EEN","PEN")){
for (j in c("PLEASE-T1","PLEASE-T2","PLEASE-T3","PLEASE-T4")){
data.temp <- subset_data_by_sample_info(data.heat,sample.info,Treatment.Specific==i & Type==j)
#dist.temp <- calculate_distance(X = data.temp, method = bacteral.distance.method)
pheatmap(t(data.temp)[ordered.labels,],
         border_color="grey60",
         #clustering_distance_rows = 'correlation',
         #clustering_distance_cols = dist.temp,
         #clustering_method = "complete",
         cluster_cols = FALSE,
         cluster_rows = FALSE, 
         annotation=sample.info[,c("Disease","Response","Cluster","FCP","Antibiotics.visit","Steroids")],
         color =hmcols, breaks=bk,
         show_rownames = TRUE,
         main=paste(i,j),
         cellwidth = 8, cellheight = 8,
         scale="none")
}}

dev.off()
```

## Treatment, antibiotics use, FCP on bacterial abundance
### Logistic regression
```{r bacteria_logistic_regression_treatment_antibiotics_fcp,cache=TRUE,eval=FALSE}
## T4
logistic.res <- list()
for (spe in colnames(taxa.data)){
  taxa.tem <- taxa.data[,spe,drop=FALSE]
  colnames(taxa.tem) <- 'Taxa'
  logistic.res[[spe]] <- add_rownames(taxa.tem,var='Sample') %>%
    left_join(add_rownames(sample.info,var='Sample'),by='Sample') %>%
    dplyr::filter(Type=='PLEASE-T4') %>%
    dplyr::select(Taxa,Treatment.Specific,Antibiotics.visit,log.FCP) %>%
    mutate(Taxa.present = ifelse(Taxa>0,'Present','Absent')) %>%
    glm(as.factor(Taxa.present) ~ Treatment.Specific + Antibiotics.visit + log.FCP,
        family=binomial(logit), data=.) %>%
    summary %>% coef  %>%  
    as.data.frame %>% add_rownames(var='Var') %>%
    dplyr::select(1,2,5) %>% dplyr::select(Var,Est=Estimate,Pval=starts_with("Pr"))%>%
    gather(value,var,Est,Pval) %>% 
    unite(Variables,Var,value, sep = ".", remove = TRUE) %>%
    add_back_rownames(row.var='Variables') %>% as.matrix
  ####
  colnames(logistic.res[[spe]]) <- spe
  ###  
  rm(taxa.tem)
}

logistic.res <- signif(t(as.data.frame(logistic.res)),2)

print(logistic.res)
######
write.csv(logistic.res,file=paste('../3_Result/2015_03_05_Bacteria_Logistic_Regression_Treatment_Antibiotics_FCP.csv',sep=''),row.names=TRUE)
```


### quantile regression on fold change
```{r bacteria_fc_quantile_regression_treatment_antibiotics_fcp,cache=TRUE}
## fold change T4/T1
rq.fc.res <- list()
for (spe in colnames(taxa.data)){
  #spe <- "Bacteroides"
  ### it seems rq used some random numbers
  set.seed(1)
  bact.tem <- taxa.data[,spe,drop=FALSE]
  colnames(bact.tem) <- 'Taxa'
    ####
    master.table <- add_rownames(as.data.frame(bact.tem),var='Sample') %>%
    left_join(add_rownames(sample.info,var='Sample'),by='Sample') %>%
    dplyr::filter(Time==1 | Time==4) %>%
    dplyr::select(Taxa,Subject,Time,Treatment.Specific,Antibiotics.visit,log.FCP,Response) %>%
    mutate(Time=ifelse(Time==1,'T1','T4'))
    ### taxa
    taxa.temp <- master.table %>%
    mutate(Taxa = Taxa+min(Taxa[Taxa>0]))  %>% 
    dplyr::select(Subject,Taxa,Time,Treatment.Specific) %>%
    spread(Time,Taxa) %>%
    mutate(Taxa.log.fc = log(T4/T1)) %>%
    dplyr::select(Subject,Treatment.Specific,Taxa.log.fc)
    ### response
    response.temp <- master.table %>%
    dplyr::select(Subject,Response) 
    ### log FCP
    log.fcp.temp <- master.table %>%
    dplyr::select(Subject,Time,log.FCP) %>%
    spread(Time,log.FCP) %>%
    mutate(log.FCP.T4 = T4) %>%
    dplyr::select(Subject,log.FCP.T4)
    ### antibiotics
    antibiotics.temp <- master.table %>%
    dplyr::select(Subject,Time,Antibiotics.visit) %>%
    spread(Time,Antibiotics.visit) %>%
    mutate(antibiotics = ifelse(is.na(T4),T1,T4)) %>%
    dplyr::select(Subject,antibiotics)
    ### merge together 
    rq.fc.res[[spe]] <- taxa.temp %>% full_join(antibiotics.temp,by='Subject') %>%
    full_join(log.fcp.temp,by='Subject') %>%
    full_join(response.temp,by='Subject') %>%
    ### !!!!!!!!!! ########
    ### After updating RStudio and some R packages,
    ### a problem has occured: the reference level in the linear model
    ### has changed for no reason
    ### Change the reference level back to EEN for consistency with
    ### previous analysis
    ###
    mutate( Treatment.Specific = relevel(as.factor(Treatment.Specific), ref = 'EEN')) %>%
    ######################
    ### quantile regression
    rq( Taxa.log.fc ~ Treatment.Specific + 
                      antibiotics + 
                      Response,
        tau=0.75,data=.,na.action=na.omit) %>%
    summary.rq(.,se="boot") %>% coef %>%
    as.data.frame %>% add_rownames(var='Var') %>%
    dplyr::select(1,2,5) %>% dplyr::select(Var,Est=Value,Pval=starts_with("Pr"))%>%
    gather(value,var,Est,Pval) %>% 
    unite(Variables,Var,value, sep = ".", remove = TRUE) %>%
    add_back_rownames(row.var='Variables') %>% as.matrix
  #### add pathway name
  colnames(rq.fc.res[[spe]]) <- spe
  ###  
  rm(bact.tem)
  rm(master.table)
  rm(taxa.temp)
  rm(antibiotics.temp)
  rm(log.fcp.temp)
  #break
}

rq.res.table <- 
  as.data.frame(t(as.data.frame(rq.fc.res))) %>% ## must use as.data.frame again 
  add_rownames(var='taxa') %>%
  mutate(
    Treatment.SpecificPEN.Qval = p.adjust(Treatment.SpecificPEN.Pval,method='fdr'),
    Treatment.SpecificantiTNF.Qval = p.adjust(Treatment.SpecificantiTNF.Pval,method='fdr'),
    Antibiotics.visitUse.Qval = p.adjust(antibioticsUse.Pval,method='fdr'),
    #log.FCP.Qval = p.adjust(log.FCP.T4.Pval,method='fdr')
    ResponseResponse.Qval = p.adjust(ResponseResponse.Pval ,method='fdr')
  ) 


######
#### row name is a column
write.csv(rq.res.table,file=paste('../3_Result/2015_03_05_Bacteria_Fold_Change_Quantile_Q75_Regression_Treatment_Antibiotics_Response.csv',sep=''),row.names=FALSE)

rm(rq.fc.res)
rm(rq.res.table)
```


## Distance-based analysis
### Calculate the distance to COMBO centroid
```{r bacteria_calculate_distance_to_combo_centroid,cache=TRUE}
## use mean
COMBO.centroid <- apply(subset_data_by_sample_info(taxa.data,sample.info,Type=='COMBO'),2,mean)
head(COMBO.centroid)
PLEASE.COMBO.centroid <- rbind(taxa.data,COMBO.centroid=COMBO.centroid)
dist.all <- calculate_distance(X = PLEASE.COMBO.centroid, method = bacteral.distance.method)

### Get the distance to the COMBO centroid for all samples
dist.all.mat <- as.matrix(dist.all)
dist.to.cen <- dist.all.mat[-which(colnames(dist.all.mat)=='COMBO.centroid'),"COMBO.centroid",drop=FALSE]
colnames(dist.to.cen) <- 'Distance'
### Check the data maniputation
####Before data manipulation
dist.all.mat["5001-01","COMBO.centroid"]
####After data manipulation
dist.to.cen["5001-01",]


dim(dist.to.cen)

### What is the distribution of the distance    
### The data looks normal, we can use t.test in the following analysis
hist(dist.to.cen,xlab='Distance to the COMBO centroid',breaks=20,main='')

### Add distance information to the sample information
sample.info <- add_rownames(sample.info,var='Sample') %>%
  left_join(add_rownames(as.data.frame(dist.to.cen),var='Sample'),by='Sample') %>%
  add_back_rownames(row.var='Sample')


##############################
### save the sample information
### again
write.csv(sample.info,file=paste('../1_Data/Processed_Data/Sample_Information/2015_02_13_Processed_Sample_Information.csv',sep=''),row.names=TRUE)
```
 
 
```{r distance_to_combo_centroid_mds_plot,cache=TRUE}
## use distance as input
## the centroid is not in the center at MDS plot
ind <- c(grep('-',rownames(as.matrix(dist.all)),invert=TRUE),
       grep('-01',rownames(as.matrix(dist.all)),invert=FALSE))
dist.all.cp1 <- as.dist(as.matrix(dist.all)[ind,ind])
set.seed(10)
meta.MDS.cen <- metaMDS(dist.all.cp1)

## use raw data as input
## the centroid is in the center at MDS plot with binary jaccard distance
## but in the center with numerical jaccard distance
distf <- function(X,dist.method,tree=NA){
  if (dist.method == 'UniFrac'){
    if (is.na(tree)){stop('Need the tree information to run UniFrac!')}
    #overlap.species <- intersect(colnames(X),tree$tip.label)
    unifracs <- GUniFrac(X,tree,alpha=0.5)
    dist <- as.dist(unifracs$unifracs[,,1])
  }
  else if (dist.method == 'Binary_Jaccard'){
    dist <- vegdist(X,method='jaccard',binary=TRUE)
  }
  else if (dist.method == 'Numerical_Jaccard'){
    dist <- vegdist(X,method='jaccard',binary=FALSE)
  }
  else if (dist.method == 'Bray'){
    dist <- vegdist(X,method='bray',binary=FALSE)
  }
  else {
    stop('method is not correct')
  }
  return(dist)
}

ind <- c(grep('-',rownames(PLEASE.COMBO.centroid),invert=TRUE),
       grep('-01',rownames(PLEASE.COMBO.centroid),invert=FALSE))
PLEASE.COMBO.centroid.cp1 <- PLEASE.COMBO.centroid[ind,]
set.seed(10)
meta.MDS.cen <- metaMDS(PLEASE.COMBO.centroid.cp1,distfun=distf,dist.method = 'Numerical_Jaccard')

## The control centroid is calculated by averaged species abundance across all
## control samples. Unless the species has zero abundance across all samples,
## the average would be non-zero. Therefore, the control centroid would have non-zero
## abundance across all species. When using the binary Jaccard distance,
## the distance between any control sample and control centroid may not give the
## the centroid as the center.
## The controid was calculated numerically, but the distance was calculated binaryly.

### Figure for Rick
### we start with the same MDS as in 1B, but don't show human DNA or near and far cluster.  Control is shown as the filled circle, Crohn's as open circles.  Current antibiotic use is shown by blue versus red coloring.  If not too hard, maybe also plot the centroid of the healties as a large yellow triangle.
p <- add_rownames(as.data.frame(meta.MDS.cen$points),var='Sample') %>%
  left_join(dplyr::select(add_rownames(sample.info,var='Sample'),
                          Sample,Type,Disease,Antibiotics.visit),
            by='Sample') %>% 
  dplyr::filter(is.na(Type) | Type != 'PLEASE-T2' & Type != 'PLEASE-T3' & Type != 'PLEASE-T4') %>%
  mutate(Disease=ifelse(is.na(Disease),'Centroid',Disease)) %>%
  mutate(Antibiotics.visit=ifelse(is.na(Antibiotics.visit),'Centroid',Antibiotics.visit)) %>%
  ggplot(aes(x=MDS1,y=MDS2))+ 
  geom_point(aes(colour = Antibiotics.visit,shape= Disease),size=3) +
  scale_color_manual(values=c("darkgoldenrod1","#0066CC", "#FF3333"),name="") +
  scale_shape_manual(values=c(3,20,1),name="") +
  #scale_size_area(range = c(1, 10))+
  #scale_size_continuous(range = c(3, 10)) +
  labs(title = 'Ordination bacterial taxa') +
  ##########################
  theme_classic()
print(p)
rm(p)
rm(ind)
rm(PLEASE.COMBO.centroid.cp1)
rm(dist.all.cp1)
rm(meta.MDS.cen)
```



### Plot the distance to the COMBO centroid
```{r plot_distance_to_combo_centroid,cache=TRUE}
### Plot distance by treatment
p<- sample.info %>% dplyr::select(Type,Distance,Treatment) %>%
  ggplot(aes(factor(Type),Distance)) + 
  geom_boxplot(aes(fill = factor(Treatment)))+
  xlab('')+ylab('Distance to the COMBO centroid')
plot(p)

p<- sample.info %>% dplyr::select(Type,Distance,Treatment.Specific) %>%
  ggplot(aes(factor(Type),Distance)) + 
  geom_boxplot(aes(fill = factor(Treatment.Specific)))+
  xlab('')+ylab('Distance to the COMBO centroid')
plot(p)

#### For publication
p<- sample.info %>% dplyr::select(Type,Distance,Treatment.Specific) %>%
  ggplot(aes(factor(Type),Distance)) + 
  geom_boxplot(aes(fill = factor(Treatment.Specific)),outlier.shape = NA)+
  xlab('')+ylab('Distance to the COMBO centroid') + 
  theme_classic() 
plot(p)



#### Plot distance by response
p<- sample.info %>% dplyr::select(Type,Distance,Response) %>%
  filter(Type=='COMBO' | (Type!='COMBO' & !is.na(Response))) %>%
  ggplot(aes(factor(Type),Distance)) + 
  geom_boxplot(aes(fill = factor(Response)))+
  xlab('')+ylab('Distance to the COMBO centroid')
plot(p)


#### Plot distance by treatment,response
p<- sample.info %>% dplyr::select(Type,Distance,Treatment.Specific,Response) %>%
  mutate(Response=ifelse(Type=='COMBO','COMBO',Response)) %>%
  mutate(Treatment.Specific=ifelse(Type=='COMBO','COMBO',Treatment.Specific)) %>%
  na.omit() %>%
  unite(Treatment.Response, Treatment.Specific,Response, sep = ".") %>%
  ggplot(aes(factor(Type),Distance)) + 
  geom_boxplot(aes(fill = factor(Treatment.Response)),outlier.shape = NA)+
  xlab('')+ylab('Distance to the COMBO centroid') + 
  theme_classic() 
plot(p)


### Plot distance by both treatment and response
p<- sample.info %>% dplyr::select(Distance,Response,Time,Treatment) %>%
  na.omit %>%
  ggplot( aes(factor(Time), Distance),na.rm=T)+
  geom_boxplot(na.rm=T)+facet_grid(Treatment ~ Response, margins=TRUE)
plot(p)

### Plot distance by both treatment(PEN vs. EEN) and response
p<- sample.info %>% dplyr::select(Distance,Response,Time,Treatment.Specific) %>%
  na.omit %>%
  ggplot(aes(factor(Time), Distance),na.rm=T)+
  geom_boxplot(na.rm=T)+facet_grid(Treatment.Specific ~ Response, margins=TRUE)
plot(p)
```


### Compare distance change between time points 
```{r test_distance_to_combo_centroid, cache=TRUE}
####  T1 vs. T2,3,4 
sample.info %>%
  dplyr::select(Subject,Time,Treatment,Distance) %>% 
  na.omit() %>% 
  mutate(Time=paste('T',Time,sep='')) %>%
  spread(Time,Distance) %>%
  summarise_each(funs(t.test(., T1, pair = TRUE)$p.value), vars = T2:T4) %>%
  rename(T1.vs.T2=vars1,T1.vs.T3=vars2,T1.vs.T4=vars3)


####  T1 vs. T2,3,4 for each treatment arm
sample.info %>%
  dplyr::select(Subject,Time,Treatment,Distance) %>% 
  na.omit() %>% 
  mutate(Time=paste('T',Time,sep='')) %>%
  spread(Time,Distance) %>%
  group_by(Treatment) %>%
  summarise_each(funs(t.test(., T1, pair = TRUE)$p.value), vars = T2:T4) %>%
  rename(T1.vs.T2=vars1,T1.vs.T3=vars2,T1.vs.T4=vars3)

####  T1 vs. T2,3,4 for each treatment arm
sample.info %>%
  dplyr::select(Subject,Time,Treatment.Specific,Distance) %>% 
  na.omit() %>% 
  mutate(Time=paste('T',Time,sep='')) %>%
  spread(Time,Distance) %>% 
  group_by(Treatment.Specific) %>%
  summarise_each(funs(t.test(., T1, pair = TRUE)$p.value), vars = T2:T4) %>%
  rename(T1.vs.T2=vars1,T1.vs.T3=vars2,T1.vs.T4=vars3)


####  T1 vs. T2,3,4 for each treatment arm & response
sample.info %>%
  dplyr::select(Subject,Time,Treatment,Response,Distance) %>% 
  na.omit() %>% 
  mutate(Time=paste('T',Time,sep='')) %>%
  spread(Time,Distance) %>%
  group_by(Treatment,Response) %>%
  summarise_each(funs(t.test(., T1, pair = TRUE)$p.value), vars = T2:T4) %>%
  rename(T1.vs.T2=vars1,T1.vs.T3=vars2,T1.vs.T4=vars3)

####  T1 vs. T2,3,4 for each treatment arm & response
####  Not enough sample for PEN
sample.info %>%
  dplyr::select(Subject,Time,Treatment.Specific,Response,Distance) %>% 
  na.omit() %>% 
  mutate(Time=paste('T',Time,sep='')) %>%
  spread(Time,Distance) %>% 
  filter(Treatment.Specific=='EEN') %>%
  group_by(Treatment.Specific,Response) %>%
  summarise_each(funs(t.test(., T1, pair = TRUE)$p.value), vars = T2:T4) %>%
  rename(T1.vs.T2=vars1,T1.vs.T3=vars2,T1.vs.T4=vars3)
```

### Compare distance change on response status
```{r test_distance_on_response, cache=TRUE}
ttest.fun = function(tdata){ 
  R.vs.NR <- NA
  try(R.vs.NR <- signif(t.test(Distance~Response,data=tdata)$p.value, 2),silent=TRUE)
  return(as.data.frame(R.vs.NR))
}


####  R vs. NR: by treatment
sample.info %>%
  dplyr::select(Subject,Time,Treatment,Treatment.Specific,Distance,Response) %>% 
  na.omit() %>%
  mutate(Time=paste('T',Time,sep='')) %>%
  group_by(Treatment) %>%
  do(ttest.fun(.)) %>%
  add_back_rownames(row.var='Treatment')

sample.info %>%
  dplyr::select(Subject,Time,Treatment,Treatment.Specific,Distance,Response) %>% 
  na.omit() %>%
  mutate(Time=paste('T',Time,sep='')) %>%
  group_by(Treatment.Specific) %>%
  do(ttest.fun(.)) %>%
  add_back_rownames(row.var='Treatment.Specific')


####  R vs. NR: by time
sample.info %>%
  dplyr::select(Subject,Time,Treatment,Treatment.Specific,Distance,Response) %>% 
  na.omit() %>%
  mutate(Time=paste('T',Time,sep='')) %>%
  group_by(Time) %>%
  do(ttest.fun(.)) %>%
  add_back_rownames(row.var='Time')


####  R vs. NR: by time & treatment
sample.info %>%
  dplyr::select(Subject,Time,Treatment,Treatment.Specific,Distance,Response) %>% 
  na.omit() %>%
  mutate(Time=paste('T',Time,sep='')) %>%
  group_by(Time,Treatment) %>%
  do(ttest.fun(.)) 

sample.info %>%
  dplyr::select(Subject,Time,Treatment,Treatment.Specific,Distance,Response) %>% 
  na.omit() %>%
  mutate(Time=paste('T',Time,sep='')) %>%
  group_by(Time,Treatment.Specific) %>%
  do(ttest.fun(.))
```

### Compare distance change on treatment
```{r test_distance_on_treatment, cache=TRUE}
####  EEN-T1 response vs. control
sample.info %>%
dplyr::select(Subject,Type,Treatment.Specific,Distance,Response,Antibiotics.visit) %>% 
dplyr::filter(Type=='COMBO'| (Type=='PLEASE-T1' & Treatment.Specific=='EEN' & Response == 'Response')) %>%
dplyr::select(Type,Distance,Antibiotics.visit) %>%
na.omit() %>%
(function(tdata){
t.test(Distance~Type,data=tdata)$p.value %>% print
lm(Distance~Type+Antibiotics.visit,data=tdata) %>% summary %>% coef %>% print
boxplot(Distance~Type,data=tdata,ylab='Distance')
})

####  EEN-T4 response vs. control
sample.info %>%
dplyr::select(Subject,Type,Treatment.Specific,Distance,Response,Antibiotics.visit) %>% 
dplyr::filter(Type=='COMBO'| (Type=='PLEASE-T4' & Treatment.Specific=='EEN' & Response == 'Response')) %>%
dplyr::select(Type,Distance,Antibiotics.visit) %>%
na.omit() %>%
(function(tdata){
t.test(Distance~Type,data=tdata)$p.value %>% print
lm(Distance~Type+Antibiotics.visit,data=tdata) %>% summary %>% coef %>% print
boxplot(Distance~Type,data=tdata,ylab='Distance')
})

####  antiTNF-T1 response vs. control
sample.info %>%
dplyr::select(Subject,Type,Treatment.Specific,Distance,Response,Antibiotics.visit) %>% 
dplyr::filter(Type=='COMBO'| (Type=='PLEASE-T1' & Treatment.Specific=='antiTNF' & Response == 'Response')) %>%
dplyr::select(Type,Distance,Antibiotics.visit) %>%
na.omit() %>%
(function(tdata){
t.test(Distance~Type,data=tdata)$p.value %>% print
lm(Distance~Type+Antibiotics.visit,data=tdata) %>% summary %>% coef %>% print
boxplot(Distance~Type,data=tdata,ylab='Distance')
})

####  antiTNF-T4 response vs. control
sample.info %>%
dplyr::select(Subject,Type,Treatment.Specific,Distance,Response,Antibiotics.visit) %>% 
dplyr::filter(Type=='COMBO'| (Type=='PLEASE-T4' & Treatment.Specific=='antiTNF' & Response == 'Response')) %>%
dplyr::select(Type,Distance,Antibiotics.visit) %>%
na.omit() %>%
(function(tdata){
t.test(Distance~Type,data=tdata)$p.value %>% print
lm(Distance~Type+Antibiotics.visit,data=tdata) %>% summary %>% coef %>% print
boxplot(Distance~Type,data=tdata,ylab='Distance')
})

####  antiTNF-T1 response vs. EEN-T1 response
sample.info %>%
dplyr::select(Subject,Type,Treatment.Specific,Distance,Response) %>% 
dplyr::filter( (Type=='PLEASE-T1' & Treatment.Specific !='PEN' & Response == 'Response')) %>%
dplyr::select(Treatment.Specific,Distance) %>%
na.omit() %>%
(function(tdata){
t.test(Distance~Treatment.Specific,data=tdata)$p.value %>% print
boxplot(Distance~Treatment.Specific,data=tdata,ylab='Distance')
})


####  antiTNF-T4 response vs. EEN-T4 response
sample.info %>%
dplyr::select(Subject,Type,Treatment.Specific,Distance,Response) %>% 
dplyr::filter( (Type=='PLEASE-T4' & Treatment.Specific !='PEN' & Response == 'Response')) %>%
dplyr::select(Treatment.Specific,Distance) %>%
na.omit() %>%
(function(tdata){
t.test(Distance~Treatment.Specific,data=tdata)$p.value %>% print
boxplot(Distance~Treatment.Specific,data=tdata,ylab='Distance')
})
```


### Compare distance to healthy controls
```{r compare_distance_to_healthy_controls,cache=TRUE}
#### I am goint to re-plot the figure before
#### Plot distance by treatment,response
p<- sample.info %>% dplyr::select(Type,Distance,Treatment.Specific,Response) %>%
  mutate(Response=ifelse(Type=='COMBO','COMBO',Response)) %>%
  mutate(Treatment.Specific=ifelse(Type=='COMBO','COMBO',Treatment.Specific)) %>%
  na.omit() %>%
  unite(Treatment.Response, Treatment.Specific,Response, sep = ".") %>%
  ggplot(aes(factor(Type),Distance)) + 
  geom_boxplot(aes(fill = factor(Treatment.Response)),outlier.shape = NA)+
  xlab('')+ylab('Distance to the COMBO centroid') + 
  theme_classic() 
plot(p)

#### Calculate the pvalues for each box in the above figure
sample.info %>%
dplyr::select(Type,Time,Treatment.Specific,Distance,Response) %>% 
mutate(Treatment.Specific=ifelse(Type=='COMBO','COMBO',Treatment.Specific)) %>%
mutate(Response=ifelse(Type=='COMBO','COMBO',Response)) %>%
mutate(Time=paste('T',Time,sep='')) %>%
mutate(Time=ifelse(Type=='COMBO','COMBO',Time)) %>%
dplyr::filter(Treatment.Specific != 'PEN') %>% 
na.omit() %>%
### group by treatment
(function(tdata){
### merge the time and treatment
temp <- tdata %>% 
unite(Time.Treatment,Time,Treatment.Specific, sep = "_") %>%
dplyr::select(Time.Treatment,Distance)
#### take the COMBO samples out, use them in the t.test
cdata <- dplyr::filter(temp,Time.Treatment=='COMBO_COMBO') %>%
         dplyr::select(Distance)
### compare each group to combo group by t.test
group_by(temp,Time.Treatment) %>%
summarise(pval.compare.to.combo=t.test(Distance,cdata$Distance)$p.value) %>%
mutate(pval.compare.to.combo=signif(pval.compare.to.combo,2)) %>%
separate(Time.Treatment, into=c("Time","Treatment"), sep = "_") %>%
print
return(tdata)
}) %>%
### group by treatment and response
unite(Time.Treatment.Response,Time, Treatment.Specific,Response, sep = "_") %>%
dplyr::select(Time.Treatment.Response,Distance) %>%
(function(tdata){
cdata <- dplyr::filter(tdata,Time.Treatment.Response=='COMBO_COMBO_COMBO') %>%
         dplyr::select(Distance)
group_by(tdata,Time.Treatment.Response) %>%
summarise(pval.compare.to.combo=t.test(Distance,cdata$Distance)$p.value) %>%
mutate(pval.compare.to.combo=signif(pval.compare.to.combo,2)) %>%
separate(Time.Treatment.Response, into=c("Time","Treatment","Response"), sep = "_")
})
```



### Linear regression 
```{r linear_regression_on_distance_to_combo_centroid,cache=TRUE}
### all, adjust for baseline distance
lm.all <- sample.info %>%
  #filter(Treatment=='antiTNF')%>% 
  dplyr::select(Subject, Time, Distance,Response,Antibiotics.visit) %>%
  na.omit() %>%
  mutate(Time=paste('Dist_T',Time,sep='')) %>%
  spread(Time,Distance) %>% 
  lm(Dist_T4~Dist_T1+Response+as.factor(Antibiotics.visit),data=.)
summary(lm.all)
coefplot(lm.all)

### antiTNF, adjust for baseline distance
lm.antiTNF <- sample.info %>%
  filter(Treatment=='antiTNF')%>% 
  dplyr::select(Subject, Time, Distance,Response,Antibiotics.visit) %>%
  na.omit() %>%
  mutate(Time=paste('Dist_T',Time,sep='')) %>%
  spread(Time,Distance) %>% 
  lm(Dist_T4~Dist_T1+Response+as.factor(Antibiotics.visit),data=.)
summary(lm.antiTNF)
coefplot(lm.antiTNF)

### Diet 
lm.Diet <- sample.info %>%
  filter(Treatment=='Diet')%>% 
  dplyr::select(Subject, Time, Distance,Response,Antibiotics.visit) %>%
  na.omit() %>%
  mutate(Time=paste('Dist_T',Time,sep='')) %>%
  spread(Time,Distance) %>% 
  lm(Dist_T4~Dist_T1+Response+as.factor(Antibiotics.visit),data=.)
summary(lm.Diet)
coefplot(lm.Diet)

### EEN
lm.EEN <- sample.info %>%
  filter(Treatment.Specific=='EEN')%>% 
  dplyr::select(Subject, Time, Distance,Response,Antibiotics.visit) %>%
  na.omit() %>%
  mutate(Time=paste('Dist_T',Time,sep='')) %>%
  spread(Time,Distance) %>% 
  lm(Dist_T4~Dist_T1+Response+as.factor(Antibiotics.visit),data=.)
summary(lm.EEN)
coefplot(lm.EEN)

### PEN 
lm.PEN <- sample.info %>%
  filter(Treatment.Specific=='PEN')%>% 
  dplyr::select(Subject, Time, Distance,Response,Antibiotics.visit) %>%
  na.omit() %>%
  mutate(Time=paste('Dist_T',Time,sep='')) %>%
  spread(Time,Distance) %>% 
  lm(Dist_T4~Dist_T1+Response+as.factor(Antibiotics.visit),data=.)
summary(lm.PEN)
coefplot(lm.PEN)


####################################
### plot all coefficients in one plot
### the order is from bottom to top
coefplot(lm.all, xlim=c(-1, 1),  col.pts="black",intercept=TRUE)
coefplot(lm.antiTNF, add=TRUE, col.pts="red",  intercept=TRUE)
coefplot(lm.EEN, add=TRUE, col.pts="blue", intercept=TRUE, offset=0.2)
legend('topright',legend=c('All','antiTNF','EEN'),col=c("black","red","blue"),pch=19)
```






## Diversity
```{r calculate_diversity,cache=TRUE,eval=FALSE,include=FALSE}
### calculate shannon diversity
###
### !!! use raw metaphlan data (raw data) to calcualte the diversity
###
bact.div <- as.data.frame(apply(taxa.rare,1,function(x){-sum(x/100*log(x/100),na.rm=TRUE)}))
colnames(bact.div) <- 'bact.div'

```

Rarefraction
```{r rarefraction,cache=TRUE}
taxa.rare <- as.data.frame(taxa.rare.raw) %>%
add_rownames(var='Name') %>%
separate(Name,into=c('Sample','Reads','Depth','Repeats','Repeat'),sep='_') %>%
dplyr::select(-Reads,-Repeats) %>%
group_by(Sample,Depth) %>%
summarise_each(funs(mean),3+(1:ncol(taxa.rare.raw)))%>%
(function(tdata){
  print(tdata %>% dplyr::select(Sample,Depth) %>% group_by(Depth) %>% summarize(Num=n()) %>% arrange(Num))
  return(tdata)
}) %>%
dplyr::filter(Depth=='100000') %>%
dplyr::select(-Depth) %>%
add_back_rownames(row.var='Sample')
```



```{r unweighted_phylogenetic_diversity,cache=FALSE,eval=FALSE,include=FALSE}
### calculate phylogenetic diversity
bact.comm <- taxa.rare
phylo.tree.pru <- prune.sample(bact.comm,phylo.tree.uniq)
### plot tree
#for (i in row.names(taxa.raw)) {
#     plot(phylo.tree.pru, show.tip.label = FALSE, main = i)
#     tiplabels(tip = which(phylo.tree.pru$tip.label %in% names(which(taxa.raw[i,]> 0))), pch = 19, cex = 2)
# }
phylo.div <- pd(bact.comm, phylo.tree.pru, include.root = TRUE)

bact.div <- phylo.div[,'PD',drop=F] ## SR: species richness
colnames(bact.div) <- 'bact.div'
```


```{r abundance_weighted_phylogenetic_diversity,cache=TRUE,eval=TRUE,include=TRUE}
### got an error when using the species level PhyloPhAn tree
###
### calculate abundance-weighted phylogenetic diversity
bact.comm <- taxa.raw
phylo.tree.pru <- prune.sample(bact.comm,phylo.tree.uniq)
branches <- as.data.frame(matrix(NA,ncol=4,nrow=nrow(phylo.tree.pru$edge)))
branches[,1:2] <- phylo.tree.pru$edge
branches[,3] <- phylo.tree.pru$edge.length
##############################
wpd <- matrix(NA,ncol=1,nrow=nrow(taxa.raw))
rownames(wpd) <- rownames(taxa.raw)
for (j in 1:nrow(wpd)){
  for (i in 1:nrow(branches)){
    #print(c(j,i))
    leaves.node <- tips(phylo.tree.pru,branches[i,2])
    branches[i,4] <- mean(bact.comm[j,leaves.node]/100) ## in x% scale
  }
  number.of.branches <- nrow(phylo.tree.pru$edge)
  denominator <- sum(branches[,4],na.rm=TRUE)
  numerator <- sum(branches[,3]*branches[,4],na.rm=TRUE)
  wpd[j,1] <- number.of.branches * numerator / denominator
  #break
}
bact.div <- as.data.frame(wpd)
colnames(bact.div) <- 'bact.div'
```


```{r plot_phylogenetic_tree,include=TRUE,eval=TRUE,fig.height=20}
#pdf('PhyloPhlAn_tree.pdf',height=500,width=10)
plot(phylo.tree.pru,cex=0.5)
#dev.off()
```



```{r add_diversity_to_sample_info,cache=TRUE}
### Add diversity to the sample information
sample.info <- add_rownames(sample.info,var='Sample') %>%
  left_join(add_rownames(as.data.frame(bact.div),var='Sample'),by='Sample') %>%
  rename(Bact.Div=bact.div) %>%
  add_back_rownames(row.var='Sample')

##############################
### save the sample information
### again
write.csv(sample.info,file=paste('../1_Data/Processed_Data/Sample_Information/2015_02_13_Processed_Sample_Information.csv',sep=''),row.names=TRUE)
```



Plot the correlation between total non-human reads vs. diversity to see if they are correlated.
```{r correlation_between_diversity_and_non_human_reads,cache=TRUE}
### check the distribution of the diversity
hist(bact.div$bact.div,xlab='Diversity',main='',breaks=20)
#hist(log(bact.div$bact.div),xlab='log phylogenetic diversity',main='',breaks=20)

####
add_rownames(bact.div,var='Sample') %>%
  left_join(add_rownames(sample.info,var='Sample'),by='Sample') %>% 
  dplyr::select(bact.div,NonHumanReads) %>%
  mutate(log10.NonHumanReads=log10(NonHumanReads)) %>%
  na.omit() %>%  
  (function(dat){
    plot(bact.div~log10.NonHumanReads,data=dat)
    mtext(paste('Rank cor =',
                signif(cor(dat$bact.div,dat$log10.NonHumanReads,method='spearman'),2),
                'P =',
                signif(cor.test(dat$bact.div,dat$log10.NonHumanReads,method='spearman')$p.value,2)
                ))
  })
```

Plot the correlation between FCP vs. diversity to see if they are correlated.
```{r correlation_between_diversity_and_fcp,cache=TRUE}
### all
add_rownames(bact.div,var='Sample') %>%
  left_join(add_rownames(sample.info,var='Sample'),by='Sample') %>% 
  dplyr::select(bact.div,log.FCP) %>%
  na.omit() %>%  
  (function(dat){
    plot(bact.div~log.FCP,data=dat)
    mtext(paste('Rank cor =',
                signif(cor(dat$bact.div,dat$log.FCP,method='spearman'),2),
                'P =',
                signif(cor.test(dat$bact.div,dat$log.FCP,method='spearman')$p.value,2)
    ))
  })

### antiTNF response
add_rownames(bact.div,var='Sample') %>%
  left_join(add_rownames(sample.info,var='Sample'),by='Sample') %>% 
  dplyr::filter(Treatment=='antiTNF',Response=='Response') %>%
  dplyr::select(bact.div,log.FCP) %>%
  na.omit() %>%  
  (function(dat){
    plot(bact.div~log.FCP,data=dat)
    mtext(paste('Rank cor =',
                signif(cor(dat$bact.div,dat$log.FCP,method='spearman'),2),
                'P =',
                signif(cor.test(dat$bact.div,dat$log.FCP,method='spearman')$p.value,2)
    ))
  })
```



```{r diversity_change_aross_time,eval=FALSE,include=FALSE}
add_rownames(bact.div,var='Sample') %>%
  left_join(add_rownames(sample.info,var='Sample'),by='Sample') %>% 
  dplyr::filter(Response=='Response') %>%
  dplyr::select(Subject,bact.div,Time,Treatment.Specific) %>%
  na.omit() %>%
  mutate(Time=paste('T',Time,sep='')) %>%
  #spread(Time,bact.div) %>%
  #head
  ggplot(aes(Time,bact.div,group=Subject,color=Treatment.Specific),data=.) + geom_line(alpha = 0.5)
```



```{r plot_diversity,cache=FALSE}
### normal vs. crohn-T1
tem <- wilcox_test(data.matrix=bact.div,label.matrix=subset(sample.info,Type=='COMBO' | (Type == 'PLEASE-T1'),select=Type),plot.heatmap=FALSE,boxplot.top=2,boxplot.y='Diversity')
 

### normal vs. cluster1
tem <- wilcox_test(data.matrix=bact.div,label.matrix=subset(sample.info,Type=='COMBO' | (Type == 'PLEASE-T1' & Cluster =='cluster 1' ),select=Type),plot.heatmap=FALSE,boxplot.top=2,boxplot.y='Diversity')

### normal vs. cluster2
tem <- wilcox_test(data.matrix=bact.div,label.matrix=subset(sample.info,Type=='COMBO' | (Type == 'PLEASE-T1' & Cluster =='cluster 2' ),select=Type),plot.heatmap=FALSE,boxplot.top=2,boxplot.y='Diversity')

### cluster 1 vs. 2 (both COMBO and PLEASE-T1)
tem <- wilcox_test(data.matrix=bact.div,label.matrix=subset(sample.info,Type=='COMBO' | (Type == 'PLEASE-T1' & Antibiotics.visit=='Not.Use'),select=Cluster),plot.heatmap=FALSE,boxplot.top=2,boxplot.y='Diversity')

### cluster 1 vs. 2 (only PLEASE-T1)
tem <- wilcox_test(data.matrix=bact.div,label.matrix=subset(sample.info,Type == 'PLEASE-T1',select=Cluster),plot.heatmap=FALSE,boxplot.top=2,boxplot.y='Diversity')


### treatment and response
p <- add_rownames(bact.div,var='Sample') %>%
  left_join(add_rownames(sample.info,var='Sample'),by='Sample') %>% 
  dplyr::select(bact.div,Treatment,Response, Time) %>%
  na.omit() %>%  
  ggplot(aes(factor(Time),bact.div),na.rm=T) +
  geom_boxplot(na.rm=T)+facet_grid(Treatment ~ Response, margins=TRUE)
plot(p)

### specific treatment and response
p <- add_rownames(bact.div,var='Sample') %>%
  left_join(add_rownames(sample.info,var='Sample'),by='Sample') %>% 
  dplyr::select(bact.div,Treatment.Specific,Response, Time) %>%
  na.omit() %>%  
  ggplot(aes(factor(Time),bact.div),na.rm=T) +
  geom_boxplot(na.rm=T)+facet_grid(Treatment.Specific ~ Response, margins=TRUE)
plot(p)
```

### Compare diversity change between time points 
```{r test_diversity_by_time, cache=TRUE,eval=FALSE}
####  T1 vs. T2,3,4 
sample.info %>%
  dplyr::select(Subject,Time,Treatment,Bact.Div) %>% 
  na.omit() %>% 
  mutate(Time=paste('T',Time,sep='')) %>%
  spread(Time,Bact.Div) %>%
  summarise_each(funs(t.test(., T1, pair = TRUE)$p.value), vars = T2:T4) %>%
  rename(T1.vs.T2=vars1,T1.vs.T3=vars2,T1.vs.T4=vars3)


####  T1 vs. T2,3,4 for each treatment arm
sample.info %>%
  dplyr::select(Subject,Time,Treatment,Bact.Div) %>% 
  na.omit() %>% 
  mutate(Time=paste('T',Time,sep='')) %>%
  spread(Time,Bact.Div) %>%
  group_by(Treatment) %>%
  summarise_each(funs(t.test(., T1, pair = TRUE)$p.value), vars = T2:T4) %>%
  rename(T1.vs.T2=vars1,T1.vs.T3=vars2,T1.vs.T4=vars3)

####  T1 vs. T2,3,4 for each treatment arm
sample.info %>%
  dplyr::select(Subject,Time,Treatment.Specific,Bact.Div) %>% 
  na.omit() %>% 
  mutate(Time=paste('T',Time,sep='')) %>%
  spread(Time,Bact.Div) %>% 
  group_by(Treatment.Specific) %>%
  summarise_each(funs(t.test(., T1, pair = TRUE)$p.value), vars = T2:T4) %>%
  rename(T1.vs.T2=vars1,T1.vs.T3=vars2,T1.vs.T4=vars3)


####  T1 vs. T2,3,4 for each treatment arm & response
sample.info %>%
  dplyr::select(Subject,Time,Treatment,Response,Bact.Div) %>% 
  na.omit() %>% 
  mutate(Time=paste('T',Time,sep='')) %>%
  spread(Time,Bact.Div) %>%
  group_by(Treatment,Response) %>%
  summarise_each(funs(t.test(., T1, pair = TRUE)$p.value), vars = T2:T4) %>%
  rename(T1.vs.T2=vars1,T1.vs.T3=vars2,T1.vs.T4=vars3)

####  T1 vs. T2,3,4 for each treatment arm & response
####  Not enough sample for PEN
sample.info %>%
  dplyr::select(Subject,Time,Treatment.Specific,Response,Bact.Div) %>% 
  na.omit() %>% 
  mutate(Time=paste('T',Time,sep='')) %>%
  spread(Time,Bact.Div) %>% 
  filter(Treatment.Specific=='EEN') %>%
  group_by(Treatment.Specific,Response) %>%
  summarise_each(funs(t.test(., T1, pair = TRUE)$p.value), vars = T2:T4) %>%
  rename(T1.vs.T2=vars1,T1.vs.T3=vars2,T1.vs.T4=vars3)
```

### Compare diversity change on response status
```{r test_diversity_on_response, cache=TRUE,eval=FALSE}
ttest.fun = function(tdata){ 
  R.vs.NR <- NA
  try(R.vs.NR <- signif(t.test(Bact.Div~Response,data=tdata)$p.value, 2),silent=TRUE)
  return(as.data.frame(R.vs.NR))
}


####  R vs. NR: by treatment
sample.info %>%
  dplyr::select(Subject,Time,Treatment,Treatment.Specific,Bact.Div,Response) %>% 
  na.omit() %>%
  mutate(Time=paste('T',Time,sep='')) %>%
  group_by(Treatment) %>%
  do(ttest.fun(.)) %>%
  add_back_rownames(row.var='Treatment')

sample.info %>%
  dplyr::select(Subject,Time,Treatment,Treatment.Specific,Bact.Div,Response) %>% 
  na.omit() %>%
  mutate(Time=paste('T',Time,sep='')) %>%
  group_by(Treatment.Specific) %>%
  do(ttest.fun(.)) %>%
  add_back_rownames(row.var='Treatment.Specific')


####  R vs. NR: by time
sample.info %>%
  dplyr::select(Subject,Time,Treatment,Treatment.Specific,Bact.Div,Response) %>% 
  na.omit() %>%
  mutate(Time=paste('T',Time,sep='')) %>%
  group_by(Time) %>%
  do(ttest.fun(.)) %>%
  add_back_rownames(row.var='Time')


####  R vs. NR: by time & treatment
sample.info %>%
  dplyr::select(Subject,Time,Treatment,Treatment.Specific,Bact.Div,Response) %>% 
  na.omit() %>%
  mutate(Time=paste('T',Time,sep='')) %>%
  group_by(Time,Treatment) %>%
  do(ttest.fun(.)) 

sample.info %>%
  dplyr::select(Subject,Time,Treatment,Treatment.Specific,Bact.Div,Response) %>% 
  na.omit() %>%
  mutate(Time=paste('T',Time,sep='')) %>%
  group_by(Time,Treatment.Specific) %>%
  do(ttest.fun(.))
```




### Linear regression 
```{r linear_regression_on_diversity,cache=TRUE,eval=FALSE}
### all, adjust for baseline 
lm.all <- sample.info %>%
  #filter(Treatment=='antiTNF')%>% 
  dplyr::select(Subject, Time, Bact.Div,Response,Antibiotics.visit) %>%
  na.omit() %>%
  mutate(Time=paste('Div_T',Time,sep='')) %>%
  spread(Time,Bact.Div) %>% 
  lm(Div_T4~Div_T1+Response+as.factor(Antibiotics.visit),data=.)
summary(lm.all)
coefplot(lm.all)

### antiTNF, adjust for baseline distance
lm.antiTNF <- sample.info %>%
  filter(Treatment=='antiTNF')%>% 
  dplyr::select(Subject, Time, Bact.Div,Response,Antibiotics.visit) %>%
  na.omit() %>%
  mutate(Time=paste('Div_T',Time,sep='')) %>%
  spread(Time,Bact.Div) %>% 
  lm(Div_T4~Div_T1+Response+as.factor(Antibiotics.visit),data=.)
summary(lm.antiTNF)
coefplot(lm.antiTNF)

### Diet 
lm.Diet <- sample.info %>%
  filter(Treatment=='Diet')%>% 
  dplyr::select(Subject, Time, Bact.Div,Response,Antibiotics.visit) %>%
  na.omit() %>%
  mutate(Time=paste('Div_T',Time,sep='')) %>%
  spread(Time,Bact.Div) %>% 
  lm(Div_T4~Div_T1+Response+as.factor(Antibiotics.visit),data=.)
summary(lm.Diet)
coefplot(lm.Diet)

### EEN
lm.EEN <- sample.info %>%
  filter(Treatment.Specific=='EEN')%>% 
  dplyr::select(Subject, Time, Bact.Div,Response,Antibiotics.visit) %>%
  na.omit() %>%
  mutate(Time=paste('Div_T',Time,sep='')) %>%
  spread(Time,Bact.Div) %>% 
  lm(Div_T4~Div_T1+Response+as.factor(Antibiotics.visit),data=.)
summary(lm.EEN)
coefplot(lm.EEN)

### PEN 
lm.PEN <- sample.info %>%
  filter(Treatment.Specific=='PEN')%>% 
  dplyr::select(Subject, Time, Bact.Div,Response,Antibiotics.visit) %>%
  na.omit() %>%
  mutate(Time=paste('Div_T',Time,sep='')) %>%
  spread(Time,Bact.Div) %>% 
  lm(Div_T4~Div_T1+Response+as.factor(Antibiotics.visit),data=.)
summary(lm.PEN)
coefplot(lm.PEN)

```


### Batch effect of the two clusters
```{r batch_effect,cache=TRUE}
Plate.file <- '../1_Data/Raw_Data/Plate_Information/Plate_Source_for_Samples.txt'
Plate <- read.table(Plate.file,sep='\t',header=TRUE,check.names=FALSE,stringsAsFactors=FALSE)

add_rownames(sample.info,var='Sample') %>%
left_join(Plate,by='Sample') %>%
group_by(Cluster,Plate) %>%
summarize(N=n()) %>%
na.omit() %>%
spread(Plate, N) %>%
add_back_rownames(row.var='Cluster') %>%
chisq.test()
## not significant
## p-value = 0.09816
```

# Analyze bacterial data (species level)

## Load data
```{r species_load_bacterial_data, cache=TRUE}
PLEASE.file <- paste('../1_Data/Raw_Data/MetaPhlAn/PLEASE/','S', '_Remove_unclassfied_Renormalized_Merge_Rel_MetaPhlAn_Result.xls',sep='')
PLEASE.raw <- read.table(PLEASE.file,sep='\t',header=TRUE,row.names = 1,check.names=FALSE,stringsAsFactors=FALSE)
PLEASE.raw <- t(PLEASE.raw)
cat('samples','taxa',dim(PLEASE.raw),'\n')
PLEASE.raw[1:3,1:3]


COMBO.file <- paste('../1_Data/Raw_Data/MetaPhlAn/COMBO/','S',
'_Remove_unclassfied_Renormalized_Merge_Rel_MetaPhlAn_Result.xls',sep='')
COMBO.raw <- read.table(COMBO.file,sep='\t',header=TRUE,row.names = 1,check.names=FALSE,stringsAsFactors=FALSE)
COMBO.raw <- t(COMBO.raw)
cat('samples','taxa',dim(COMBO.raw),'\n')
COMBO.raw[1:3,1:3]

rm(PLEASE.file)
rm(COMBO.file)
```

### Merge PLEASE data and COMBO data. 
```{r species_merge_please_combo_data, cache=TRUE}
species.raw <- merge_two_tables(COMBO.raw,PLEASE.raw,
                             fill.zero = TRUE,rnames = 'union',cnames = 'union')
cat('samples','species',dim(species.raw),'\n')
species.raw[1:3,1:3]
```

### Filter low depth samples
Filter out low sequencing depth samples.Filter out samples with non-humand reads < 30000.
```{r species_filter_low_depth_samples,cache=TRUE}
low.depth.samples <- subset(human.read,NonHumanReads<30000)
low.depth.samples[,1:5]
### Delete these samples from PLEASE data.
### Due to the low sequencing depth, some of the samples have no MetaPhlAn output
### These are the samples with MetaPhlAn output but also low sequencing
rownames(species.raw)[which(rownames(species.raw) %in% rownames(low.depth.samples))]
### Before deletion
dim(species.raw)
### After deletion
species.raw <- species.raw[-which(rownames(species.raw) %in% rownames(low.depth.samples)),]
dim(species.raw)

rm(low.depth.samples)
```

### Filter low abundant bacterial data
Filter low abundant bacteria. I'm going to filter out the low abundant species.The species must present in at least 10% of the samples and the max abundance across samples must large than 1%. Renormalize the species abundance so that the sum of the species abundance is 100 (%)
```{r species_filter_low_abundant_bacteria, cache=TRUE, fig.width=4,fig.height=4}
filter.index1 <- apply(species.raw,2,function(X){sum(X>0)>0.1*length(X)})
filter.index2 <- apply(species.raw,2,function(X){max(X)>1})
species.filter <- species.raw[,filter.index1 & filter.index2]
species.filter <- 100*sweep(species.filter, 1, rowSums(species.filter), FUN="/")
cat('after filter:','samples','taxa',dim(species.filter),'\n')
head(rowSums(species.filter))

## Did the filtering process change the abundance ?
plot(species.raw[,colnames(species.filter)],species.filter,
xlab='before filtering',ylab='after filtering')

rm(filter.index1)
rm(filter.index2)
```


```{r species_save_processed_taxa_abundance_data,cache=TRUE}
### save the processed taxa abundance
write.csv(species.filter,file=paste('../1_Data/Processed_Data/MetaPhlAn/','S','_Processed_MetaPhlAn_Abundance.csv',sep=''))

### Remove s__ before the taxa name
colnames(species.filter) <- colsplit(colnames(species.filter), '__', c('V1','V2'))[,'V2']

####### change the variable name
species.data <- as.data.frame(species.filter)

rm(species.filter)
```


## Normal vs. crohn-T1
```{r species_bacteria_normal_vs_crohn_t1,cache=TRUE}
### I don't use dplyr filter 
### because it will remove the row names
species.cp1 <- subset_data_by_sample_info(species.data,sample.info,
                                       Type=='COMBO' | Type=='PLEASE-T1')
```

### Calculate distance
```{r species_bacteria_calculate_distance,cache=TRUE}
species.dist <- calculate_distance(X = species.data, method = bacteral.distance.method, tree = NA)
```

```{r species_bacteria_calculate_distance_cp1,cache=TRUE}
species.dist.cp1 <- calculate_distance(X = species.cp1, method = bacteral.distance.method, tree = NA)
```

### Find the best clustering
```{r species_bacteria_find_the_best_clustering,cache=TRUE}
#pdf('best_cluster.pdf')
pamk.best <- find_best_clustering(species.dist.cp1,plot.results=TRUE)
#dev.off()
```

```{r species_update_cluster_info_and_save_sample_informaion,cache=TRUE}
### rename does not work
clustering.res <- as.data.frame(pamk.best$pamobject$clustering)
colnames(clustering.res) <- 'Species.Cluster'
### add clustering result to sample information
### only T1 samples are in the clustering result
sample.info <- 
  add_rownames(clustering.res,var='Sample') %>%
  separate(col=Sample,into=c("Subject", "Time"),sep='-',extra='drop' ) %>%
  dplyr::select(Subject,Species.Cluster) %>%
  right_join(add_rownames(sample.info,var='Sample'),by = c('Subject')) %>% #head  
  mutate(Species.Cluster=ifelse(Species.Cluster==1,'cluster 1','cluster 2')) %>% #head
  add_back_rownames(row.var='Sample') 


##############################
### save the sample information
write.csv(sample.info,file=paste('../1_Data/Processed_Data/Sample_Information/2015_02_13_Processed_Sample_Information.csv',sep=''),row.names=TRUE)

rm(clustering.res)
rm(pamk.best)
```

### MDS analysis
```{r species_bacteria_MDS,cache=TRUE,results='hide'}
### MDS
set.seed(10)
meta.MDS <- metaMDS(species.dist.cp1)

### Normal/Disease
p <- add_rownames(as.data.frame(meta.MDS$points),var='Sample') %>%
  left_join(dplyr::select(add_rownames(sample.info,var='Sample'),
                          Sample,Disease),
            by='Sample') %>% 
  ggplot(aes(x=MDS1,y=MDS2))+ geom_point(aes(colour = Disease)) +
  scale_color_manual(values=c("#0066CC", "#FF3333"),name="") +
  labs(title = 'metaMDS') 
print(p)

### Normal/Disease, cluster
p <- add_rownames(as.data.frame(meta.MDS$points),var='Sample') %>%
  left_join(dplyr::select(add_rownames(sample.info,var='Sample'),
                          Sample,Disease,Species.Cluster),
            by='Sample') %>% 
  ggplot(aes(x=MDS1,y=MDS2))+ geom_point(aes(colour = Disease,shape=Species.Cluster)) +
  scale_color_manual(values=c("#0066CC", "#FF3333"),name="") +
  scale_shape_manual(values=c(20,4),name="") +
  labs(title = 'metaMDS') 
print(p)

#### Fugal percentage 
p <- add_rownames(as.data.frame(meta.MDS$points),var='Sample') %>%
  left_join(dplyr::select(add_rownames(sample.info,var='Sample'),
                          Sample,Disease,Species.Cluster,Fungi.Per),
            by='Sample') %>% 
  mutate(Fungi.Per.sqrt=sqrt(Fungi.Per)) %>%
  ggplot(aes(x=MDS1,y=MDS2))+ 
  geom_point(aes(colour = Disease,shape=Species.Cluster,size = Fungi.Per.sqrt)) +
  scale_color_manual(values=c("#0066CC", "#FF3333"),name="") +
  scale_shape_manual(values=c(20,1),name="") +
  scale_size_area()+
  labs(title = 'metaMDS') 
print(p)


#### Human DNA percentage
p <- add_rownames(as.data.frame(meta.MDS$points),var='Sample') %>%
  left_join(dplyr::select(add_rownames(sample.info,var='Sample'),
                          Sample,Disease,Species.Cluster,Human.Per),
            by='Sample') %>% 
  ggplot(aes(x=MDS1,y=MDS2))+ 
  geom_point(aes(colour = Disease,shape=Species.Cluster,size = Human.Per)) +
  scale_color_manual(values=c("#0066CC", "#FF3333"),name="") +
  scale_shape_manual(values=c(20,1),name="") +
  scale_size_area()+
  labs(title = 'metaMDS') 
print(p)

#### Human DNA percentage (for publication)
p <- add_rownames(as.data.frame(meta.MDS$points),var='Sample') %>%
  left_join(dplyr::select(add_rownames(sample.info,var='Sample'),
                          Sample,Disease,Species.Cluster,Human.Per),
            by='Sample') %>% 
  ggplot(aes(x=MDS1,y=MDS2))+ 
  geom_point(aes(colour = Species.Cluster,shape= Disease,size = Human.Per)) +
  scale_color_manual(values=c("#0066CC", "#FF3333"),name="") +
  scale_shape_manual(values=c(20,1),name="") +
  #scale_size_area(range = c(1, 10))+
  scale_size_continuous(range = c(3, 10)) +
  labs(title = 'Ordination bacterial taxa') +
  ##########################
  theme_classic()
print(p)

### Figure for Rick
### we start with the same MDS as in 1B, but don't show human DNA or near and far cluster.  Control is shown as the filled circle, Crohn's as open circles.  Current antibiotic use is shown by blue versus red coloring.  If not too hard, maybe also plot the centroid of the healties as a large yellow triangle.
p <- add_rownames(as.data.frame(meta.MDS$points),var='Sample') %>%
  left_join(dplyr::select(add_rownames(sample.info,var='Sample'),
                          Sample,Disease,Antibiotics.visit),
            by='Sample') %>% 
  ggplot(aes(x=MDS1,y=MDS2))+ 
  geom_point(aes(colour = Antibiotics.visit,shape= Disease),size=3) +
  scale_color_manual(values=c("#0066CC", "#FF3333"),name="") +
  scale_shape_manual(values=c(20,1),name="") +
  #scale_size_area(range = c(1, 10))+
  #scale_size_continuous(range = c(3, 10)) +
  labs(title = 'Ordination bacterial taxa') +
  ##########################
  theme_classic()
print(p)


#### Fugal percentage and Human DNA percentage
p <- add_rownames(as.data.frame(meta.MDS$points),var='Sample') %>%
  left_join(dplyr::select(add_rownames(sample.info,var='Sample'),
                          Sample,Disease,Species.Cluster,Human.Per,Fungi.Per),
            by='Sample') %>% 
  mutate(Fungi.Per.sqrt=sqrt(Fungi.Per)) %>%
  ggplot(aes(x=MDS1,y=MDS2,colour=Human.Per,shape=Species.Cluster))+ 
  geom_point(aes(size = Fungi.Per.sqrt)) + scale_size_area()+
  scale_colour_gradient(low = "#43a2ca",high="#de2d26")+ 
  scale_shape_manual(values=c(20,1),name="") + labs(title = 'metaMDS')
print(p)

### switch human and fungi 
p <- add_rownames(as.data.frame(meta.MDS$points),var='Sample') %>%
  left_join(dplyr::select(add_rownames(sample.info,var='Sample'),
                          Sample,Disease,Species.Cluster,Human.Per,Fungi.Per),
            by='Sample') %>% 
  mutate(Fungi.Per.sqrt=sqrt(Fungi.Per)) %>%
  ggplot(aes(x=MDS1,y=MDS2,colour=Fungi.Per.sqrt,shape=Species.Cluster))+ 
  geom_point(aes(size = Human.Per)) + scale_size_area()+
  scale_colour_gradient(low = "#43a2ca",high="#de2d26")+ 
  scale_shape_manual(values=c(20,1),name="") + labs(title = 'metaMDS')
print(p)

rm(meta.MDS)
```



### Wilcoxon rank test of normal vs. Crohn
```{r species_bacteria_rank_test_of_normal_vs_crohn,cache=TRUE}
### only plot the top 5 significant ones
species.cp1.wil <- wilcox_test(species.cp1,sample.info[,'Disease',drop=FALSE],
                            plot.heatmap=FALSE,boxplot.top=0.05,boxplot.y='Abundance(%)')
###
## write.csv(species.cp1.wil,file=paste('../3_Result/2015_03_05_Species_Bacteria_COMBO_vs_PLEASET1_Wilcox.csv',sep=''),row.names=TRUE)
```

### Wilcoxon rank test of normal vs. Crohn without antibiotics use
```{r species_bacteria_rank_test_of_normal_vs_crohn_without_antibiotics_use,cache=TRUE}
### only plot the top 5 significant ones
species.cp1.noabx.wil <- wilcox_test(species.cp1,
              subset(sample.info,Antibiotics.visit=='Not.Use',select=Disease),
              plot.heatmap=FALSE,boxplot.top=0.05,boxplot.y='Abundance(%)')
###
### write.csv(species.cp1.noabx.wil,file=paste('../3_Result/2015_03_05_Species_Bacteria_COMBO_vs_PLEASET1_No_Abx_Wilcox.csv',sep=''),row.names=TRUE)
```


## Cluster 1 vs. cluster 2

### Wilcoxon rank test 
```{r species_bacteria_rank_test_of_cluster_1_vs_2,cache=TRUE}
### plot all the significant ones
species.c1c2.wil <- wilcox_test(species.cp1,
                             subset(sample.info,Group=='PLEASE',select=Species.Cluster),
                             plot.heatmap=FALSE,boxplot.top=0.05,boxplot.y='Abundance(%)')      
###
## write.csv(species.c1c2.wil,file=paste('../3_Result/2015_03_05_Species_Bacteria_PLEASE_Cluster_1_2_Wilcox.csv',sep=''),row.names=TRUE)
```


### Heatmap: normal vs. disease and cluster 1 vs. 2 (new label)
```{r species_bacteria_heatmap_manually_order_samples_normal_disease_cluster1_2,cache=TRUE,fig.width=8,fig.height=8}
#### manually re-order the samples
data.heat <- 
  add_rownames(as.data.frame(species.cp1.wil),var='Taxa') %>%
  left_join(add_rownames(as.data.frame( species.c1c2.wil),var='Taxa'),by='Taxa') %>%
  mutate(Disease=ifelse(Disease.qvalue<0.05,'*',' '),
         Species.Cluster=ifelse(Species.Cluster.qvalue<0.05,'+',' ')
  ) %>%
  mutate(Tnames=paste(Taxa,Disease,Species.Cluster,sep=''))%>%
  dplyr::select(Taxa,Tnames) %>%
  right_join( add_rownames(as.data.frame(t(species.cp1)),var='Taxa'),by='Taxa') %>%
  dplyr::select(-Taxa) %>%
  add_back_rownames(row.var='Tnames') %>%
  t() %>%
  as.data.frame %>%
  add_rownames(var='Sample') %>%
  ##
  left_join(add_rownames(sample.info,var='Sample'),by='Sample') %>% 
  arrange(Disease,Species.Cluster,Response,Treatment.Specific,Antibiotics.visit,FCP,Steroids) %>%
  dplyr::select(1:(ncol(species.cp1)+1)) %>%  ## first column is sample name
  add_back_rownames(row.var='Sample') 


### * disease.q < 0.05
### + cluster.q < 0.05
bk <- unique(c(
        0,
        min(min(species.data[species.data>0])/2,0.1),
        seq(min(min(species.data[species.data>0])/2,1),0.1,length=10),
        seq(0.1,1,length=10),
        seq(1,5,length=10),
        seq(5,20,length=20),
        seq(20,50,length=20),
        seq(50,100,length=20)
        ))
hmcols<- colorRampPalette(c('white','#1f78b4','#ff7f00','#e41a1c'))(length(bk)-1)

pheatmap(t(data.heat),border_color="grey60",
         cluster_cols = FALSE,
         clustering_distance_rows = 'correlation',
         #clustering_distance_cols = bact.dist.cp1,
         clustering_method = "complete",
         annotation=sample.info[,c("Disease","Species.Cluster","Response","Treatment.Specific","FCP","Antibiotics.visit","Steroids")],
         color =hmcols, breaks=bk,
         show_rownames = TRUE,
         scale="none")

rm(data.heat)
```



## Distance-based analysis
### Calculate the distance to COMBO centroid
```{r species_bacteria_calculate_distance_to_combo_centroid,cache=TRUE}
## use mean
species.COMBO.centroid <- apply(subset_data_by_sample_info(species.data,sample.info,Type=='COMBO'),2,mean)
head(species.COMBO.centroid)
species.PLEASE.COMBO.centroid <- rbind(species.data, species.COMBO.centroid=species.COMBO.centroid)
species.dist.all <- calculate_distance(X = species.PLEASE.COMBO.centroid, method = bacteral.distance.method)

### Get the distance to the COMBO centroid for all samples
species.dist.all.mat <- as.matrix(species.dist.all)
species.dist.to.cen <- species.dist.all.mat[-which(colnames(species.dist.all.mat)=='species.COMBO.centroid'),"species.COMBO.centroid",drop=FALSE]
colnames(species.dist.to.cen) <- 'Species.Distance'
### Check the data maniputation
####Before data manipulation
species.dist.all.mat["5001-01","species.COMBO.centroid"]
####After data manipulation
species.dist.to.cen["5001-01",]


dim(species.dist.to.cen)

### What is the distribution of the distance    
### The data looks normal, we can use t.test in the following analysis
hist(species.dist.to.cen,xlab='Distance to the COMBO centroid',breaks=20,main='')

### Add distance information to the sample information
sample.info <- add_rownames(sample.info,var='Sample') %>%
  left_join(add_rownames(as.data.frame(species.dist.to.cen),var='Sample'),by='Sample') %>%
  add_back_rownames(row.var='Sample')


##############################
### save the sample information
### again
write.csv(sample.info,file=paste('../1_Data/Processed_Data/Sample_Information/2015_02_13_Processed_Sample_Information.csv',sep=''),row.names=TRUE)
```
 


### Plot the distance to the COMBO centroid
```{r species_plot_distance_to_combo_centroid,cache=TRUE}
### Plot distance by treatment
p<- sample.info %>% dplyr::select(Type,Species.Distance,Treatment) %>%
  ggplot(aes(factor(Type),Species.Distance)) + 
  geom_boxplot(aes(fill = factor(Treatment)))+
  xlab('')+ylab('Distance to the COMBO centroid')
plot(p)

p<- sample.info %>% dplyr::select(Type,Species.Distance,Treatment.Specific) %>%
  ggplot(aes(factor(Type),Species.Distance)) + 
  geom_boxplot(aes(fill = factor(Treatment.Specific)))+
  xlab('')+ylab('Distance to the COMBO centroid')
plot(p)

#### For publication
p<- sample.info %>% dplyr::select(Type,Species.Distance,Treatment.Specific) %>%
  ggplot(aes(factor(Type),Species.Distance)) + 
  geom_boxplot(aes(fill = factor(Treatment.Specific)),outlier.shape = NA)+
  xlab('')+ylab('Distance to the COMBO centroid') + 
  theme_classic() 
plot(p)


#### Plot distance by response
p<- sample.info %>% dplyr::select(Type,Species.Distance,Response) %>%
  filter(Type=='COMBO' | (Type!='COMBO' & !is.na(Response))) %>%
  ggplot(aes(factor(Type),Species.Distance)) + 
  geom_boxplot(aes(fill = factor(Response)))+
  xlab('')+ylab('Distance to the COMBO centroid')
plot(p)


#### Plot distance by treatment,response
p<- sample.info %>% dplyr::select(Type,Species.Distance,Treatment.Specific,Response) %>%
  mutate(Response=ifelse(Type=='COMBO','COMBO',Response)) %>%
  mutate(Treatment.Specific=ifelse(Type=='COMBO','COMBO',Treatment.Specific)) %>%
  na.omit() %>%
  unite(Treatment.Response, Treatment.Specific,Response, sep = ".") %>%
  ggplot(aes(factor(Type),Species.Distance)) + 
  geom_boxplot(aes(fill = factor(Treatment.Response)),outlier.shape = NA)+
  xlab('')+ylab('Distance to the COMBO centroid') + 
  theme_classic() 
plot(p)


### Plot distance by both treatment and response
p<- sample.info %>% dplyr::select(Species.Distance,Response,Time,Treatment) %>%
  na.omit %>%
  ggplot( aes(factor(Time), Species.Distance),na.rm=T)+
  geom_boxplot(na.rm=T)+facet_grid(Treatment ~ Response, margins=TRUE)
plot(p)

### Plot distance by both treatment(PEN vs. EEN) and response
p<- sample.info %>% dplyr::select(Species.Distance,Response,Time,Treatment.Specific) %>%
  na.omit %>%
  ggplot(aes(factor(Time), Species.Distance),na.rm=T)+
  geom_boxplot(na.rm=T)+facet_grid(Treatment.Specific ~ Response, margins=TRUE)
plot(p)
```



### Compare distance change between time points 
```{r species_test_distance_to_combo_centroid, cache=TRUE}
####  T1 vs. T2,3,4 
sample.info %>%
  dplyr::select(Subject,Time,Treatment,Species.Distance) %>% 
  na.omit() %>% 
  mutate(Time=paste('T',Time,sep='')) %>%
  spread(Time,Species.Distance) %>%
  summarise_each(funs(t.test(., T1, pair = TRUE)$p.value), vars = T2:T4) %>%
  rename(T1.vs.T2=vars1,T1.vs.T3=vars2,T1.vs.T4=vars3)


####  T1 vs. T2,3,4 for each treatment arm
sample.info %>%
  dplyr::select(Subject,Time,Treatment,Species.Distance) %>% 
  na.omit() %>% 
  mutate(Time=paste('T',Time,sep='')) %>%
  spread(Time,Species.Distance) %>%
  group_by(Treatment) %>%
  summarise_each(funs(t.test(., T1, pair = TRUE)$p.value), vars = T2:T4) %>%
  rename(T1.vs.T2=vars1,T1.vs.T3=vars2,T1.vs.T4=vars3)

####  T1 vs. T2,3,4 for each treatment arm
sample.info %>%
  dplyr::select(Subject,Time,Treatment.Specific,Species.Distance) %>% 
  na.omit() %>% 
  mutate(Time=paste('T',Time,sep='')) %>%
  spread(Time,Species.Distance) %>% 
  group_by(Treatment.Specific) %>%
  summarise_each(funs(t.test(., T1, pair = TRUE)$p.value), vars = T2:T4) %>%
  rename(T1.vs.T2=vars1,T1.vs.T3=vars2,T1.vs.T4=vars3)


####  T1 vs. T2,3,4 for each treatment arm & response
sample.info %>%
  dplyr::select(Subject,Time,Treatment,Response,Species.Distance) %>% 
  na.omit() %>% 
  mutate(Time=paste('T',Time,sep='')) %>%
  spread(Time,Species.Distance) %>%
  group_by(Treatment,Response) %>%
  summarise_each(funs(t.test(., T1, pair = TRUE)$p.value), vars = T2:T4) %>%
  rename(T1.vs.T2=vars1,T1.vs.T3=vars2,T1.vs.T4=vars3)

####  T1 vs. T2,3,4 for each treatment arm & response
####  Not enough sample for PEN
####  t-test
sample.info %>%
  dplyr::select(Subject,Time,Treatment.Specific,Response,Species.Distance) %>% 
  na.omit() %>% 
  mutate(Time=paste('T',Time,sep='')) %>%
  spread(Time,Species.Distance) %>% 
  filter(Treatment.Specific=='EEN') %>%
  group_by(Treatment.Specific,Response) %>%
  summarise_each(funs(t.test(., T1, pair = TRUE)$p.value), vars = T2:T4) %>%
  rename(T1.vs.T2=vars1,T1.vs.T3=vars2,T1.vs.T4=vars3)


####  T1 vs. T2,3,4 for each treatment arm & response
####  Not enough sample for PEN
####  wilcox
sample.info %>%
  dplyr::select(Subject,Time,Treatment.Specific,Response,Species.Distance) %>% 
  na.omit() %>% 
  mutate(Time=paste('T',Time,sep='')) %>%
  spread(Time,Species.Distance) %>% 
  filter(Treatment.Specific=='EEN') %>%
  group_by(Treatment.Specific,Response) %>%
  summarise_each(funs(wilcox.test(., T1, pair = TRUE)$p.value), vars = T2:T4) %>%
  rename(T1.vs.T2=vars1,T1.vs.T3=vars2,T1.vs.T4=vars3)


sample.info %>%
  dplyr::select(Subject,Time,Treatment.Specific,Response,Species.Distance) %>% 
  na.omit() %>% 
  mutate(Time=paste('T',Time,sep='')) %>%
  spread(Time,Species.Distance) %>% 
  filter(Treatment.Specific=='EEN',Response=='Non.Response')
```

### Compare distance change on response status
```{r species_test_distance_on_response, cache=TRUE}
ttest.fun = function(tdata){ 
  R.vs.NR <- NA
  try(R.vs.NR <- signif(t.test(Species.Distance~Response,data=tdata)$p.value, 2),silent=TRUE)
  return(as.data.frame(R.vs.NR))
}


####  R vs. NR: by treatment
sample.info %>%
  dplyr::select(Subject,Time,Treatment,Treatment.Specific,Species.Distance,Response) %>% 
  na.omit() %>%
  mutate(Time=paste('T',Time,sep='')) %>%
  group_by(Treatment) %>%
  do(ttest.fun(.)) %>%
  add_back_rownames(row.var='Treatment')

sample.info %>%
  dplyr::select(Subject,Time,Treatment,Treatment.Specific,Species.Distance,Response) %>% 
  na.omit() %>%
  mutate(Time=paste('T',Time,sep='')) %>%
  group_by(Treatment.Specific) %>%
  do(ttest.fun(.)) %>%
  add_back_rownames(row.var='Treatment.Specific')


####  R vs. NR: by time
sample.info %>%
  dplyr::select(Subject,Time,Treatment,Treatment.Specific,Species.Distance,Response) %>% 
  na.omit() %>%
  mutate(Time=paste('T',Time,sep='')) %>%
  group_by(Time) %>%
  do(ttest.fun(.)) %>%
  add_back_rownames(row.var='Time')


####  R vs. NR: by time & treatment
sample.info %>%
  dplyr::select(Subject,Time,Treatment,Treatment.Specific,Species.Distance,Response) %>% 
  na.omit() %>%
  mutate(Time=paste('T',Time,sep='')) %>%
  group_by(Time,Treatment) %>%
  do(ttest.fun(.)) 

sample.info %>%
  dplyr::select(Subject,Time,Treatment,Treatment.Specific,Species.Distance,Response) %>% 
  na.omit() %>%
  mutate(Time=paste('T',Time,sep='')) %>%
  group_by(Time,Treatment.Specific) %>%
  do(ttest.fun(.))
```

### Compare distance change on treatment
```{r species_test_distance_on_treatment, cache=TRUE}
####  EEN-T1 response vs. control
sample.info %>%
dplyr::select(Subject,Type,Treatment.Specific,Species.Distance,Response,Antibiotics.visit) %>% 
dplyr::filter(Type=='COMBO'| (Type=='PLEASE-T1' & Treatment.Specific=='EEN' & Response == 'Response')) %>%
dplyr::select(Type,Species.Distance,Antibiotics.visit) %>%
na.omit() %>%
(function(tdata){
t.test(Species.Distance~Type,data=tdata)$p.value %>% print
lm(Species.Distance~Type+Antibiotics.visit,data=tdata) %>% summary %>% coef %>% print
boxplot(Species.Distance~Type,data=tdata,ylab='Distance')
})

####  EEN-T4 response vs. control
sample.info %>%
dplyr::select(Subject,Type,Treatment.Specific,Species.Distance,Response,Antibiotics.visit) %>% 
dplyr::filter(Type=='COMBO'| (Type=='PLEASE-T4' & Treatment.Specific=='EEN' & Response == 'Response')) %>%
dplyr::select(Type,Species.Distance,Antibiotics.visit) %>%
na.omit() %>%
(function(tdata){
t.test(Species.Distance~Type,data=tdata)$p.value %>% print
lm(Species.Distance~Type+Antibiotics.visit,data=tdata) %>% summary %>% coef %>% print
boxplot(Species.Distance~Type,data=tdata,ylab='Distance')
})

####  antiTNF-T1 response vs. control
sample.info %>%
dplyr::select(Subject,Type,Treatment.Specific,Species.Distance,Response,Antibiotics.visit) %>% 
dplyr::filter(Type=='COMBO'| (Type=='PLEASE-T1' & Treatment.Specific=='antiTNF' & Response == 'Response')) %>%
dplyr::select(Type,Species.Distance,Antibiotics.visit) %>%
na.omit() %>%
(function(tdata){
t.test(Species.Distance~Type,data=tdata)$p.value %>% print
lm(Species.Distance~Type+Antibiotics.visit,data=tdata) %>% summary %>% coef %>% print
boxplot(Species.Distance~Type,data=tdata,ylab='Distance')
})

####  antiTNF-T4 response vs. control
sample.info %>%
dplyr::select(Subject,Type,Treatment.Specific,Species.Distance,Response,Antibiotics.visit) %>% 
dplyr::filter(Type=='COMBO'| (Type=='PLEASE-T4' & Treatment.Specific=='antiTNF' & Response == 'Response')) %>%
dplyr::select(Type,Species.Distance,Antibiotics.visit) %>%
na.omit() %>%
(function(tdata){
t.test(Species.Distance~Type,data=tdata)$p.value %>% print
lm(Species.Distance~Type+Antibiotics.visit,data=tdata) %>% summary %>% coef %>% print
boxplot(Species.Distance~Type,data=tdata,ylab='Distance')
})

####  antiTNF-T1 response vs. EEN-T1 response
sample.info %>%
dplyr::select(Subject,Type,Treatment.Specific,Species.Distance,Response) %>% 
dplyr::filter( (Type=='PLEASE-T1' & Treatment.Specific !='PEN' & Response == 'Response')) %>%
dplyr::select(Treatment.Specific,Species.Distance) %>%
na.omit() %>%
(function(tdata){
t.test(Species.Distance~Treatment.Specific,data=tdata)$p.value %>% print
boxplot(Species.Distance~Treatment.Specific,data=tdata,ylab='Distance')
})


####  antiTNF-T4 response vs. EEN-T4 response
sample.info %>%
dplyr::select(Subject,Type,Treatment.Specific,Species.Distance,Response) %>% 
dplyr::filter( (Type=='PLEASE-T4' & Treatment.Specific !='PEN' & Response == 'Response')) %>%
dplyr::select(Treatment.Specific,Species.Distance) %>%
na.omit() %>%
(function(tdata){
t.test(Species.Distance~Treatment.Specific,data=tdata)$p.value %>% print
boxplot(Species.Distance~Treatment.Specific,data=tdata,ylab='Distance')
})
```


### Compare distance to healthy controls
```{r species_compare_distance_to_healthy_controls,cache=TRUE}
#### I am goint to re-plot the figure before
#### Plot distance by treatment,response
p<- sample.info %>% dplyr::select(Type,Species.Distance,Treatment.Specific,Response) %>%
  mutate(Response=ifelse(Type=='COMBO','COMBO',Response)) %>%
  mutate(Treatment.Specific=ifelse(Type=='COMBO','COMBO',Treatment.Specific)) %>%
  na.omit() %>%
  unite(Treatment.Response, Treatment.Specific,Response, sep = ".") %>%
  ggplot(aes(factor(Type),Species.Distance)) + 
  geom_boxplot(aes(fill = factor(Treatment.Response)),outlier.shape = NA)+
  xlab('')+ylab('Distance to the COMBO centroid') + 
  theme_classic() 
plot(p)

#### Calculate the pvalues for each box in the above figure
sample.info %>%
dplyr::select(Type,Time,Treatment.Specific,Species.Distance,Response) %>% 
mutate(Treatment.Specific=ifelse(Type=='COMBO','COMBO',Treatment.Specific)) %>%
mutate(Response=ifelse(Type=='COMBO','COMBO',Response)) %>%
mutate(Time=paste('T',Time,sep='')) %>%
mutate(Time=ifelse(Type=='COMBO','COMBO',Time)) %>%
dplyr::filter(Treatment.Specific != 'PEN') %>% 
na.omit() %>%
### group by treatment
(function(tdata){
### merge the time and treatment
temp <- tdata %>% 
unite(Time.Treatment,Time,Treatment.Specific, sep = "_") %>%
dplyr::select(Time.Treatment,Species.Distance)
#### take the COMBO samples out, use them in the t.test
cdata <- dplyr::filter(temp,Time.Treatment=='COMBO_COMBO') %>%
         dplyr::select(Species.Distance)
### compare each group to combo group by t.test
group_by(temp,Time.Treatment) %>%
summarise(pval.compare.to.combo=t.test(Species.Distance,cdata$Species.Distance)$p.value) %>%
mutate(pval.compare.to.combo=signif(pval.compare.to.combo,2)) %>%
separate(Time.Treatment, into=c("Time","Treatment"), sep = "_") %>%
print
return(tdata)
}) %>%
### group by treatment and response
unite(Time.Treatment.Response,Time, Treatment.Specific,Response, sep = "_") %>%
dplyr::select(Time.Treatment.Response,Species.Distance) %>%
(function(tdata){
cdata <- dplyr::filter(tdata,Time.Treatment.Response=='COMBO_COMBO_COMBO') %>%
         dplyr::select(Species.Distance)
group_by(tdata,Time.Treatment.Response) %>%
summarise(pval.compare.to.combo=t.test(Species.Distance,cdata$Species.Distance)$p.value) %>%
mutate(pval.compare.to.combo=signif(pval.compare.to.combo,2)) %>%
separate(Time.Treatment.Response, into=c("Time","Treatment","Response"), sep = "_")
})
```


### Linear regression 
```{r species_linear_regression_on_distance_to_combo_centroid,cache=TRUE}
### all, adjust for baseline distance
lm.all <- sample.info %>%
  #filter(Treatment=='antiTNF')%>% 
  dplyr::select(Subject, Time, Species.Distance,Response,Antibiotics.visit) %>%
  na.omit() %>%
  mutate(Time=paste('Dist_T',Time,sep='')) %>%
  spread(Time, Species.Distance) %>% 
  lm(Dist_T4~Dist_T1+Response+as.factor(Antibiotics.visit),data=.)
summary(lm.all)
coefplot(lm.all)

### antiTNF, adjust for baseline distance
lm.antiTNF <- sample.info %>%
  filter(Treatment=='antiTNF')%>% 
  dplyr::select(Subject, Time, Species.Distance,Response,Antibiotics.visit) %>%
  na.omit() %>%
  mutate(Time=paste('Dist_T',Time,sep='')) %>%
  spread(Time, Species.Distance) %>% 
  lm(Dist_T4~Dist_T1+Response+as.factor(Antibiotics.visit),data=.)
summary(lm.antiTNF)
coefplot(lm.antiTNF)

### Diet 
lm.Diet <- sample.info %>%
  filter(Treatment=='Diet')%>% 
  dplyr::select(Subject, Time, Species.Distance,Response,Antibiotics.visit) %>%
  na.omit() %>%
  mutate(Time=paste('Dist_T',Time,sep='')) %>%
  spread(Time,Species.Distance) %>% 
  lm(Dist_T4~Dist_T1+Response+as.factor(Antibiotics.visit),data=.)
summary(lm.Diet)
coefplot(lm.Diet)

### EEN
lm.EEN <- sample.info %>%
  filter(Treatment.Specific=='EEN')%>% 
  dplyr::select(Subject, Time, Species.Distance,Response,Antibiotics.visit) %>%
  na.omit() %>%
  mutate(Time=paste('Dist_T',Time,sep='')) %>%
  spread(Time,Species.Distance) %>% 
  lm(Dist_T4~Dist_T1+Response+as.factor(Antibiotics.visit),data=.)
summary(lm.EEN)
coefplot(lm.EEN)

### PEN 
lm.PEN <- sample.info %>%
  filter(Treatment.Specific=='PEN')%>% 
  dplyr::select(Subject, Time, Species.Distance,Response,Antibiotics.visit) %>%
  na.omit() %>%
  mutate(Time=paste('Dist_T',Time,sep='')) %>%
  spread(Time,Species.Distance) %>% 
  lm(Dist_T4~Dist_T1+Response+as.factor(Antibiotics.visit),data=.)
summary(lm.PEN)
coefplot(lm.PEN)


####################################
### plot all coefficients in one plot
### the order is from bottom to top
coefplot(lm.all, xlim=c(-1, 1),  col.pts="black",intercept=TRUE)
coefplot(lm.antiTNF, add=TRUE, col.pts="red",  intercept=TRUE)
coefplot(lm.EEN, add=TRUE, col.pts="blue", intercept=TRUE, offset=0.2)
legend('topright',legend=c('All','antiTNF','EEN'),col=c("black","red","blue"),pch=19)

rm(lm.all)
rm(lm.PEN)
rm(lm.Diet)
rm(lm.antiTNF)
rm(lm.EEN)
```





# Analyze fungal data





## Normal vs. Crohn-T1
```{r fungi_normal_vs_crohn_t1,cache=TRUE}
### I don't use dplyr filter 
### because it will remove the row names
fungi.cp1 <- subset_data_by_sample_info(fungi.RPKM,sample.info,
                                        Type=='COMBO' | Type=='PLEASE-T1')

### remove empty rows, "6007-01"
rownames(fungi.cp1)[rowSums(fungi.cp1) == 0]
fungi.cp1 <- fungi.cp1[rowSums(fungi.cp1) > 0,]
### there are other samples with empty rows
### remember to remove them in the latter analysis

dim(fungi.cp1)
```

### Calculate distance
We only have five fungi in the data, binary jaccard may not be a good choice. Try bray.
```{r fungi_calculate_distance,cache=TRUE}
fungi.distance.method <- "Bray"
fungi.dist.cp1 <- calculate_distance(X = fungi.cp1, method = fungi.distance.method, tree = NA)

### distribution
hist(fungi.dist.cp1,xlab='Distance',main='Fungi (COMBO and PLEASE-T1)')
#hist(log(fungi.dist.cp1),xlab='log distance',main='Fungi (COMBO and PLEASE-T1)')
```

### Find the best clustering
```{r fungi_find_the_best_clustering,cache=TRUE}
fungi.pamk.best <- find_best_clustering(fungi.dist.cp1,plot.results=TRUE)
```

### MDS analysis
```{r fungi_MDS,cache=TRUE,results='hide'}
### MDS
fungi.meta.MDS <- metaMDS(fungi.dist.cp1)

### Normal/Disease
p <- add_rownames(as.data.frame(fungi.meta.MDS$points),var='Sample') %>%
  left_join(dplyr::select(add_rownames(sample.info,var='Sample'),
                          Sample,Disease),
            by='Sample') %>% 
  ggplot(aes(x=MDS1,y=MDS2))+ geom_point(aes(colour = Disease)) +
  scale_color_manual(values=c("#0066CC", "#FF3333"),name="") +
  labs(title = 'metaMDS') 
print(p)

### Normal/Disease, cluster
### These are bacterial clusters !!
p <- add_rownames(as.data.frame(fungi.meta.MDS$points),var='Sample') %>%
  left_join(dplyr::select(add_rownames(sample.info,var='Sample'),
                          Sample,Disease,Cluster),
            by='Sample') %>% 
  ggplot(aes(x=MDS1,y=MDS2))+ geom_point(aes(colour = Disease,shape=Cluster)) +
  scale_color_manual(values=c("#0066CC", "#FF3333"),name="") +
  scale_shape_manual(values=c(20,4),name="") +
  labs(title = 'metaMDS') 
print(p)
```

### Heatmap of normal vs Crohn-T1
```{r fungi_heatmap_of_normal_vs_crohn_t1,cache=TRUE,fig.width=8,fig.height=8}
bk <- unique(c(
        0,
        min(min(fungi.RPKM[fungi.RPKM>0])/2,0.00001),
        seq(min(min(fungi.RPKM[fungi.RPKM>0])/2,1),0.00001,length=10),
        seq(0.00001,0.0001,length=10),
        seq(0.0001,0.01,length=10),
        seq(0.01,1,length=20),
        seq(1,10,length=20),
        seq(10,max(fungi.RPKM),length=20)
        ))
hmcols<- colorRampPalette(c('white','#1f78b4','#ff7f00','#e41a1c'))(length(bk)-1)


#bk = unique(c(seq(0,quantile(as.matrix(fungi.cp1),0.2), length=100),
#              seq(quantile(as.matrix(fungi.cp1),0.2),quantile(as.matrix(fungi.cp1),0.5),length=100),
#              seq(quantile(as.matrix(fungi.cp1),0.5),quantile(as.matrix(fungi.cp1),1.0),length=100)))
#hmcols<- colorRampPalette(c("white","firebrick3","yellow"))(length(bk)-1) 


### the color lables must be columns,so I have to transform the data
pheatmap(t(fungi.cp1),border_color="grey60",
         clustering_distance_rows = 'correlation',
         clustering_distance_cols = fungi.dist.cp1,clustering_method = "complete",
         annotation=sample.info[,c("Type","Cluster","Treatment.Specific","FCP","Response","Antibiotics.visit","Steroids")],
         color =hmcols, breaks=bk,
         show_rownames = TRUE,
         scale="none")
```


```{r fungi_heatmap_manually_order_samples,cache=TRUE,fig.width=8,fig.height=8}
#### manually re-order the samples
data.heat <- 
  add_rownames(fungi.cp1,var='Sample') %>%
  left_join(add_rownames(sample.info,var='Sample'),by='Sample') %>% 
  arrange(Disease,Cluster,Response,Treatment.Specific,Antibiotics.visit,FCP,Steroids) %>%
  dplyr::select(1:(ncol(fungi.cp1)+1)) %>% 
  add_back_rownames(row.var='Sample')

pheatmap(t(data.heat),border_color="grey60",
         cluster_cols = FALSE,
         clustering_distance_rows = 'correlation',
         #clustering_distance_cols =,
         clustering_method = "complete",
         annotation=sample.info[,c("Disease","Cluster","Response","Treatment.Specific","FCP","Antibiotics.visit","Steroids")],
         color =hmcols, breaks=bk,
         show_rownames = TRUE,
         scale="none")

## for publication
pheatmap(t(data.heat),border_color="grey60",
         cluster_cols = FALSE,
         clustering_distance_rows = 'correlation',
         #clustering_distance_cols =,
         clustering_method = "complete",
         annotation=sample.info[,c("Disease","Cluster","Response","Treatment.Specific","FCP","Antibiotics.visit","Steroids")],
         color =hmcols, breaks=bk,
         cellwidth = 8, cellheight = 8 ,
         show_rownames = TRUE,
         scale="none")

rm(data.heat)
```


### Wilcoxon rank test of normal vs. Crohn
```{r fungi_rank_test_of_normal_vs_crohn,cache=TRUE}
### only plot the top 5 significant ones
fungi.cp1.wil <- wilcox_test(fungi.cp1,sample.info[,'Disease',drop=FALSE],
                             plot.heatmap=FALSE,boxplot.top=5,boxplot.y='RPKM')

```

### Random Forest 
```{r fungi_random_forest_of_normal_vs_crohn,cache=TRUE}
random_forest(fungi.cp1,sample.info[,'Disease',drop=FALSE],
              plot.MDS=TRUE,plot.importance=TRUE,print.result=FALSE,
              guess.num=50,boxplot.top=0,
              title="Normal vs. Crohn",boxplot.y="RPKM",
              seed=100)
```


### Wilcoxon rank test of normal vs. Crohn (no antibiotics use)
```{r fungi_rank_test_of_normal_vs_crohn_no_antibio,cache=TRUE}
### only plot the top 5 significant ones
fungi.cp1.wil.no.anti <- wilcox_test(fungi.cp1,
         subset(sample.info,Antibiotics.visit=='Not.Use',select=Disease),
        plot.heatmap=FALSE,boxplot.top=0,boxplot.y='RPKM')
###
write.csv(fungi.cp1.wil.no.anti,file=paste('../3_Result/2015_03_05_Fungi_COMBO_vs_PLEASET1_No_Abx_Wilcox.csv',sep=''),row.names=TRUE)
```


## Antibiotics use

### Wilcoxon rank test 
```{r fungi_rank_test_of_antibiotics_use,cache=TRUE}
### plot all the significant ones
fungi.antibio.wil <- wilcox_test(fungi.cp1,
                    subset(sample.info,Group=='PLEASE',select=Antibiotics.visit),
                             plot.heatmap=FALSE,boxplot.top=5,boxplot.y='RPKM')      
###
write.csv(fungi.antibio.wil,file=paste('../3_Result/2015_03_05_Fungi_PLEASE_Antibiotics_Use_Wilcox.csv',sep=''),row.names=TRUE)

```

### Random Forest 
```{r fungi_random_forest_of_antibiotics_use,cache=TRUE}
random_forest(fungi.cp1,
              subset(sample.info,Group=='PLEASE',select=Antibiotics.visit),
              plot.MDS=TRUE,plot.importance=TRUE,print.result=FALSE,
              guess.num=50,boxplot.top=5,
              title="Antibiotics use",boxplot.y="RPKM",
              seed=100)
```


## Cluster 1 vs. cluster 2 (becterial clusters)

### Boxplot fungi abundance in becterial cluster 1 vs. 2
```{r fungi_boxplot_of_cluster_1_vs_2,cache=TRUE}
for (fg in colnames(fungi.RPKM)){
  print(fg)
  ggdata <- add_rownames(fungi.RPKM[,fg,drop=F],var='Sample') %>% 
  select(Sample,RPKM=2)  %>% ## the 2nd column -> Abu
  left_join(add_rownames(sample.info,var='Sample'),by='Sample') %>%
  dplyr::select(Type,Cluster,RPKM) %>%
  dplyr::filter(Type=='COMBO' | Type=='PLEASE-T1') %>%
  na.omit() %>%
  unite(Group,Type,Cluster, sep = ' ')
  p <- ggplot(ggdata,aes(factor(Group), RPKM)) +
  geom_boxplot(aes(fill = Group),outlier.shape = NA) + 
  #geom_jitter() +
  theme_classic() +
  ggtitle(fg) +
  ylim(0,quantile(ggdata$RPKM,0.8)) ## remove outliers
  print(p)
}

```

### Wilcoxon rank test of bacterial cluster 1 vs. 2
```{r fungi_rank_test_of_cluster_1_vs_2,cache=TRUE}
fungi.cluster.wil <- wilcox_test(fungi.cp1,sample.info[,'Cluster',drop=FALSE],
                                 plot.heatmap=FALSE,boxplot.top=5,boxplot.y='RPKM')
#### only five fungi
#### random forest may not be a good idea
```

## Compare fungal abundance by treaments, time points and response status

### Overall fungal abundance
```{r fungi_compare_abundance_by_time_treatment_response,cache=TRUE}
### Sum all fungal abundance together
fungi.data <- fungi.RPKM %>% rowSums %>% as.matrix %>%
## rename header
as.data.frame %>% dplyr::select(RPKM=1) %>%
add_rownames(var='Sample') %>%
mutate(log.RPKM=log10(RPKM+min(RPKM[RPKM>0]))) %>%
### add sample information 
left_join(add_rownames(sample.info,var='Sample'),by='Sample')
 
p <- fungi.data %>% dplyr::select(log.RPKM,Response,Time,Treatment.Specific) %>%
  na.omit %>%
  ggplot(aes(factor(Time), log.RPKM),na.rm=T)+
  geom_boxplot(na.rm=T)+facet_grid(Treatment.Specific ~ Response, margins=TRUE)
plot(p)

####  T1 vs. T2,3,4 
fungi.data %>%
  dplyr::select(Subject,Time,Treatment,RPKM) %>% 
  na.omit() %>% 
  mutate(Time=paste('T',Time,sep='')) %>%
  spread(Time,RPKM) %>%
  summarise_each(funs(wilcox.test(., T1, pair = TRUE)$p.value), vars = T2:T4) %>%
  rename(T1.vs.T2=vars1,T1.vs.T3=vars2,T1.vs.T4=vars3)

####  T1 vs. T2,3,4 for each treatment arm
fungi.data %>%
  dplyr::select(Subject,Time,Treatment,RPKM) %>% 
  na.omit() %>% 
  mutate(Time=paste('T',Time,sep='')) %>%
  spread(Time,RPKM) %>%
  group_by(Treatment) %>%
  summarise_each(funs(wilcox.test(., T1, pair = TRUE)$p.value), vars = T2:T4) %>%
  rename(T1.vs.T2=vars1,T1.vs.T3=vars2,T1.vs.T4=vars3)


####  T1 vs. T2,3,4 for each treatment arm
fungi.data %>%
  dplyr::select(Subject,Time,Treatment.Specific,RPKM) %>% 
  na.omit() %>% 
  mutate(Time=paste('T',Time,sep='')) %>%
  spread(Time,RPKM) %>% 
  group_by(Treatment.Specific) %>%
  summarise_each(funs(wilcox.test(., T1, pair = TRUE)$p.value), vars = T2:T4) %>%
  rename(T1.vs.T2=vars1,T1.vs.T3=vars2,T1.vs.T4=vars3)

####  T1 vs. T2,3,4 for each treatment arm & response
fungi.data %>%
  dplyr::select(Subject,Time,Treatment,Response,RPKM) %>% 
  na.omit() %>% 
  mutate(Time=paste('T',Time,sep='')) %>%
  spread(Time,RPKM) %>%
  group_by(Treatment,Response) %>%
  summarise_each(funs(wilcox.test(., T1, pair = TRUE)$p.value), vars = T2:T4) %>%
  rename(T1.vs.T2=vars1,T1.vs.T3=vars2,T1.vs.T4=vars3)

####  T1 vs. T2,3,4 for each treatment arm & response
####  Not enough sample for PEN
fungi.data %>%
  dplyr::select(Subject,Time,Treatment.Specific,Response,RPKM) %>% 
  na.omit() %>% 
  mutate(Time=paste('T',Time,sep='')) %>%
  spread(Time,RPKM) %>% 
  filter(Treatment.Specific=='EEN') %>%
  group_by(Treatment.Specific,Response) %>%
  summarise_each(funs(wilcox.test(., T1, pair = TRUE)$p.value), vars = T2:T4) %>%
  rename(T1.vs.T2=vars1,T1.vs.T3=vars2,T1.vs.T4=vars3)

###########################
##### R vs. NR
wilcoxtest.fun = function(tdata){ 
  R.vs.NR <- NA
  try(R.vs.NR <- signif(wilcox.test(RPKM~Response,data=tdata)$p.value, 2),silent=TRUE)
  return(as.data.frame(R.vs.NR))
}


####  R vs. NR: by treatment
fungi.data %>%
  dplyr::select(Subject,Time,Treatment,Treatment.Specific,RPKM,Response) %>% 
  na.omit() %>%
  mutate(Time=paste('T',Time,sep='')) %>%
  group_by(Treatment) %>%
  do(wilcoxtest.fun(.)) %>%
  add_back_rownames(row.var='Treatment')

fungi.data %>%
  dplyr::select(Subject,Time,Treatment,Treatment.Specific,RPKM,Response) %>% 
  na.omit() %>%
  mutate(Time=paste('T',Time,sep='')) %>%
  group_by(Treatment.Specific) %>%
  do(wilcoxtest.fun(.)) %>%
  add_back_rownames(row.var='Treatment.Specific')

####  R vs. NR: by time
fungi.data %>%
  dplyr::select(Subject,Time,Treatment,Treatment.Specific,RPKM,Response) %>% 
  na.omit() %>%
  mutate(Time=paste('T',Time,sep='')) %>%
  group_by(Time) %>%
  do(wilcoxtest.fun(.)) %>%
  add_back_rownames(row.var='Time')

####  R vs. NR: by time & treatment
fungi.data %>%
  dplyr::select(Subject,Time,Treatment,Treatment.Specific,RPKM,Response) %>% 
  na.omit() %>%
  mutate(Time=paste('T',Time,sep='')) %>%
  group_by(Time,Treatment) %>%
  do(wilcoxtest.fun(.)) 


fungi.data %>%
  dplyr::select(Subject,Time,Treatment,Treatment.Specific,RPKM,Response) %>% 
  na.omit() %>%
  mutate(Time=paste('T',Time,sep='')) %>%
  group_by(Time,Treatment.Specific) %>%
  do(wilcoxtest.fun(.))

###### EEN T1 R vs. NR
fungi.data %>%
  dplyr::select(Subject,Time,Treatment,Treatment.Specific,log.RPKM,Response) %>% 
  dplyr::filter(Time=='1'&Treatment.Specific=='EEN') %>%
  na.omit() %>%
  ggplot(aes(factor(Response), log.RPKM),na.rm=T)+
  geom_boxplot(na.rm=T)

fungi.data %>%
  dplyr::select(Subject,Time,Treatment,Treatment.Specific,RPKM,Response) %>% 
  dplyr::filter(Time=='1'&Treatment.Specific=='EEN') %>%
  wilcox.test(as.numeric(RPKM)~as.factor(Response),data=.)

##
rm(fungi.data)
```


### EEN  T1 vs. T2
```{r fungi_t1_vs_t2_een,cache=TRUE}
### Individual fungi comparison
fungi.wil.t1.t2.een <- add_rownames(fungi.RPKM,var='Sample') %>%
  gather(Taxa,Abu,-Sample) %>%
  left_join(dplyr::select(add_rownames(sample.info,var='Sample'),Subject,Sample,Time,Treatment.Specific,Response), by='Sample') %>%  
  filter(!is.na(Time)) %>% 
  mutate(Time=paste('T',Time,sep='')) %>%
  dplyr::select(-Sample) %>% 
  spread(Time,Abu) %>% 
  dplyr::filter(Treatment.Specific=='EEN') %>% 
  group_by(Taxa) %>%
  summarise(median.T1=median(T1,na.rm=TRUE),mad.T1=mad(T1,na.rm=TRUE),
            median.T2=median(T2,na.rm=TRUE),mad.T2=mad(T2,na.rm=TRUE),
            mean.T1=mean(T1,na.rm=TRUE),sd.T1=sd(T1,na.rm=TRUE),
            mean.T2=mean(T2,na.rm=TRUE),sd.T2=sd(T2,na.rm=TRUE),
            T1.vs.T2.p=wilcox.test(T1, T2, pair = TRUE)$p.value
            ) %>%
  ## raw pvalues
  mutate(T1.vs.T2.q = p.adjust(T1.vs.T2.p,method='fdr'))
  
 
########### save the results
write.csv(fungi.wil.t1.t2.een,file=paste('../3_Result/2015_03_05_Fungi_EEN_T1_vs_T2_Wilcox.csv',sep=''),row.names=FALSE)
```

###  T1 vs. T4 for each treatment
```{r fungi_t1_vs_t4,cache=TRUE}
### T1 vs. T4: EEN
fungi.wil.t1.t4.een <- add_rownames(fungi.RPKM,var='Sample') %>%
  gather(Taxa,Abu,-Sample) %>%
  left_join(dplyr::select(add_rownames(sample.info,var='Sample'),Subject,Sample,Time,Treatment.Specific,Response), by='Sample') %>%  
  filter(!is.na(Time)) %>% 
  mutate(Time=paste('T',Time,sep='')) %>%
  dplyr::select(-Sample) %>% 
  spread(Time,Abu) %>% 
  dplyr::filter(Treatment.Specific=='EEN') %>% 
  group_by(Taxa) %>%
  summarise(median.T1=median(T1,na.rm=TRUE),mad.T1=mad(T1,na.rm=TRUE),
            median.T4=median(T4,na.rm=TRUE),mad.T4=mad(T4,na.rm=TRUE),
            mean.T1=mean(T1,na.rm=TRUE),sd.T1=sd(T1,na.rm=TRUE),
            mean.T4=mean(T4,na.rm=TRUE),sd.T4=sd(T4,na.rm=TRUE),
            T1.vs.T4.p=wilcox.test(T1, T4, pair = TRUE)$p.value
            ) %>%
  ## raw pvalues
  mutate(T1.vs.T4.q = p.adjust(T1.vs.T4.p,method='fdr'))
  
########### save the results
write.csv(fungi.wil.t1.t4.een,file=paste('../3_Result/2015_03_05_Fungi_EEN_T1_vs_T4_Wilcox.csv',sep=''),row.names=FALSE)

rm(fungi.wil.t1.t4.een)


### T1 vs. T4: PEN
fungi.wil.t1.t4.pen <- add_rownames(fungi.RPKM,var='Sample') %>%
  gather(Taxa,Abu,-Sample) %>%
  left_join(dplyr::select(add_rownames(sample.info,var='Sample'),Subject,Sample,Time,Treatment.Specific,Response), by='Sample') %>%  
  filter(!is.na(Time)) %>% 
  mutate(Time=paste('T',Time,sep='')) %>%
  dplyr::select(-Sample) %>% 
  spread(Time,Abu) %>% 
  dplyr::filter(Treatment.Specific=='PEN') %>% 
  group_by(Taxa) %>%
  summarise(median.T1=median(T1,na.rm=TRUE),mad.T1=mad(T1,na.rm=TRUE),
            median.T4=median(T4,na.rm=TRUE),mad.T4=mad(T4,na.rm=TRUE),
            mean.T1=mean(T1,na.rm=TRUE),sd.T1=sd(T1,na.rm=TRUE),
            mean.T4=mean(T4,na.rm=TRUE),sd.T4=sd(T4,na.rm=TRUE),
            T1.vs.T4.p=wilcox.test(T1, T4, pair = TRUE)$p.value
            ) %>%
  ## raw pvalues
  mutate(T1.vs.T4.q = p.adjust(T1.vs.T4.p,method='fdr'))
  
 
########### save the results
write.csv(fungi.wil.t1.t4.pen,file=paste('../3_Result/2015_03_05_Fungi_PEN_T1_vs_T4_Wilcox.csv',sep=''),row.names=FALSE)

rm(fungi.wil.t1.t4.pen)


### T1 vs. T4: antiTNF
fungi.wil.t1.t4.antitnf <- add_rownames(fungi.RPKM,var='Sample') %>%
  gather(Taxa,Abu,-Sample) %>%
  left_join(dplyr::select(add_rownames(sample.info,var='Sample'),Subject,Sample,Time,Treatment.Specific,Response), by='Sample') %>%  
  filter(!is.na(Time)) %>% 
  mutate(Time=paste('T',Time,sep='')) %>%
  dplyr::select(-Sample) %>% 
  spread(Time,Abu) %>% 
  dplyr::filter(Treatment.Specific=='antiTNF') %>% 
  group_by(Taxa) %>%
  summarise(median.T1=median(T1,na.rm=TRUE),mad.T1=mad(T1,na.rm=TRUE),
            median.T4=median(T4,na.rm=TRUE),mad.T4=mad(T4,na.rm=TRUE),
            mean.T1=mean(T1,na.rm=TRUE),sd.T1=sd(T1,na.rm=TRUE),
            mean.T4=mean(T4,na.rm=TRUE),sd.T4=sd(T4,na.rm=TRUE),
            T1.vs.T4.p=wilcox.test(T1, T4, pair = TRUE)$p.value
            ) %>%
  ## raw pvalues
  mutate(T1.vs.T4.q = p.adjust(T1.vs.T4.p,method='fdr'))
  
 
########### save the results
write.csv(fungi.wil.t1.t4.antitnf,file=paste('../3_Result/2015_03_05_Fungi_antiTNF_T1_vs_T4_Wilcox.csv',sep=''),row.names=FALSE)

rm(fungi.wil.t1.t4.antitnf)
```



## quantile regression (baseline, antibiotics)
RPKM ~ baseline + antibiotics for each treatment
```{r fungi_quantile_regression,cache=TRUE}
antibiotics.temp <- add_rownames(sample.info,var='Sample') %>%
    dplyr::select(Subject,Time,Antibiotics.visit) %>%
    dplyr::filter(Time==1 | Time==4) %>%
    mutate(Time=ifelse(Time==1,'T1','T4')) %>%
    spread(Time,Antibiotics.visit) %>%
    mutate(antibiotics = ifelse(is.na(T4),T1,T4)) %>%
    dplyr::select(Subject,antibiotics)

## sum up of fungal abundance
qr.data <- fungi.RPKM %>% rowSums %>% as.matrix %>%
## rename header
as.data.frame %>% dplyr::select(RPKM=1) %>%
add_rownames(var='Sample') %>%
### add sample information 
left_join(add_rownames(sample.info,var='Sample'),by='Sample') %>% 
dplyr::filter(Time==1 | Time==4) %>%
mutate(Time=ifelse(Time==1,'T1','T4')) %>%
dplyr::select(Subject,Time,RPKM,Treatment.Specific,Response) %>% 
# mutate(antibiotics = ifelse(is.na(T4),T1,T4))
tidyr::spread(Time,RPKM) %>%
left_join(antibiotics.temp,by='Subject')

## all
qr.data %>% 
rq(T4~ T1 + Treatment.Specific + Response + antibiotics,
        tau=0.75,data=.,na.action=na.omit) %>%
summary.rq(.,se="boot") %>% coef 

## antiTNF
qr.data %>% dplyr::filter(Treatment.Specific == 'antiTNF') %>%
rq(T4~ T1 + Response + antibiotics,
        tau=0.75,data=.,na.action=na.omit) %>%
summary.rq(.,se="boot") %>% coef 


## EEN
qr.data %>% dplyr::filter(Treatment.Specific == 'EEN') %>%
rq(T4~ T1 + Response + antibiotics,
        tau=0.75,data=.,na.action=na.omit) %>%
summary.rq(.,se="boot") %>% coef 
```



### quantile regression on fold change
```{r fungi_fc_quantile_regression_treatment_antibiotics_response,cache=TRUE}
### add overall fungal abundance to the data matrix
fungi.data <- fungi.RPKM %>% rowSums %>% as.matrix %>%
## rename header
as.data.frame %>% dplyr::select(All=1) %>%
add_rownames(var='Sample') %>%
### add sample information 
left_join(add_rownames(as.data.frame(fungi.RPKM),var='Sample'),by='Sample') %>%
add_back_rownames(row.var='Sample')
  
  
## fold change T4/T1
rq.fc.res <- list()
for (fg in colnames(fungi.data)){
  #print(pw)
  ### it seems rq used some random numbers
  set.seed(1)
  fungi.tem <- fungi.data[,fg,drop=FALSE]
  colnames(fungi.tem) <- 'Fungi'
    ####
    master.table <- add_rownames(as.data.frame(fungi.tem),var='Sample') %>%
    left_join(add_rownames(sample.info,var='Sample'),by='Sample') %>%
    dplyr::filter(Time==1 | Time==4) %>%
    dplyr::select(Fungi,Subject,Time,Treatment.Specific,Antibiotics.visit,log.FCP,Response) %>%
    mutate(Time=ifelse(Time==1,'T1','T4'))
    ### fungi
    fg.temp <- master.table %>%
    mutate(Fungi = Fungi+min(Fungi[Fungi>0]))  %>% 
    dplyr::select(Subject,Fungi,Time,Treatment.Specific) %>%
    spread(Time,Fungi) %>%
    mutate(Fungi.log.fc = log(T4/T1)) %>%
    dplyr::select(Subject,Treatment.Specific,Fungi.log.fc)
    ### response
    response.temp <- master.table %>%
    dplyr::select(Subject,Response) 
    ### log FCP
    log.fcp.temp <- master.table %>%
    dplyr::select(Subject,Time,log.FCP) %>%
    spread(Time,log.FCP) %>%
    mutate(log.FCP.T4 = T4) %>%
    dplyr::select(Subject,log.FCP.T4)
    ### antibiotics
    antibiotics.temp <- master.table %>%
    dplyr::select(Subject,Time,Antibiotics.visit) %>%
    spread(Time,Antibiotics.visit) %>%
    mutate(antibiotics = ifelse(is.na(T4),T1,T4)) %>%
    dplyr::select(Subject,antibiotics)
    ### merge together 
    rq.fc.res[[fg]] <- fg.temp %>% full_join(antibiotics.temp,by='Subject') %>%
    full_join(response.temp,by='Subject') %>%
    ### !!!!!!!!!! ########
    ### After updating RStudio and some R packages,
    ### a problem has occured: the reference level in the linear model
    ### has changed for no reason
    ### Change the reference level back to EEN for consistency with
    ### previous analysis
    ###
    mutate( Treatment.Specific = relevel(as.factor(Treatment.Specific), ref = 'EEN')) %>%
    ######################
    ### quantile regression
    rq( Fungi.log.fc ~ Treatment.Specific + antibiotics + Response,
        tau=0.75,data=.,na.action=na.omit) %>%
    summary.rq(.,se="boot") %>% coef %>%
    as.data.frame %>% add_rownames(var='Var') %>%
    dplyr::select(1,2,5) %>% dplyr::select(Var,Est=Value,Pval=starts_with("Pr"))%>%
    gather(value,var,Est,Pval) %>% 
    unite(Variables,Var,value, sep = ".", remove = TRUE) %>%
    add_back_rownames(row.var='Variables') %>% as.matrix
  #### add fungi name
  colnames(rq.fc.res[[fg]]) <- fg
  ###  
  rm(fungi.tem)
  rm(master.table)
  rm(fg.temp)
  rm(antibiotics.temp)
  rm(log.fcp.temp)
  rm(response.temp)
  rm(fg)
  #break
}

rq.res.table <- 
  as.data.frame(t(as.data.frame(rq.fc.res))) %>% ## must use as.data.frame again 
  add_rownames(var='fungi') %>%
  mutate(
    Treatment.SpecificPEN.Qval = p.adjust(Treatment.SpecificPEN.Pval,method='fdr'),
    Treatment.SpecificantiTNF.Qval = p.adjust(Treatment.SpecificantiTNF.Pval,method='fdr'),
    Antibiotics.visitUse.Qval = p.adjust(antibioticsUse.Pval,method='fdr'),
    ResponseResponse.Qval = p.adjust(ResponseResponse.Pval ,method='fdr')
  ) 

rm(rq.fc.res)
######
#### row name is a column
write.csv(rq.res.table,file=paste('../3_Result/2015_03_05_Fungi_Fold_Change_Quantile_Q75_Regression_Treatment_Antibiotics_Response.csv',sep=''),row.names=FALSE)

rm(rq.res.table)
rm(fungi.data)
```


## Linear regression on response status
```{r fungi_linear_regression_on_response,cache=TRUE, eval=FALSE, fig.width=10}
### FIX THIS: some subjects have different antibiotics.visit information at 
### different time points. THe data transformation will result in duplicated rows
### check data.mat 5007
for (fg in colnames(fungi.RPKM)){
  print(fg)
  ########!!!!!!!!!!!!!!!!!!!!!!!!
  #### consider log transfomation of the abundance
  ########
  lm.dat <- add_rownames(fungi.RPKM[,fg,drop=F],var='Sample') %>% 
    select(Sample,Abu=2)  %>% ## the 2nd column -> Abu
    left_join(add_rownames(sample.info,var='Sample'),by='Sample') %>%
    dplyr::select(Subject,Abu,Time,Treatment,Treatment.Specific,Response,Antibiotics.visit) %>%
    na.omit() %>%
    mutate(Time=paste('T',Time,sep='')) %>%
    spread(Time,Abu) 
  
  ### antiTNF
  cat('\n','antiTNF','\n')
  data.mat <- lm.dat %>% filter(Treatment =='antiTNF') 
  lm.antiTNF <- linear_regression_with_permutation(T4~T1+Response+Antibiotics.visit, data.mat=data.mat, perm.var='T4',perm.num=1000)
  print(lm.antiTNF)
  
  
  ### Diet
  cat('\n','Diet','\n')
  data.mat <- lm.dat %>% filter(Treatment =='Diet') 
  lm.Diet <- linear_regression_with_permutation(T4~T1+Response+Antibiotics.visit, data.mat=data.mat, perm.var='T4',perm.num=1000)
  print(lm.Diet)
  
  
  
  ### EEN (change model to T4~T1+Response, otherwise get an error)
  cat('\n','EEN','\n')
  data.mat <- lm.dat %>% filter(Treatment.Specific =='EEN')
  lm.EEN <- linear_regression_with_permutation(T4~T1+Response, data.mat=data.mat, perm.var='T4',perm.num=1000)
  print(lm.EEN)
  
  #break
  cat('\n')
}

```


### Compare to healthy controls
```{r fungi_to_healthy_controls,cache=TRUE,eval=FALSE}
#### Plot distance by treatment,response
p<- sample.info %>% dplyr::select(Type,Fungi.Per,Treatment.Specific,Response) %>%
  mutate(Response=ifelse(Type=='COMBO','COMBO',Response)) %>%
  mutate(Treatment.Specific=ifelse(Type=='COMBO','COMBO',Treatment.Specific)) %>%
  na.omit() %>%
  unite(Treatment.Response, Treatment.Specific,Response, sep = ".") %>%
  ggplot(aes(factor(Type),log(Fungi.Per+min(Fungi.Per[Fungi.Per>0])))) + 
  geom_boxplot(aes(fill = factor(Treatment.Response)),outlier.shape = NA)+
  xlab('')+ylab('log Fungi Per') + 
  theme_classic() 
plot(p)

#### Calculate the pvalues for each box in the above figure
sample.info %>%
dplyr::select(Type,Time,Treatment.Specific,Human.Per,Response) %>% 
mutate(Treatment.Specific=ifelse(Type=='COMBO','COMBO',Treatment.Specific)) %>%
mutate(Response=ifelse(Type=='COMBO','COMBO',Response)) %>%
mutate(Time=paste('T',Time,sep='')) %>%
mutate(Time=ifelse(Type=='COMBO','COMBO',Time)) %>%
dplyr::filter(Treatment.Specific != 'PEN') %>% 
na.omit() %>%
unite(Time.Treatment.Response,Time, Treatment.Specific,Response, sep = "_") %>%
dplyr::select(Time.Treatment.Response,Human.Per) %>%
(function(tdata){
cdata <- dplyr::filter(tdata,Time.Treatment.Response=='COMBO_COMBO_COMBO') %>%
         dplyr::select(Human.Per) %>%
         mutate(Human.Per=as.numeric(Human.Per))
#print(cdata)
group_by(tdata,Time.Treatment.Response) %>%
summarise(pval.compare.to.combo=wilcox.test(Human.Per,cdata$Human.Per)$p.value) %>%
mutate(pval.compare.to.combo=signif(pval.compare.to.combo,2)) %>%
separate(Time.Treatment.Response, into=c("Time","Treatment","Response"), sep = "_")
})
```


## Abundance change across time
```{r fungi_abundance_change_aross_time,cache=TRUE,eval=FALSE,include=FALSE}
sample.info %>%
  dplyr::select(Subject,Time, Fungi.Per,Response,Treatment.Specific) %>%
  #dplyr::filter(Treatment.Specific != 'PEN') %>%
  dplyr::filter(Time==1 | Time==4) %>%
  na.omit() %>%
  ### not right, need non-zero min value
  mutate(Fungi.Per.log=log(Fungi.Per+min(Fungi.Per))) %>%
  #spread(Time,bact.div) %>%
  #head
  ggplot(aes(Time,Fungi.Per.log,group=Subject,color=Response),data=.) + geom_line(alpha = 0.5) +facet_grid(.~Treatment.Specific)
```


## Correlation of fungi abundance
```{r fungi_correlation_please_t1,cache=TRUE}
fungi.p1 <- subset_data_by_sample_info(fungi.RPKM,sample.info,Type=='PLEASE-T1')
signif(cor(fungi.p1 ,method='spearman'),2)
```


## Correlation of fungi and bacteria
```{r fungi_correlation_with_bacteria,cache=TRUE,fig.height=10}
### scale the colors so that all heatmaps are comparable.
bk <- unique(c(seq(-1,-0.4,length=10),seq(-0.4,0.4,length=50),seq(0.4,1,length=10)))
hmcols<- colorRampPalette(c('#4575b4','#91bfdb','#e0f3f8',
                            '#ffffbf','#fee090','#fc8d59','#d73027'
                            ))(length(bk)-1)

### all
ind <- intersect(rownames(fungi.RPKM),rownames(taxa.data))
fungi.bact.cor <- cor(fungi.RPKM[ind,],taxa.data[ind,],method='spearman')
pheatmap(t(fungi.bact.cor), 
           scale='none',fontsize_row = 10, 
           clustering_distance_cols = "manhattan",
           clustering_distance_rows = "manhattan",
           cluster_cols = TRUE,
           color =hmcols, breaks=bk,
           cellwidth = 12, cellheight = 8
           )
rm(ind)

### control
ind <- intersect(intersect(rownames(fungi.RPKM),rownames(taxa.data)),rownames(subset(sample.info,Type=='COMBO')))
fungi.bact.cor <- cor(fungi.RPKM[ind,],taxa.data[ind,],method='spearman')
pheatmap(na.omit(t(fungi.bact.cor)), 
           scale='none',fontsize_row = 10, 
           clustering_distance_cols = "manhattan",
           clustering_distance_rows = "manhattan",
           cluster_cols = TRUE,
         color =hmcols, breaks=bk,
           cellwidth = 12, cellheight = 8
           )
rm(ind)

### PLEASE Cluster 1
ind <- intersect(intersect(rownames(fungi.RPKM),rownames(taxa.data)),rownames(subset(sample.info,Type=='PLEASE-T1' & Cluster=='cluster 1')))
fungi.bact.cor <- cor(fungi.RPKM[ind,],taxa.data[ind,],method='spearman')
pheatmap(na.omit(t(fungi.bact.cor)), 
           scale='none',fontsize_row = 10, 
           clustering_distance_cols = "manhattan",
           clustering_distance_rows = "manhattan",
           cluster_cols = TRUE,
         color =hmcols, breaks=bk,
           cellwidth = 12, cellheight = 8
           )
rm(ind)

### PLEASE Cluster 2
### some of the pairs are NA (sd is zero)
### so the rows in the final figures are different in cluster 1 vs. 2
ind <- intersect(intersect(rownames(fungi.RPKM),rownames(taxa.data)),rownames(subset(sample.info,Type=='PLEASE-T1' & Cluster=='cluster 2')))
fungi.bact.cor <- cor(fungi.RPKM[ind,],taxa.data[ind,],method='spearman')
pheatmap(na.omit(t(fungi.bact.cor)), 
           scale='none',fontsize_row = 10, 
           clustering_distance_cols = "manhattan",
           clustering_distance_rows = "manhattan",
           cluster_cols = TRUE,
         color =hmcols, breaks=bk,
           cellwidth = 12, cellheight = 8
           )
rm(ind)

### PLEASE-T1
ind <- intersect(intersect(rownames(fungi.RPKM),rownames(taxa.data)),rownames(subset(sample.info,Type=='PLEASE-T1')))
fungi.bact.cor <- cor(fungi.RPKM[ind,],taxa.data[ind,],method='spearman')
pheatmap(na.omit(t(fungi.bact.cor)), 
           scale='none',fontsize_row = 10, 
           clustering_distance_cols = "manhattan",
           clustering_distance_rows = "manhattan",
           cluster_cols = TRUE,
         color =hmcols, breaks=bk,
           cellwidth = 12, cellheight = 8
           )
rm(ind)

### PLEASE-T4
ind <- intersect(intersect(rownames(fungi.RPKM),rownames(taxa.data)),rownames(subset(sample.info,Type=='PLEASE-T4')))
fungi.bact.cor <- cor(fungi.RPKM[ind,],taxa.data[ind,],method='spearman')
pheatmap(na.omit(t(fungi.bact.cor)), 
           scale='none',fontsize_row = 10, 
           clustering_distance_cols = "manhattan",
           clustering_distance_rows = "manhattan",
           cluster_cols = TRUE,
         color =hmcols, breaks=bk,
           cellwidth = 12, cellheight = 8
           )
rm(ind)

### antiTNF
ind <- intersect(intersect(rownames(fungi.RPKM),rownames(taxa.data)),rownames(subset(sample.info,Treatment=='antiTNF')))
fungi.bact.cor <- cor(fungi.RPKM[ind,],taxa.data[ind,],method='spearman')
pheatmap(t(fungi.bact.cor), 
           scale='none',fontsize_row = 10, 
           clustering_distance_cols = "manhattan",
           clustering_distance_rows = "manhattan",
           cluster_cols = TRUE,
         color =hmcols, breaks=bk,
           cellwidth = 12, cellheight = 8
           )
rm(ind)

### EEN
ind <- intersect(intersect(rownames(fungi.RPKM),rownames(taxa.data)),rownames(subset(sample.info,Treatment.Specific=='EEN')))
fungi.bact.cor <- cor(fungi.RPKM[ind,],taxa.data[ind,],method='spearman')
pheatmap(na.omit(t(fungi.bact.cor)), 
           scale='none',fontsize_row = 10, 
           clustering_distance_cols = "manhattan",
           clustering_distance_rows = "manhattan",
           cluster_cols = TRUE,
         color =hmcols, breaks=bk,
           cellwidth = 12, cellheight = 8
           )
rm(ind)

### response
ind <- intersect(intersect(rownames(fungi.RPKM),rownames(taxa.data)),rownames(subset(sample.info,Response=='Response')))
fungi.bact.cor <- cor(fungi.RPKM[ind,],taxa.data[ind,],method='spearman')
pheatmap(t(fungi.bact.cor), 
           scale='none',fontsize_row = 10, 
           clustering_distance_cols = "manhattan",
           clustering_distance_rows = "manhattan",
           cluster_cols = TRUE,
         color =hmcols, breaks=bk,
           cellwidth = 12, cellheight = 8
           )
rm(ind)

### non-response
ind <- intersect(intersect(rownames(fungi.RPKM),rownames(taxa.data)),rownames(subset(sample.info,Response=='Non.Response')))
fungi.bact.cor <- cor(fungi.RPKM[ind,],taxa.data[ind,],method='spearman')
pheatmap(t(fungi.bact.cor), 
           scale='none',fontsize_row = 10, 
           clustering_distance_cols = "manhattan",
           clustering_distance_rows = "manhattan",
           cluster_cols = TRUE,
         color =hmcols, breaks=bk,
           cellwidth = 12, cellheight = 8
           )
rm(ind)


### PLEASE-T4 Response
ind <- intersect(intersect(rownames(fungi.RPKM),rownames(taxa.data)),rownames(subset(sample.info,Type=='PLEASE-T4' & Response =='Response')))
fungi.bact.cor <- cor(fungi.RPKM[ind,],taxa.data[ind,],method='spearman')
pheatmap(na.omit(t(fungi.bact.cor)), 
           scale='none',fontsize_row = 10, 
           clustering_distance_cols = "manhattan",
           clustering_distance_rows = "manhattan",
           cluster_cols = TRUE,
         color =hmcols, breaks=bk,
           cellwidth = 12, cellheight = 8
           )
rm(ind)
```

# Analyze gene data

## Normal vs. Crohn-T1
```{r gene_normal_vs_crohn_t1,cache=TRUE}
### I don't use dplyr filter 
### because it will remove the row names
gene.cp1 <- subset_data_by_sample_info(gene.data,sample.info,
                                       Type=='COMBO' | Type=='PLEASE-T1')

### remove empty rows, "6007-01"
rownames(gene.cp1)[rowSums(gene.cp1) == 0]
# gene.cp1 <- gene.cp1[rowSums(gene.cp1) > 0,]
### there are other samples with empty rows
### remember to remove them in the latter analysis

dim(gene.cp1)
```

### Calculate distance
```{r gene_calculate_distance,cache=TRUE}
gene.distance.method <- "Bray"
gene.dist.cp1 <- calculate_distance(X = gene.cp1, method = gene.distance.method, tree = NA)

### distribution
hist(gene.dist.cp1,xlab='Distance',main='Gene (COMBO and PLEASE-T1)')

```

### Find the best clustering
```{r gene_find_the_best_clustering,cache=TRUE}
gene.pamk.best <- find_best_clustering(gene.dist.cp1,plot.results=TRUE)
```

### MDS analysis
```{r gene_MDS,cache=TRUE,results='hide'}
### MDS
gene.meta.MDS <- metaMDS(gene.dist.cp1)

### Normal/Disease
p <- add_rownames(as.data.frame(gene.meta.MDS$points),var='Sample') %>%
  left_join(dplyr::select(add_rownames(sample.info,var='Sample'),
                          Sample,Disease),
            by='Sample') %>% 
  ggplot(aes(x=MDS1,y=MDS2))+ geom_point(aes(colour = Disease)) +
  scale_color_manual(values=c("#0066CC", "#FF3333"),name="") +
  labs(title = 'metaMDS') 
print(p)

### Normal/Disease, cluster
### These are bacterial clusters !!
p <- add_rownames(as.data.frame(gene.meta.MDS$points),var='Sample') %>%
  left_join(dplyr::select(add_rownames(sample.info,var='Sample'),
                          Sample,Disease,Cluster),
            by='Sample') %>% 
  ggplot(aes(x=MDS1,y=MDS2))+ geom_point(aes(colour = Disease,shape=Cluster)) +
  scale_color_manual(values=c("#0066CC", "#FF3333"),name="") +
  scale_shape_manual(values=c(20,4),name="") +
  labs(title = 'metaMDS') 
print(p)


#### Fugal percentage 
p <- add_rownames(as.data.frame(gene.meta.MDS$points),var='Sample') %>%
  left_join(dplyr::select(add_rownames(sample.info,var='Sample'),
                          Sample,Disease,Cluster,Fungi.Per),
            by='Sample') %>% 
  mutate(Fungi.Per.sqrt=sqrt(Fungi.Per)) %>%
  ggplot(aes(x=MDS1,y=MDS2))+ 
  geom_point(aes(colour = Disease,shape=Cluster,size = Fungi.Per.sqrt)) +
  scale_color_manual(values=c("#0066CC", "#FF3333"),name="") +
  scale_shape_manual(values=c(20,1),name="") +
  scale_size_area()+
  labs(title = 'metaMDS') 
print(p)

#### Human DNA percentage
p <- add_rownames(as.data.frame(gene.meta.MDS$points),var='Sample') %>%
  left_join(dplyr::select(add_rownames(sample.info,var='Sample'),
                          Sample,Disease,Cluster,Human.Per),
            by='Sample') %>% 
  ggplot(aes(x=MDS1,y=MDS2))+ 
  geom_point(aes(colour = Disease,shape=Cluster,size = Human.Per)) +
  scale_color_manual(values=c("#0066CC", "#FF3333"),name="") +
  scale_shape_manual(values=c(20,1),name="") +
  scale_size_area()+
  labs(title = 'metaMDS') 
print(p)
```


### Random Forest 
```{r gene_random_forest_of_normal_vs_crohn,cache=TRUE}
random_forest(gene.cp1,sample.info[,'Disease',drop=FALSE],
              plot.MDS=TRUE,plot.importance=TRUE,print.result=FALSE,
              guess.num=50,boxplot.top=5,
              title="Normal vs. Crohn",boxplot.y="Abundance(%)",
              seed=100)
```


### Wilcoxon rank test of normal vs. Crohn
```{r gene_rank_test_of_normal_vs_crohn,cache=TRUE}
### only plot the top 5 significant ones
gene.cp1.wil <- wilcox_test(gene.cp1,sample.info[,'Disease',drop=FALSE],
                            plot.heatmap=FALSE,boxplot.top=5,boxplot.y='Abundance(%)')

#### 
write.csv(gene.cp1.wil,file=paste('../3_Result/2015_03_05_Gene_COMBO_PLEASET1_Wilcox.csv',sep=''),row.names=TRUE)
```



## Antibiotics use

### Wilcoxon rank test 
```{r gene_rank_test_of_antibiotics_use,cache=TRUE}
### plot top ones
gene.antibio.wil <- wilcox_test(gene.cp1,
                    subset(sample.info,Group=='PLEASE',select=Antibiotics.visit),
                             plot.heatmap=FALSE,boxplot.top=10,boxplot.y='Abundance(%)')      
###
write.csv(gene.antibio.wil,file=paste('../3_Result/2015_03_05_Gene_PLEASE_Antibiotics_Use_Wilcox.csv',sep=''),row.names=TRUE)

```

### Random Forest 
```{r gene_random_forest_of_antibiotics_use,cache=TRUE}
random_forest(gene.cp1,
              subset(sample.info,Group=='PLEASE',select=Antibiotics.visit),
              plot.MDS=TRUE,plot.importance=TRUE,print.result=FALSE,
              guess.num=50,boxplot.top=6,
              title="Antibiotics use",boxplot.y="Abundance(%)",
              seed=100)
```

## Cluster 1 vs. cluster 2

### Wilcoxon rank test 
```{r gene_rank_test_of_cluster_1_vs_2,cache=TRUE}
### plot top 5
gene.c1c2.wil <- wilcox_test(gene.cp1,
                             subset(sample.info,Group=='PLEASE',select=Cluster),
                             plot.heatmap=FALSE,boxplot.top=5,boxplot.y='Abundance(%)')

#### 
write.csv(gene.c1c2.wil,file=paste('../3_Result/2015_03_05_Gene_PLEASE_Cluster_1_2_Wilcox.csv',sep=''),row.names=TRUE)
```

### Random Forest 
```{r gene_random_forest_of_cluster_1_vs_2,cache=TRUE}
random_forest(gene.cp1,subset(sample.info,Group=='PLEASE',select=Cluster),
              plot.MDS=TRUE,plot.importance=TRUE,print.result=FALSE,
              guess.num=50,boxplot.top=5,
              title="Normal vs. Crohn",boxplot.y="Abundance(%)",
              seed=100)
```


# Analyze pathway data

## Normal vs. Crohn-T1
```{r pathway_normal_vs_crohn_t1,cache=TRUE}
### I don't use dplyr filter 
### because it will remove the row names
pathway.cp1 <- subset_data_by_sample_info(pathway.data,sample.info,
                                          Type=='COMBO' | Type=='PLEASE-T1')

### remove empty rows, "6007-01"
rownames(pathway.cp1)[rowSums(pathway.cp1) == 0]
# pathway.cp1 <- pathway.cp1[rowSums(pathway.cp1) > 0,]
### there are other samples with empty rows
### remember to remove them in the latter analysis

dim(pathway.cp1)
```

### Calculate distance
```{r pathway_calculate_distance,cache=TRUE}
pathway.distance.method <- "Bray"
pathway.dist.cp1 <- calculate_distance(X = pathway.cp1, method = pathway.distance.method, tree = NA)

### distribution
hist(pathway.dist.cp1,xlab='Distance',main='Pathway (COMBO and PLEASE-T1)')

```

### Find the best clustering
```{r pathway_find_the_best_clustering,cache=TRUE}
pathway.pamk.best <- find_best_clustering(pathway.dist.cp1,plot.results=TRUE)
```

### MDS analysis
```{r pathway_MDS,cache=TRUE,results='hide'}
### MDS
set.seed(10)
pathway.meta.MDS <- metaMDS(pathway.dist.cp1)

### Normal/Disease
p <- add_rownames(as.data.frame(pathway.meta.MDS$points),var='Sample') %>%
  left_join(dplyr::select(add_rownames(sample.info,var='Sample'),
                          Sample,Disease),
            by='Sample') %>% 
  ggplot(aes(x=MDS1,y=MDS2))+ geom_point(aes(colour = Disease)) +
  scale_color_manual(values=c("#0066CC", "#FF3333"),name="") +
  labs(title = 'metaMDS') 
print(p)

### Normal/Disease, cluster
### These are bacterial clusters !!
p <- add_rownames(as.data.frame(pathway.meta.MDS$points),var='Sample') %>%
  left_join(dplyr::select(add_rownames(sample.info,var='Sample'),
                          Sample,Disease,Cluster),
            by='Sample') %>% 
  ggplot(aes(x=MDS1,y=MDS2))+ geom_point(aes(colour = Disease,shape=Cluster)) +
  scale_color_manual(values=c("#0066CC", "#FF3333"),name="") +
  scale_shape_manual(values=c(20,4),name="") +
  labs(title = 'metaMDS') 
print(p)

### Normal/Disease, cluster (for publication)
### These are bacterial clusters !!
p <- add_rownames(as.data.frame(pathway.meta.MDS$points),var='Sample') %>%
  left_join(dplyr::select(add_rownames(sample.info,var='Sample'),
                          Sample,Disease,Cluster),
            by='Sample') %>% 
  ggplot(aes(x=MDS1,y=MDS2))+ geom_point(aes(colour = Cluster,shape=Disease)) +
  scale_color_manual(values=c("#0066CC", "#FF3333"),name="") +
  scale_shape_manual(values=c(20,1),name="") +
  labs(title = 'Ordination pathways') +
  theme_classic()
print(p)


#### Fugal percentage 
p <- add_rownames(as.data.frame(pathway.meta.MDS$points),var='Sample') %>%
  left_join(dplyr::select(add_rownames(sample.info,var='Sample'),
                          Sample,Disease,Cluster,Fungi.Per),
            by='Sample') %>% 
  mutate(Fungi.Per.sqrt=sqrt(Fungi.Per)) %>%
  ggplot(aes(x=MDS1,y=MDS2))+ 
  geom_point(aes(colour = Disease,shape=Cluster,size = Fungi.Per.sqrt)) +
  scale_color_manual(values=c("#0066CC", "#FF3333"),name="") +
  scale_shape_manual(values=c(20,1),name="") +
  scale_size_area()+
  labs(title = 'metaMDS') 
print(p)

#### Human DNA percentage
p <- add_rownames(as.data.frame(pathway.meta.MDS$points),var='Sample') %>%
  left_join(dplyr::select(add_rownames(sample.info,var='Sample'),
                          Sample,Disease,Cluster,Human.Per),
            by='Sample') %>% 
  ggplot(aes(x=MDS1,y=MDS2))+ 
  geom_point(aes(colour = Disease,shape=Cluster,size = Human.Per)) +
  scale_color_manual(values=c("#0066CC", "#FF3333"),name="") +
  scale_shape_manual(values=c(20,1),name="") +
  scale_size_area()+
  labs(title = 'metaMDS') 
print(p)
```


### Random Forest 
```{r pathway_random_forest_of_normal_vs_crohn,cache=TRUE}
### (3) for the random forest Importance score plots in Figure 2C, we have a list of pathways. For this list, can you get results on which pathways have high abundances in Crohn's disease individual based on medians in normal and Crohn's children?

#pdf('pathway_random_forest_of_normal_vs_crohn.pdf',width=10)
random_forest(pathway.cp1,sample.info[,'Disease',drop=FALSE],
              plot.MDS=TRUE,plot.importance=TRUE,print.result=FALSE,
              guess.num=50,boxplot.top=30,
              title="Normal vs. Crohn",boxplot.y="Abundance(%)",
              seed=100)
#dev.off()
```


### Wilcoxon rank test of normal vs. Crohn
```{r pathway_rank_test_of_normal_vs_crohn,cache=TRUE}
### only plot the top 5 significant ones
pathway.cp1.wil <- wilcox_test(pathway.cp1,sample.info[,'Disease',drop=FALSE],
                               plot.heatmap=FALSE,boxplot.top=0.05,boxplot.y='Abundance(%)')

###
write.csv(pathway.cp1.wil,file=paste('../3_Result/2015_03_05_Pathway_COMBO_PLEASET1_Wilcox.csv',sep=''),row.names=TRUE)

```

### Heatmap of normal vs Crohn-T1 (significant pathways)
```{r pathway_heatmap_of_normal_vs_crohn_t1_significant,cache=TRUE,fig.width=15,fig.height=10}
pathway.cp1.sig <- t(subset_data_by_sample_info(t(pathway.cp1), as.data.frame(pathway.cp1.wil),Disease.qvalue<0.05))
pathway.dist.cp1.sig <- calculate_distance(X = pathway.cp1.sig, method = pathway.distance.method, tree = NA)

bk <- unique(c(
        seq(-3,-2,length=100),
        seq(-2,-1.5,length=100),
        seq(-1.5,0,length=100),
        seq(0,1.5,length=100),
        seq(1.5,2,length=100),
        seq(2,3,length=100)
        ))
hmcols<- colorRampPalette(c('#d73027','#fc8d59','#fee090',
                            '#ffffbf','#e0f3f8','#91bfdb','#4575b4'
                            ))(length(bk)-1)

dat.tmp <- log(pathway.cp1.sig+min(pathway.cp1.sig[pathway.cp1.sig>0]))
dat.nor <- dat.tmp
for (pw in colnames(dat.tmp)){
  pw.dat <- dat.tmp[,pw,drop=TRUE]
  #hist(pw.dat,breaks=100)
  Me <- median(pw.dat)
  Ma <- mad(pw.dat)
  if (Ma==0){Me <- mean(pw.dat);Ma <- sd(pw.dat)}
  pw.dat <- (pw.dat-Me)/Ma
  pw.dat[pw.dat < -3] <- -3
  pw.dat[pw.dat > 3] <- 3
  dat.nor[,pw] <- pw.dat
}

### the color lables must be columns,so I have to transform the data
pheatmap(t(dat.nor),border_color="grey60",
         clustering_distance_rows = 'euclidean',
         clustering_distance_cols = pathway.dist.cp1.sig,
         clustering_method = "average",
         annotation=sample.info[,c("Type","FCP","Treatment.Specific","Cluster","Response","Antibiotics.visit","Steroids")],
         #color =hmcols, breaks=bk,
         show_rownames = TRUE,
         scale="none")
```


```{r pathway_heatmap_manually_order_samples,cache=TRUE,fig.width=15,fig.height=10}
#### manually re-order the samples
data.heat <- 
  add_rownames(as.data.frame(pathway.cp1.sig),var='Sample') %>%
  left_join(add_rownames(sample.info,var='Sample'),by='Sample') %>% 
  arrange(Disease,Cluster,Response,Treatment.Specific,Antibiotics.visit,FCP) %>%
  dplyr::select(1:(ncol(pathway.cp1.sig)+1)) %>% ##first column: sample
  add_back_rownames(row.var='Sample')

dat.tmp <- log(data.heat+min(data.heat[data.heat>0]))
dat.nor <- dat.tmp
for (pw in colnames(dat.tmp)){
  pw.dat <- dat.tmp[,pw,drop=TRUE]
  #hist(pw.dat,breaks=100)
  Me <- median(pw.dat)
  Ma <- mad(pw.dat)
  if (Ma==0){Me <- mean(pw.dat);Ma <- sd(pw.dat)}
  pw.dat <- (pw.dat-Me)/Ma
  pw.dat[pw.dat < -3] <- -3
  pw.dat[pw.dat > 3] <- 3
  dat.nor[,pw] <- pw.dat
}

bk <- unique(c(
        seq(-3.5,-2,length=100),
        seq(-2,-1.5,length=100),
        seq(-1.5,0,length=100),
        seq(0,1.5,length=100),
        seq(1.5,2,length=100),
        seq(2,3.5,length=100)
        ))
hmcols<- colorRampPalette(c('#d73027','#fc8d59','#fee090',
                            '#ffffbf','#e0f3f8','#91bfdb','#4575b4'
                            ))(length(bk)-1)

pheatmap(t(dat.nor),border_color="grey60",
         cluster_cols = FALSE,
         ##### the pathway abundances are very small
         ##### very difficult to visualize the difference without scaling
         scale = "none",
         #####
         clustering_distance_rows = 'euclidean',
         #clustering_distance_cols = pathway.dist.cp1.sig,
         clustering_method = "average",
         annotation=sample.info[,c("Disease","Cluster","Response","Treatment.Specific","FCP","Antibiotics.visit","Steroids")],
         #color =hmcols, breaks=bk,
         show_rownames = TRUE
         )


rm(data.heat)
```

## Antibiotics use

### Wilcoxon rank test 
```{r pathway_rank_test_of_antibiotics_use,cache=TRUE}
### plot top ones
pathway.antibio.wil <- wilcox_test(pathway.cp1,
                    subset(sample.info,Group=='PLEASE',select=Antibiotics.visit),
                             plot.heatmap=FALSE,boxplot.top=10,boxplot.y='Abundance(%)')      
###
write.csv(pathway.antibio.wil,file=paste('../3_Result/2015_03_05_Pathway_PLEASE_Antibiotics_Use_Wilcox.csv',sep=''),row.names=TRUE)

```

### Random Forest 
```{r pathway_random_forest_of_antibiotics_use,cache=TRUE}
random_forest(pathway.cp1,
              subset(sample.info,Group=='PLEASE',select=Antibiotics.visit),
              plot.MDS=TRUE,plot.importance=TRUE,print.result=FALSE,
              guess.num=50,boxplot.top=6,
              title="Antibiotics use",boxplot.y="Abundance(%)",
              seed=100)
```

## Cluster 1 vs. cluster 2

### Wilcoxon rank test 
```{r pathway_rank_test_of_cluster_1_vs_2,cache=TRUE}
### plot all the significant ones
pathway.c1c2.wil <- wilcox_test(pathway.cp1,
                                subset(sample.info,Group=='PLEASE',select=Cluster),
                                plot.heatmap=FALSE,boxplot.top=0.05,boxplot.y='Abundance(%)')       
###
write.csv(pathway.c1c2.wil,file=paste('../3_Result/2015_03_05_Pathway_PLEASE_Cluster_1_vs_2_Wilcox.csv',sep=''),row.names=TRUE)
```

### Random Forest 
```{r pathway_random_forest_of_cluster_1_vs_2,cache=TRUE}
random_forest(pathway.cp1,subset(sample.info,Group=='PLEASE',select=Cluster),
              plot.MDS=TRUE,plot.importance=TRUE,print.result=FALSE,
              guess.num=50,boxplot.top=5,
              title="Normal vs. Crohn",boxplot.y="Abundance(%)",
              seed=100)
```


## Responders vs. non-responders

### Random Forest 
```{r pathway_random_forest_of_r_vs_nr,cache=TRUE,fig.width=10}
### EEN T1
random_forest(pathway.data,
              subset(sample.info,Time==1 & Treatment.Specific=='EEN' & !is.na(Response),
                     select=Response),
              plot.MDS=TRUE,plot.importance=TRUE,print.result=FALSE,
              guess.num=50,boxplot.top=5,
              title="Response vs. non-response",boxplot.y="Abundance(%)",
              seed=100)
### EEN T4
random_forest(pathway.data,
              subset(sample.info,Time==4 & Treatment.Specific=='EEN' & !is.na(Response),
                     select=Response),
              plot.MDS=TRUE,plot.importance=TRUE,print.result=FALSE,
              guess.num=50,boxplot.top=5,
              title="Response vs. non-response",boxplot.y="Abundance(%)",
              seed=100)


### antiTNF T1
random_forest(pathway.data,
              subset(sample.info,Time==1 & Treatment.Specific=='antiTNF' & !is.na(Response),
                     select=Response),
              plot.MDS=TRUE,plot.importance=TRUE,print.result=FALSE,
              guess.num=50,boxplot.top=5,
              title="Response vs. non-response",boxplot.y="Abundance(%)",
              seed=100)

### antiTNF T4
random_forest(pathway.data,
              subset(sample.info,Time==4 & Treatment.Specific=='antiTNF' & !is.na(Response),
                     select=Response),
              plot.MDS=TRUE,plot.importance=TRUE,print.result=FALSE,
              guess.num=50,boxplot.top=5,
              title="Response vs. non-response",boxplot.y="Abundance(%)",
              seed=100)
```

### Wilcoxon rank test
```{r pathway_rank_test_of_r_vs_nr,cache=TRUE,fig.width=10}
### EEN T1
bact.EEN.T1 <- wilcox_test(pathway.data,
                           subset(sample.info,Time==1 & Treatment.Specific=='EEN' & !is.na(Response),
                                  select=Response),
                           plot.heatmap=FALSE,boxplot.top=5,boxplot.y='Abundance(%)')

### EEN T4
bact.EEN.T4 <- wilcox_test(pathway.data,
                           subset(sample.info,Time==4 & Treatment.Specific=='EEN' & !is.na(Response),
                                  select=Response),
                           plot.heatmap=FALSE,boxplot.top=5,boxplot.y='Abundance(%)')

```


## T1 vs. T2 in EEN
```{r pathway_T1_vs_T2_in_EEN,cache=TRUE,fig.width=10}
add_rownames(as.data.frame(pathway.data),var='Sample') %>%
  gather(Pathway,Abu,-Sample) %>%
  left_join(dplyr::select(add_rownames(sample.info,var='Sample'),Subject,Sample,Time,Treatment.Specific,Response), by='Sample') %>%  
  filter(!is.na(Time)) %>% 
  mutate(Time=paste('T',Time,sep='')) %>%
  dplyr::select(-Sample) %>% 
  spread(Time,Abu) %>%
  group_by(Treatment.Specific,Pathway) %>%
  summarise_each(funs(wilcox.test(., T1, pair = TRUE)$p.value), vars = T2:T4) %>%
  #mutate_each(funs(p.adjust(.,method='fdr')),vars1:vars3) %>%
  ## raw pvalues
  mutate_each(funs(signif(.,digits=2)),vars1:vars3) %>%
  rename(T1.vs.T2=vars1,T1.vs.T3=vars2,T1.vs.T4=vars3) %>%
  arrange(Treatment.Specific,T1.vs.T2)


```


## Treatment, antibiotics use, FCP on pathway abundance
### linear regression
```{r pathway_linear_regression_treatment_antibiotics_fcp,cache=TRUE}
## T4
linear.res <- list()
for (pw in colnames(pathway.data)){
  pathway.tem <- pathway.data[,pw,drop=FALSE]
  colnames(pathway.tem) <- 'Pathway'
  linear.res[[pw]] <- add_rownames(as.data.frame(pathway.tem),var='Sample') %>%
    left_join(add_rownames(sample.info,var='Sample'),by='Sample') %>%
    dplyr::filter(Type=='PLEASE-T4') %>%
    dplyr::select(Pathway,Treatment.Specific,Antibiotics.visit,log.FCP) %>%
    mutate(Pathway.log = log(Pathway+min(Pathway[Pathway>0])/2)) %>%
    lm( Pathway.log ~ Treatment.Specific + Antibiotics.visit + log.FCP,
        data=.) %>%
    summary %>% coef  %>%  
    as.data.frame %>% add_rownames(var='Var') %>%
    dplyr::select(1,2,5) %>% dplyr::select(Var,Est=Estimate,Pval=starts_with("Pr"))%>%
    gather(value,var,Est,Pval) %>% 
    unite(Variables,Var,value, sep = ".", remove = TRUE) %>%
    add_back_rownames(row.var='Variables') %>% as.matrix
  ####
  colnames(linear.res[[pw]]) <- pw
  ###  
  rm(pathway.tem)
}

linear.res <- signif(t(as.data.frame(linear.res)),2)

#print(linear.res)
######
write.csv(linear.res,file=paste('../3_Result/2015_03_05_Pathway_Linear_Regression_Treatment_Antibiotics_FCP.csv',sep=''),row.names=TRUE)
```


### quantile regression
```{r pathway_quantile_regression_treatment_antibiotics_fcp,cache=TRUE}
## T4
rq.res <- list()
for (pw in colnames(pathway.data)){
  #print(pw)
  ### it seems rq used some random numbers
  set.seed(1)
  pathway.tem <- pathway.data[,pw,drop=FALSE]
  colnames(pathway.tem) <- 'Pathway'
  rq.res[[pw]] <- add_rownames(as.data.frame(pathway.tem),var='Sample') %>%
    left_join(add_rownames(sample.info,var='Sample'),by='Sample') %>%
    dplyr::filter(Type=='PLEASE-T4') %>%
    dplyr::select(Pathway,Treatment.Specific,Antibiotics.visit,log.FCP) %>%
    ### log transform the data
    mutate(Pathway.log = log(Pathway+min(Pathway[Pathway>0])/2)) %>%
    ### !!!!!!!!!! ########
    ### After updating RStudio and some R packages,
    ### a problem has occured: the reference level in the linear model
    ### has changed for no reason
    ### Change the reference level back to EEN for consistency with
    ### previous analysis
    ###
    mutate( Treatment.Specific = relevel(as.factor(Treatment.Specific), ref = 'EEN')) %>%
    ######################
    rq( Pathway.log ~ Treatment.Specific + Antibiotics.visit + log.FCP,
        tau=0.5,data=.,na.action=na.omit) %>%
    ### since using quantile regression here
    ### no need to log transform the data
    ### freeze at some pathway
    #rq( Pathway ~ Treatment.Specific + Antibiotics.visit + log.FCP,
    #    tau=0.5,data=.) %>%
    summary.rq(.,se="boot") %>% coef %>%
    as.data.frame %>% add_rownames(var='Var') %>%
    dplyr::select(1,2,5) %>% dplyr::select(Var,Est=Value,Pval=starts_with("Pr"))%>%
    gather(value,var,Est,Pval) %>% 
    unite(Variables,Var,value, sep = ".", remove = TRUE) %>%
    add_back_rownames(row.var='Variables') %>% as.matrix
  #### add pathway name
  colnames(rq.res[[pw]]) <- pw
  ###  
  rm(pathway.tem)
  #break
}

rq.res.table <- 
  as.data.frame(t(as.data.frame(rq.res))) %>% ## must use as.data.frame again 
  add_rownames(var='pathway') %>%
  mutate(
    Treatment.SpecificPEN.Qval = p.adjust(Treatment.SpecificPEN.Pval,method='fdr'),
    Treatment.SpecificantiTNF.Qval = p.adjust(Treatment.SpecificantiTNF.Pval,method='fdr'),
    Antibiotics.visitUse.Qval = p.adjust(Antibiotics.visitUse.Pval,method='fdr'),
    log.FCP.Qval = p.adjust(log.FCP.Pval,method='fdr')
  ) 

rm(rq.res)
######
#### row name is a column
write.csv(rq.res.table,file=paste('../3_Result/2015_03_05_Pathway_Quantile_Median_Regression_Treatment_Antibiotics_FCP.csv',sep=''),row.names=FALSE)

```

### quantile regression on fold change
```{r pathway_fc_quantile_regression_treatment_antibiotics_fcp,cache=TRUE}
## fold change T4/T1
rq.fc.res <- list()
for (pw in colnames(pathway.data)){
  #print(pw)
  ### it seems rq used some random numbers
  set.seed(1)
  pathway.tem <- pathway.data[,pw,drop=FALSE]
  colnames(pathway.tem) <- 'Pathway'
    ####
    master.table <- add_rownames(as.data.frame(pathway.tem),var='Sample') %>%
    left_join(add_rownames(sample.info,var='Sample'),by='Sample') %>%
    dplyr::filter(Time==1 | Time==4) %>%
    dplyr::select(Pathway,Subject,Time,Treatment.Specific,Antibiotics.visit,log.FCP,Response) %>%
    mutate(Time=ifelse(Time==1,'T1','T4'))
    ### pathway
    path.temp <- master.table %>%
    mutate(Pathway = Pathway+min(Pathway[Pathway>0]))  %>% 
    dplyr::select(Subject,Pathway,Time,Treatment.Specific) %>%
    spread(Time,Pathway) %>%
    mutate(Path.log.fc = log(T4/T1)) %>%
    dplyr::select(Subject,Treatment.Specific,Path.log.fc)
    ### response
    response.temp <- master.table %>%
    dplyr::select(Subject,Response) 
    ### log FCP
    log.fcp.temp <- master.table %>%
    dplyr::select(Subject,Time,log.FCP) %>%
    spread(Time,log.FCP) %>%
    mutate(log.FCP.T4 = T4) %>%
    dplyr::select(Subject,log.FCP.T4)
    ### antibiotics
    antibiotics.temp <- master.table %>%
    dplyr::select(Subject,Time,Antibiotics.visit) %>%
    spread(Time,Antibiotics.visit) %>%
    mutate(antibiotics = ifelse(is.na(T4),T1,T4)) %>%
    dplyr::select(Subject,antibiotics)
    ### merge together 
    rq.fc.res[[pw]] <- path.temp %>% full_join(antibiotics.temp,by='Subject') %>%
    full_join(response.temp,by='Subject') %>%
    ### !!!!!!!!!! ########
    ### After updating RStudio and some R packages,
    ### a problem has occured: the reference level in the linear model
    ### has changed for no reason
    ### Change the reference level back to EEN for consistency with
    ### previous analysis
    ###
    mutate( Treatment.Specific = relevel(as.factor(Treatment.Specific), ref = 'EEN')) %>%
    ######################
    ### quantile regression
    rq( Path.log.fc ~ Treatment.Specific + antibiotics + Response,
        tau=0.75,data=.,na.action=na.omit) %>%
    summary.rq(.,se="boot") %>% coef %>%
    as.data.frame %>% add_rownames(var='Var') %>%
    dplyr::select(1,2,5) %>% dplyr::select(Var,Est=Value,Pval=starts_with("Pr"))%>%
    gather(value,var,Est,Pval) %>% 
    unite(Variables,Var,value, sep = ".", remove = TRUE) %>%
    add_back_rownames(row.var='Variables') %>% as.matrix
  #### add pathway name
  colnames(rq.fc.res[[pw]]) <- pw
  ###  
  rm(pathway.tem)
  rm(master.table)
  rm(path.temp)
  rm(antibiotics.temp)
  rm(log.fcp.temp)
  rm(response.temp)
  #break
}

rq.res.table <- 
  as.data.frame(t(as.data.frame(rq.fc.res))) %>% ## must use as.data.frame again 
  add_rownames(var='pathway') %>%
  mutate(
    Treatment.SpecificPEN.Qval = p.adjust(Treatment.SpecificPEN.Pval,method='fdr'),
    Treatment.SpecificantiTNF.Qval = p.adjust(Treatment.SpecificantiTNF.Pval,method='fdr'),
    Antibiotics.visitUse.Qval = p.adjust(antibioticsUse.Pval,method='fdr'),
    ResponseResponse.Qval = p.adjust(ResponseResponse.Pval ,method='fdr')
  ) 

rm(rq.fc.res)
######
#### row name is a column
write.csv(rq.res.table,file=paste('../3_Result/2015_03_05_Pathway_Fold_Change_Quantile_Q75_Regression_Treatment_Antibiotics_Response.csv',sep=''),row.names=FALSE)

```

## Correlation of pathways and bacteria
```{r pathway_correlation_with_bacteria,cache=TRUE,fig.height=18,fig.width=15}

plot.correlation.heatmap = function(cor.matrix,
                                    row.filter=0.5,
                                    col.filter=0.1
                                    ){
  cor.matrix <- cor.matrix[
  apply(cor.matrix,1,function(X){max(abs(X),na.rm=TRUE)>row.filter}),
  apply(cor.matrix,2,function(X){max(abs(X),na.rm=TRUE)>col.filter})
  ]
  pheatmap(cor.matrix, 
           scale='none',fontsize_row = 10, 
           clustering_distance_cols = "manhattan",
           clustering_distance_rows = "manhattan",
           cluster_cols = TRUE,
           cellwidth = 10, cellheight = 10
  )                              
}


### all
ind <- intersect(rownames(pathway.data),rownames(taxa.data))
pathway.bact.cor <- cor(pathway.data[ind,],taxa.data[ind,],method='spearman')
plot.correlation.heatmap(pathway.bact.cor)
rm(ind)

### control
ind <- intersect(intersect(rownames(pathway.data),rownames(taxa.data)),rownames(subset(sample.info,Type=='COMBO')))
pathway.bact.cor <- cor(pathway.data[ind,colSums(pathway.data[ind,])>0],
                        taxa.data[ind,colSums(taxa.data[ind,])>0],
                        method='spearman')
plot.correlation.heatmap(pathway.bact.cor)
rm(ind)

### PLEASE-T1
ind <- intersect(intersect(rownames(pathway.data),rownames(taxa.data)),rownames(subset(sample.info,Type=='PLEASE-T1')))
pathway.bact.cor <- cor(pathway.data[ind,colSums(pathway.data[ind,])>0],
                        taxa.data[ind,colSums(taxa.data[ind,])>0],
                        method='spearman')
plot.correlation.heatmap(pathway.bact.cor)
rm(ind)

### PLEASE-T4
ind <- intersect(intersect(rownames(pathway.data),rownames(taxa.data)),rownames(subset(sample.info,Type=='PLEASE-T4')))
pathway.bact.cor <- cor(pathway.data[ind,colSums(pathway.data[ind,])>0],
                        taxa.data[ind,colSums(taxa.data[ind,])>0],
                        method='spearman')
plot.correlation.heatmap(pathway.bact.cor)
rm(ind)

### antiTNF
ind <- intersect(intersect(rownames(pathway.data),rownames(taxa.data)),rownames(subset(sample.info,Treatment=='antiTNF')))
pathway.bact.cor <- cor(pathway.data[ind,],taxa.data[ind,],method='spearman')
plot.correlation.heatmap(pathway.bact.cor)
rm(ind)

### EEN
ind <- intersect(intersect(rownames(pathway.data),rownames(taxa.data)),rownames(subset(sample.info,Treatment.Specific=='EEN')))
pathway.bact.cor <- cor(pathway.data[ind,colSums(pathway.data[ind,])>0],
                        taxa.data[ind,colSums(taxa.data[ind,])>0],
                        method='spearman')
plot.correlation.heatmap(pathway.bact.cor)
rm(ind)

### response
ind <- intersect(intersect(rownames(pathway.data),rownames(taxa.data)),rownames(subset(sample.info,Response=='Response')))
pathway.bact.cor <- cor(pathway.data[ind,],taxa.data[ind,],method='spearman')
plot.correlation.heatmap(pathway.bact.cor)
rm(ind)

### non-response
ind <- intersect(intersect(rownames(pathway.data),rownames(taxa.data)),rownames(subset(sample.info,Response=='Non.Response')))
pathway.bact.cor <- cor(pathway.data[ind,],taxa.data[ind,],method='spearman')
plot.correlation.heatmap(pathway.bact.cor)
rm(ind)


### PLEASE-T4 Response
ind <- intersect(intersect(rownames(pathway.data),rownames(taxa.data)),rownames(subset(sample.info,Type=='PLEASE-T4' & Response =='Response')))
pathway.bact.cor <- cor(pathway.data[ind,colSums(pathway.data[ind,])>0],
                        taxa.data[ind,colSums(taxa.data[ind,])>0],
                        method='spearman')
plot.correlation.heatmap(pathway.bact.cor)
rm(ind)
```

## Network
```{r network_merge_all_data,cache=TRUE,eval=FALSE}
network.data <- add_rownames(sample.info,var='Sample') %>%
dplyr::select(Sample,Type,Antibiotics.visit) %>%
left_join(add_rownames(taxa.data,var='Sample'),by='Sample') %>%
left_join(add_rownames(fungi.RPKM,var='Sample'),by='Sample') %>%
left_join(add_rownames(as.data.frame(pathway.data),var='Sample'),by='Sample') %>% 
left_join(add_rownames(as.data.frame(t(mdata.nor)),var='Sample'),by='Sample') 

### Control samples
control.cor <- network.data %>% dplyr::filter(Type=='COMBO') %>%
dplyr::select(-Sample,-Type,-Antibiotics.visit) %>%
#dplyr::select(1:3) %>%
cor(,method='spearman',use='pairwise.complete.obs') %>%
(function(tdata){
tdata[lower.tri(tdata,diag=TRUE)] <- NA
return(tdata)
}) %>%
as.data.frame %>%
add_rownames(var='var1') %>%
gather(var2,correlation,-1)

## please-T1 no antibiotics use
pleaset1.no.anti.cor <- network.data %>% dplyr::filter(Type=='PLEASE-T1' & Antibiotics.visit=='Not.Use') %>%
dplyr::select(-Sample,-Type,-Antibiotics.visit) %>%
#dplyr::select(1:3) %>%
cor(,method='spearman',use='pairwise.complete.obs') %>%
(function(tdata){
tdata[lower.tri(tdata,diag=TRUE)] <- NA
return(tdata)
}) %>%
as.data.frame %>%
add_rownames(var='var1') %>%
gather(var2,correlation,-1)

bk <- unique(c(seq(-1,-0.4,length=10),seq(-0.4,0.4,length=50),seq(0.4,1,length=10)))
hmcols<- colorRampPalette(c('#4575b4','#91bfdb','#e0f3f8',
                            '#ffffbf','#fee090','#fc8d59','#d73027'
                            ))(length(bk)-1)
data.frame(ctrl=control.cor,please=pleaset1.no.anti.cor) %>%
mutate(cor.diff=abs(ctrl.correlation-please.correlation)) %>%
dplyr::filter(cor.diff>1) %>%
mutate(name=paste(ctrl.var1,ctrl.var2,sep='--')) %>%
dplyr::select(name,ctrl.correlation,please.correlation) %>%
na.omit %>%
add_back_rownames(row.var='name') %>%
pheatmap(., 
           scale='none',fontsize_row = 10, 
           clustering_distance_cols = "manhattan",
           clustering_distance_rows = "manhattan",
           cluster_cols = TRUE,
           color =hmcols, breaks=bk,
           cellwidth = 12, cellheight = 8
           )


######
write.csv(control.cor.sig,file=paste('../3_Result/2015_03_05_Network_Correlation_Sig_COMBO.csv',sep=''),row.names=TRUE)
```

# Analyze UCLA 84 metabolite data


```{r m84metabolite_check_the_data,cache=TRUE}
##############
# From Gary:
# I just got off the phone with Jon Braun at UCLA.  Jon directs the CCFA Microbiome Initiative and is very involved with both our PLEASE metabolomic study and FARMM.  His group has identified 26 small molecules, imputed from microbiome sequencing data, that influence immune function in vitro.  He would like to know if any of these small molecules might have interesting associations in our PLEASE metabolomic dataset.  He would like to send us the name of these molecules and their KEGG annotations.  I was thinking that we could do a quick analysis to see if these metabolites are differentially expressed in the following fashion:
# 
# 1) T1 vs. T4
# 2) EEN vs. anti-TNF
# 3) Response vs. Nonresponse

# Gary: these are bacterial metabolites
#############################

### How many metabolites in total
print(colnames(m84data.nor))
```


## Normal vs. Crohn-T1
```{r m84metabolites_normal_vs_crohn_t1,cache=TRUE}
###################### Normal and disease T1 #####################
m84data.cp1 <- subset_data_by_sample_info(m84data.nor,sample.info,
                                       Type=='COMBO' | Type=='PLEASE-T1')
dim(m84data.cp1)
```

### Calculate distance
```{r m84metabolites_calculate_distance,cache=TRUE}
m84.distance.method <- "Bray"
#### distance for taxa
dist.m84.cp1 <- calculate_distance(X = m84data.cp1, method = m84.distance.method)

### distribution
hist(dist.m84.cp1,xlab='Distance',main='Metabolites (COMBO and PLEASE-T1)',breaks=20)

```

### Find the best clustering
```{r m84metabolites_find_the_best_clustering,cache=TRUE}
m84.pamk.best <- find_best_clustering(dist.m84.cp1,plot.results=TRUE)
```

### MDS analysis
```{r m84metabolites_MDS,cache=TRUE,results='hide'}
### MDS
set.seed(10)
m84.meta.MDS <- metaMDS(dist.m84.cp1)

### Normal/Disease
p <- add_rownames(as.data.frame(m84.meta.MDS$points),var='Sample') %>%
  left_join(dplyr::select(add_rownames(sample.info,var='Sample'),
                          Sample,Disease),
            by='Sample') %>% 
  ggplot(aes(x=MDS1,y=MDS2))+ geom_point(aes(colour = Disease)) +
  scale_color_manual(values=c("#0066CC", "#FF3333"),name="") +
  labs(title = 'metaMDS') 
print(p)

### Normal/Disease, cluster
### These are bacterial clusters !!
p <- add_rownames(as.data.frame(m84.meta.MDS$points),var='Sample') %>%
  left_join(dplyr::select(add_rownames(sample.info,var='Sample'),
                          Sample,Disease,Cluster),
            by='Sample') %>% 
  ggplot(aes(x=MDS1,y=MDS2))+ geom_point(aes(colour = Disease,shape=Cluster)) +
  scale_color_manual(values=c("#0066CC", "#FF3333"),name="") +
  scale_shape_manual(values=c(20,4),name="") +
  labs(title = 'metaMDS') 
print(p)



#### Fugal percentage 
p <- add_rownames(as.data.frame(m84.meta.MDS$points),var='Sample') %>%
  left_join(dplyr::select(add_rownames(sample.info,var='Sample'),
                          Sample,Disease,Cluster,Fungi.Per),
            by='Sample') %>% 
  mutate(Fungi.Per.sqrt=sqrt(Fungi.Per)) %>%
  ggplot(aes(x=MDS1,y=MDS2))+ 
  geom_point(aes(colour = Disease,shape=Cluster,size = Fungi.Per.sqrt)) +
  scale_color_manual(values=c("#0066CC", "#FF3333"),name="") +
  scale_shape_manual(values=c(20,1),name="") +
  scale_size_area()+
  labs(title = 'metaMDS') 
print(p)

#### Human DNA percentage
p <- add_rownames(as.data.frame(m84.meta.MDS$points),var='Sample') %>%
  left_join(dplyr::select(add_rownames(sample.info,var='Sample'),
                          Sample,Disease,Cluster,Human.Per),
            by='Sample') %>% 
  ggplot(aes(x=MDS1,y=MDS2))+ 
  geom_point(aes(colour = Disease,shape=Cluster,size = Human.Per)) +
  scale_color_manual(values=c("#0066CC", "#FF3333"),name="") +
  scale_shape_manual(values=c(20,1),name="") +
  scale_size_area()+
  labs(title = 'metaMDS') 
print(p)
```


### Random Forest 
```{r m84metabolites_random_forest_of_normal_vs_crohn,cache=TRUE}
### (3) for the random forest Importance score plots in Figure 2C, we have a list of pathways. For this list, can you get results on which pathways have high abundances in Crohn's disease individual based on medians in normal and Crohn's children?

random_forest(m84data.cp1,sample.info[,'Disease',drop=FALSE],
              plot.MDS=TRUE,plot.importance=TRUE,print.result=FALSE,
              guess.num=50,boxplot.top=10,
              title="Normal vs. Crohn",boxplot.y="Abundance",
              seed=100)

```

### Wilcoxon rank test of normal vs. normal
```{r m84metabolites_rank_test_of_normal_vs_normal,cache=TRUE}
### "Control" vs. "control"
sam.tmp <- subset(sample.info,Disease=='Control',select=Disease)
sam.tmp[1:13,1] <- 'control' 
###  
m84.cc.wil <- wilcox_test(m84data.cp1,
                           sam.tmp,                       plot.heatmap=FALSE,boxplot.top=0.05,boxplot.y='Abundance')

cat("signifcant metabolites",sum(m84.cc.wil[,2]<0.05))
###
#write.csv(pathway.cp1.wil,file=paste('../3_Result/2015_03_05_Pathway_COMBO_PLEASET1_Wilcox.csv',sep=''),row.names=TRUE)
rm(sam.tmp)
```


### Wilcoxon rank test of normal vs. Crohn
```{r m84metabolites_rank_test_of_normal_vs_crohn,cache=TRUE}
#######################################################
### it is well known that the inflamed gastrointestinal mucosa of patients with IBD has a massive infiltration of Th17 cells and that Th17-related cytokines are produced in excess in both CD and UC tissues 
### In addition to IL-17A and IL-17F, various studies have reported an increased production of other Th17 cytokines in the inflamed gut of IBD patients, such as IL-22 and IL-26 [48], or IL-21 and IL-23, which in turn exacerbate inflammation by promoting Th1 and Th17 responses
########################################################
### only plot the top 5 significant ones
m84.cp1.wil <- wilcox_test(m84data.cp1,sample.info[,'Disease',drop=FALSE],
              plot.heatmap=FALSE,boxplot.top=0.05,boxplot.y='Abundance')
cat("signifcant metabolites",sum(m84.cp1.wil[,2]<0.05))
### add Th17 column and write into file
add_rownames(as.data.frame(m84.cp1.wil),var='ID') %>% 
left_join(dplyr::select(add_rownames(minfo,var='ID'),ID,CO),by='ID') %>%
left_join(dplyr::select(m84id,KEGGNumber,EffectOnTh17),by=c("CO"="KEGGNumber")) %>%
dplyr::select(-CO) %>%
write.csv(file=paste('../3_Result/2015_03_05_Metabolites84_COMBO_PLEASET1_Wilcox.csv',sep=''),row.names=FALSE)
```

### Wilcoxon rank test of normal vs. Cluster 1
```{r m84metabolites_rank_test_of_normal_vs_cluster_1,cache=TRUE}
m84.normal.cluster1.wil <- wilcox_test(m84data.cp1,
       subset(sample.info,Disease=='Control'|(Disease=='Crohn'&Cluster=='cluster 1'),select=Disease),
                               plot.heatmap=FALSE,boxplot.top=0.05,boxplot.y='Abundance')

cat("signifcant metabolites",sum(m84.normal.cluster1.wil[,2]<0.05))
###
### add Th17 column and write into file
add_rownames(as.data.frame(m84.normal.cluster1.wil),var='ID') %>% 
left_join(dplyr::select(add_rownames(minfo,var='ID'),ID,CO),by='ID') %>%
left_join(dplyr::select(m84id,KEGGNumber,EffectOnTh17),by=c("CO"="KEGGNumber")) %>%
dplyr::select(-CO) %>%
write.csv(file=paste('../3_Result/2015_03_05_Metabolites84_COMBO_vs_Cluster1_Wilcox.csv',sep=''),row.names=FALSE)
```

### Wilcoxon rank test of normal vs. Cluster 2
```{r m84metabolites_rank_test_of_normal_vs_cluster_2,cache=TRUE}
m84.normal.cluster2.wil <- wilcox_test(m84data.cp1,
      subset(sample.info,Disease=='Control'|(Disease=='Crohn'&Cluster=='cluster 2'),select=Disease),
      plot.heatmap=FALSE,boxplot.top=0.05,boxplot.y='Abundance')

cat("signifcant metabolites",sum(m84.normal.cluster2.wil[,2]<0.05))
###
### add Th17 column and write into file
add_rownames(as.data.frame(m84.normal.cluster2.wil),var='ID') %>% 
left_join(dplyr::select(add_rownames(minfo,var='ID'),ID,CO),by='ID') %>%
left_join(dplyr::select(m84id,KEGGNumber,EffectOnTh17),by=c("CO"="KEGGNumber")) %>%
dplyr::select(-CO) %>%
write.csv(file=paste('../3_Result/2015_03_05_Metabolites84_COMBO_vs_Cluster2_Wilcox.csv',sep=''),row.names=FALSE)
```


### Wilcoxon rank test of EEN vs. antiTNF (T1)
```{r m84metabolites_rank_test_of_een_vs_antitnf_t1,cache=TRUE}
m84.een.antitnf.t1.wil <- wilcox_test(m84data.nor,
      subset(sample.info,Treatment.Specific!='PEN' & Time==1,select=Treatment.Specific),
      plot.heatmap=FALSE,boxplot.top=0.05,boxplot.y='Abundance')

cat("signifcant metabolites",sum(m84.een.antitnf.t1.wil[,2]<0.05))
###
### add Th17 column and write into file
add_rownames(as.data.frame(m84.een.antitnf.t1.wil),var='ID') %>% 
left_join(dplyr::select(add_rownames(minfo,var='ID'),ID,CO),by='ID') %>%
left_join(dplyr::select(m84id,KEGGNumber,EffectOnTh17),by=c("CO"="KEGGNumber")) %>%
dplyr::select(-CO) %>%
write.csv(file=paste('../3_Result/2015_03_05_Metabolites84_EEN_vs_antiTNF_T1_Wilcox.csv',sep=''),row.names=FALSE)
```


### Wilcoxon rank test of EEN vs. antiTNF (T4)
```{r m84metabolites_rank_test_of_een_vs_antitnf_t4,cache=TRUE}
m84.een.antitnf.t4.wil <- wilcox_test(m84data.nor,
      subset(sample.info,Treatment.Specific!='PEN' & Time==4,select=Treatment.Specific),
      plot.heatmap=FALSE,boxplot.top=0.05,boxplot.y='Abundance')

cat("signifcant metabolites",sum(m84.een.antitnf.t4.wil[,2]<0.05))
###
### add Th17 column and write into file
add_rownames(as.data.frame(m84.een.antitnf.t4.wil),var='ID') %>% 
left_join(dplyr::select(add_rownames(minfo,var='ID'),ID,CO),by='ID') %>%
left_join(dplyr::select(m84id,KEGGNumber,EffectOnTh17),by=c("CO"="KEGGNumber")) %>%
dplyr::select(-CO) %>%
write.csv(file=paste('../3_Result/2015_03_05_Metabolites84_EEN_vs_antiTNF_T4_Wilcox.csv',sep=''),row.names=FALSE)
```


### Wilcoxon rank test of Response vs. non-Response (T1)
```{r m84metabolites_rank_test_of_r_vs_nr_t1,cache=TRUE}
m84.r.nr.t1.wil <- wilcox_test(m84data.nor,
      subset(sample.info, !is.na(Response) & Time==1, select=Response),
      plot.heatmap=FALSE,boxplot.top=0.05,boxplot.y='Abundance')

cat("signifcant metabolites",sum(m84.r.nr.t1.wil[,2]<0.05))
###
### add Th17 column and write into file
add_rownames(as.data.frame(m84.r.nr.t1.wil),var='ID') %>% 
left_join(dplyr::select(add_rownames(minfo,var='ID'),ID,CO),by='ID') %>%
left_join(dplyr::select(m84id,KEGGNumber,EffectOnTh17),by=c("CO"="KEGGNumber")) %>%
dplyr::select(-CO) %>%
write.csv(file=paste('../3_Result/2015_03_05_Metabolites84_Response_vs_NonResponse_T1_Wilcox.csv',sep=''),row.names=FALSE)
```


### Wilcoxon rank test of Response vs. non-Response (T4)
```{r m84metabolites_rank_test_of_r_vs_nr_t4,cache=TRUE}
m84.r.nr.t4.wil <- wilcox_test(m84data.nor,
      subset(sample.info, !is.na(Response) & Time==4, select=Response),
      plot.heatmap=FALSE,boxplot.top=0.05,boxplot.y='Abundance')

cat("signifcant metabolites",sum(m84.r.nr.t4.wil[,2]<0.05))
###
### add Th17 column and write into file
add_rownames(as.data.frame(m84.r.nr.t4.wil),var='ID') %>% 
left_join(dplyr::select(add_rownames(minfo,var='ID'),ID,CO),by='ID') %>%
left_join(dplyr::select(m84id,KEGGNumber,EffectOnTh17),by=c("CO"="KEGGNumber")) %>%
dplyr::select(-CO) %>%
write.csv(file=paste('../3_Result/2015_03_05_Metabolites84_Response_vs_NonResponse_T4_Wilcox.csv',sep=''),row.names=FALSE)
```



### Heatmap of normal vs. Crohn's disease (T1)
```{r m84metabolites_heatmap_cp1,cache=TRUE,fig.width=15,fig.height=10}

#####
dat.nor <- m84data.cp1
for (pw in colnames(m84data.cp1)){
  pw.dat <- m84data.cp1[,pw,drop=TRUE]
  Me <- median(pw.dat)
  Ma <- mad(pw.dat)
  if (Ma==0){Me <- mean(pw.dat);Ma <- sd(pw.dat)}
  pw.dat <- (pw.dat-Me)/Ma
  pw.dat[pw.dat < -3] <- -3
  pw.dat[pw.dat > 3] <- 3
  dat.nor[,pw] <- pw.dat
}

pheatmap(t(dat.nor),
         border_color="grey60",
         cluster_cols = calculate_distance(X = m84data.cp1, method = "Bray"),
         clustering_distance_rows = 'euclidean',#

         #clustering_method = "complete",
         annotation=sample.info[,c("Disease","Time","Response","Treatment.Specific","Cluster","FCP","Antibiotics.visit","Steroids")],
         #color =hmcols, breaks=bk,
         show_rownames = TRUE
         #scale="row",
         #gaps_col = c(26, 111,191,273)
         )

#rm(data.heat)
rm(pw)
rm(dat.nor)
rm(pw.dat)
rm(Me)
rm(Ma)
```



```{r m84metabolites_heatmap_all_time,cache=TRUE,eval=FALSE,fig.width=15,fig.height=10}
## Heatmap of normal vs. Crohn's disease (all time points)
##### hard to see something
##### NOT RUN
#####
dat.nor <- m84data.nor
for (pw in colnames(m84data.nor)){
  pw.dat <- m84data.nor[,pw,drop=TRUE]
  Me <- median(pw.dat)
  Ma <- mad(pw.dat)
  if (Ma==0){Me <- mean(pw.dat);Ma <- sd(pw.dat)}
  pw.dat <- (pw.dat-Me)/Ma
  pw.dat[pw.dat < -3] <- -3
  pw.dat[pw.dat > 3] <- 3
  dat.nor[,pw] <- pw.dat
}

pheatmap(t(dat.nor),
         border_color="grey60",
         cluster_cols = calculate_distance(X = m84data.nor, method = "Bray") ,
         clustering_distance_rows = 'euclidean', #c
         #clustering_method = "complete",
         annotation=sample.info[,c("Disease","Time","Response","Treatment.Specific","Cluster","FCP","Antibiotics.visit","Steroids")],
         #color =hmcols, breaks=bk,
         show_rownames = TRUE
         #scale="row",
         #gaps_col = c(26, 111,191,273)
         )

#rm(data.heat)
rm(pw)
rm(dat.nor)
rm(pw.dat)
rm(Me)
rm(Ma)
```

## Correlation between metabolites and FCP
```{r m84metabolites_correlation_with_fcp,cache=TRUE,fig.height=10,fig.width=12,warning=FALSE}

### FCP
for (meta in colnames(m84data.nor)){
  meta.dat <- m84data.nor[,meta,drop=FALSE]
  colnames(meta.dat) <- 'Meta'
  ####
  p <- add_rownames(as.data.frame(meta.dat),var='Sample') %>%
  left_join(add_rownames(sample.info,var='Sample'),by='Sample') %>% 
  dplyr::filter(Time!='3') %>%
  # dplyr::select(Meta,FCP,Time) %>%
  ggplot(aes(x=Meta,y=FCP,colour = Cluster,shape=Cluster))+ stat_smooth(method = "rlm",se=FALSE,fullrange=FALSE, alpha = 0.1) + geom_point() + 
  scale_color_manual(values=c("#0066CC", "#FF3333")) +
  scale_shape_manual(values=c(20,1)) +
  labs(title = meta) +
  facet_grid(Treatment.Specific ~ Time, margins=TRUE) +
  ylab('FCP')+xlab('Metabolites')+ #ylim(c(0,2500))+
  theme_classic()
  plot(p)
}


### By antibiotics
for (meta in colnames(m84data.nor)){
  meta.dat <- m84data.nor[,meta,drop=FALSE]
  colnames(meta.dat) <- 'Meta'
  ####
  p <- add_rownames(as.data.frame(meta.dat),var='Sample') %>%
  left_join(add_rownames(sample.info,var='Sample'),by='Sample') %>% 
  dplyr::filter(Time!='3') %>%
  # dplyr::select(Meta,FCP,Time) %>%
  ggplot(aes(x=Meta,y=FCP,colour = Antibiotics.visit,shape=Antibiotics.visit))+ stat_smooth(method = "rlm",se=FALSE,fullrange=FALSE, alpha = 0.1) + geom_point() + 
  scale_color_manual(values=c("#0066CC", "#FF3333")) +
  scale_shape_manual(values=c(20,1)) +
  labs(title = meta) +
  facet_grid(Treatment.Specific ~ Time, margins=TRUE) +
  ylab('FCP')+xlab('Metabolites')+ #ylim(c(0,2500))+
  theme_classic()
  plot(p)
}

### antibiotics only
for (meta in colnames(m84data.nor)){
  meta.dat <- m84data.nor[,meta,drop=FALSE]
  colnames(meta.dat) <- 'Meta'
  ####
  p <- add_rownames(as.data.frame(meta.dat),var='Sample') %>%
  left_join(add_rownames(sample.info,var='Sample'),by='Sample') %>% 
  dplyr::filter(Time!='3' & Antibiotics.visit=='Not.Use') %>%
  # dplyr::select(Meta,FCP,Time) %>%
  ggplot(aes(x=Meta,y=FCP,colour = Antibiotics.visit))+ stat_smooth(method = "rlm",se=FALSE,fullrange=FALSE, alpha = 0.1) + geom_point() + 
  scale_color_manual(values=c("#0066CC", "#FF3333")) +
  scale_shape_manual(values=c(20,1)) +
  labs(title = meta) +
  facet_grid(Treatment.Specific ~ Time, margins=TRUE) +
  ylab('FCP')+xlab('Metabolites')+ #ylim(c(0,2500))+
  theme_classic()
  plot(p)
}
```

## Correlation of pathways and metabolites
```{r m84metabolite_correlation_with_pathways,cache=TRUE,fig.height=18,fig.width=20}

plot.correlation.heatmap = function(cor.matrix,
                                    row.filter=0.5,
                                    col.filter=0.5
                                    ){
  cor.matrix <- cor.matrix[
  apply(cor.matrix,1,function(X){max(abs(X),na.rm=TRUE)>row.filter}),
  apply(cor.matrix,2,function(X){max(abs(X),na.rm=TRUE)>col.filter})
  ]
  
  bk <- unique(c(seq(-1,-0.4,length=10),
                 seq(-0.4,0.4,length=50),
                 seq(0.4,1,length=10)))
  hmcols<- colorRampPalette(c('#4575b4','#91bfdb','#e0f3f8',
                            '#ffffbf','#fee090','#fc8d59','#d73027'
                            ))(length(bk)-1)

  pheatmap(cor.matrix, 
           scale='none',fontsize_row = 10, 
           clustering_distance_cols = "manhattan",
           clustering_distance_rows = "manhattan",
           cluster_cols = TRUE,
           color =hmcols, breaks=bk,
           cellwidth = 10, cellheight = 10
  )                              
}


### all
ind <- intersect(rownames(pathway.data),rownames(m84data.nor))
pathway.m84.cor <- cor(pathway.data[ind,],
                              m84data.nor[ind,apply(m84data.nor[ind,],2,sd)>0],
                              method='spearman')
plot.correlation.heatmap(pathway.m84.cor,row.filter=0.4,col.filter=0)
rm(ind)
rm(pathway.m84.cor)

### control
ind <- intersect(intersect(rownames(pathway.data),rownames(m84data.nor)),rownames(subset(sample.info,Type=='COMBO')))
pathway.m84.cor <- cor(pathway.data[ind,colSums(pathway.data[ind,])>0],
                        m84data.nor[ind,apply(m84data.nor[ind,],2,sd)>0],
                        method='spearman')
plot.correlation.heatmap(pathway.m84.cor,row.filter=0.4,col.filter=0)
rm(ind)
rm(pathway.m84.cor)


### PLEASE-T1
ind <- intersect(intersect(rownames(pathway.data),rownames(m84data.nor)),rownames(subset(sample.info,Type=='PLEASE-T1')))
pathway.m84.cor <- cor(pathway.data[ind,colSums(pathway.data[ind,])>0],
                        m84data.nor[ind,apply(m84data.nor[ind,],2,sd)>0],
                        method='spearman')
plot.correlation.heatmap(pathway.m84.cor,row.filter=0.5,col.filter=0)
rm(ind)
rm(pathway.m84.cor)


### PLEASE-T4
ind <- intersect(intersect(rownames(pathway.data),rownames(m84data.nor)),rownames(subset(sample.info,Type=='PLEASE-T4')))
pathway.m84.cor <- cor(pathway.data[ind,colSums(pathway.data[ind,])>0],
                        m84data.nor[ind,apply(m84data.nor[ind,],2,sd)>0],
                        method='spearman')
plot.correlation.heatmap(pathway.m84.cor,row.filter=0.5,col.filter=0)
rm(ind)
rm(pathway.m84.cor)


### antiTNF
ind <- intersect(intersect(rownames(pathway.data),rownames(m84data.nor)),rownames(subset(sample.info,Treatment=='antiTNF')))
pathway.m84.cor <- cor(pathway.data[ind,colSums(pathway.data[ind,])>0],
                        m84data.nor[ind,apply(m84data.nor[ind,],2,sd)>0],
                        method='spearman')
plot.correlation.heatmap(pathway.m84.cor,row.filter=0.5,col.filter=0)
rm(ind)
rm(pathway.m84.cor)


### EEN
ind <- intersect(intersect(rownames(pathway.data),rownames(m84data.nor)),rownames(subset(sample.info,Treatment.Specific=='EEN')))
pathway.m84.cor <- cor(pathway.data[ind,colSums(pathway.data[ind,])>0],
                        m84data.nor[ind,apply(m84data.nor[ind,],2,sd)>0],
                        method='spearman')
plot.correlation.heatmap(pathway.m84.cor,row.filter=0.5,col.filter=0)
rm(ind)
rm(pathway.m84.cor)

### response
ind <- intersect(intersect(rownames(pathway.data),rownames(m84data.nor)),rownames(subset(sample.info,Response=='Response')))
pathway.m84.cor <- cor(pathway.data[ind,colSums(pathway.data[ind,])>0],
                        m84data.nor[ind,apply(m84data.nor[ind,],2,sd)>0],
                        method='spearman')
plot.correlation.heatmap(pathway.m84.cor,row.filter=0.5,col.filter=0)
rm(ind)
rm(pathway.m84.cor)


### non-response
ind <- intersect(intersect(rownames(pathway.data),rownames(m84data.nor)),rownames(subset(sample.info,Response=='Non.Response')))
pathway.m84.cor <- cor(pathway.data[ind,colSums(pathway.data[ind,])>0],
                        m84data.nor[ind,apply(m84data.nor[ind,],2,sd)>0],
                        method='spearman')
plot.correlation.heatmap(pathway.m84.cor,row.filter=0.5,col.filter=0)
rm(ind)
rm(pathway.m84.cor)


### PLEASE-T4 Response
ind <- intersect(intersect(rownames(pathway.data),rownames(m84data.nor)),rownames(subset(sample.info,Type=='PLEASE-T4' & Response =='Response')))
pathway.m84.cor <- cor(pathway.data[ind,colSums(pathway.data[ind,])>0],
                        m84data.nor[ind,apply(m84data.nor[ind,],2,sd)>0],
                        method='spearman')
plot.correlation.heatmap(pathway.m84.cor,row.filter=0.6,col.filter=0)
rm(ind)
rm(pathway.m84.cor)

### PLEASE-T4 Non-Response
ind <- intersect(intersect(rownames(pathway.data),rownames(m84data.nor)),rownames(subset(sample.info,Type=='PLEASE-T4' & Response =='Non.Response')))
pathway.m84.cor <- cor(pathway.data[ind,colSums(pathway.data[ind,])>0],
                        m84data.nor[ind,apply(m84data.nor[ind,],2,sd)>0],
                        method='spearman')
plot.correlation.heatmap(pathway.m84.cor,row.filter=0.6,col.filter=0)
rm(ind)
rm(pathway.m84.cor)
```


## Antibiotics use (T1)

### Wilcoxon rank test 
```{r m84metabolites_rank_test_of_antibiotics_use,cache=TRUE}
### plot top ones
m84.antibio.wil <- wilcox_test(m84data.cp1,
                    subset(sample.info,Group=='PLEASE',select=Antibiotics.visit),
                             plot.heatmap=FALSE,boxplot.top=10,boxplot.y='Abundance', parallel=FALSE)      
###
## write.csv(m84.antibio.wil,file=paste('../3_Result/2015_03_05_UCLA84Metabolites_PLEASE_Antibiotics_Use_Wilcox.csv',sep=''),row.names=TRUE)

```

### Random Forest 
```{r m84metabolites_random_forest_of_antibiotics_use,cache=TRUE}
random_forest(m84data.cp1,
              subset(sample.info,Group=='PLEASE',select=Antibiotics.visit),
              plot.MDS=TRUE,plot.importance=TRUE,print.result=FALSE,
              guess.num=50,boxplot.top=6,
              title="Antibiotics use",boxplot.y="Abundance",
              seed=100)
```

## Cluster 1 vs. cluster 2 (Bacterial clusters)

### Wilcoxon rank test 
```{r m84metabolites_rank_test_of_cluster_1_vs_2,cache=TRUE}
### plot all the significant ones
m84.c1c2.wil <- wilcox_test(m84data.cp1,
                                subset(sample.info,Group=='PLEASE',select=Cluster),
                                plot.heatmap=FALSE,boxplot.top=0.05,boxplot.y='Abundance')       
###
##write.csv(m84.c1c2.wil,file=paste('../3_Result/2015_03_05_UCLA84Metabolites_PLEASE_Cluster_1_vs_2_Wilcox.csv',sep=''),row.names=TRUE)
```

### Random Forest 
```{r m84metabolites_random_forest_of_cluster_1_vs_2,cache=TRUE}
random_forest(m84data.cp1,subset(sample.info,Group=='PLEASE' & !is.na(Cluster),select=Cluster),
              plot.MDS=TRUE,plot.importance=TRUE,print.result=FALSE,
              guess.num=50,boxplot.top=5,
              title="Normal vs. Crohn",boxplot.y="Abundance",
              seed=100)
```


## Responders vs. non-responders

### Random Forest 
```{r m84metabolites_random_forest_of_r_vs_nr,cache=TRUE,fig.width=10}
### EEN T1
random_forest(m84data.nor,
              subset(sample.info,Time==1 & Treatment.Specific=='EEN' & !is.na(Response),
                     select=Response),
              plot.MDS=TRUE,plot.importance=TRUE,print.result=FALSE,
              guess.num=50,boxplot.top=5,
              title="Response vs. non-response",boxplot.y="Abundance",
              seed=100)
### EEN T4
random_forest(m84data.nor,
              subset(sample.info,Time==4 & Treatment.Specific=='EEN' & !is.na(Response),
                     select=Response),
              plot.MDS=TRUE,plot.importance=TRUE,print.result=FALSE,
              guess.num=50,boxplot.top=5,
              title="Response vs. non-response",boxplot.y="Abundance",
              seed=100)


### antiTNF T1
random_forest(m84data.nor,
              subset(sample.info,Time==1 & Treatment.Specific=='antiTNF' & !is.na(Response),
                     select=Response),
              plot.MDS=TRUE,plot.importance=TRUE,print.result=FALSE,
              guess.num=50,boxplot.top=5,
              title="Response vs. non-response",boxplot.y="Abundance",
              seed=100)

### antiTNF T4
random_forest(m84data.nor,
              subset(sample.info,Time==4 & Treatment.Specific=='antiTNF' & !is.na(Response),
                     select=Response),
              plot.MDS=TRUE,plot.importance=TRUE,print.result=FALSE,
              guess.num=50,boxplot.top=5,
              title="Response vs. non-response",boxplot.y="Abundance",
              seed=100)
```

### Wilcoxon rank test
```{r m84metabolites_compare_response_vs_nonresponse,cache=TRUE}
### compare response vs. non-response

### T1: R vs. NR
m84.t1.RvsNR <- wilcox_test(m84data.nor,subset(sample.info,Time==1 &  !is.na(Response),
                                  select=Response),plot.heatmap=FALSE,boxplot.top=5,boxplot.y='Abundance')
### T4: R vs. NR
m84.t4.RvsNR <- wilcox_test(m84data.nor,subset(sample.info,Time==4 & !is.na(Response),
                                  select=Response),plot.heatmap=FALSE,boxplot.top=5,boxplot.y='Abundance')


### EEN T1: R vs. NR
m84.een.t1.RvsNR <- wilcox_test(m84data.nor,subset(sample.info,Time==1 & Treatment.Specific=='EEN' & !is.na(Response),
                                  select=Response),plot.heatmap=FALSE,boxplot.top=5,boxplot.y='Abundance')
### EEN T4: R vs. NR
m84.een.t4.RvsNR <- wilcox_test(m84data.nor,subset(sample.info,Time==4 & Treatment.Specific=='EEN' & !is.na(Response),
                                  select=Response),plot.heatmap=FALSE,boxplot.top=5,boxplot.y='Abundance')
### aniTNF T1: R vs. NR
m84.antiTNF.t1.RvsNR <- wilcox_test(m84data.nor,subset(sample.info,Time==1 & Treatment.Specific=='antiTNF' & !is.na(Response),
                                  select=Response),plot.heatmap=FALSE,boxplot.top=5,boxplot.y='Abundance')
### antiTNF T4: R vs. NR
m84.antiTNF.t4.RvsNR <- wilcox_test(m84data.nor,subset(sample.info,Time==4 & Treatment.Specific=='antiTNF' & !is.na(Response),
                                  select=Response),plot.heatmap=FALSE,boxplot.top=5,boxplot.y='Abundance')

```


## T1 vs. T2/T4
```{r m84metabolites_compare_time_points,cache=TRUE,warning=TRUE}
options(width=200)
### compare time points
### !!!!! No T3 data !!!!!!!

#### by treatment only
add_rownames(as.data.frame(m84data.nor),var='Sample') %>%
  gather(Meta,Abu,-Sample) %>%
  left_join(dplyr::select(add_rownames(sample.info,var='Sample'),Subject,Sample,Time,Treatment.Specific,Response), by='Sample') %>%  
  filter(!is.na(Time) & Time!='3') %>% 
  mutate(Time=paste('T',Time,sep='')) %>%
  dplyr::select(-Sample) %>% 
  spread(Time,Abu) %>%
  group_by(Treatment.Specific,Meta) %>%
  summarise_each(funs(wilcox.test(., T1, pair = TRUE)$p.value), vars = T2:T4) %>%
  ## raw pvalues
  rename(T1.vs.T2.p=vars1,T1.vs.T4.p=vars2) %>%
  mutate(T1.vs.T2.q = p.adjust(T1.vs.T2.p,method='fdr'),
         T1.vs.T4.q = p.adjust(T1.vs.T4.p,method='fdr')
         ) %>%
  mutate_each(funs(signif(.,digits=2)),T1.vs.T2.p:T1.vs.T4.q) %>%
  arrange(Treatment.Specific,T1.vs.T4.p) %>%
  dplyr::filter(T1.vs.T2.p<0.05 | T1.vs.T4.p<0.05) %>%
  as.data.frame


#### by treatment and response
add_rownames(as.data.frame(m84data.nor),var='Sample') %>%
  gather(Meta,Abu,-Sample) %>%
  left_join(dplyr::select(add_rownames(sample.info,var='Sample'),Subject,Sample,Time,Treatment.Specific,Response), by='Sample') %>%  
  filter(!is.na(Time) & Time!='3') %>% 
  mutate(Time=paste('T',Time,sep='')) %>%
  dplyr::select(-Sample) %>% 
  spread(Time,Abu) %>%
  group_by(Treatment.Specific,Response,Meta) %>%
  summarise_each(funs(wilcox.test(., T1, pair = TRUE)$p.value), vars = T2:T4) %>%
  na.omit() %>%
  ## raw pvalues
  rename(T1.vs.T2.p=vars1,T1.vs.T4.p=vars2) %>%
  mutate(T1.vs.T2.q = p.adjust(T1.vs.T2.p,method='fdr'),
         T1.vs.T4.q = p.adjust(T1.vs.T4.p,method='fdr')
         ) %>%
  mutate_each(funs(signif(.,digits=2)),T1.vs.T2.p:T1.vs.T4.q) %>%
  arrange(Treatment.Specific,Response,T1.vs.T4.p) %>%
  dplyr::filter(T1.vs.T2.p<0.05 | T1.vs.T4.p<0.05) %>%
  as.data.frame
options(width=90)
```


## T1 vs. T4 (with median in output)
```{r m84metabolites_compare_time_points_with_median,cache=TRUE,warning=TRUE}
options(width=200)
### compare time points
### !!!!! No T3 data !!!!!!!

#### by treatment only
add_rownames(as.data.frame(m84data.nor),var='Sample') %>%
  gather(Meta,Abu,-Sample) %>%
  left_join(dplyr::select(add_rownames(sample.info,var='Sample'),Subject,Sample,Time,Treatment.Specific,Response), by='Sample') %>%  
  filter(!is.na(Time) & Time!='3') %>% 
  mutate(Time=paste('T',Time,sep='')) %>%
  dplyr::select(-Sample) %>% 
  spread(Time,Abu) %>%
  group_by(Treatment.Specific,Meta) %>%
  summarise_each(funs(p=wilcox.test(T1,., pair = TRUE)$p.value,m=median(.,na.rm=TRUE)), T1,T2,T4) %>%
  ## raw pvalues
  rename(T1.vs.T4.p=T4_p, T1.vs.T2.p=T2_p, T1.median=T1_m, T2.median=T2_m, T4.median=T4_m) %>%
  mutate(T1.vs.T2.q = p.adjust(T1.vs.T2.p,method='fdr'),
         T1.vs.T4.q = p.adjust(T1.vs.T4.p,method='fdr')
         ) %>%
  mutate_each(funs(signif(.,digits=4)),T1.vs.T2.p:T1.vs.T4.q) %>%
  arrange(Treatment.Specific,T1.vs.T4.p) %>%
  dplyr::select(-T1.vs.T2.p,-T1.vs.T4.p,-T1_p) %>%
  #dplyr::filter(T1.vs.T2.p<0.05 | T1.vs.T4.p<0.05) %>%
  write.csv(file=paste('../3_Result/2015_03_05_Metabolites84_T1_T2_T4_by_Treatment_Wilcox.csv',sep=''),row.names=FALSE)


#### by treatment and response
add_rownames(as.data.frame(m84data.nor),var='Sample') %>%
  gather(Meta,Abu,-Sample) %>%
  left_join(dplyr::select(add_rownames(sample.info,var='Sample'),Subject,Sample,Time,Treatment.Specific,Response), by='Sample') %>%  
  filter(!is.na(Time) & Time!='3') %>% 
  mutate(Time=paste('T',Time,sep='')) %>%
  dplyr::select(-Sample) %>% 
  spread(Time,Abu) %>%
  group_by(Treatment.Specific,Response,Meta) %>%
  summarise_each(funs(p=wilcox.test(T1,., pair = TRUE)$p.value,m=median(.,na.rm=TRUE)), T1,T2,T4) %>%
  ## raw pvalues
  rename(T1.vs.T4.p=T4_p, T1.vs.T2.p=T2_p, T1.median=T1_m, T2.median=T2_m, T4.median=T4_m) %>%
  mutate(T1.vs.T2.q = p.adjust(T1.vs.T2.p,method='fdr'),
         T1.vs.T4.q = p.adjust(T1.vs.T4.p,method='fdr')
         ) %>%
  mutate_each(funs(signif(.,digits=4)),T1.vs.T2.p:T1.vs.T4.q) %>%
  arrange(Treatment.Specific,Response,T1.vs.T4.p) %>%
  dplyr::select(-T1.vs.T2.p,-T1.vs.T4.p,-T1_p) %>%
  #dplyr::filter(T1.vs.T2.p<0.05 | T1.vs.T4.p<0.05) %>%
  write.csv(file=paste('../3_Result/2015_03_05_Metabolites84_T1_T2_T4_by_Treatment_Response_Wilcox.csv',sep=''),row.names=FALSE)


options(width=90)
```

## EEN vs. antiTNF
```{r m84metabolites_compare_treatments,cache=TRUE}
### compare treatments

### T4: EEN vs. antiTNF
m84.t4.EENvsAntiTNF <- wilcox_test(m84data.nor,subset(sample.info,Time==4 & Treatment.Specific!='PEN',
                                  select=Treatment.Specific),plot.heatmap=FALSE,boxplot.top=0.05,boxplot.y='Abundance')
### T1: EEN vs. antiTNF as a control test 
### Check the baseline, since no treatment yet, they should not be different
m84.t1.EENvsAntiTNF <- wilcox_test(m84data.nor,subset(sample.info,Time==1 & Treatment.Specific!='PEN',
                                  select=Treatment.Specific),plot.heatmap=FALSE,boxplot.top=5,boxplot.y='Abundance')


### T4 NR: EEN vs. antiTNF
m84.t4.nr.EENvsAntiTNF <- wilcox_test(m84data.nor,
                       subset(sample.info,Time==4 & Treatment.Specific!='PEN' & Response=='Non.Response',
                         select=Treatment.Specific),plot.heatmap=FALSE,boxplot.top=5,boxplot.y='Abundance')
### T1 NR: EEN vs. antiTNF as a control test 
### Check the baseline, since no treatment yet, they should not be different
m84.t1.nr.EENvsAntiTNF <- wilcox_test(m84data.nor,
                       subset(sample.info,Time==1 & Treatment.Specific!='PEN' & Response=='Non.Response',
                         select=Treatment.Specific),plot.heatmap=FALSE,boxplot.top=5,boxplot.y='Abundance')

### T4 R: EEN vs. antiTNF
m84.t4.r.EENvsAntiTNF <- wilcox_test(m84data.nor,
                       subset(sample.info,Time==4 & Treatment.Specific!='PEN' & Response=='Response',
                         select=Treatment.Specific),plot.heatmap=FALSE,boxplot.top=0.05,boxplot.y='Abundance')
### T1 R: EEN vs. antiTNF as a control test 
### Check the baseline, since no treatment yet, they should not be different
m84.t1.r.EENvsAntiTNF <- wilcox_test(m84data.nor,
                       subset(sample.info,Time==1 & Treatment.Specific!='PEN' & Response=='Response',
                         select=Treatment.Specific),plot.heatmap=FALSE,boxplot.top=5,boxplot.y='Abundance')


### T4: EEN vs. PEN
m84.t4.EENvsPEN <- wilcox_test(m84data.nor,subset(sample.info,Time==4 & Treatment.Specific!='antiTNF',
                                  select=Treatment.Specific),plot.heatmap=FALSE,boxplot.top=0.05,boxplot.y='Abundance')

### T1: EEN vs. PEN as a control test 
### Check the baseline, since no treatment yet, they should not be different
m84.t1.EENvsPEN <- wilcox_test(m84data.nor,subset(sample.info,Time==1 & Treatment.Specific!='antiTNF',
                                  select=Treatment.Specific),plot.heatmap=FALSE,boxplot.top=5,boxplot.y='Abundance')
### T2: EEN vs. PEN as a control test 
m84.t2.EENvsPEN <- wilcox_test(m84data.nor,subset(sample.info,Time==2 & Treatment.Specific!='antiTNF',
                                  select=Treatment.Specific),plot.heatmap=FALSE,boxplot.top=5,boxplot.y='Abundance')

### T4: PEN vs. antiTNF
m84.t4.PENvsAntiTNF <- wilcox_test(m84data.nor,subset(sample.info,Time==4 & Treatment.Specific!='EEN',
                                  select=Treatment.Specific),plot.heatmap=FALSE,boxplot.top=5,boxplot.y='Abundance')

### T1: PEN vs. antiTNF as a control test 
### Check the baseline, since no treatment yet, they should not be different
m84.t1.PENvsAntiTNF <- wilcox_test(m84data.nor,subset(sample.info,Time==1 & Treatment.Specific!='EEN',
                                  select=Treatment.Specific),plot.heatmap=FALSE,boxplot.top=5,boxplot.y='Abundance')
```



# Histidine
```{r histidine,cache=TRUE,warning=TRUE}
histidine.abu <- as.matrix(kmdata.nor[,"histidine"])
colnames(histidine.abu) <- 'histidine.abu'
#### plot 
p<- add_rownames(as.data.frame(histidine.abu),var='Sample') %>% 
  left_join(add_rownames(sample.info,var='Sample'),by='Sample') %>%
  dplyr::select(histidine.abu,Response,Time,Treatment.Specific) %>%
  dplyr::filter(Time!='3') %>%
  na.omit %>%
  ggplot( aes(factor(Time), histidine.abu),na.rm=T)+
  geom_boxplot(na.rm=T)+facet_grid(Treatment.Specific ~ Response, margins=TRUE)
plot(p)

#### normal vs. IBD
wilcox_test(histidine.abu,sample.info[,'Disease',drop=FALSE],                         plot.heatmap=FALSE,boxplot.top=0.05,boxplot.y='Abundance')

#### normal vs. IBD without antibiotics use
#### sample size is different compared to the previous test
wilcox_test(histidine.abu,
            subset(sample.info, Type=='COMBO' | (Type=='PLEASE-T1' & Antibiotics.visit=='Not.Use'), select=Disease),                        plot.heatmap=FALSE,boxplot.top=0.05,boxplot.y='Abundance')

#### cluster 1 vs. cluster 2
wilcox_test(histidine.abu, subset(sample.info,Group=='PLEASE',select=Cluster),
      plot.heatmap=FALSE,boxplot.top=0.05,boxplot.y='Abundance')  

#### antibiotics
wilcox_test(histidine.abu,subset(sample.info,Group=='PLEASE',select=Antibiotics.visit),
      plot.heatmap=FALSE,boxplot.top=10,boxplot.y='Abundance')      

##########################
options(width=200)
### compare time points
### !!!!! No T3 data !!!!!!!

#### by treatment only
add_rownames(as.data.frame(histidine.abu),var='Sample') %>%
  left_join(dplyr::select(add_rownames(sample.info,var='Sample'),Subject,Sample,Time,Treatment.Specific,Response), by='Sample') %>%  
  filter(!is.na(Time) & Time!='3') %>% 
  mutate(Time=paste('T',Time,sep='')) %>%
  dplyr::select(-Sample) %>% 
  spread(Time,histidine.abu) %>%
  group_by(Treatment.Specific) %>%
  summarise_each(funs(wilcox.test(., T1, pair = TRUE)$p.value), vars = T2:T4) %>%
  ## raw pvalues
  rename(T1.vs.T2.p=vars1,T1.vs.T4.p=vars2) %>%
  mutate(T1.vs.T2.q = p.adjust(T1.vs.T2.p,method='fdr'),
         T1.vs.T4.q = p.adjust(T1.vs.T4.p,method='fdr')
         ) %>%
  mutate_each(funs(signif(.,digits=2)),T1.vs.T2.p:T1.vs.T4.q) %>%
  arrange(Treatment.Specific,T1.vs.T4.p) %>%
  as.data.frame


### by treatment, response
add_rownames(as.data.frame(histidine.abu),var='Sample') %>%
  left_join(dplyr::select(add_rownames(sample.info,var='Sample'),Subject,Sample,Time,Treatment.Specific,Response), by='Sample') %>%  
  filter(!is.na(Time) & Time!='3' & !is.na(Response)) %>% 
  mutate(Time=paste('T',Time,sep='')) %>%
  dplyr::select(-Sample) %>% 
  spread(Time,histidine.abu) %>%
  group_by(Treatment.Specific, Response) %>%
  summarise_each(funs(wilcox.test(., T1, pair = TRUE)$p.value), vars = T2:T4) %>%
  ## raw pvalues
  rename(T1.vs.T2.p=vars1,T1.vs.T4.p=vars2) %>%
  mutate(T1.vs.T2.q = p.adjust(T1.vs.T2.p,method='fdr'),
         T1.vs.T4.q = p.adjust(T1.vs.T4.p,method='fdr')
         ) %>%
  mutate_each(funs(signif(.,digits=2)),T1.vs.T2.p:T1.vs.T4.q) %>%
  arrange(Treatment.Specific,Response,T1.vs.T4.p) %>%
  as.data.frame
```



# Analyze known metabolite data


## Normal vs. Crohn-T1
```{r known_metabolite_normal_vs_crohn_t1,cache=TRUE}
### I don't use dplyr filter 
### because it will remove the row names
kmetabolite.cp1 <- subset_data_by_sample_info(kmdata.nor,sample.info,
                                          Type=='COMBO' | Type=='PLEASE-T1')

### remove empty rows, "6007-01"
# rownames(pathway.cp1)[rowSums(pathway.cp1) == 0]
### there are other samples with empty rows
### remember to remove them in the latter analysis

dim(kmetabolite.cp1)
```


### Calculate distance
```{r known_metabolite_calculate_distance,cache=TRUE,eval=FALSE}
kmetabolite.distance.method <- "Bray"
kmetabolite.dist.cp1 <- calculate_distance(X = kmdata.cp1, method = kmetabolite.distance.method, tree = NA)

### distribution
hist(kmetabolite.dist.cp1,xlab='Distance',main='Known metabolites (COMBO and PLEASE-T1)')

```

### Find the best clustering
```{r known_metabolite_find_the_best_clustering,cache=TRUE,eval=FALSE}
kmetabolite.pamk.best <- find_best_clustering(kmetabolite.dist.cp1,plot.results=TRUE)
```

### MDS analysis
```{r known_metabolite_MDS,cache=TRUE,results='hide',eval=FALSE}
### MDS
set.seed(10)
pathway.meta.MDS <- metaMDS(pathway.dist.cp1)

### Normal/Disease
p <- add_rownames(as.data.frame(pathway.meta.MDS$points),var='Sample') %>%
  left_join(dplyr::select(add_rownames(sample.info,var='Sample'),
                          Sample,Disease),
            by='Sample') %>% 
  ggplot(aes(x=MDS1,y=MDS2))+ geom_point(aes(colour = Disease)) +
  scale_color_manual(values=c("#0066CC", "#FF3333"),name="") +
  labs(title = 'metaMDS') 
print(p)

### Normal/Disease, cluster
### These are bacterial clusters !!
p <- add_rownames(as.data.frame(pathway.meta.MDS$points),var='Sample') %>%
  left_join(dplyr::select(add_rownames(sample.info,var='Sample'),
                          Sample,Disease,Cluster),
            by='Sample') %>% 
  ggplot(aes(x=MDS1,y=MDS2))+ geom_point(aes(colour = Disease,shape=Cluster)) +
  scale_color_manual(values=c("#0066CC", "#FF3333"),name="") +
  scale_shape_manual(values=c(20,4),name="") +
  labs(title = 'metaMDS') 
print(p)

### Normal/Disease, cluster (for publication)
### These are bacterial clusters !!
p <- add_rownames(as.data.frame(pathway.meta.MDS$points),var='Sample') %>%
  left_join(dplyr::select(add_rownames(sample.info,var='Sample'),
                          Sample,Disease,Cluster),
            by='Sample') %>% 
  ggplot(aes(x=MDS1,y=MDS2))+ geom_point(aes(colour = Cluster,shape=Disease)) +
  scale_color_manual(values=c("#0066CC", "#FF3333"),name="") +
  scale_shape_manual(values=c(20,1),name="") +
  labs(title = 'Ordination pathways') +
  theme_classic()
print(p)


#### Fugal percentage 
p <- add_rownames(as.data.frame(pathway.meta.MDS$points),var='Sample') %>%
  left_join(dplyr::select(add_rownames(sample.info,var='Sample'),
                          Sample,Disease,Cluster,Fungi.Per),
            by='Sample') %>% 
  mutate(Fungi.Per.sqrt=sqrt(Fungi.Per)) %>%
  ggplot(aes(x=MDS1,y=MDS2))+ 
  geom_point(aes(colour = Disease,shape=Cluster,size = Fungi.Per.sqrt)) +
  scale_color_manual(values=c("#0066CC", "#FF3333"),name="") +
  scale_shape_manual(values=c(20,1),name="") +
  scale_size_area()+
  labs(title = 'metaMDS') 
print(p)

#### Human DNA percentage
p <- add_rownames(as.data.frame(pathway.meta.MDS$points),var='Sample') %>%
  left_join(dplyr::select(add_rownames(sample.info,var='Sample'),
                          Sample,Disease,Cluster,Human.Per),
            by='Sample') %>% 
  ggplot(aes(x=MDS1,y=MDS2))+ 
  geom_point(aes(colour = Disease,shape=Cluster,size = Human.Per)) +
  scale_color_manual(values=c("#0066CC", "#FF3333"),name="") +
  scale_shape_manual(values=c(20,1),name="") +
  scale_size_area()+
  labs(title = 'metaMDS') 
print(p)
```



### Random Forest 
```{r known_metabolite_random_forest_of_normal_vs_crohn,cache=TRUE}
random_forest(kmetabolite.cp1,sample.info[,'Disease',drop=FALSE],
              plot.MDS=TRUE,plot.importance=TRUE,print.result=FALSE,
              guess.num=50,boxplot.top=5,
              title="Normal vs. Crohn",boxplot.y="Abundance",
              seed=100)
```

### Wilcoxon rank test of normal vs. normal
```{r known_metabolite_rank_test_of_normal_vs_normal,cache=TRUE}
### "Control" vs. "control"
sam.tmp <- subset(sample.info,Disease=='Control',select=Disease)
sam.tmp[1:13,1] <- 'control' 
###  
kmetabolite.cc.wil <- wilcox_test(kmetabolite.cp1, sam.tmp,                       
                          plot.heatmap=FALSE,boxplot.top=0.05,boxplot.y='Abundance')

cat("signifcant metabolites",sum(kmetabolite.cc.wil[,2]<0.05))
###
#write.csv(pathway.cp1.wil,file=paste('../3_Result/2015_03_05_Pathway_COMBO_PLEASET1_Wilcox.csv',sep=''),row.names=TRUE)
rm(sam.tmp)
```


### Wilcoxon rank test of normal vs. Crohn
```{r known_metabolite_rank_test_of_normal_vs_crohn,cache=TRUE}
### only plot the top 5 significant ones
kmetabolite.cp1.wil <- wilcox_test(kmetabolite.cp1,sample.info[,'Disease',drop=FALSE],
                            plot.heatmap=FALSE,boxplot.top=10,boxplot.y='Abundance')
cat("signifcant metabolites",sum(kmetabolite.cp1.wil[,2]<0.05))
###
write.csv(kmetabolite.cp1.wil,file=paste('../3_Result/2015_03_05_Known_Metabolite_COMBO_PLEASET1_Wilcox.csv',sep=''),row.names=TRUE)

### Enrichment
add_rownames(minfo,var='ID') %>% 
dplyr::filter(!is.na(Metabolite)) %>%
right_join(dplyr::select(add_rownames(as.data.frame(kmetabolite.cp1.wil),var="Metabolite"),Metabolite,Disease.qvalue),by='Metabolite') %>%
dplyr::select(Metabolite,Super_class,Class,Direct_Parent,Disease.qvalue) %>%
dplyr::filter(!is.na(Super_class)) %>%
## count how many metabolites
mutate(total_meta=n()) %>%
(function(pdata){
  print(nrow(pdata))
  return(pdata)
}) %>%
## Significant metabolites
dplyr::mutate(Sig=ifelse(Disease.qvalue<0.05,'Significant','Not-Significant')) %>%
### count how many signicant ones
(function(pdata){
  print(sum(pdata$Sig=='Significant'))
  return(pdata)
}) %>%  
###############################
### plot the count distribution
(function(pdata){
p <- pdata %>% ggplot(aes(factor(Super_class),fill=Sig)) + geom_bar(position="dodge") + coord_flip()
plot(p)
return(pdata)
}) %>% 
(function(pdata){
p <- pdata %>% ggplot(aes(factor(Class),fill=Sig)) + geom_bar(position="dodge") + coord_flip()
plot(p)
return(pdata)
}) %>% 
#################
mutate(n_sig=sum(Sig=='Significant')) %>%
#################
(function(pdata){
pdata %>%
group_by(Super_class) %>%
summarize(total=unique(total_meta), n_meta=n(),n_sig=unique(n_sig), n_overlap=sum(Sig=='Significant')) %>%
### Fisher's exact text
group_by(Super_class) %>%
do(
  data.frame(fisher.p = fisher.test(matrix(c(.$n_overlap,.$n_sig,.$n_meta,.$total),nrow=2,ncol=2))$p.value)
) %>% print
return(pdata)
}) %>%
(function(pdata){
pdata %>%
group_by(Class) %>%
summarize(total=unique(total_meta), n_meta=n(),n_sig=unique(n_sig), n_overlap=sum(Sig=='Significant')) %>%
### Fisher's exact text
group_by(Class) %>%
do(
  data.frame(fisher.p = fisher.test(matrix(c(.$n_overlap,.$n_sig,.$n_meta,.$total),nrow=2,ncol=2))$p.value)
) %>% print(n=50)
#return(pdata)
}) 
```



### Wilcoxon rank test of normal vs. Cluster 1
```{r known_metabolite_rank_test_of_normal_vs_cluster_1,cache=TRUE}
kmetabolite.normal.cluster1.wil <- wilcox_test(kmetabolite.cp1,
       subset(sample.info,Disease=='Control'|(Disease=='Crohn'&Cluster=='cluster 1'),select=Disease),
                               plot.heatmap=FALSE,boxplot.top=0.05,boxplot.y='Abundance')

cat("signifcant metabolites",sum(kmetabolite.normal.cluster1.wil[,2]<0.05))
###
write.csv(kmetabolite.normal.cluster1.wil,file=paste('../3_Result/2015_03_05_Known_Metabolite_COMBO_vs_Cluster1_Wilcox.csv',sep=''),row.names=FALSE)

### Enrichment
add_rownames(minfo,var='ID') %>% 
dplyr::filter(!is.na(Metabolite)) %>%
right_join(dplyr::select(add_rownames(as.data.frame(kmetabolite.normal.cluster1.wil),var="Metabolite"),Metabolite,Disease.qvalue),by='Metabolite') %>%
dplyr::select(Metabolite,Super_class,Class,Direct_Parent,Disease.qvalue) %>%
dplyr::filter(!is.na(Super_class)) %>%
## count how many metabolites
mutate(total_meta=n()) %>%
(function(pdata){
  print(nrow(pdata))
  return(pdata)
}) %>%
## Significant metabolites
dplyr::mutate(Sig=ifelse(Disease.qvalue<0.05,'Significant','Not-Significant')) %>%
### count how many signicant ones
(function(pdata){
  print(sum(pdata$Sig=='Significant'))
  return(pdata)
}) %>%  
###############################
### plot the count distribution
(function(pdata){
p <- pdata %>% ggplot(aes(factor(Super_class),fill=Sig)) + geom_bar(position="dodge") + coord_flip()
plot(p)
return(pdata)
}) %>% 
(function(pdata){
p <- pdata %>% ggplot(aes(factor(Class),fill=Sig)) + geom_bar(position="dodge") + coord_flip()
plot(p)
return(pdata)
}) %>% 
#################
mutate(n_sig=sum(Sig=='Significant')) %>%
#################
(function(pdata){
pdata %>%
group_by(Super_class) %>%
summarize(total=unique(total_meta), n_meta=n(),n_sig=unique(n_sig), n_overlap=sum(Sig=='Significant')) %>%
### Fisher's exact text
group_by(Super_class) %>%
do(
  data.frame(fisher.p = fisher.test(matrix(c(.$n_overlap,.$n_sig,.$n_meta,.$total),nrow=2,ncol=2))$p.value)
) %>% print
return(pdata)
}) %>%
(function(pdata){
pdata %>%
group_by(Class) %>%
summarize(total=unique(total_meta), n_meta=n(),n_sig=unique(n_sig), n_overlap=sum(Sig=='Significant')) %>%
### Fisher's exact text
group_by(Class) %>%
do(
  data.frame(fisher.p = fisher.test(matrix(c(.$n_overlap,.$n_sig,.$n_meta,.$total),nrow=2,ncol=2))$p.value)
) %>% print(n=50)
#return(pdata)
}) 
```

### Wilcoxon rank test of normal vs. Cluster 2
```{r known_metabolite_rank_test_of_normal_vs_cluster_2,cache=TRUE}
kmetabolite.normal.cluster2.wil <- wilcox_test(kmetabolite.cp1,
      subset(sample.info,Disease=='Control'|(Disease=='Crohn'&Cluster=='cluster 2'),select=Disease),
      plot.heatmap=FALSE,boxplot.top=0.05,boxplot.y='Abundance')

cat("signifcant metabolites",sum(kmetabolite.normal.cluster2.wil[,2]<0.05))
###
write.csv(kmetabolite.normal.cluster2.wil,file=paste('../3_Result/2015_03_05_Known_Metabolite_COMBO_vs_Cluster2_Wilcox.csv',sep=''),row.names=FALSE)

### Enrichment
add_rownames(minfo,var='ID') %>% 
dplyr::filter(!is.na(Metabolite)) %>%
right_join(dplyr::select(add_rownames(as.data.frame(kmetabolite.normal.cluster2.wil),var="Metabolite"),Metabolite,Disease.qvalue),by='Metabolite') %>%
dplyr::select(Metabolite,Super_class,Class,Direct_Parent,Disease.qvalue) %>%
dplyr::filter(!is.na(Super_class)) %>%
## count how many metabolites
mutate(total_meta=n()) %>%
(function(pdata){
  print(nrow(pdata))
  return(pdata)
}) %>%
## Significant metabolites
dplyr::mutate(Sig=ifelse(Disease.qvalue<0.05,'Significant','Not-Significant')) %>%
### count how many signicant ones
(function(pdata){
  print(sum(pdata$Sig=='Significant'))
  return(pdata)
}) %>%  
###############################
### plot the count distribution
(function(pdata){
p <- pdata %>% ggplot(aes(factor(Super_class),fill=Sig)) + geom_bar(position="dodge") + coord_flip()
plot(p)
return(pdata)
}) %>% 
(function(pdata){
p <- pdata %>% ggplot(aes(factor(Class),fill=Sig)) + geom_bar(position="dodge") + coord_flip()
plot(p)
return(pdata)
}) %>% 
#################
mutate(n_sig=sum(Sig=='Significant')) %>%
#################
(function(pdata){
pdata %>%
group_by(Super_class) %>%
summarize(total=unique(total_meta), n_meta=n(),n_sig=unique(n_sig), n_overlap=sum(Sig=='Significant')) %>%
### Fisher's exact text
group_by(Super_class) %>%
do(
  data.frame(fisher.p = fisher.test(matrix(c(.$n_overlap,.$n_sig,.$n_meta,.$total),nrow=2,ncol=2))$p.value)
) %>% print
return(pdata)
}) %>%
(function(pdata){
pdata %>%
group_by(Class) %>%
summarize(total=unique(total_meta), n_meta=n(),n_sig=unique(n_sig), n_overlap=sum(Sig=='Significant')) %>%
### Fisher's exact text
group_by(Class) %>%
do(
  data.frame(fisher.p = fisher.test(matrix(c(.$n_overlap,.$n_sig,.$n_meta,.$total),nrow=2,ncol=2))$p.value)
) %>% print(n=50)
#return(pdata)
}) 
```


### Heatmap of normal vs. Crohn's disease (T1)
```{r known_metabolite_heatmap_cp1,cache=TRUE,fig.width=15,fig.height=20}
#####
dat.nor <- kmetabolite.cp1
for (pw in colnames(kmetabolite.cp1)){
  pw.dat <- kmetabolite.cp1[,pw,drop=TRUE]
  Me <- median(pw.dat)
  Ma <- mad(pw.dat)
  if (Ma==0){Me <- mean(pw.dat);Ma <- sd(pw.dat)}
  pw.dat <- (pw.dat-Me)/Ma
  pw.dat[pw.dat < -3] <- -3
  pw.dat[pw.dat > 3] <- 3
  dat.nor[,pw] <- pw.dat
}
row.ann <- subset(minfo,!is.na(Metabolite),select=c("Metabolite","Kingdom",  "Super_class", "Class", "Direct_Parent"))
rownames(row.ann) <- row.ann$Metabolite

pheatmap(t(dat.nor),
         border_color="grey60",
         cluster_cols = calculate_distance(X = kmetabolite.cp1, method = "Bray"),
         clustering_distance_rows = 'euclidean',#
         #clustering_method = "complete",
         annotation_col=sample.info[,c("Disease","Time","Response","Treatment.Specific","Cluster","FCP","Antibiotics.visit","Steroids")],
         annotation_row=row.ann[, c("Super_class"),drop=F],
         #color =hmcols, breaks=bk,
         fontsize_row = 5,
         show_rownames = TRUE
         #scale="row",
         #gaps_col = c(26, 111,191,273)
         )

#rm(data.heat)
rm(pw)
rm(dat.nor)
rm(pw.dat)
rm(Me)
rm(Ma)
rm(row.ann)
```


### Heatmap of normal vs. Crohn's disease (T1), significant metabolites
```{r known_metabolite_heatmap_cp1_sig,cache=TRUE,fig.width=15,fig.height=18}
#####
sig.ind <- rownames(subset(as.data.frame(kmetabolite.cp1.wil), Disease.qvalue <0.05))
dat.nor <- kmetabolite.cp1[,sig.ind]
for (pw in sig.ind){
  pw.dat <- kmetabolite.cp1[,pw,drop=TRUE]
  Me <- median(pw.dat)
  Ma <- mad(pw.dat)
  if (Ma==0){Me <- mean(pw.dat);Ma <- sd(pw.dat)}
  pw.dat <- (pw.dat-Me)/Ma
  pw.dat[pw.dat < -3] <- -3
  pw.dat[pw.dat > 3] <- 3
  dat.nor[,pw] <- pw.dat
}
row.ann <- subset(minfo,!is.na(Metabolite),select=c("Metabolite","Kingdom",  "Super_class", "Class", "Direct_Parent"))
rownames(row.ann) <- row.ann$Metabolite

pheatmap(t(dat.nor),
         border_color="grey60",
         cluster_cols = calculate_distance(X = kmetabolite.cp1, method = "Bray"),
         clustering_distance_rows = 'euclidean',#
         #clustering_method = "complete",
         annotation_col=sample.info[,c("Disease","Time","Response","Treatment.Specific","Cluster","FCP","Antibiotics.visit","Steroids")],
         annotation_row=row.ann[, c("Super_class"),drop=F],
         #color =hmcols, breaks=bk,
         fontsize_row = 5.5,
         show_rownames = TRUE
         #scale="row",
         #gaps_col = c(26, 111,191,273)
         )

#rm(data.heat)
rm(pw)
rm(dat.nor)
rm(pw.dat)
rm(Me)
rm(Ma)
rm(row.ann)
```


### Heatmap of normal vs. Crohn's disease (T1), amino acid
```{r known_metabolite_heatmap_cp1_aa,cache=TRUE,fig.width=15,fig.height=18}
#####
row.ann <- subset(minfo,!is.na(Metabolite),select=c("Metabolite","Kingdom",  "Super_class", "Class", "Direct_Parent"))
rownames(row.ann) <- row.ann$Metabolite
aa.ind <- rownames(subset(row.ann,Class=="Amino Acids and Derivatives"))

dat.nor <- kmetabolite.cp1[,aa.ind]
for (pw in aa.ind){
  pw.dat <- kmetabolite.cp1[,pw,drop=TRUE]
  Me <- median(pw.dat)
  Ma <- mad(pw.dat)
  if (Ma==0){Me <- mean(pw.dat);Ma <- sd(pw.dat)}
  pw.dat <- (pw.dat-Me)/Ma
  pw.dat[pw.dat < -3] <- -3
  pw.dat[pw.dat > 3] <- 3
  dat.nor[,pw] <- pw.dat
}


pheatmap(t(dat.nor),
         border_color="grey60",
         cluster_cols = calculate_distance(X = kmetabolite.cp1, method = "Bray"),
         clustering_distance_rows = 'euclidean',#
         #clustering_method = "complete",
         annotation_col=sample.info[,c("Disease","Response","Treatment.Specific","Cluster","FCP","Antibiotics.visit","Steroids")],
         annotation_row=row.ann[, c("Super_class"),drop=F],
         #color =hmcols, breaks=bk,
         fontsize_row = 12,
         show_rownames = TRUE,
         scale="none"
         #gaps_col = c(26, 111,191,273)
         )

rm(aa.ind)
rm(pw)
rm(dat.nor)
rm(pw.dat)
rm(Me)
rm(Ma)
rm(row.ann)
```



## Normal vs. Crohn-T1 without antibiotic use
```{r known_metabolite_normal_vs_crohn_t1_no_antibiotics,cache=TRUE}
### I don't use dplyr filter 
### because it will remove the row names
kmetabolite.cp1.no.anti <- subset_data_by_sample_info(kmetabolite.cp1,sample.info,
           Type=='COMBO' | (Type=='PLEASE-T1' & Antibiotics.visit=='Not.Use'))

dim(kmetabolite.cp1.no.anti)

```

### Random Forest 
```{r known_metabolite_random_forest_of_normal_vs_crohn_no_antibiotics,cache=TRUE}
random_forest(kmetabolite.cp1.no.anti,sample.info[,'Disease',drop=FALSE],
              plot.MDS=TRUE,plot.importance=TRUE,print.result=FALSE,
              guess.num=50,boxplot.top=5,
              title="Normal vs. Crohn (no antibiotics)",boxplot.y="Abundance",
              seed=100)
```


### Wilcoxon rank test of normal vs. Crohn
```{r known_metabolite_rank_test_of_normal_vs_crohn_no_antibiotics,cache=TRUE}
### only plot the top 5 significant ones
kmetabolite.cp1.no.anti.wil <- wilcox_test(kmetabolite.cp1.no.anti,sample.info[,'Disease',drop=FALSE],
                            plot.heatmap=FALSE,boxplot.top=10,boxplot.y='Abundance')

###
write.csv(kmetabolite.cp1.no.anti.wil,file=paste('../3_Result/2015_03_05_Known_Metabolite_COMBO_PLEASET1_Without_Antibiotics_Wilcox.csv',sep=''),row.names=TRUE)


### Enrichment
add_rownames(minfo,var='ID') %>% 
dplyr::filter(!is.na(Metabolite)) %>%
right_join(dplyr::select(add_rownames(as.data.frame(kmetabolite.cp1.no.anti.wil),var="Metabolite"),Metabolite,Disease.qvalue),by='Metabolite') %>%
dplyr::select(Metabolite,Super_class,Class,Direct_Parent,Disease.qvalue) %>%
dplyr::filter(!is.na(Super_class)) %>%
## count how many metabolites
mutate(total_meta=n()) %>%
(function(pdata){
  print(nrow(pdata))
  return(pdata)
}) %>%
## Significant metabolites
dplyr::mutate(Sig=ifelse(Disease.qvalue<0.05,'Significant','Not-Significant')) %>%
### count how many signicant ones
(function(pdata){
  print(sum(pdata$Sig=='Significant'))
  return(pdata)
}) %>%  
###############################
### plot the count distribution
(function(pdata){
p <- pdata %>% ggplot(aes(factor(Super_class),fill=Sig)) + geom_bar(position="dodge") + coord_flip()
plot(p)
return(pdata)
}) %>% 
(function(pdata){
p <- pdata %>% ggplot(aes(factor(Class),fill=Sig)) + geom_bar(position="dodge") + coord_flip()
plot(p)
return(pdata)
}) %>% 
#################
mutate(n_sig=sum(Sig=='Significant')) %>%
#################
(function(pdata){
pdata %>%
group_by(Super_class) %>%
summarize(total=unique(total_meta), n_meta=n(),n_sig=unique(n_sig), n_overlap=sum(Sig=='Significant')) %>%
### Fisher's exact text
group_by(Super_class) %>%
do(
  data.frame(fisher.p = fisher.test(matrix(c(.$n_overlap,.$n_sig,.$n_meta,.$total),nrow=2,ncol=2))$p.value)
) %>% print
return(pdata)
}) %>%
(function(pdata){
pdata %>%
group_by(Class) %>%
summarize(total=unique(total_meta), n_meta=n(),n_sig=unique(n_sig), n_overlap=sum(Sig=='Significant')) %>%
### Fisher's exact text
group_by(Class) %>%
do(
  data.frame(fisher.p = fisher.test(matrix(c(.$n_overlap,.$n_sig,.$n_meta,.$total),nrow=2,ncol=2))$p.value)
) %>% print(n=50)
#return(pdata)
}) 
```



## Cluster 1 vs. cluster 2

### Wilcoxon rank test 
```{r known_metabolite_rank_test_of_cluster_1_vs_2,cache=TRUE}
### plot all the significant ones
kmetabolite.c1c2.wil <- wilcox_test(kmetabolite.cp1,
                             subset(sample.info,Group=='PLEASE',select=Cluster),
                             plot.heatmap=FALSE,boxplot.top=10,boxplot.y='Abundance')      
###
write.csv(kmetabolite.c1c2.wil,file=paste('../3_Result/2015_03_05_Known_Metabolite_PLEASE_Cluster_1_2_Wilcox.csv',sep=''),row.names=TRUE)
```

### Random Forest 
```{r known_metabolite_random_forest_of_cluster_1_vs_2,cache=TRUE,eval=FALSE}
random_forest(kmetabolite.cp1,
              na.omit(subset(sample.info,Group=='PLEASE',select=Cluster)),
              plot.MDS=TRUE,plot.importance=TRUE,print.result=FALSE,
              guess.num=50,boxplot.top=5,
              title="Cluster 1 vs. Cluster 2",boxplot.y="Abundance",
              seed=100)
```



## Correlation of pathways and metabolites
```{r known_metabolite_correlation_with_pathways,cache=TRUE,fig.height=18,fig.width=20}

plot.correlation.heatmap = function(cor.matrix,
                                    row.filter=0.5,
                                    col.filter=0.5
                                    ){
  cor.matrix <- cor.matrix[
  apply(cor.matrix,1,function(X){max(abs(X),na.rm=TRUE)>row.filter}),
  apply(cor.matrix,2,function(X){max(abs(X),na.rm=TRUE)>col.filter})
  ]
  
  bk <- unique(c(seq(-1,-0.4,length=10),
                 seq(-0.4,0.4,length=50),
                 seq(0.4,1,length=10)))
  hmcols<- colorRampPalette(c('#4575b4','#91bfdb','#e0f3f8',
                            '#ffffbf','#fee090','#fc8d59','#d73027'
                            ))(length(bk)-1)

  pheatmap(cor.matrix, 
           scale='none',fontsize_row = 10, 
           clustering_distance_cols = "manhattan",
           clustering_distance_rows = "manhattan",
           cluster_cols = TRUE,
           color =hmcols, breaks=bk,
           cellwidth = 10, cellheight = 10
  )                              
}


### all
ind <- intersect(rownames(pathway.data),rownames(kmdata.nor))
pathway.metabolite.cor <- cor(pathway.data[ind,],
                              kmdata.nor[ind,apply(kmdata.nor[ind,],2,sd)>0],
                              method='spearman')
plot.correlation.heatmap(pathway.metabolite.cor,row.filter=0.4,col.filter=0.4)
rm(ind)
rm(pathway.metabolite.cor)

### control
ind <- intersect(intersect(rownames(pathway.data),rownames(kmdata.nor)),rownames(subset(sample.info,Type=='COMBO')))
pathway.metabolite.cor <- cor(pathway.data[ind,colSums(pathway.data[ind,])>0],
                        kmdata.nor[ind,apply(kmdata.nor[ind,],2,sd)>0],
                        method='spearman')
plot.correlation.heatmap(pathway.metabolite.cor,row.filter=0.6,col.filter=0.6)
rm(ind)
rm(pathway.metabolite.cor)


### PLEASE-T1
ind <- intersect(intersect(rownames(pathway.data),rownames(kmdata.nor)),rownames(subset(sample.info,Type=='PLEASE-T1')))
pathway.metabolite.cor <- cor(pathway.data[ind,colSums(pathway.data[ind,])>0],
                        kmdata.nor[ind,apply(kmdata.nor[ind,],2,sd)>0],
                        method='spearman')
plot.correlation.heatmap(pathway.metabolite.cor,row.filter=0.5,col.filter=0.5)
rm(ind)
rm(pathway.metabolite.cor)


### PLEASE-T4
ind <- intersect(intersect(rownames(pathway.data),rownames(kmdata.nor)),rownames(subset(sample.info,Type=='PLEASE-T4')))
pathway.metabolite.cor <- cor(pathway.data[ind,colSums(pathway.data[ind,])>0],
                        kmdata.nor[ind,apply(kmdata.nor[ind,],2,sd)>0],
                        method='spearman')
plot.correlation.heatmap(pathway.metabolite.cor,row.filter=0.5,col.filter=0.5)
rm(ind)
rm(pathway.metabolite.cor)


### antiTNF
ind <- intersect(intersect(rownames(pathway.data),rownames(kmdata.nor)),rownames(subset(sample.info,Treatment=='antiTNF')))
pathway.metabolite.cor <- cor(pathway.data[ind,colSums(pathway.data[ind,])>0],
                        kmdata.nor[ind,apply(kmdata.nor[ind,],2,sd)>0],
                        method='spearman')
plot.correlation.heatmap(pathway.metabolite.cor,row.filter=0.5,col.filter=0.5)
rm(ind)
rm(pathway.metabolite.cor)


### EEN
ind <- intersect(intersect(rownames(pathway.data),rownames(kmdata.nor)),rownames(subset(sample.info,Treatment.Specific=='EEN')))
pathway.metabolite.cor <- cor(pathway.data[ind,colSums(pathway.data[ind,])>0],
                        kmdata.nor[ind,apply(kmdata.nor[ind,],2,sd)>0],
                        method='spearman')
plot.correlation.heatmap(pathway.metabolite.cor,row.filter=0.5,col.filter=0.5)
rm(ind)
rm(pathway.metabolite.cor)

### response
ind <- intersect(intersect(rownames(pathway.data),rownames(kmdata.nor)),rownames(subset(sample.info,Response=='Response')))
pathway.metabolite.cor <- cor(pathway.data[ind,colSums(pathway.data[ind,])>0],
                        kmdata.nor[ind,apply(kmdata.nor[ind,],2,sd)>0],
                        method='spearman')
plot.correlation.heatmap(pathway.metabolite.cor,row.filter=0.5,col.filter=0.5)
rm(ind)
rm(pathway.metabolite.cor)


### non-response
ind <- intersect(intersect(rownames(pathway.data),rownames(kmdata.nor)),rownames(subset(sample.info,Response=='Non.Response')))
pathway.metabolite.cor <- cor(pathway.data[ind,colSums(pathway.data[ind,])>0],
                        kmdata.nor[ind,apply(kmdata.nor[ind,],2,sd)>0],
                        method='spearman')
plot.correlation.heatmap(pathway.metabolite.cor,row.filter=0.5,col.filter=0.5)
rm(ind)
rm(pathway.metabolite.cor)


### PLEASE-T4 Response
ind <- intersect(intersect(rownames(pathway.data),rownames(kmdata.nor)),rownames(subset(sample.info,Type=='PLEASE-T4' & Response =='Response')))
pathway.metabolite.cor <- cor(pathway.data[ind,colSums(pathway.data[ind,])>0],
                        kmdata.nor[ind,apply(kmdata.nor[ind,],2,sd)>0],
                        method='spearman')
plot.correlation.heatmap(pathway.metabolite.cor,row.filter=0.6,col.filter=0.6)
rm(ind)
rm(pathway.metabolite.cor)

### PLEASE-T4 Non-Response
ind <- intersect(intersect(rownames(pathway.data),rownames(kmdata.nor)),rownames(subset(sample.info,Type=='PLEASE-T4' & Response =='Non.Response')))
pathway.metabolite.cor <- cor(pathway.data[ind,colSums(pathway.data[ind,])>0],
                        kmdata.nor[ind,apply(kmdata.nor[ind,],2,sd)>0],
                        method='spearman')
plot.correlation.heatmap(pathway.metabolite.cor,row.filter=0.6,col.filter=0.6)
rm(ind)
rm(pathway.metabolite.cor)
```


## Correlation of fungi and metabolites
```{r known_metabolite_correlation_with_fungi,cache=TRUE,fig.height=18,fig.width=20}

plot.correlation.heatmap = function(cor.matrix,
                                    row.filter=0.5,
                                    col.filter=0.5
                                    ){
  cor.matrix <- cor.matrix[
  apply(cor.matrix,1,function(X){max(abs(X),na.rm=TRUE)>row.filter}),
  apply(cor.matrix,2,function(X){max(abs(X),na.rm=TRUE)>col.filter})
  ]
  
  bk <- unique(c(seq(-1,-0.4,length=10),
                 seq(-0.4,0.4,length=50),
                 seq(0.4,1,length=10)))
  hmcols<- colorRampPalette(c('#4575b4','#91bfdb','#e0f3f8',
                            '#ffffbf','#fee090','#fc8d59','#d73027'
                            ))(length(bk)-1)

  pheatmap(cor.matrix, 
           scale='none',fontsize_row = 10, 
           clustering_distance_cols = "manhattan",
           clustering_distance_rows = "manhattan",
           cluster_cols = TRUE,
           color =hmcols, breaks=bk,
           cellwidth = 10, cellheight = 10
  )                              
}


### all
ind <- intersect(rownames(fungi.RPKM),rownames(kmdata.nor))
fungi.metabolite.cor <- cor(fungi.RPKM[ind,],
                              kmdata.nor[ind,apply(kmdata.nor[ind,],2,sd)>0],
                              method='spearman')
plot.correlation.heatmap(fungi.metabolite.cor,row.filter=0,col.filter=0.4)
rm(ind)
rm(fungi.metabolite.cor)

### control
ind <- intersect(intersect(rownames(fungi.RPKM),rownames(kmdata.nor)),rownames(subset(sample.info,Type=='COMBO')))
fungi.metabolite.cor <- cor(fungi.RPKM[ind,],
                              kmdata.nor[ind,apply(kmdata.nor[ind,],2,sd)>0],
                              method='spearman')
plot.correlation.heatmap(fungi.metabolite.cor,row.filter=0,col.filter=0.4)
rm(ind)
rm(fungi.metabolite.cor)


### PLEASE-T1
ind <- intersect(intersect(rownames(fungi.RPKM),rownames(kmdata.nor)),rownames(subset(sample.info,Type=='PLEASE-T1')))
fungi.metabolite.cor <- cor(fungi.RPKM[ind,],
                              kmdata.nor[ind,apply(kmdata.nor[ind,],2,sd)>0],
                              method='spearman')
plot.correlation.heatmap(fungi.metabolite.cor,row.filter=0,col.filter=0.4)
rm(ind)
rm(fungi.metabolite.cor)


### PLEASE-T4
ind <- intersect(intersect(rownames(fungi.RPKM),rownames(kmdata.nor)),rownames(subset(sample.info,Type=='PLEASE-T4')))
fungi.metabolite.cor <- cor(fungi.RPKM[ind,],
                              kmdata.nor[ind,apply(kmdata.nor[ind,],2,sd)>0],
                              method='spearman')
plot.correlation.heatmap(fungi.metabolite.cor,row.filter=0,col.filter=0.4)
rm(ind)
rm(fungi.metabolite.cor)

### antiTNF
ind <- intersect(intersect(rownames(fungi.RPKM),rownames(kmdata.nor)),rownames(subset(sample.info,Treatment=='antiTNF')))
fungi.metabolite.cor <- cor(fungi.RPKM[ind,],
                              kmdata.nor[ind,apply(kmdata.nor[ind,],2,sd)>0],
                              method='spearman')
plot.correlation.heatmap(fungi.metabolite.cor,row.filter=0,col.filter=0.4)
rm(ind)
rm(fungi.metabolite.cor)


### EEN
ind <- intersect(intersect(rownames(fungi.RPKM),rownames(kmdata.nor)),rownames(subset(sample.info,Treatment.Specific=='EEN')))
fungi.metabolite.cor <- cor(fungi.RPKM[ind,],
                              kmdata.nor[ind,apply(kmdata.nor[ind,],2,sd)>0],
                              method='spearman')
plot.correlation.heatmap(fungi.metabolite.cor,row.filter=0,col.filter=0.4)
rm(ind)
rm(fungi.metabolite.cor)

### response
ind <- intersect(intersect(rownames(fungi.RPKM),rownames(kmdata.nor)),rownames(subset(sample.info,Response=='Response')))
fungi.metabolite.cor <- cor(fungi.RPKM[ind,],
                              kmdata.nor[ind,apply(kmdata.nor[ind,],2,sd)>0],
                              method='spearman')
plot.correlation.heatmap(fungi.metabolite.cor,row.filter=0,col.filter=0.4)
rm(ind)
rm(fungi.metabolite.cor)


### non-response
ind <- intersect(intersect(rownames(fungi.RPKM),rownames(kmdata.nor)),rownames(subset(sample.info,Response=='Non.Response')))
fungi.metabolite.cor <- cor(fungi.RPKM[ind,],
                              kmdata.nor[ind,apply(kmdata.nor[ind,],2,sd)>0],
                              method='spearman')
plot.correlation.heatmap(fungi.metabolite.cor,row.filter=0,col.filter=0.4)
rm(ind)
rm(fungi.metabolite.cor)


### PLEASE-T4 Response
ind <- intersect(intersect(rownames(fungi.RPKM),rownames(kmdata.nor)),rownames(subset(sample.info,Type=='PLEASE-T4' & Response =='Response')))
fungi.metabolite.cor <- cor(fungi.RPKM[ind,],
                              kmdata.nor[ind,apply(kmdata.nor[ind,],2,sd)>0],
                              method='spearman')
plot.correlation.heatmap(fungi.metabolite.cor,row.filter=0,col.filter=0.4)
rm(ind)
rm(fungi.metabolite.cor)


### PLEASE-T4 Non-Response
ind <- intersect(intersect(rownames(fungi.RPKM),rownames(kmdata.nor)),rownames(subset(sample.info,Type=='PLEASE-T4' & Response =='Non.Response')))
fungi.metabolite.cor <- cor(fungi.RPKM[ind,],
                              kmdata.nor[ind,apply(kmdata.nor[ind,],2,sd)>0],
                              method='spearman')
plot.correlation.heatmap(fungi.metabolite.cor,row.filter=0,col.filter=0.4)
rm(ind)
rm(fungi.metabolite.cor)

```


## Correlation of microbiome and amino acid
```{r known_metabolite_correlation_aa_microbiome,cache=TRUE,fig.height=18,fig.width=20}

plot.correlation.heatmap = function(cor.matrix,
                                    row.filter=0.5,
                                    col.filter=0.5
                                    ){
  cor.matrix <- cor.matrix[
  apply(cor.matrix,1,function(X){max(abs(X),na.rm=TRUE)>row.filter}),
  apply(cor.matrix,2,function(X){max(abs(X),na.rm=TRUE)>col.filter})
  ]
  
  bk <- unique(c(seq(-1,-0.4,length=10),
                 seq(-0.4,0.4,length=50),
                 seq(0.4,1,length=10)))
  hmcols<- colorRampPalette(c('#4575b4','#91bfdb','#e0f3f8',
                            '#ffffbf','#fee090','#fc8d59','#d73027'
                            ))(length(bk)-1)

  pheatmap(cor.matrix, 
           scale='none',fontsize_row = 10, 
           clustering_distance_cols = "manhattan",
           clustering_distance_rows = "manhattan",
           cluster_cols = TRUE,
           color =hmcols, breaks=bk,
           cellwidth = 10, cellheight = 10
  )                              
}

###
row.ann <- subset(minfo,!is.na(Metabolite),select=c("Metabolite","Kingdom",  "Super_class", "Class", "Direct_Parent"))
rownames(row.ann) <- row.ann$Metabolite
aa.ind <- rownames(subset(row.ann,Class=="Amino Acids and Derivatives"))


### all
ind <- intersect(rownames(kmdata.nor),rownames(taxa.data))
aa.bact.cor <- cor(taxa.data[ind,],
                              kmdata.nor[ind,aa.ind],
                              method='spearman')
plot.correlation.heatmap(aa.bact.cor,row.filter=0,col.filter=0)
rm(ind)
rm(aa.bact.cor)


### control
ind <- intersect(intersect(rownames(kmdata.nor),rownames(taxa.data)),rownames(subset(sample.info,Type=='COMBO')))
aa.bact.cor <- cor(taxa.data[ind,colSums(taxa.data[ind,])>0],
                  kmdata.nor[ind,aa.ind],
                  method='spearman')
plot.correlation.heatmap(aa.bact.cor,row.filter=0,col.filter=0)
rm(ind)
rm(aa.bact.cor)


### PLEASE-T1
ind <- intersect(intersect(rownames(kmdata.nor),rownames(taxa.data)),rownames(subset(sample.info,Type=='PLEASE-T1')))
aa.bact.cor <- cor(taxa.data[ind,colSums(taxa.data[ind,])>0],
                  kmdata.nor[ind,aa.ind],
                  method='spearman')
plot.correlation.heatmap(aa.bact.cor,row.filter=0,col.filter=0)
rm(ind)
rm(aa.bact.cor)

### PLEASE-T1, no antibiotics
ind <- intersect(intersect(rownames(kmdata.nor),rownames(taxa.data)),
                 rownames(subset(sample.info,Antibiotics.visit=='Not.Use',Type=='PLEASE-T1')))
aa.bact.cor <- cor(taxa.data[ind,colSums(taxa.data[ind,])>0],
                  kmdata.nor[ind,aa.ind],
                  method='spearman')
plot.correlation.heatmap(aa.bact.cor,row.filter=0,col.filter=0)
rm(ind)
rm(aa.bact.cor)

### PLEASE-T4
ind <- intersect(intersect(rownames(kmdata.nor),rownames(taxa.data)),rownames(subset(sample.info,Type=='PLEASE-T4')))
aa.bact.cor <- cor(taxa.data[ind,colSums(taxa.data[ind,])>0],
                  kmdata.nor[ind,aa.ind],
                  method='spearman')
plot.correlation.heatmap(aa.bact.cor,row.filter=0,col.filter=0)
rm(ind)
rm(aa.bact.cor)
```


## Correlation of gene and amino acid
```{r known_metabolite_correlation_aa_gene,cache=TRUE,fig.height=18,fig.width=20}

plot.correlation.heatmap = function(cor.matrix,
                                    row.filter=0.5,
                                    col.filter=0.5
                                    ){
  cor.matrix <- cor.matrix[
  apply(cor.matrix,1,function(X){max(abs(X),na.rm=TRUE)>row.filter}),
  apply(cor.matrix,2,function(X){max(abs(X),na.rm=TRUE)>col.filter})
  ]
  
  bk <- unique(c(seq(-1,-0.4,length=10),
                 seq(-0.4,0.4,length=50),
                 seq(0.4,1,length=10)))
  hmcols<- colorRampPalette(c('#4575b4','#91bfdb','#e0f3f8',
                            '#ffffbf','#fee090','#fc8d59','#d73027'
                            ))(length(bk)-1)

  pheatmap(cor.matrix, 
           scale='none',fontsize_row = 10, 
           clustering_distance_cols = "manhattan",
           clustering_distance_rows = "manhattan",
           cluster_cols = TRUE,
           color =hmcols, breaks=bk,
           cellwidth = 10, cellheight = 10
  )                              
}

###
row.ann <- subset(minfo,!is.na(Metabolite),select=c("Metabolite","Kingdom",  "Super_class", "Class", "Direct_Parent"))
rownames(row.ann) <- row.ann$Metabolite
aa.ind <- rownames(subset(row.ann,Class=="Amino Acids and Derivatives"))


### all
ind <- intersect(rownames(kmdata.nor),rownames(gene.data))
aa.gene.cor <- cor(gene.data[ind,],
                              kmdata.nor[ind,aa.ind],
                              method='spearman')
plot.correlation.heatmap(aa.gene.cor,row.filter=0.55,col.filter=0)
rm(ind)
rm(aa.gene.cor)


### control
ind <- intersect(intersect(rownames(kmdata.nor),rownames(gene.data)),rownames(subset(sample.info,Type=='COMBO')))
aa.gene.cor <- cor(gene.data[ind,colSums(gene.data[ind,])>0],
                  kmdata.nor[ind,aa.ind],
                  method='spearman')
plot.correlation.heatmap(aa.gene.cor,row.filter=0.6,col.filter=0)
rm(ind)
rm(aa.gene.cor)


### PLEASE-T1
ind <- intersect(intersect(rownames(kmdata.nor),rownames(gene.data)),rownames(subset(sample.info,Type=='PLEASE-T1')))
aa.gene.cor <- cor(gene.data[ind,colSums(gene.data[ind,])>0],
                  kmdata.nor[ind,aa.ind],
                  method='spearman')
plot.correlation.heatmap(aa.gene.cor,row.filter=0.6,col.filter=0)
rm(ind)
rm(aa.gene.cor)

### PLEASE-T1, no antibiotics
ind <- intersect(intersect(rownames(kmdata.nor),rownames(gene.data)),
                 rownames(subset(sample.info,Antibiotics.visit=='Not.Use',Type=='PLEASE-T1')))
aa.gene.cor <- cor(gene.data[ind,colSums(gene.data[ind,])>0],
                  kmdata.nor[ind,aa.ind],
                  method='spearman')
plot.correlation.heatmap(aa.gene.cor,row.filter=0.45,col.filter=0)
rm(ind)
rm(aa.gene.cor)

### PLEASE-T4
ind <- intersect(intersect(rownames(kmdata.nor),rownames(gene.data)),rownames(subset(sample.info,Type=='PLEASE-T4')))
aa.gene.cor <- cor(gene.data[ind,colSums(gene.data[ind,])>0],
                  kmdata.nor[ind,aa.ind],
                  method='spearman')
plot.correlation.heatmap(aa.gene.cor,row.filter=0.5,col.filter=0)
rm(ind)
rm(aa.gene.cor)
```



## Correlation of significant genes and amino acids
```{r known_metabolite_correlation_aa_sig_gene,cache=TRUE,fig.height=18,fig.width=20,eval=FALSE}

#### not finish yet
#### need to use significant genes to calculate the correlation

plot.correlation.heatmap = function(cor.matrix,
                                    row.filter=0.5,
                                    col.filter=0.5
                                    ){
  cor.matrix <- cor.matrix[
  apply(cor.matrix,1,function(X){max(abs(X),na.rm=TRUE)>row.filter}),
  apply(cor.matrix,2,function(X){max(abs(X),na.rm=TRUE)>col.filter})
  ]
  
  bk <- unique(c(seq(-1,-0.4,length=10),
                 seq(-0.4,0.4,length=50),
                 seq(0.4,1,length=10)))
  hmcols<- colorRampPalette(c('#4575b4','#91bfdb','#e0f3f8',
                            '#ffffbf','#fee090','#fc8d59','#d73027'
                            ))(length(bk)-1)

  pheatmap(cor.matrix, 
           scale='none',fontsize_row = 10, 
           clustering_distance_cols = "manhattan",
           clustering_distance_rows = "manhattan",
           cluster_cols = TRUE,
           color =hmcols, breaks=bk,
           cellwidth = 10, cellheight = 10
  )                              
}

###
row.ann <- subset(minfo,!is.na(Metabolite),select=c("Metabolite","Kingdom",  "Super_class", "Class", "Direct_Parent"))
rownames(row.ann) <- row.ann$Metabolite
aa.ind <- rownames(subset(row.ann,Class=="Amino Acids and Derivatives"))

gene.ind <- rownames(gene.cp1.wil[gene.cp1.wil[,'Disease.qvalue']<0.01,])

### all
ind <- intersect(rownames(kmdata.nor),rownames(gene.data))
aa.gene.cor <- cor(gene.data[ind,gene.ind],
                              kmdata.nor[ind,aa.ind],
                              method='spearman')
plot.correlation.heatmap(aa.gene.cor,row.filter=0.5,col.filter=0)
rm(ind)
rm(aa.gene.cor)


### control
ind <- intersect(intersect(rownames(kmdata.nor),rownames(gene.data)),rownames(subset(sample.info,Type=='COMBO')))
aa.gene.cor <- cor(gene.data[ind,colSums(gene.data[ind,])>0],
                  kmdata.nor[ind,aa.ind],
                  method='spearman')
plot.correlation.heatmap(aa.gene.cor,row.filter=0.6,col.filter=0)
rm(ind)
rm(aa.gene.cor)


### PLEASE-T1
ind <- intersect(intersect(rownames(kmdata.nor),rownames(gene.data)),rownames(subset(sample.info,Type=='PLEASE-T1')))
aa.gene.cor <- cor(gene.data[ind,colSums(gene.data[ind,])>0],
                  kmdata.nor[ind,aa.ind],
                  method='spearman')
plot.correlation.heatmap(aa.gene.cor,row.filter=0.6,col.filter=0)
rm(ind)
rm(aa.gene.cor)

### PLEASE-T1, no antibiotics
ind <- intersect(intersect(rownames(kmdata.nor),rownames(gene.data)),
                 rownames(subset(sample.info,Antibiotics.visit=='Not.Use',Type=='PLEASE-T1')))
aa.gene.cor <- cor(gene.data[ind,colSums(gene.data[ind,])>0],
                  kmdata.nor[ind,aa.ind],
                  method='spearman')
plot.correlation.heatmap(aa.gene.cor,row.filter=0.45,col.filter=0)
rm(ind)
rm(aa.gene.cor)

### PLEASE-T4
ind <- intersect(intersect(rownames(kmdata.nor),rownames(gene.data)),rownames(subset(sample.info,Type=='PLEASE-T4')))
aa.gene.cor <- cor(gene.data[ind,colSums(gene.data[ind,])>0],
                  kmdata.nor[ind,aa.ind],
                  method='spearman')
plot.correlation.heatmap(aa.gene.cor,row.filter=0.6,col.filter=0)
rm(ind)
rm(aa.gene.cor)
```


## Correlation of pathway and amino acid
```{r known_metabolite_correlation_aa_pathway,cache=TRUE,fig.height=18,fig.width=20}

plot.correlation.heatmap = function(cor.matrix,
                                    row.filter=0.5,
                                    col.filter=0.5
                                    ){
  cor.matrix <- cor.matrix[
  apply(cor.matrix,1,function(X){max(abs(X),na.rm=TRUE)>row.filter}),
  apply(cor.matrix,2,function(X){max(abs(X),na.rm=TRUE)>col.filter})
  ]
  
  bk <- unique(c(seq(-1,-0.4,length=10),
                 seq(-0.4,0.4,length=50),
                 seq(0.4,1,length=10)))
  hmcols<- colorRampPalette(c('#4575b4','#91bfdb','#e0f3f8',
                            '#ffffbf','#fee090','#fc8d59','#d73027'
                            ))(length(bk)-1)

  pheatmap(cor.matrix, 
           scale='none',fontsize_row = 10, 
           clustering_distance_cols = "manhattan",
           clustering_distance_rows = "manhattan",
           cluster_cols = TRUE,
           color =hmcols, breaks=bk,
           cellwidth = 10, cellheight = 10
  )                              
}

###
row.ann <- subset(minfo,!is.na(Metabolite),select=c("Metabolite","Kingdom",  "Super_class", "Class", "Direct_Parent"))
rownames(row.ann) <- row.ann$Metabolite
aa.ind <- rownames(subset(row.ann,Class=="Amino Acids and Derivatives"))


### all
ind <- intersect(rownames(kmdata.nor),rownames(pathway.data))
aa.path.cor <- cor(pathway.data[ind,],
                              kmdata.nor[ind,aa.ind],
                              method='spearman')
plot.correlation.heatmap(aa.path.cor,row.filter=0.4,col.filter=0)
rm(ind)
rm(aa.path.cor)


### control
ind <- intersect(intersect(rownames(kmdata.nor),rownames(pathway.data)),rownames(subset(sample.info,Type=='COMBO')))
aa.path.cor <- cor(pathway.data[ind,colSums(pathway.data[ind,])>0],
                  kmdata.nor[ind,aa.ind],
                  method='spearman')
plot.correlation.heatmap(aa.path.cor,row.filter=0.5,col.filter=0)
rm(ind)
rm(aa.path.cor)


### PLEASE-T1
ind <- intersect(intersect(rownames(kmdata.nor),rownames(pathway.data)),rownames(subset(sample.info,Type=='PLEASE-T1')))
aa.path.cor <- cor(pathway.data[ind,colSums(pathway.data[ind,])>0],
                  kmdata.nor[ind,aa.ind],
                  method='spearman')
plot.correlation.heatmap(aa.path.cor,row.filter=0.4,col.filter=0)
rm(ind)
rm(aa.path.cor)

### PLEASE-T1, no antibiotics
ind <- intersect(intersect(rownames(kmdata.nor),rownames(pathway.data)),
                 rownames(subset(sample.info,Antibiotics.visit=='Not.Use',Type=='PLEASE-T1')))
aa.path.cor <- cor(pathway.data[ind,colSums(pathway.data[ind,])>0],
                  kmdata.nor[ind,aa.ind],
                  method='spearman')
plot.correlation.heatmap(aa.path.cor,row.filter=0.4,col.filter=0)
rm(ind)
rm(aa.path.cor)

### PLEASE-T4
ind <- intersect(intersect(rownames(kmdata.nor),rownames(pathway.data)),rownames(subset(sample.info,Type=='PLEASE-T4')))
aa.path.cor <- cor(pathway.data[ind,colSums(pathway.data[ind,])>0],
                  kmdata.nor[ind,aa.ind],
                  method='spearman')
plot.correlation.heatmap(aa.path.cor,row.filter=0.4,col.filter=0)
rm(ind)
rm(aa.path.cor)
```


## Correlation of nitrogen pathway and metabolites
```{r known_metabolite_correlation_nitrogen_pathway,cache=TRUE,fig.height=18,fig.width=20}
###
Nitrogen <- pathway.data[,grep('Nitrogen',colnames(pathway.data),ignore.case=TRUE,value=TRUE),drop=FALSE]

### combo
ind <- intersect(intersect(rownames(kmdata.nor),rownames(pathway.data)),rownames(subset(sample.info,Type=='COMBO')))
aa.path.cor.combo <- cor(kmdata.nor[ind,],Nitrogen[ind,],method='spearman')
### please-t1
ind <- intersect(intersect(rownames(kmdata.nor),rownames(pathway.data)),rownames(subset(sample.info,Type=='PLEASE-T1')))
aa.path.cor.pleaset1 <- cor(kmdata.nor[ind,],Nitrogen[ind,],method='spearman')
### please-t4
ind <- intersect(intersect(rownames(kmdata.nor),rownames(pathway.data)),rownames(subset(sample.info,Type=='PLEASE-T4')))
aa.path.cor.pleaset4 <- cor(kmdata.nor[ind,],Nitrogen[ind,],method='spearman')
### please-t1 reponse
ind <- intersect(intersect(rownames(kmdata.nor),rownames(pathway.data)),rownames(subset(sample.info,Type=='PLEASE-T1' & Response == 'Response') ))
aa.path.cor.pleaset1.resp <- cor(kmdata.nor[ind,],Nitrogen[ind,],method='spearman')
### please-t4 reponse
ind <- intersect(intersect(rownames(kmdata.nor),rownames(pathway.data)),rownames(subset(sample.info,Type=='PLEASE-T4' & Response == 'Response') ))
aa.path.cor.pleaset4.resp <- cor(kmdata.nor[ind,],Nitrogen[ind,],method='spearman')
### please-t1 non-reponse
ind <- intersect(intersect(rownames(kmdata.nor),rownames(pathway.data)),rownames(subset(sample.info,Type=='PLEASE-T1' & Response == 'Non.Response') ))
aa.path.cor.pleaset1.nonresp <- cor(kmdata.nor[ind,],Nitrogen[ind,],method='spearman')
### please-t4 non-reponse
ind <- intersect(intersect(rownames(kmdata.nor),rownames(pathway.data)),rownames(subset(sample.info,Type=='PLEASE-T4' & Response == 'Non.Response') ))
aa.path.cor.pleaset4.nonresp <- cor(kmdata.nor[ind,],Nitrogen[ind,],method='spearman')

### merge
data.frame(COMBO=aa.path.cor.combo,
           PLEASET1=aa.path.cor.pleaset1,
           PLEASET4=aa.path.cor.pleaset4,
           PLEASET1.Resp=aa.path.cor.pleaset1.resp,
           PLEASET4.Resp=aa.path.cor.pleaset4.resp,
           PLEASET1.NonResp=aa.path.cor.pleaset1.nonresp,
           PLEASET4.NonResp=aa.path.cor.pleaset4.nonresp) %>% as.data.frame %>%
add_rownames(var='pathway') %>%
#dplyr::arrange(PLEASET1,PLEASET4,COMBO) %>%
dplyr::filter(abs(COMBO)>0.3 | abs(PLEASET1)>0.3 | abs(PLEASET4)>0.3 | abs(PLEASET1.Resp)>0.3 | abs(PLEASET4.Resp)>0.3 | abs(PLEASET1.NonResp)>0.3 | abs(PLEASET4.NonResp)>0.3) %>%
add_back_rownames(row.var = 'pathway') %>% t() %>%
pheatmap(.,scale='none',#fontsize_row = 10, 
           #clustering_distance_cols = "manhattan",
           clustering_distance_rows = "manhattan",
           cluster_rows = FALSE,
           cluster_cols = TRUE,
           #color =hmcols, breaks=bk,
           cellwidth = 8, cellheight = 10 
)     

rm(ind)
rm(aa.path.cor.combo)
rm(aa.path.cor.pleaset1)
rm(aa.path.cor.pleaset4)
rm(aa.path.cor.pleaset1.resp)
rm(aa.path.cor.pleaset4.resp)
rm(aa.path.cor.pleaset1.nonresp)
rm(aa.path.cor.pleaset4.nonresp)
```




## Correlation of nitrogen genes and metabolites
```{r known_metabolite_correlation_nitrogen_genes,cache=TRUE,fig.height=18,fig.width=20}

plot.correlation.heatmap = function(cor.matrix,
                                    row.filter=0.5,
                                    col.filter=0.5
                                    ){
  cor.matrix <- cor.matrix[
  apply(cor.matrix,1,function(X){max(abs(X),na.rm=TRUE)>row.filter}),
  apply(cor.matrix,2,function(X){max(abs(X),na.rm=TRUE)>col.filter})
  ]
  
  bk <- unique(c(seq(-1,-0.4,length=10),
                 seq(-0.4,0.4,length=50),
                 seq(0.4,1,length=10)))
  hmcols<- colorRampPalette(c('#4575b4','#91bfdb','#e0f3f8',
                            '#ffffbf','#fee090','#fc8d59','#d73027'
                            ))(length(bk)-1)

  pheatmap(cor.matrix, 
           scale='none',fontsize_row = 10, 
           clustering_distance_cols = "manhattan",
           clustering_distance_rows = "manhattan",
           cluster_cols = TRUE,
           color =hmcols, breaks=bk,
           cellwidth = 10, cellheight = 10
  )                              
}



nitrogen.gene <- c('K00370__nitrate_reductase_1_alpha_subunit',
  'K00371__nitrate_reductase_1_beta_subunit','K03635__molybdenum_cofactor_biosynthesis_protein_E','K03831__molybdopterin_biosynthesis_protein_Mog','K07715__two_component_system_NtrC_family_response_regulator_YfhA','K04752__nitrogen_regulatory_protein_P_II_2','K04751__nitrogen_regulatory_protein_P_II_1','K07687__two_component_system_NarL_family_captular_synthesis_response','K00363__nitrite_reductase_NAD_P_H_small_subunit','K01990__ABC_2_type_transport_system_ATP_binding_protein','K03637__molybdenum_cofactor_biosynthesis_protein_C','K03753__molybdopterin_guanine_dinucleotide_biosynthesis_protein_B','K03638__molybdenum_cofactor_biosynthesis_protein_B','K02806__PTS_system_nitrogen_regulatory_IIA_component')

###
Nitrogen <- gene.data[,nitrogen.gene]


########################################################
##### Nitrogen and microbiome


### all
ind <- intersect(rownames(Nitrogen),rownames(taxa.data))
pathway.bact.cor <- cor(Nitrogen[ind,],taxa.data[ind,],method='spearman')
plot.correlation.heatmap(pathway.bact.cor,row.filter=0,col.filter=0)
rm(ind)

### control
ind <- intersect(intersect(rownames(Nitrogen),rownames(taxa.data)),rownames(subset(sample.info,Type=='COMBO')))
pathway.bact.cor <- cor(Nitrogen[ind,colSums(Nitrogen[ind,])>0],
                        taxa.data[ind,colSums(taxa.data[ind,])>0],
                        method='spearman')
plot.correlation.heatmap(pathway.bact.cor,row.filter=0,col.filter=0)
rm(ind)

### PLEASE-T1
ind <- intersect(intersect(rownames(Nitrogen),rownames(taxa.data)),rownames(subset(sample.info,Type=='PLEASE-T1')))
pathway.bact.cor <- cor(Nitrogen[ind,colSums(Nitrogen[ind,])>0],
                        taxa.data[ind,colSums(taxa.data[ind,])>0],
                        method='spearman')
plot.correlation.heatmap(pathway.bact.cor,row.filter=0,col.filter=0)
rm(ind)

### PLEASE-T4
ind <- intersect(intersect(rownames(Nitrogen),rownames(taxa.data)),rownames(subset(sample.info,Type=='PLEASE-T4')))
pathway.bact.cor <- cor(Nitrogen[ind,colSums(Nitrogen[ind,])>0],
                        taxa.data[ind,colSums(taxa.data[ind,])>0],
                        method='spearman')
plot.correlation.heatmap(pathway.bact.cor,row.filter=0,col.filter=0)
rm(ind)

### antiTNF
ind <- intersect(intersect(rownames(Nitrogen),rownames(taxa.data)),rownames(subset(sample.info,Treatment=='antiTNF')))
pathway.bact.cor <- cor(Nitrogen[ind,],taxa.data[ind,],method='spearman')
plot.correlation.heatmap(pathway.bact.cor,row.filter=0,col.filter=0)
rm(ind)

### EEN
ind <- intersect(intersect(rownames(Nitrogen),rownames(taxa.data)),rownames(subset(sample.info,Treatment.Specific=='EEN')))
pathway.bact.cor <- cor(Nitrogen[ind,colSums(pathway.data[ind,])>0],
                        taxa.data[ind,colSums(taxa.data[ind,])>0],
                        method='spearman')
plot.correlation.heatmap(pathway.bact.cor,row.filter=0,col.filter=0)
rm(ind)

### response
ind <- intersect(intersect(rownames(Nitrogen),rownames(taxa.data)),rownames(subset(sample.info,Response=='Response')))
pathway.bact.cor <- cor(Nitrogen[ind,],taxa.data[ind,],method='spearman')
plot.correlation.heatmap(pathway.bact.cor,row.filter=0,col.filter=0)
rm(ind)

### non-response
ind <- intersect(intersect(rownames(Nitrogen),rownames(taxa.data)),rownames(subset(sample.info,Response=='Non.Response')))
pathway.bact.cor <- cor(Nitrogen[ind,],taxa.data[ind,],method='spearman')
plot.correlation.heatmap(pathway.bact.cor,row.filter=0,col.filter=0)
rm(ind)

### PLEASE-T1 Response
ind <- intersect(intersect(rownames(Nitrogen),rownames(taxa.data)),rownames(subset(sample.info,Type=='PLEASE-T1' & Response =='Response')))
pathway.bact.cor <- cor(Nitrogen[ind,colSums(Nitrogen[ind,])>0],
                        taxa.data[ind,colSums(taxa.data[ind,])>0],
                        method='spearman')
plot.correlation.heatmap(pathway.bact.cor,row.filter=0,col.filter=0)
rm(ind)


### PLEASE-T4 Response
ind <- intersect(intersect(rownames(Nitrogen),rownames(taxa.data)),rownames(subset(sample.info,Type=='PLEASE-T4' & Response =='Response')))
pathway.bact.cor <- cor(Nitrogen[ind,colSums(Nitrogen[ind,])>0],
                        taxa.data[ind,colSums(taxa.data[ind,])>0],
                        method='spearman')
plot.correlation.heatmap(pathway.bact.cor,row.filter=0,col.filter=0)
rm(ind)



########################################################
##### Nitrogen and metablites


### all
ind <- intersect(rownames(Nitrogen),rownames(kmdata.nor))
pathway.bact.cor <- cor(Nitrogen[ind,],kmdata.nor[ind,],method='spearman')
plot.correlation.heatmap(pathway.bact.cor,row.filter=0,col.filter=0.35)
rm(ind)

### control
ind <- intersect(intersect(rownames(Nitrogen),rownames(kmdata.nor)),rownames(subset(sample.info,Type=='COMBO')))
pathway.bact.cor <- cor(Nitrogen[ind,colSums(Nitrogen[ind,])>0],
                        kmdata.nor[ind,colSums(kmdata.nor[ind,])>0],
                        method='spearman')
plot.correlation.heatmap(pathway.bact.cor,row.filter=0,col.filter=0.45)
rm(ind)

### PLEASE-T1
ind <- intersect(intersect(rownames(Nitrogen),rownames(kmdata.nor)),rownames(subset(sample.info,Type=='PLEASE-T1')))
pathway.bact.cor <- cor(Nitrogen[ind,colSums(Nitrogen[ind,])>0],
                        kmdata.nor[ind,colSums(kmdata.nor[ind,])>0],
                        method='spearman')
plot.correlation.heatmap(pathway.bact.cor,row.filter=0,col.filter=0.4)
rm(ind)

### PLEASE-T4
ind <- intersect(intersect(rownames(Nitrogen),rownames(kmdata.nor)),rownames(subset(sample.info,Type=='PLEASE-T4')))
pathway.bact.cor <- cor(Nitrogen[ind,colSums(Nitrogen[ind,])>0],
                        kmdata.nor[ind,colSums(kmdata.nor[ind,])>0],
                        method='spearman')
plot.correlation.heatmap(pathway.bact.cor,row.filter=0,col.filter=0.4)
rm(ind)

### antiTNF
ind <- intersect(intersect(rownames(Nitrogen),rownames(kmdata.nor)),rownames(subset(sample.info,Treatment=='antiTNF')))
pathway.bact.cor <- cor(Nitrogen[ind,],kmdata.nor[ind,],method='spearman')
plot.correlation.heatmap(pathway.bact.cor,row.filter=0,col.filter=0)
rm(ind)

### EEN
ind <- intersect(intersect(rownames(Nitrogen),rownames(kmdata.nor)),rownames(subset(sample.info,Treatment.Specific=='EEN')))
pathway.bact.cor <- cor(Nitrogen[ind,colSums(Nitrogen[ind,])>0],
                        kmdata.nor[ind,colSums(kmdata.nor[ind,])>0],
                        method='spearman')
plot.correlation.heatmap(pathway.bact.cor,row.filter=0,col.filter=0)
rm(ind)

### response
ind <- intersect(intersect(rownames(Nitrogen),rownames(kmdata.nor)),rownames(subset(sample.info,Response=='Response')))
pathway.bact.cor <- cor(Nitrogen[ind,],kmdata.nor[ind,],method='spearman')
plot.correlation.heatmap(pathway.bact.cor,row.filter=0,col.filter=0)
rm(ind)

### non-response
ind <- intersect(intersect(rownames(Nitrogen),rownames(kmdata.nor)),rownames(subset(sample.info,Response=='Non.Response')))
pathway.bact.cor <- cor(Nitrogen[ind,],kmdata.nor[ind,],method='spearman')
plot.correlation.heatmap(pathway.bact.cor,row.filter=0,col.filter=0)
rm(ind)

### PLEASE-T1 Response
ind <- intersect(intersect(rownames(Nitrogen),rownames(kmdata.nor)),rownames(subset(sample.info,Type=='PLEASE-T1' & Response =='Response')))
pathway.bact.cor <- cor(Nitrogen[ind,colSums(Nitrogen[ind,])>0],
                        kmdata.nor[ind,colSums(kmdata.nor[ind,])>0],
                        method='spearman')
plot.correlation.heatmap(pathway.bact.cor,row.filter=0,col.filter=0)
rm(ind)


### PLEASE-T4 Response
ind <- intersect(intersect(rownames(Nitrogen),rownames(kmdata.nor)),rownames(subset(sample.info,Type=='PLEASE-T4' & Response =='Response')))
pathway.bact.cor <- cor(Nitrogen[ind,colSums(Nitrogen[ind,])>0],
                        kmdata.nor[ind,colSums(kmdata.nor[ind,])>0],
                        method='spearman')
plot.correlation.heatmap(pathway.bact.cor,row.filter=0,col.filter=0)
rm(ind)

rm(Nitrogen)

```


## Correlation of metabolites and FCP
```{r known_metabolite_correlation_with_fcp,cache=TRUE,fig.height=18,fig.width=20}

#####################################################################
############# PLEASE-T1 ###################################################

### antibiotic use 
fcp.t1 <- subset(sample.info,Type=='PLEASE-T1' & Antibiotics.visit == 'Use' & !is.na(FCP),select=FCP)
ind <- intersect(rownames(fcp.t1),rownames(kmdata.nor))
fcp.t1 <- fcp.t1[ind,]
cor.antibio <- cor(kmdata.nor[ind,],fcp.t1,method='spearman') 

### no antibiotic use
fcp.t1 <- subset(sample.info,Type=='PLEASE-T1' & Antibiotics.visit == 'Not.Use' & !is.na(FCP),select=FCP)
ind <- intersect(rownames(fcp.t1),rownames(kmdata.nor))
fcp.t1 <- fcp.t1[ind,]
cor.no.antibio <- cor(kmdata.nor[ind,],fcp.t1,method='spearman') 

### heatmap, no antibiotic use
bk <- unique(c(seq(-1,-0.4,length=10),
                 seq(-0.4,0.4,length=50),
                 seq(0.4,1,length=10)))
hmcols<- colorRampPalette(c('#4575b4','#91bfdb','#e0f3f8',
                            '#ffffbf','#fee090','#fc8d59','#d73027'
                            ))(length(bk)-1)

row.ann <- subset(minfo,!is.na(Metabolite),select=c("Metabolite","Kingdom",  "Super_class", "Class", "Direct_Parent"))
rownames(row.ann) <- row.ann$Metabolite

### cluster metabolites
pheatmap(cor.no.antibio, 
           scale='none',fontsize_row = 10, 
           #clustering_distance_cols = "manhattan",
           clustering_distance_rows = "manhattan",
           show_rownames = FALSE,
           cluster_cols = FALSE,
           color =hmcols, breaks=bk,
           cellwidth = 5, cellheight = 1.5,
           annotation_row = row.ann[,c('Super_class'),drop=FALSE]
) 

### no cluster, order by metabolite type
pdata <- right_join(add_rownames(row.ann,var='metabolites'),
add_rownames(as.data.frame(cor.no.antibio),var='metabolites'),by='metabolites') %>%
arrange(Super_class)  %>% 
add_back_rownames(row.var = 'metabolites')


pheatmap(pdata[,'V1',drop=FALSE], 
           scale='none',fontsize_row = 10, 
           #clustering_distance_cols = "manhattan",
           #clustering_distance_rows = "manhattan",
           show_rownames = FALSE,
           cluster_cols = FALSE,
           cluster_rows = FALSE,
           color =hmcols, breaks=bk,
           cellwidth = 5, cellheight = 1.5,
           annotation_row = pdata[,c('Super_class'),drop=FALSE]
) 


### check if the metabolites in the two data frames are in the same order
if(!identical(rownames(cor.antibio),rownames(cor.no.antibio))){
  print('Row names (metabolites) are not identical')
}

### merge two correlation data frames
### Class
data.frame(antibio=cor.antibio,no.antibio=cor.no.antibio) %>%
add_rownames(var='Metabolite') %>%
right_join(dplyr::select(minfo,Metabolite,Super_class,Class,Direct_Parent),by='Metabolite') %>%
dplyr::filter(!is.na(Class)) %>%
group_by(Class) %>%
dplyr::mutate(count=n()) %>%
dplyr::filter(count>10) %>%
(function(tdata){
  tdata %>% group_by(Class) %>%
  summarise_each(funs(wilcox.test(., antibio, pair = TRUE)$p.value), vars = no.antibio) %>%
  dplyr::rename(pval=vars) %>%
  dplyr::mutate(qval = p.adjust(pval,method='fdr')) %>%
  dplyr::arrange(qval) %>%
  print
  return(tdata)
}) %>%
tidyr::gather(antibiotic,cor,-Metabolite,-Super_class,-Class,-Direct_Parent, -count) %>%
(function(tdata){
  p<-tdata %>% dplyr::filter(Class=='Fatty Acids and Conjugates') %>% 
  ggplot(aes(x=factor(Metabolite),y=cor,fill=antibiotic)) + geom_bar(position="dodge",stat = "identity") + 
  theme(axis.text=element_text(size=15)) +
  coord_flip() + xlab('')+ylab('Correlation with FCP')+ggtitle('PLEASE-T1 Class')
  plot(p)
  return(tdata)
}) %>%
ggplot(aes(Class,cor)) + 
geom_boxplot(aes(fill = antibiotic))+
coord_flip()+ theme(axis.text=element_text(size=15)) +
xlab('')+ylab('Correlation with FCP')+ggtitle('PLEASE-T1 Class')



### merge two correlation data frames
### Direct_Parent
data.frame(antibio=cor.antibio,no.antibio=cor.no.antibio) %>%
add_rownames(var='Metabolite') %>%
right_join(dplyr::select(minfo,Metabolite,Super_class,Class,Direct_Parent),by='Metabolite') %>%
dplyr::filter(!is.na(Direct_Parent)) %>%
group_by(Direct_Parent) %>%
dplyr::mutate(count=n()) %>%
dplyr::filter(count>10) %>%
(function(tdata){
  tdata %>% group_by(Direct_Parent) %>%
  summarise_each(funs(wilcox.test(., antibio, pair = TRUE)$p.value), vars = no.antibio) %>%
  dplyr::rename(pval=vars) %>%
  dplyr::mutate(qval = p.adjust(pval,method='fdr')) %>%
  dplyr::arrange(qval) %>%
  print
  return(tdata)
}) %>%
tidyr::gather(antibiotic,cor,-Metabolite,-Super_class,-Class,-Direct_Parent,-count) %>%
ggplot(aes(Direct_Parent,cor)) + 
geom_boxplot(aes(fill = antibiotic))+
coord_flip()+ theme(axis.text=element_text(size=15)) +
xlab('')+ylab('Correlation with FCP')+ggtitle('PLEASE-T1 Direct Parent')

rm(fcp.t1)
rm(ind)
rm(cor.antibio)
rm(cor.no.antibio)



#####################################################################
############# PLEASE-T4 #############################################
#####################################################################

### antibiotic use 
fcp.t4 <- subset(sample.info,Type=='PLEASE-T4' & Antibiotics.visit == 'Use' & !is.na(FCP),select=FCP)
ind <- intersect(rownames(fcp.t4),rownames(kmdata.nor))
fcp.t4 <- fcp.t4[ind,]
cor.antibio <- cor(kmdata.nor[ind,],fcp.t4,method='spearman') 

### no antibiotic use
fcp.t4 <- subset(sample.info,Type=='PLEASE-T4' & Antibiotics.visit == 'Not.Use' & !is.na(FCP),select=FCP)
ind <- intersect(rownames(fcp.t4),rownames(kmdata.nor))
fcp.t4 <- fcp.t4[ind,]
cor.no.antibio <- cor(kmdata.nor[ind,],fcp.t4,method='spearman') 

### check if the metabolites in the two data frames are in the same order
if(!identical(rownames(cor.antibio),rownames(cor.no.antibio))){
  print('Row names (metabolites) are not identical')
}

### merge two correlation data frames
### Class
data.frame(antibio=cor.antibio,no.antibio=cor.no.antibio) %>%
add_rownames(var='Metabolite') %>%
right_join(dplyr::select(minfo,Metabolite,Super_class,Class,Direct_Parent),by='Metabolite') %>%
dplyr::filter(!is.na(Class)) %>%
group_by(Class) %>%
dplyr::mutate(count=n()) %>%
dplyr::filter(count>10) %>%
(function(tdata){
  tdata %>% group_by(Class) %>%
  summarise_each(funs(wilcox.test(., antibio, pair = TRUE)$p.value), vars = no.antibio) %>%
  dplyr::rename(pval=vars) %>%
  dplyr::mutate(qval = p.adjust(pval,method='fdr')) %>%
  dplyr::arrange(qval) %>%
  print
  return(tdata)
}) %>%
tidyr::gather(antibiotic,cor,-Metabolite,-Super_class,-Class,-Direct_Parent, -count) %>%
(function(tdata){
  p<-tdata %>% dplyr::filter(Class=='Fatty Acids and Conjugates') %>% 
  ggplot(aes(x=factor(Metabolite),y=cor,fill=antibiotic)) + geom_bar(position="dodge",stat = "identity") + 
  theme(axis.text=element_text(size=15)) +
  coord_flip() + xlab('')+ylab('Correlation with FCP')+ggtitle('PLEASE-T4 Class')
  plot(p)
  return(tdata)
}) %>%
ggplot(aes(Class,cor)) + 
geom_boxplot(aes(fill = antibiotic))+
coord_flip()+ theme(axis.text=element_text(size=15)) +
xlab('')+ylab('Correlation with FCP')+ggtitle('PLEASE-T4 Class')


### merge two correlation data frames
### Direct_Parent
data.frame(antibio=cor.antibio,no.antibio=cor.no.antibio) %>%
add_rownames(var='Metabolite') %>%
right_join(dplyr::select(minfo,Metabolite,Super_class,Class,Direct_Parent),by='Metabolite') %>%
dplyr::filter(!is.na(Direct_Parent)) %>%
group_by(Direct_Parent) %>%
dplyr::mutate(count=n()) %>%
dplyr::filter(count>10) %>%
(function(tdata){
  tdata %>% group_by(Direct_Parent) %>%
  summarise_each(funs(wilcox.test(., antibio, pair = TRUE)$p.value), vars = no.antibio) %>%
  dplyr::rename(pval=vars) %>%
  dplyr::mutate(qval = p.adjust(pval,method='fdr')) %>%
  dplyr::arrange(qval) %>%
  print
  return(tdata)
}) %>%
tidyr::gather(antibiotic,cor,-Metabolite,-Super_class,-Class,-Direct_Parent,-count) %>%
ggplot(aes(Direct_Parent,cor)) + 
geom_boxplot(aes(fill = antibiotic))+
coord_flip()+ theme(axis.text=element_text(size=15)) +
xlab('')+ylab('Correlation with FCP')+ggtitle('PLEASE-T4 Direct Parent')

rm(fcp.t4)
rm(ind)
rm(cor.antibio)
rm(cor.no.antibio)
```

## Correlation of microbiome and amino acid by MIC
```{r known_metabolite_correlation_aa_microbiome_by_mic,cache=TRUE,fig.height=18,fig.width=20}

plot.correlation.heatmap = function(cor.matrix,
                                    row.filter=0.5,
                                    col.filter=0.5
                                    ){
  cor.matrix <- cor.matrix[
  apply(cor.matrix,1,function(X){max(abs(X),na.rm=TRUE)>row.filter}),
  apply(cor.matrix,2,function(X){max(abs(X),na.rm=TRUE)>col.filter})
  ]
  
  bk <- unique(c(seq(0,0.2,length=10),
                 seq(0.2,0.4,length=50),
                 seq(0.4,1,length=10)))
  hmcols<- colorRampPalette(c('#4575b4','#91bfdb','#e0f3f8',
                            '#ffffbf','#fee090','#fc8d59','#d73027'
                            ))(length(bk)-1)

  pheatmap(cor.matrix, 
           scale='none',fontsize_row = 10, 
           clustering_distance_cols = "manhattan",
           clustering_distance_rows = "manhattan",
           cluster_cols = TRUE,
           color =hmcols, breaks=bk,
           cellwidth = 10, cellheight = 10
  )                              
}

###
row.ann <- subset(minfo,!is.na(Metabolite),select=c("Metabolite","Kingdom",  "Super_class", "Class", "Direct_Parent"))
rownames(row.ann) <- row.ann$Metabolite
aa.ind <- rownames(subset(row.ann,Class=="Amino Acids and Derivatives"))


### all
ind <- intersect(rownames(kmdata.nor),rownames(taxa.data))
aa.bact.cor <- mine(data.frame(taxa.data[ind,],kmdata.nor[ind,aa.ind]) )
plot.correlation.heatmap(
  aa.bact.cor$MIC[1:ncol(kmdata.nor[ind,aa.ind]),(ncol(kmdata.nor[ind,aa.ind])+1):ncol(aa.bact.cor$MIC)],
  row.filter=0,col.filter=0)
rm(ind)
rm(aa.bact.cor)


### control
ind <- intersect(intersect(rownames(kmdata.nor),rownames(taxa.data)),rownames(subset(sample.info,Type=='COMBO')))
aa.bact.cor <- mine(data.frame(taxa.data[ind,colSums(taxa.data[ind,])>0],kmdata.nor[ind,aa.ind]) )
plot.correlation.heatmap(
  aa.bact.cor$MIC[1:ncol(kmdata.nor[ind,aa.ind]),(ncol(kmdata.nor[ind,aa.ind])+1):ncol(aa.bact.cor$MIC)],
  row.filter=0,col.filter=0)
rm(ind)
rm(aa.bact.cor)


### PLEASE-T1
ind <- intersect(intersect(rownames(kmdata.nor),rownames(taxa.data)),rownames(subset(sample.info,Type=='PLEASE-T1')))
aa.bact.cor <- mine(data.frame(taxa.data[ind,colSums(taxa.data[ind,])>0],kmdata.nor[ind,aa.ind]) )
plot.correlation.heatmap(
  aa.bact.cor$MIC[1:ncol(kmdata.nor[ind,aa.ind]),(ncol(kmdata.nor[ind,aa.ind])+1):ncol(aa.bact.cor$MIC)],
  row.filter=0,col.filter=0)
rm(ind)
rm(aa.bact.cor)

### PLEASE-T1, no antibiotic use
ind <- intersect(intersect(rownames(kmdata.nor),rownames(taxa.data)),
                 rownames(subset(sample.info,Antibiotics.visit=='Not.Use',Type=='PLEASE-T1')))
aa.bact.cor <- mine(data.frame(taxa.data[ind,colSums(taxa.data[ind,])>0],kmdata.nor[ind,aa.ind]) )
plot.correlation.heatmap(
  aa.bact.cor$MIC[1:ncol(kmdata.nor[ind,aa.ind]),(ncol(kmdata.nor[ind,aa.ind])+1):ncol(aa.bact.cor$MIC)],
  row.filter=0,col.filter=0)
rm(ind)
rm(aa.bact.cor)

### PLEASE-T4
ind <- intersect(intersect(rownames(kmdata.nor),rownames(taxa.data)),rownames(subset(sample.info,Type=='PLEASE-T4')))
aa.bact.cor <- mine(data.frame(taxa.data[ind,colSums(taxa.data[ind,])>0],kmdata.nor[ind,aa.ind]) )
plot.correlation.heatmap(
  aa.bact.cor$MIC[1:ncol(kmdata.nor[ind,aa.ind]),(ncol(kmdata.nor[ind,aa.ind])+1):ncol(aa.bact.cor$MIC)],
  row.filter=0,col.filter=0)
rm(ind)
rm(aa.bact.cor)
```




## Responders vs. non-responders

### Random Forest 
```{r known_metabolite_random_forest_of_r_vs_nr,cache=TRUE,fig.width=10}
### EEN T1
random_forest(kmdata.nor,
              subset(sample.info,Time==1 & Treatment.Specific=='EEN' & !is.na(Response),
                     select=Response),
              plot.MDS=TRUE,plot.importance=TRUE,print.result=FALSE,
              guess.num=50,boxplot.top=5,
              title="Response vs. non-response",boxplot.y="Abundance",
              seed=100)
### EEN T4
random_forest(kmdata.nor,
              subset(sample.info,Time==4 & Treatment.Specific=='EEN' & !is.na(Response),
                     select=Response),
              plot.MDS=TRUE,plot.importance=TRUE,print.result=FALSE,
              guess.num=50,boxplot.top=5,
              title="Response vs. non-response",boxplot.y="Abundance",
              seed=100)


### antiTNF T1
random_forest(kmdata.nor,
              subset(sample.info,Time==1 & Treatment.Specific=='antiTNF' & !is.na(Response),
                     select=Response),
              plot.MDS=TRUE,plot.importance=TRUE,print.result=FALSE,
              guess.num=50,boxplot.top=5,
              title="Response vs. non-response",boxplot.y="Abundance",
              seed=100)

### antiTNF T4
random_forest(kmdata.nor,
              subset(sample.info,Time==4 & Treatment.Specific=='antiTNF' & !is.na(Response),
                     select=Response),
              plot.MDS=TRUE,plot.importance=TRUE,print.result=FALSE,
              guess.num=50,boxplot.top=5,
              title="Response vs. non-response",boxplot.y="Abundance",
              seed=100)
```

### Wilcoxon rank test
```{r known_metabolites_compare_response_vs_nonresponse,cache=TRUE}
### compare response vs. non-response

### T1: R vs. NR
kmeta.t1.RvsNR <- wilcox_test(kmdata.nor,subset(sample.info,Time==1 &  !is.na(Response),
                                  select=Response),plot.heatmap=FALSE,boxplot.top=10,boxplot.y='Abundance')
### T4: R vs. NR
kmeta.t4.RvsNR <- wilcox_test(kmdata.nor,subset(sample.info,Time==4 & !is.na(Response),
                                  select=Response),plot.heatmap=FALSE,boxplot.top=10,boxplot.y='Abundance')


### EEN T1: R vs. NR
kmeta.een.t1.RvsNR <- wilcox_test(kmdata.nor,subset(sample.info,Time==1 & Treatment.Specific=='EEN' & !is.na(Response),
                                  select=Response),plot.heatmap=FALSE,boxplot.top=10,boxplot.y='Abundance')
### EEN T4: R vs. NR
kmeta.een.t4.RvsNR <- wilcox_test(kmdata.nor,subset(sample.info,Time==4 & Treatment.Specific=='EEN' & !is.na(Response),
                                  select=Response),plot.heatmap=FALSE,boxplot.top=10,boxplot.y='Abundance')
### aniTNF T1: R vs. NR
kmeta.antiTNF.t1.RvsNR <- wilcox_test(kmdata.nor,subset(sample.info,Time==1 & Treatment.Specific=='antiTNF' & !is.na(Response),
                                  select=Response),plot.heatmap=FALSE,boxplot.top=10,boxplot.y='Abundance')
### antiTNF T4: R vs. NR
kmeta.antiTNF.t4.RvsNR <- wilcox_test(kmdata.nor,subset(sample.info,Time==4 & Treatment.Specific=='antiTNF' & !is.na(Response),
                                  select=Response),plot.heatmap=FALSE,boxplot.top=10,boxplot.y='Abundance')

```


## T1 vs. T2/T4
```{r known_metabolites_compare_time_points,cache=TRUE}
options(width=200)
### compare time points
### !!!!! No T3 data !!!!!!!

#### by treatment only
add_rownames(as.data.frame(kmdata.nor),var='Sample') %>%
  gather(Meta,Abu,-Sample) %>%
  left_join(dplyr::select(add_rownames(sample.info,var='Sample'),Subject,Sample,Time,Treatment.Specific,Response), by='Sample') %>%  
  filter(!is.na(Time) & Time!='3') %>% 
  mutate(Time=paste('T',Time,sep='')) %>%
  dplyr::select(-Sample) %>% 
  spread(Time,Abu) %>%
  group_by(Treatment.Specific,Meta) %>%
  summarise_each(funs(wilcox.test(., T1, pair = TRUE)$p.value), vars = T2:T4) %>%
  ## raw pvalues
  rename(T1.vs.T2.p=vars1,T1.vs.T4.p=vars2) %>%
  mutate(T1.vs.T2.q = p.adjust(T1.vs.T2.p,method='fdr'),
         T1.vs.T4.q = p.adjust(T1.vs.T4.p,method='fdr')
         ) %>%
  mutate_each(funs(signif(.,digits=2)),T1.vs.T2.p:T1.vs.T4.q) %>%
  arrange(Treatment.Specific,T1.vs.T4.p) %>%
  #dplyr::filter(T1.vs.T2.p<0.05 | T1.vs.T4.p<0.05) %>%
  #as.data.frame
  (function(pdata){
    pdata %>% group_by(Treatment.Specific) %>%
    summarize(total=n(),
              T1.vs.T2.sig.count=sum(T1.vs.T2.q<0.05),
              T1.vs.T4.sig.count=sum(T1.vs.T4.q<0.05)) %>%
    print
    return(pdata)
  }) %>%
  write.csv(file=paste('../3_Result/2015_03_05_Known_Metabolites_Compare_Time_Points_by_Treatment_Wilcox.csv',sep=''),row.names=FALSE)


#### by treatment and response
add_rownames(as.data.frame(kmdata.nor),var='Sample') %>%
  gather(Meta,Abu,-Sample) %>%
  left_join(dplyr::select(add_rownames(sample.info,var='Sample'),Subject,Sample,Time,Treatment.Specific,Response), by='Sample') %>%  
  filter(!is.na(Time) & Time!='3') %>% 
  mutate(Time=paste('T',Time,sep='')) %>%
  dplyr::select(-Sample) %>% 
  spread(Time,Abu) %>%
  group_by(Treatment.Specific,Response,Meta) %>%
  summarise_each(funs(wilcox.test(., T1, pair = TRUE)$p.value), vars = T2:T4) %>%
  na.omit() %>%
  ## raw pvalues
  rename(T1.vs.T2.p=vars1,T1.vs.T4.p=vars2) %>%
  mutate(T1.vs.T2.q = p.adjust(T1.vs.T2.p,method='fdr'),
         T1.vs.T4.q = p.adjust(T1.vs.T4.p,method='fdr')
         ) %>%
  mutate_each(funs(signif(.,digits=2)),T1.vs.T2.p:T1.vs.T4.q) %>%
  arrange(Treatment.Specific,Response,T1.vs.T4.p) %>%
  dplyr::filter(T1.vs.T2.p<0.05 | T1.vs.T4.p<0.05) %>%
  as.data.frame
options(width=90)
```



## EEN vs. antiTNF
```{r known_metabolites_compare_treatments,cache=TRUE}
### compare treatments

### T4: EEN vs. antiTNF
kmeta.t4.EENvsAntiTNF <- wilcox_test(kmdata.nor,subset(sample.info,Time==4 & Treatment.Specific!='PEN',
                                  select=Treatment.Specific),plot.heatmap=FALSE,boxplot.top=0.05,boxplot.y='Abundance')
### T1: EEN vs. antiTNF as a control test 
### Check the baseline, since no treatment yet, they should not be different
kmeta.t1.EENvsAntiTNF <- wilcox_test(kmdata.nor,subset(sample.info,Time==1 & Treatment.Specific!='PEN',
                                  select=Treatment.Specific),plot.heatmap=FALSE,boxplot.top=5,boxplot.y='Abundance')


### T4 NR: EEN vs. antiTNF
kmeta.t4.nr.EENvsAntiTNF <- wilcox_test(kmdata.nor,
                       subset(sample.info,Time==4 & Treatment.Specific!='PEN' & Response=='Non.Response',
                         select=Treatment.Specific),plot.heatmap=FALSE,boxplot.top=5,boxplot.y='Abundance')
### T1 NR: EEN vs. antiTNF as a control test 
### Check the baseline, since no treatment yet, they should not be different
kmeta.t1.nr.EENvsAntiTNF <- wilcox_test(kmdata.nor,
                       subset(sample.info,Time==1 & Treatment.Specific!='PEN' & Response=='Non.Response',
                         select=Treatment.Specific),plot.heatmap=FALSE,boxplot.top=5,boxplot.y='Abundance')

### T4 R: EEN vs. antiTNF
kmeta.t4.r.EENvsAntiTNF <- wilcox_test(kmdata.nor,
                       subset(sample.info,Time==4 & Treatment.Specific!='PEN' & Response=='Response',
                         select=Treatment.Specific),plot.heatmap=FALSE,boxplot.top=0.05,boxplot.y='Abundance')
### T1 R: EEN vs. antiTNF as a control test 
### Check the baseline, since no treatment yet, they should not be different
kmeta.t1.r.EENvsAntiTNF <- wilcox_test(kmdata.nor,
                       subset(sample.info,Time==1 & Treatment.Specific!='PEN' & Response=='Response',
                         select=Treatment.Specific),plot.heatmap=FALSE,boxplot.top=5,boxplot.y='Abundance')


### T4: EEN vs. PEN
kmeta.t4.EENvsPEN <- wilcox_test(kmdata.nor,subset(sample.info,Time==4 & Treatment.Specific!='antiTNF',
                                  select=Treatment.Specific),plot.heatmap=FALSE,boxplot.top=0.05,boxplot.y='Abundance')

### T1: EEN vs. PEN as a control test 
### Check the baseline, since no treatment yet, they should not be different
kmeta.t1.EENvsPEN <- wilcox_test(kmdata.nor,subset(sample.info,Time==1 & Treatment.Specific!='antiTNF',
                                  select=Treatment.Specific),plot.heatmap=FALSE,boxplot.top=5,boxplot.y='Abundance')
### T2: EEN vs. PEN as a control test 
kmeta.t2.EENvsPEN <- wilcox_test(kmdata.nor,subset(sample.info,Time==2 & Treatment.Specific!='antiTNF',
                                  select=Treatment.Specific),plot.heatmap=FALSE,boxplot.top=5,boxplot.y='Abundance')

### T4: PEN vs. antiTNF
kmeta.t4.PENvsAntiTNF <- wilcox_test(kmdata.nor,subset(sample.info,Time==4 & Treatment.Specific!='EEN',
                                  select=Treatment.Specific),plot.heatmap=FALSE,boxplot.top=5,boxplot.y='Abundance')

### T1: PEN vs. antiTNF as a control test 
### Check the baseline, since no treatment yet, they should not be different
kmeta.t1.PENvsAntiTNF <- wilcox_test(kmdata.nor,subset(sample.info,Time==1 & Treatment.Specific!='EEN',
                                  select=Treatment.Specific),plot.heatmap=FALSE,boxplot.top=5,boxplot.y='Abundance')
```


## Human DNA, bile acids and acyl-Carnitines

### Merge data
Merge the following data: human DNA, primary bile acids, secondary bile acids, and acyl-carnitines. Also calculate the ratio of primary / secondary bile acids.
```{r known_metabolite_human_dna_bile_acid_acyl_carnitine,cache=TRUE}
ba.carnitine <- 
## Human DNA
add_rownames(sample.info,var='Sample') %>%
dplyr::select(Sample,Human.Per) %>%
## Secondary bile acids
left_join(dplyr::select(add_rownames(as.data.frame(kmdata.nor),var='Sample'),
              Sample,`Deoxycholic acid`,`lithocholate`,
              `Glycolithocholic acid`,`Lithocholic acid`),
          by='Sample') %>%
## Primary bile acids
left_join(dplyr::select(add_rownames(as.data.frame(kmdata.nor),var='Sample'),
              Sample,`Glycochenodeoxycholic acid`,
              `Glycocholic acid`,`Taurochenodesoxycholic acid`,
              `Taurocholic acid`),
          by='Sample') %>%
## acyl-Carnitines
left_join(dplyr::select(add_rownames(as.data.frame(kmdata.nor),var='Sample'),
              Sample,contains('carnitine')),
          by='Sample') %>%
## primary/secondary ratio
rowwise() %>%
#mutate(pri.sec.ratio=`Glycochenodeoxycholic acid`)
mutate(pri.sec.ratio=
              median(c(`Glycochenodeoxycholic acid`,
              `Glycocholic acid`,`Taurochenodesoxycholic acid`,
              `Taurocholic acid`),na.rm=TRUE) /
               median(c(`Deoxycholic acid`,`lithocholate`,
              `Glycolithocholic acid`,`Lithocholic acid`),na.rm=TRUE)
       ) %>%
## remove NAs
na.omit %>%
## 
add_back_rownames(row.var='Sample')
```



### Correlation among human DNA, FCP, BA, and cyl-Carnitines
```{r known_metabolite_ba_carnitine_correlation,cache=TRUE,fig.height=18,fig.width=20}

plot.correlation.heatmap = function(cor.matrix,
                                    row.filter=0.5,
                                    col.filter=0.5
                                    ){
  cor.matrix <- cor.matrix[
  apply(cor.matrix,1,function(X){max(abs(X),na.rm=TRUE)>row.filter}),
  apply(cor.matrix,2,function(X){max(abs(X),na.rm=TRUE)>col.filter})
  ]
  
  bk <- unique(c(seq(-1,-0.4,length=10),
                 seq(-0.4,0.4,length=50),
                 seq(0.4,1,length=10)))
  hmcols<- colorRampPalette(c('#4575b4','#91bfdb','#e0f3f8',
                            '#ffffbf','#fee090','#fc8d59','#d73027'
                            ))(length(bk)-1)

  pheatmap(cor.matrix, 
           scale='none',fontsize_row = 10, 
           clustering_distance_cols = "manhattan",
           clustering_distance_rows = "manhattan",
           cluster_cols = TRUE,
           color =hmcols, breaks=bk,
           cellwidth = 10, cellheight = 10
  )                              
}

###
ba.carnitine.cor <-
ba.carnitine %>% add_rownames(var='Sample') %>%
left_join(add_rownames(sample.info,var='Sample') %>% dplyr::select(Sample,FCP),by='Sample') %>%
add_back_rownames(row.var = 'Sample')


### all
ba.cor <- cor(ba.carnitine.cor,method='spearman',use="pairwise.complete.obs")
plot.correlation.heatmap(ba.cor,row.filter=0,col.filter=0)
rm(ind)
rm(ba.cor)


### control
### remove FCP
ind <- intersect(rownames(ba.carnitine.cor),rownames(subset(sample.info,Type=='COMBO')))
ba.cor <- cor(ba.carnitine.cor[ind,] %>% dplyr::select(-FCP),method='spearman',use="pairwise.complete.obs")
plot.correlation.heatmap(ba.cor,row.filter=0,col.filter=0)
rm(ind)
rm(ba.cor)



### PLEASE-T1
ind <- intersect(rownames(ba.carnitine.cor),rownames(subset(sample.info,Type=='PLEASE-T1')))
ba.cor <- cor(ba.carnitine.cor[ind,],method='spearman',use="pairwise.complete.obs")
plot.correlation.heatmap(ba.cor,row.filter=0,col.filter=0)
rm(ind)
rm(ba.cor)

### PLEASE-T4, no antibiotics
ind <- intersect(rownames(ba.carnitine.cor),
                 rownames(subset(sample.info,Antibiotics.visit=='Not.Use',Type=='PLEASE-T1')))
ba.cor <- cor(ba.carnitine.cor[ind,],method='spearman',use="pairwise.complete.obs")
plot.correlation.heatmap(ba.cor,row.filter=0,col.filter=0)
rm(ind)
rm(ba.cor)

### PLEASE-T4
ind <- intersect(rownames(ba.carnitine.cor),rownames(subset(sample.info,Type=='PLEASE-T4')))
ba.cor <- cor(ba.carnitine.cor[ind,],method='spearman',use="pairwise.complete.obs")
plot.correlation.heatmap(ba.cor,row.filter=0,col.filter=0)
rm(ind)
rm(ba.cor)
```



###  Random Forest: Normal vs. IBD
```{r known_metabolite_ba_carnitine_random_forest_of_normal_vs_crohn,cache=TRUE}
random_forest(ba.carnitine,
              subset(sample.info,Type=='COMBO' | Type=='PLEASE-T1',select=Disease),
              #sample.info[,'Disease',drop=FALSE],
              plot.MDS=TRUE,plot.importance=TRUE,print.result=FALSE,
              guess.num=50,boxplot.top=10,
              title="Normal vs. Crohn",boxplot.y="Abundance",
              seed=100)
```

###  Random Forest: Normal vs. IBD, carnitine, bile acid and human DNA
```{r known_metabolite_ba_carnitine_random_forest_of_normal_vs_crohn_carnitine,cache=TRUE}

ca.ba.dna.rf <- random_forest(ba.carnitine,
              subset(sample.info,Type=='COMBO' | Type=='PLEASE-T1',select=Disease),
              #sample.info[,'Disease',drop=FALSE],
              plot.MDS=FALSE,plot.importance=FALSE,print.result=FALSE,
              guess.num=0,boxplot.top=0, 
              title="Normal vs. Crohn",boxplot.y="Abundance",
              seed=100)

ca.ba.rf <- random_forest(ba.carnitine %>% dplyr::select(-Human.Per),
              subset(sample.info,Type=='COMBO' | Type=='PLEASE-T1',select=Disease),
              #sample.info[,'Disease',drop=FALSE],
              plot.MDS=FALSE,plot.importance=FALSE,print.result=FALSE,
              guess.num=0,boxplot.top=0, 
              title="Normal vs. Crohn",boxplot.y="Abundance",
              seed=100)

ca.rf <- random_forest(ba.carnitine %>% dplyr::select(contains("carnitine")),
              subset(sample.info,Type=='COMBO' | Type=='PLEASE-T1',select=Disease),
              #sample.info[,'Disease',drop=FALSE],
              plot.MDS=FALSE,plot.importance=FALSE,print.result=FALSE,
              guess.num=0,boxplot.top=0, 
              title="Normal vs. Crohn",boxplot.y="Abundance",
              seed=100)

ca.dna.rf<- random_forest(ba.carnitine %>% dplyr::select(contains("carnitine"),Human.Per),
              subset(sample.info,Type=='COMBO' | Type=='PLEASE-T1',select=Disease),
              #sample.info[,'Disease',drop=FALSE],
              plot.MDS=FALSE,plot.importance=FALSE,print.result=FALSE,
              guess.num=0,boxplot.top=0, 
              title="Normal vs. Crohn",boxplot.y="Abundance",
              seed=100)

ba.rf <- random_forest(ba.carnitine %>% dplyr::select(-contains("carnitine"),-Human.Per),
              subset(sample.info,Type=='COMBO' | Type=='PLEASE-T1',select=Disease),
              #sample.info[,'Disease',drop=FALSE],
              plot.MDS=FALSE,plot.importance=FALSE,print.result=FALSE,
              guess.num=0,boxplot.top=0, 
              title="Normal vs. Crohn",boxplot.y="Abundance",
              seed=100)

ba.dna.rf <- random_forest(ba.carnitine %>% dplyr::select(-contains("carnitine"),Human.Per),
              subset(sample.info,Type=='COMBO' | Type=='PLEASE-T1',select=Disease),
              #sample.info[,'Disease',drop=FALSE],
              plot.MDS=FALSE,plot.importance=FALSE,print.result=FALSE,
              guess.num=0,boxplot.top=0, 
              title="Normal vs. Crohn",boxplot.y="Abundance",
              seed=100)

####################################################
roc.data <- 
### Human DNA + Primary and Secondary BA + Carnitines
dplyr::rename(add_rownames(as.data.frame(ca.ba.dna.rf$votes[,1,drop=F]),var='Sample'),Carnitines.BA.DNA=Control) %>%
### Carnitines, BA
dplyr::full_join(dplyr::rename(add_rownames(as.data.frame(ca.ba.rf$votes[,1,drop=F]),var='Sample'),Carnitines.BA=Control),by='Sample') %>%
### Carnitines 
dplyr::full_join(dplyr::rename(add_rownames(as.data.frame(ca.rf$votes[,1,drop=F]),var='Sample'),Carnitines=Control),by='Sample') %>%
### Carnitines + Human DNA
dplyr::full_join(dplyr::rename(add_rownames(as.data.frame(ca.dna.rf$votes[,1,drop=F]),var='Sample'),Carnitines.DNA =Control),by='Sample') %>%
### Bile acid
dplyr::full_join(dplyr::rename(add_rownames(as.data.frame(ba.rf$votes[,1,drop=F]),var='Sample'),BA=Control),by='Sample') %>%
### Bile acid + Human DNA
dplyr::full_join(dplyr::rename(add_rownames(as.data.frame(ba.dna.rf$votes[,1,drop=F]),var='Sample'),BA.DNA=Control),by='Sample') %>%
### sample infor
dplyr::left_join(dplyr::select(add_rownames(sample.info,var='Sample'),Sample,Disease),by='Sample') %>%
as.data.frame


#### ROC plot with AUC and smoothing
roc <- plot.roc(roc.data[,"Disease",drop=TRUE], 
         roc.data[,"Carnitines.BA.DNA",drop=TRUE], 
         main="ROC comparison", percent=TRUE, col="#e41a1c",type="n",
         print.auc=TRUE,print.auc.pattern="Carnitines.BA.DNA AUC:%.1f%%",               print.auc.col="#e41a1c",print.auc.y=80)
lines(smooth(roc, method = "binormal"), col="#e41a1c")

roc <- plot.roc(roc.data[,"Disease",drop=TRUE],                                        roc.data[,"Carnitines.BA",drop=TRUE],add=TRUE,  
              percent=TRUE, col="#377eb8",type="n",
              print.auc=TRUE,print.auc.pattern="Carnitines.BA AUC:%.1f%%",             print.auc.col="#377eb8",print.auc.y=77) 
lines(smooth(roc, method = "binormal"), col="#377eb8")

roc <- plot.roc(roc.data[,"Disease",drop=TRUE],                                                roc.data[,"Carnitines",drop=TRUE],add=TRUE,   
               percent=TRUE, col="#4daf4a",type="n",
              print.auc=TRUE,print.auc.pattern="Carnitines AUC:%.1f%%",               print.auc.col="#4daf4a",print.auc.y=74) 
lines(smooth(roc, method = "binormal"), col="#4daf4a")

roc<-plot.roc(roc.data[,"Disease",drop=TRUE], roc.data[,"Carnitines.DNA",drop=TRUE],add=TRUE,
          percent=TRUE, col="#d95f02",print.thres=TRUE,type="n",
          print.auc=TRUE,print.auc.pattern="Carnitines.DNA AUC:%.1f%%",                  print.auc.col="#d95f02",print.auc.y=71) 
lines.roc(smooth(roc, method = "binormal"), col="#d95f02",print.thres=FALSE)

roc<-plot.roc(roc.data[,"Disease",drop=TRUE], roc.data[,"BA",drop=TRUE],add=TRUE,   
               percent=TRUE, col="#984ea3",type="n",
              print.auc=TRUE,print.auc.pattern="BA AUC:%.1f%%",                print.auc.col="#984ea3",print.auc.y=68) 
lines(smooth(roc, method = "binormal"), col="#984ea3")

roc<-plot.roc(roc.data[,"Disease",drop=TRUE],                                           roc.data[,"BA.DNA",drop=TRUE],add=TRUE,  
         percent=TRUE, col="black",print.thres=FALSE,type="n",
         print.auc=TRUE,print.auc.pattern="BA.DNA AUC:%.1f%%",                          print.auc.col="black",print.auc.y=65)
lines(smooth(roc, method = "binormal"), col="black")


#
legend("bottomright", legend = c("Carnitines.BA.DNA","Carnitines.BA", "Carnitines", "Carnitines.DNA","BA","BA.DNA"), 
       col = c("#e41a1c", "#377eb8","#4daf4a","#d95f02", "#984ea3", "black"),lwd = 2,y.intersp=0.5)

rm(ca.ba.dna.rf)
rm(ca.ba.rf)
rm(ca.rf)
rm(ca.dna.rf)
rm(ba.dna.rf)
```

### Random Forest: Normal vs. IBD (no antibiotics)
```{r known_metabolite_ba_carnitine_random_forest_of_normal_vs_crohn_no_antibiotics,cache=TRUE}
random_forest(ba.carnitine,
              subset(sample.info,Antibiotics.visit=='Not.Use' & (Type=='COMBO' | Type=='PLEASE-T1'),select=Disease),
              plot.MDS=TRUE,plot.importance=TRUE,print.result=FALSE,
              guess.num=50,boxplot.top=10,
              title="Normal vs. Crohn (no antibiotics)",boxplot.y="Abundance",
              seed=100)
```

### Random Forest: Cluster1 vs. Cluster2 (baseline)
```{r known_metabolite_ba_carnitine_random_forest_of_clusters_baseline,cache=TRUE}
random_forest(ba.carnitine,
              subset(sample.info,(Type=='COMBO' | Type=='PLEASE-T1') & !is.na(Cluster),select=Cluster),
              plot.MDS=TRUE,plot.importance=TRUE,print.result=FALSE,
              guess.num=50,boxplot.top=10,
              title="Clusters",boxplot.y="Abundance",
              seed=100)
```

### Random Forest: EEN vs. antiTNF (T4)
```{r known_metabolite_ba_carnitine_random_forest_of_een_antitnf_t4,cache=TRUE}
random_forest(ba.carnitine,
              subset(sample.info,Type=='PLEASE-T4' & (Treatment.Specific=='EEN' | Treatment.Specific=='antiTNF'),select=Treatment.Specific),
              plot.MDS=TRUE,plot.importance=TRUE,print.result=FALSE,
              guess.num=50,boxplot.top=10,
              title="Treatment",boxplot.y="Abundance",
              seed=100)
```

### Random Forest: Response vs. Non-Response (T1) (exclude PEN subjects)
```{r known_metabolite_ba_carnitine_random_forest_of_response_t1,cache=TRUE}
random_forest(ba.carnitine,
              subset(sample.info,Type=='PLEASE-T1' & !is.na(Response) & (Treatment.Specific=='EEN' | Treatment.Specific=='antiTNF'),select=Response),
              plot.MDS=TRUE,plot.importance=TRUE,print.result=FALSE,
              guess.num=50,boxplot.top=10,
              title="Response",boxplot.y="Abundance",
              seed=100)
```


### Random Forest: Response vs. Non-Response (T4) (exclude PEN subjects)
```{r known_metabolite_ba_carnitine_random_forest_of_response_t4,cache=TRUE}
random_forest(ba.carnitine,
              subset(sample.info,Type=='PLEASE-T4' & !is.na(Response) & (Treatment.Specific=='EEN' | Treatment.Specific=='antiTNF'),select=Response),
              plot.MDS=TRUE,plot.importance=TRUE,print.result=FALSE,
              guess.num=50,boxplot.top=10,
              title="Response",boxplot.y="Abundance",
              seed=100)
```

### Random Forest: Response vs. Non-Response (T1) (EEN only)
```{r known_metabolite_ba_carnitine_random_forest_of_een_response_t1,cache=TRUE}
random_forest(ba.carnitine,
              subset(sample.info,Type=='PLEASE-T1' & !is.na(Response) & Treatment.Specific=='EEN',select=Response),
              plot.MDS=TRUE,plot.importance=TRUE,print.result=FALSE,
              guess.num=50,boxplot.top=10,
              title="Response",boxplot.y="Abundance",
              seed=100)
```


### Random Forest: Response vs. Non-Response (T4) (EEN only)
```{r known_metabolite_ba_carnitine_random_forest_of_een_response_t4,cache=TRUE}
random_forest(ba.carnitine,
              subset(sample.info,Type=='PLEASE-T4' & !is.na(Response) & Treatment.Specific=='EEN',select=Response),
              plot.MDS=TRUE,plot.importance=TRUE,print.result=FALSE,
              guess.num=50,boxplot.top=10,
              title="Response",boxplot.y="Abundance",
              seed=100)
```

### Random Forest: Response vs. Non-Response (T1) (antiTNF only)
```{r known_metabolite_ba_carnitine_random_forest_of_antitnf_response_t1,cache=TRUE}
random_forest(ba.carnitine,
              subset(sample.info,Type=='PLEASE-T1' & !is.na(Response) & Treatment.Specific=='antiTNF',select=Response),
              plot.MDS=TRUE,plot.importance=TRUE,print.result=FALSE,
              guess.num=50,boxplot.top=10,
              title="Response",boxplot.y="Abundance",
              seed=100)
```


### Random Forest: Response vs. Non-Response (T4) (antiTNF only)
```{r known_metabolite_ba_carnitine_random_forest_of_antitnf_response_t4,cache=TRUE}
random_forest(ba.carnitine,
              subset(sample.info,Type=='PLEASE-T4' & !is.na(Response) & Treatment.Specific=='antiTNF',select=Response),
              plot.MDS=TRUE,plot.importance=TRUE,print.result=FALSE,
              guess.num=50,boxplot.top=10,
              title="Response",boxplot.y="Abundance",
              seed=100)
```

## Compare Random Forest prediction of (1) Human DNA, bile acids and acyl-Carnitines (2) 84 metabolites (3) all known (4) all (5) microbiome (6) pathway
```{r known_metabolite_compare_random_forst_prediction,cache=TRUE,fig.width=10,fig.height=10}

##################################################
###### normal vs. IBD
sind <- Reduce(intersect, 
       list(
         rownames(allmdata.nor),
         rownames(ba.carnitine),
         rownames(kmetabolite.cp1),
         rownames(m84data.cp1),
         rownames(pathway.cp1),
         rownames(bact.cp1),
         rownames(kmer.nor)
         ))


#### make sure all the random forest use the same sample set
rf.all.metabolites <- random_forest(allmdata.nor[sind,],
              subset(sample.info,Type=='COMBO' | Type=='PLEASE-T1',select=Disease),
              plot.MDS=FALSE,plot.importance=FALSE,print.result=FALSE,
              guess.num=0,boxplot.top=0, #cutoff=c(0.2,0.8),
              title="Normal vs. Crohn",boxplot.y="Abundance",
              seed=100)
rf.ba.carnitine <- random_forest(ba.carnitine[sind,],
              subset(sample.info,Type=='COMBO' | Type=='PLEASE-T1',select=Disease),
              plot.MDS=FALSE,plot.importance=FALSE,print.result=FALSE,
              guess.num=0,boxplot.top=0,
              title="Normal vs. Crohn",boxplot.y="Abundance",
              seed=100)
rf.known.metabolites <- random_forest(kmetabolite.cp1[sind,],
              sample.info[,'Disease',drop=FALSE],
              plot.MDS=FALSE,plot.importance=FALSE,print.result=FALSE,
              guess.num=0,boxplot.top=0,
              title="Normal vs. Crohn (no antibiotics)",boxplot.y="Abundance",
              seed=100)
rf.m84 <- random_forest(m84data.cp1[sind,],
              sample.info[,'Disease',drop=FALSE],
              plot.MDS=FALSE,plot.importance=FALSE,print.result=FALSE,
              guess.num=0,boxplot.top=0,
              title="Normal vs. Crohn",boxplot.y="Abundance",
              seed=100)
rf.pathway <- random_forest(pathway.cp1[sind,],
              sample.info[,'Disease',drop=FALSE],
              plot.MDS=FALSE,plot.importance=FALSE,print.result=FALSE,
              guess.num=0,boxplot.top=0,
              title="Normal vs. Crohn",boxplot.y="Abundance(%)",
              seed=100)
rf.bact <- random_forest(bact.cp1[sind,],
              sample.info[,'Disease',drop=FALSE],
              plot.MDS=FALSE,plot.importance=FALSE,print.result=FALSE,
              guess.num=0,boxplot.top=0,
              title="Normal vs. Crohn",boxplot.y="Abundance(%)",
              seed=100)
rf.kmer <- random_forest(kmer.nor[sind,], 
              subset(sample.info,Type=='COMBO' | Type=='PLEASE-T1',select=Disease),
              plot.MDS=FALSE,plot.importance=FALSE,print.result=FALSE,
              guess.num=0,boxplot.top=0,
              title="Normal vs. Crohn",boxplot.y="Abundance(%)",
              seed=100)
####
roc.data <- 
### Human DNA + Primary and Secondary BA + Carnitines
dplyr::rename(add_rownames(as.data.frame(rf.ba.carnitine$votes[,1,drop=F]),var='Sample'),Carnitines=Control) %>%
### UCLA.Metabolites
dplyr::full_join(dplyr::rename(add_rownames(as.data.frame(rf.m84$votes[,1,drop=F]),var='Sample'),UCLA.Metabolites=Control),by='Sample') %>%
### Known.Metabolites 
dplyr::full_join(dplyr::rename(add_rownames(as.data.frame(rf.known.metabolites$votes[,1,drop=F]),var='Sample'),Known.Metabolites=Control),by='Sample') %>%
### All.Metabolites
dplyr::full_join(dplyr::rename(add_rownames(as.data.frame(rf.all.metabolites$votes[,1,drop=F]),var='Sample'),All.Metabolites=Control),by='Sample') %>%
### Pathway
dplyr::full_join(dplyr::rename(add_rownames(as.data.frame(rf.pathway$votes[,1,drop=F]),var='Sample'),Pathway=Control),by='Sample') %>%
### Bact
dplyr::full_join(dplyr::rename(add_rownames(as.data.frame(rf.bact$votes[,1,drop=F]),var='Sample'),Bact=Control),by='Sample') %>%
### kmer
dplyr::full_join(dplyr::rename(add_rownames(as.data.frame(rf.kmer $votes[,1,drop=F]),var='Sample'),Kmer6=Control),by='Sample') %>%
### sample infor
dplyr::left_join(dplyr::select(add_rownames(sample.info,var='Sample'),Sample,Disease),by='Sample') %>%
as.data.frame
#### ROC plot
#### I manually checked serveral thresholds, the direction seems correct
roc<-plot.roc(roc.data[,"Disease",drop=TRUE],print.thres=TRUE, roc.data[,"Carnitines",drop=TRUE], 
         main="ROC comparison", percent=TRUE, col="#e41a1c") 
roc<-lines.roc(roc.data[,"Disease",drop=TRUE], roc.data[,"UCLA.Metabolites",drop=TRUE],   percent=TRUE, col="#377eb8") 
roc<-lines.roc(roc.data[,"Disease",drop=TRUE],print.thres=TRUE, roc.data[,"Known.Metabolites",drop=TRUE],   percent=TRUE, col="#4daf4a") 
roc<-plot.roc(roc.data[,"Disease",drop=TRUE], roc.data[,"All.Metabolites",drop=TRUE],add=TRUE,
          percent=TRUE, col="#d95f02",print.thres=TRUE) 
roc<-lines.roc(roc.data[,"Disease",drop=TRUE], roc.data[,"Pathway",drop=TRUE],   percent=TRUE, col="#984ea3") 
roc<-plot.roc(roc.data[,"Disease",drop=TRUE], roc.data[,"Bact",drop=TRUE],add=TRUE,  
         percent=TRUE, col="black",print.thres=FALSE)
roc<-plot.roc(roc.data[,"Disease",drop=TRUE], roc.data[,"Kmer6",drop=TRUE],add=TRUE,  
         percent=TRUE, col="#8dd3c7",print.thres=FALSE)
#lines(smooth(roc, method = "density"))
legend("bottomright", legend = c("Human DNA + Primary and Secondary BA + Carnitines","UCLA.Metabolites", "Known.Metabolites", "All.Metabolites","Pathway","Microbiome","Kmer6"), col = c("#e41a1c", "#377eb8","#4daf4a","#d95f02", "#984ea3", "black","#8dd3c7"),lwd = 2)


#### ROC plot with AUC and smoothing
roc<-plot.roc(roc.data[,"Disease",drop=TRUE], roc.data[,"Carnitines",drop=TRUE], 
         main="ROC comparison", percent=TRUE, col="#e41a1c",type="n",
         print.auc=TRUE,print.auc.pattern="Human+BA+Carnitine AUC:%.1f%%", print.auc.col="#e41a1c",print.auc.y=80)
lines(smooth(roc, method = "binormal"), col="#e41a1c")
roc<-plot.roc(roc.data[,"Disease",drop=TRUE], roc.data[,"UCLA.Metabolites",drop=TRUE],add=TRUE,  
              percent=TRUE, col="#377eb8",type="n",
              print.auc=TRUE,print.auc.pattern="UCLA.Metabolites AUC:%.1f%%", print.auc.col="#377eb8",print.auc.y=77) 
lines(smooth(roc, method = "binormal"), col="#377eb8")
roc<-plot.roc(roc.data[,"Disease",drop=TRUE], roc.data[,"Known.Metabolites",drop=TRUE],add=TRUE,   
               percent=TRUE, col="#4daf4a",type="n",
              print.auc=TRUE,print.auc.pattern="Known.Metabolites AUC:%.1f%%", print.auc.col="#4daf4a",print.auc.y=74) 
lines(smooth(roc, method = "binormal"), col="#4daf4a")
roc<-plot.roc(roc.data[,"Disease",drop=TRUE], roc.data[,"All.Metabolites",drop=TRUE],add=TRUE,
          percent=TRUE, col="#d95f02",print.thres=TRUE,type="n",
          print.auc=TRUE,print.auc.pattern="All.Metabolites AUC:%.1f%%", print.auc.col="#d95f02",print.auc.y=71) 
lines.roc(smooth(roc, method = "binormal"), col="#d95f02",print.thres=FALSE)
roc<-plot.roc(roc.data[,"Disease",drop=TRUE], roc.data[,"Pathway",drop=TRUE],add=TRUE,   
               percent=TRUE, col="#984ea3",type="n",
              print.auc=TRUE,print.auc.pattern="Pathway AUC:%.1f%%", print.auc.col="#984ea3",print.auc.y=68) 
lines(smooth(roc, method = "binormal"), col="#984ea3")
roc<-plot.roc(roc.data[,"Disease",drop=TRUE], roc.data[,"Bact",drop=TRUE],add=TRUE,  
         percent=TRUE, col="black",print.thres=FALSE,type="n",
         print.auc=TRUE,print.auc.pattern="Microbiome AUC:%.1f%%", print.auc.col="black",print.auc.y=65)
lines(smooth(roc, method = "binormal"), col="black")
roc<-plot.roc(roc.data[,"Disease",drop=TRUE], roc.data[,"Kmer6",drop=TRUE],add=TRUE,  
         percent=TRUE, col="#8dd3c7",print.thres=FALSE,type="n",
         print.auc=TRUE,print.auc.pattern="Kmer6 AUC:%.1f%%", print.auc.col="#8dd3c7",print.auc.y=62)
lines(smooth(roc, method = "binormal"), col="#8dd3c7")
#
legend("bottomright", legend = c("Human DNA + Primary and Secondary BA + Carnitines","UCLA.Metabolites", "Known.Metabolites", "All.Metabolites","Pathway","Microbiome","Kmer6"), col = c("#e41a1c", "#377eb8","#4daf4a","#d95f02", "#984ea3", "black","#8dd3c7"),lwd = 2)

#rm(roc)
#rm(roc.data)
#rm(rf.all.metabolites)
#rm(rf.ba.carnitine)
#rm(rf.known.metabolites)
#rm(rf.m84)
#rm(rf.pathway)
#rm(rf.bact)
```


## Acyl-Carnitines
```{r known_metabolite_acyl_carnitines,cache=TRUE,fig.width=10,fig.height=10}
carnitine.dat <- dplyr::select(add_rownames(as.data.frame(kmdata.nor),var='Sample'),
              Sample,contains('carnitine')) %>% 
              add_back_rownames(row.var = 'Sample')

#######################
### Normal + PLEASE.T1
sind <- rownames(subset(sample.info,Type=='COMBO' | Type=='PLEASE-T1')) %>%
        intersect(.,rownames(carnitine.dat))

dat.nor <- carnitine.dat[sind,]
for (pw in colnames(dat.nor)){
  pw.dat <- dat.nor[,pw,drop=TRUE]
  Me <- median(pw.dat)
  Ma <- mad(pw.dat)
  if (Ma==0){Me <- mean(pw.dat);Ma <- sd(pw.dat)}
  pw.dat <- (pw.dat-Me)/Ma
  pw.dat[pw.dat < -3] <- -3
  pw.dat[pw.dat > 3] <- 3
  dat.nor[,pw] <- pw.dat
}

pheatmap(t(dat.nor),
         border_color="grey60",
         cluster_cols = calculate_distance(X = m84data.cp1, method = "Bray"),
         clustering_distance_rows = 'euclidean',
         #clustering_method = "complete",
         annotation=sample.info[,c("Disease","Time","Response","Treatment.Specific","Cluster","log.FCP","Antibiotics.visit","Steroids")],
         #color =hmcols, breaks=bk,
         main='COMBO+PLEASE-T1',
         show_rownames = TRUE
         #scale="row",
         #gaps_col = c(26, 111,191,273)
         )
rm(sind)
rm(dat.nor)

#######################
### Normal + PLEASE.T4
sind <- rownames(subset(sample.info,Type=='COMBO' | Type=='PLEASE-T4')) %>%
        intersect(.,rownames(carnitine.dat))

dat.nor <- carnitine.dat[sind,]
for (pw in colnames(dat.nor)){
  pw.dat <- dat.nor[,pw,drop=TRUE]
  Me <- median(pw.dat)
  Ma <- mad(pw.dat)
  if (Ma==0){Me <- mean(pw.dat);Ma <- sd(pw.dat)}
  pw.dat <- (pw.dat-Me)/Ma
  pw.dat[pw.dat < -3] <- -3
  pw.dat[pw.dat > 3] <- 3
  dat.nor[,pw] <- pw.dat
}

pheatmap(t(dat.nor),
         border_color="grey60",
         cluster_cols = calculate_distance(X = m84data.cp1, method = "Bray"),
         clustering_distance_rows = 'euclidean',
         #clustering_method = "complete",
         annotation=sample.info[,c("Disease","Time","Response","Treatment.Specific","Cluster","log.FCP","Antibiotics.visit","Steroids")],
         #color =hmcols, breaks=bk,
         main='COMBO+PLEASE-T4',
         show_rownames = TRUE
         #scale="row",
         #gaps_col = c(26, 111,191,273)
         )
rm(sind)
rm(dat.nor)

######################################################################
######################################################################
#### correlation plot
plot.correlation.heatmap = function(cor.matrix,
                                    row.filter=0.5,
                                    col.filter=0.5
                                    ){
  cor.matrix <- cor.matrix[
  apply(cor.matrix,1,function(X){max(abs(X),na.rm=TRUE)>row.filter}),
  apply(cor.matrix,2,function(X){max(abs(X),na.rm=TRUE)>col.filter})
  ]
  
  bk <- unique(c(seq(-1,-0.4,length=10),
                 seq(-0.4,0.4,length=50),
                 seq(0.4,1,length=10)))
  hmcols<- colorRampPalette(c('#4575b4','#91bfdb','#e0f3f8',
                            '#ffffbf','#fee090','#fc8d59','#d73027'
                            ))(length(bk)-1)

  pheatmap(cor.matrix, 
           scale='none',fontsize_row = 10, 
           clustering_distance_cols = "manhattan",
           clustering_distance_rows = "manhattan",
           cluster_cols = TRUE,
           color =hmcols, breaks=bk,
           cellwidth = 10, cellheight = 10
  )                              
}

### Control
sind <- rownames(subset(sample.info,Type=='COMBO')) %>%
        intersect(.,rownames(carnitine.dat)) %>%
        intersect(.,rownames(taxa.data))

ca.bact.cor <- cor(taxa.data[sind,colSums(taxa.data[sind,])>0],
                              carnitine.dat[sind,],
                              method='spearman')
plot.correlation.heatmap(ca.bact.cor,row.filter=0,col.filter=0)
rm(sind)
rm(ca.bact.cor)

### PLEASE-T1
sind <- rownames(subset(sample.info,Type=='PLEASE-T1')) %>%
        intersect(.,rownames(carnitine.dat)) %>%
        intersect(.,rownames(taxa.data))

ca.bact.cor <- cor(taxa.data[sind,colSums(taxa.data[sind,])>0],
                              carnitine.dat[sind,],
                              method='spearman')
plot.correlation.heatmap(ca.bact.cor,row.filter=0,col.filter=0)
rm(sind)
rm(ca.bact.cor)


### PLEASE-T4
sind <- rownames(subset(sample.info,Type=='PLEASE-T4')) %>%
        intersect(.,rownames(carnitine.dat)) %>%
        intersect(.,rownames(taxa.data))

ca.bact.cor <- cor(taxa.data[sind,colSums(taxa.data[sind,])>0],
                              carnitine.dat[sind,],
                              method='spearman')
plot.correlation.heatmap(ca.bact.cor,row.filter=0,col.filter=0)
rm(sind)
rm(ca.bact.cor)
####
rm(carnitine.dat)
```


# Analyze all metabolite data

```{r all_metabolite_check_the_data,cache=TRUE}
### How many metabolites in total
dim(allmdata.nor)
```


## Normal vs. Crohn-T1
```{r all_metabolites_normal_vs_crohn_t1,cache=TRUE}
###################### Normal and disease T1 #####################
allmdata.cp1 <- subset_data_by_sample_info(allmdata.nor,sample.info,
                                       Type=='COMBO' | Type=='PLEASE-T1')
dim(allmdata.cp1)
```

### Calculate distance
```{r all_metabolites_calculate_distance,cache=TRUE}
allmeta.distance.method <- "Bray"
#### distance for taxa
dist.allmeta.cp1 <- calculate_distance(X = allmdata.cp1, method = allmeta.distance.method )

### distribution
hist(dist.allmeta.cp1,xlab='Distance',main='Metabolites (COMBO and PLEASE-T1)',breaks=20)
```

### Find the best clustering
```{r all_metabolites_find_the_best_clustering,cache=TRUE}
allmeta.pamk.best <- find_best_clustering(dist.allmeta.cp1,plot.results=TRUE)
```

### MDS analysis
```{r all_metabolites_MDS,cache=TRUE,results='hide'}
### MDS
set.seed(10)
allmeta.MDS <- metaMDS(dist.allmeta.cp1)

### Normal/Disease
p <- add_rownames(as.data.frame(allmeta.MDS$points),var='Sample') %>%
  left_join(dplyr::select(add_rownames(sample.info,var='Sample'),
                          Sample,Disease),
            by='Sample') %>% 
  ggplot(aes(x=MDS1,y=MDS2))+ geom_point(aes(colour = Disease)) +
  scale_color_manual(values=c("#0066CC", "#FF3333"),name="") +
  labs(title = 'metaMDS') 
print(p)

### Normal/Disease, cluster
### These are bacterial clusters !!
p <- add_rownames(as.data.frame(allmeta.MDS$points),var='Sample') %>%
  left_join(dplyr::select(add_rownames(sample.info,var='Sample'),
                          Sample,Disease,Cluster),
            by='Sample') %>% 
  ggplot(aes(x=MDS1,y=MDS2))+ geom_point(aes(colour = Disease,shape=Cluster)) +
  scale_color_manual(values=c("#0066CC", "#FF3333"),name="") +
  scale_shape_manual(values=c(20,4),name="") +
  labs(title = 'metaMDS') 
print(p)



#### Fugal percentage 
p <- add_rownames(as.data.frame(allmeta.MDS$points),var='Sample') %>%
  left_join(dplyr::select(add_rownames(sample.info,var='Sample'),
                          Sample,Disease,Cluster,Fungi.Per),
            by='Sample') %>% 
  mutate(Fungi.Per.sqrt=sqrt(Fungi.Per)) %>%
  ggplot(aes(x=MDS1,y=MDS2))+ 
  geom_point(aes(colour = Disease,shape=Cluster,size = Fungi.Per.sqrt)) +
  scale_color_manual(values=c("#0066CC", "#FF3333"),name="") +
  scale_shape_manual(values=c(20,1),name="") +
  scale_size_area()+
  labs(title = 'metaMDS') 
print(p)

#### Human DNA percentage
p <- add_rownames(as.data.frame(allmeta.MDS$points),var='Sample') %>%
  left_join(dplyr::select(add_rownames(sample.info,var='Sample'),
                          Sample,Disease,Cluster,Human.Per),
            by='Sample') %>% 
  ggplot(aes(x=MDS1,y=MDS2))+ 
  geom_point(aes(colour = Disease,shape=Cluster,size = Human.Per)) +
  scale_color_manual(values=c("#0066CC", "#FF3333"),name="") +
  scale_shape_manual(values=c(20,1),name="") +
  scale_size_area()+
  labs(title = 'metaMDS') 
print(p)

rm(p)
```


### Random Forest 
```{r all_metabolites_random_forest_of_normal_vs_crohn,cache=TRUE}
random_forest(allmdata.cp1,sample.info[,'Disease',drop=FALSE],
              plot.MDS=TRUE,plot.importance=TRUE,print.result=FALSE,
              guess.num=0,boxplot.top=10,
              title="Normal vs. Crohn",boxplot.y="Abundance",
              seed=100)

```

### Wilcoxon rank test of normal vs. normal
```{r all_metabolites_rank_test_of_normal_vs_normal,cache=TRUE}
### "Control" vs. "control"  (as a control test)
sam.tmp <- subset(sample.info,Disease=='Control',select=Disease)
sam.tmp[1:13,1] <- 'control' 
###  
allmeta.cc.wil <- wilcox_test(allmdata.cp1,sam.tmp,                      
                          plot.heatmap=FALSE,boxplot.top=10,boxplot.y='Abundance',parallel=TRUE)
cat("signifcant metabolites",sum(allmeta.cc.wil[,2]<0.05))
###
#write.csv(pathway.cp1.wil,file=paste('../3_Result/2015_03_05_Pathway_COMBO_PLEASET1_Wilcox.csv',sep=''),row.names=TRUE)
rm(sam.tmp)
```


### Wilcoxon rank test of normal vs. Crohn
```{r all_metabolites_rank_test_of_normal_vs_crohn,cache=TRUE}
### only plot the top 10 significant ones
allmeta.cp1.wil <- wilcox_test(allmdata.cp1,sample.info[,'Disease',drop=FALSE],
              plot.heatmap=FALSE,boxplot.top=10,boxplot.y='Abundance',parallel=TRUE)
cat("signifcant metabolites",sum(allmeta.cp1.wil[,2]<0.05))
### 
write.csv(allmeta.cp1.wil, file=paste('../3_Result/2015_03_05_All_Metabolites_COMBO_PLEASET1_Wilcox.csv',sep=''),row.names=TRUE)
```

### Wilcoxon rank test of normal vs. Cluster 1
```{r all_metabolites_rank_test_of_normal_vs_cluster_1,cache=TRUE}
allmeta.normal.cluster1.wil <- wilcox_test(allmdata.cp1,
       subset(sample.info,Disease=='Control'|(Disease=='Crohn'&Cluster=='cluster 1'),select=Disease),
                               plot.heatmap=FALSE,boxplot.top=10,boxplot.y='Abundance',parallel=TRUE)

cat("signifcant metabolites",sum(allmeta.normal.cluster1.wil[,2]<0.05))
###
write.csv(allmeta.normal.cluster1.wil, file=paste('../3_Result/2015_03_05_All_Metabolite_COMBO_vs_Cluster1_Wilcox.csv',sep=''),row.names=FALSE)
```

### Wilcoxon rank test of normal vs. Cluster 2
```{r all_metabolites_rank_test_of_normal_vs_cluster_2,cache=TRUE}
allmeta.normal.cluster2.wil <- wilcox_test(allmdata.cp1,
      subset(sample.info,Disease=='Control'|(Disease=='Crohn'&Cluster=='cluster 2'),select=Disease),
      plot.heatmap=FALSE,boxplot.top=10,boxplot.y='Abundance',parallel=TRUE)

cat("signifcant metabolites",sum(allmeta.normal.cluster2.wil[,2]<0.05))
###
write.csv(allmeta.normal.cluster2.wil, file=paste('../3_Result/2015_03_05_All_Metabolites_COMBO_vs_Cluster2_Wilcox.csv',sep=''),row.names=FALSE)
```

